\section{Nucleotide tasks}\label{sec:nucleotide}

The genome contains a vast amount of information regarding protein-coding genes and the regulatory DNA and RNA sequences that control their expression. In this section, we evaluated our model on nucleotide sequence generation tasks, including both unconditional generation and cross-domain generation, specifically DNA to RNA generation (guide RNA design) and protein to RNA generation.

\subsection{Unconditional RNA generation}\label{sec:unconditional_rna_generation}
Designing RNA molecules is crucial for advancing RNA vaccines, nucleic acid therapies, and various biotechnological applications. In this section, we evaluate the proficiency of NatureLM in generating RNA sequences without any conditional prompts. For evaluation purposes, we constrained the generated RNA sequences to a maximum length of 1024 nucleotides. An example of an unconditionally generated sequence is provided below:
\begin{example}
\noindent\textbf{Instruction}: \rna{} \\
\noindent\textbf{Response}: \rna{} C C A C G G A G C C \erna{}
\end{example}

We assessed the quality of the generated RNA sequences by calculating their Minimum Free Energy (MFE) using RNAfold~\cite{Lorenz2011} (see Section \ref{app:rna_generation} for details). A lower MFE value indicates a potentially more stable RNA secondary structure. For each model, we generated 5,000 sequences and computed their MFE values. To establish a baseline for comparison, we generated control sequences and computed their average MFE values. Specifically, for each generated sequence, we created:

\noindent(1) {\em Shuffled Sequences}: For each generated sequence, we created a new sequence by randomly shuffling its nucleotides, thereby preserving the original nucleotide composition and length but potentially disrupting any inherent structural motifs. 

\noindent(2) {\em Random Sequences}: For each generated sequence, we created an entirely random sequence of the same length, where each nucleotide position was independently sampled from the four nucleotides (A, G, C, U) with equal probability. This baseline represents sequences with no designed structure or composition bias.

As a reference for the MFE values of natural RNA sequences, we randomly sampled 5,000 sequences of length up to 1,024 nucleotides from RNAcentral\footnote{\url{https://rnacentral.org/}} and computed their MFE values.

The average MFE values are reported in Table \ref{tab:unconditional_rna_generation_mfe}.

\begin{table}[!htpb]
\centering
\begin{tabular}{lccccccc}
\toprule
               & MFE (kcal/mol) & Retrieved Rfam Families \\
\midrule
RNAcentral     & -165.4 \\
Shuffled sequences & -156.4 \\
Random sequences &  -142.0 \\
\ourM{} (1B)   & -160.6 & 23 \\
\ourM{} (8B)   & -170.6 & 38 \\
\ourM{} (8x7B) & -177.1 & 165 \\
\bottomrule
\end{tabular}
\caption{Average MFE values (in kcal/mol) of RNA sequences generated by different methods and the number of unique Rfam families retrieved by different models. \textbf{MFE} denotes the mean Minimum Free Energy of the sequences. \textbf{Retrieved Rfam Families} represents the count of unique RNA families identified in the generated sequences using \texttt{cmscan}. }
\label{tab:unconditional_rna_generation_mfe}
\end{table}

From the results, we observe that larger models tend to generate RNA sequences with lower (more negative) MFE values, indicating potentially more stable secondary structures. Additionally, shuffling and randomizing the sequences result in higher (less negative) MFE values, suggesting that the original sequences generated by our models have structural features that contribute to stability.

To evaluate the diversity of the RNA sequences generated by NatureLM, we compared them to known RNA families in Rfam~\cite{10.1093/nar/gkaa1047}. We used \texttt{cmscan} from the Infernal toolkit~\cite{10.1093/bioinformatics/btt509} to search for structural similarities between our generated sequences and the Rfam database (see Section \ref{app:rna_generation} for details). As shown in Table~\ref{tab:unconditional_rna_generation_mfe}, larger models retrieved a significantly higher number of unique Rfam families than smaller models: the 1B, 8B, and 8x7B models retrieved 23, 38, and 165 unique families, respectively, covering a wider range of RNA functions. These results suggest that larger models not only generate more stable sequences but also encompass a more diverse set of RNA structures and functions.

% Furthermore, to evaluate the diversity of the RNA sequences generated by our models, we compared them to known RNA families in Rfam~\cite{10.1093/nar/gkaa1047}. We used \texttt{cmscan} from the Infernal toolkit~\cite{10.1093/bioinformatics/btt509} to search for matches between our generated sequences and Rfam covariance models (see Section \ref{app:rna_generation} for details). We calculated the number of unique Rfam families identified in the generated sequences. The \ourM{} (1B), \ourM{} (8B), and \ourM{} (8$\times$7B) models retrieved 23, 38, and 165 unique Rfam families, respectively. This result suggests that larger models not only generate more stable sequences but also cover a more diverse set of RNA types and structures.

\subsection{Guide RNA design}
Guided RNA, commonly referred to as guide RNA (gRNA), is a key element in CRISPR-Cas9 gene-editing technology. It is essential for directing the Cas9 enzyme to a precise location within the genome where genetic modifications are intended. We evaluate \ourM{} on two gRNA design tasks: the first is designing gRNAs for a given DNA sequence, and the second is selecting the more effective gRNA from two candidates. Examples are provided below:

\begin{example}
\noindent$\rhd$gRNA generation\\
\noindent{\textbf{Instruction}}: \\
\texttt{Generate a guide RNA for targeting the DNA sequence} \\
\dna{}GACTGGCACCAG$\cdots$CCCTCGC\edna{}.\\	
\noindent{\textbf{Response}}: \rna{}AGACACAGCGGGTGCTCTGC\erna{}\\


\noindent$\rhd$More effective gRNA identification\\
\noindent{\textbf{Instruction}}: 
\noindent \texttt{Investigate which of} \rna{}ATGTAGAAGAATCCACC\\
ATA\erna{} 
\texttt{or} \rna{}GGAAGGGGTCAATATTCTCA\erna{} \texttt{results in better wild-type efficiency for the DNA sequence} \\
\dna{}AAGGGGTGGCA$\cdots$AGTGC\edna{}.	

\noindent{\textbf{Response}}: \rna{}ATGTAGAAGAATCCACCATA\erna{}
\end{example}

A valid gRNA is defined by the following criteria: (1) the length of the generated gRNA must be between 17 to 24 nucleotides;  (2) the generated sequences should match a region in the given DNA sequences; (3) the matched sequence in the DNA template must be directly followed by an ``NGG'' sequence, known as the Protospacer Adjacent Motif (PAM).

\ourM{} demonstrates a strong ability to generate valid gRNA sequences compared with  generalist models like GPT4, accurately targeting the specified DNA while maintaining the PAM sequence feature. Furthermore, \ourM{} shows proficiency in assessing the effectiveness of gRNAs, enabling it to select the more efficient gRNA from a given pair (Table \ref{tab:RNA:RNA generation tasks}). 
\begin{table}[!htbp]
\centering
\begin{tabular}{lccc}
\toprule
Model & Validity & Top 1 accuracy\\
\midrule
GPT-4 & 0.272 & 0.597 \\
Llama 3 8B & 0 & 0.38 \\
Mixtral 8x7B & 0 & 0.46 \\
\midrule
\ourM{} (1B) & 0.95 & 0.681 \\
\ourM{} (8B) & 0.765 & 0.657 \\
\ourM{} (8x7B) & 0.957 & 0.699 \\
%\bottomrule
%SFM-Seq (1B w/o text) & 0.96 &  \\
%SFM-Seq (8B w/o text) & 0.973 &  \\
\bottomrule
\end{tabular}
\caption{The performance of guide RNA design.}
\label{tab:RNA:RNA generation tasks}
\end{table}




% \begin{table}[!h]
%     \caption{Guide RNA Design Performance in a Few-Shot Setting}
%     \centering
%     \begin{tabular}{c|c|c}
%          \hline
%          Model & Valid & Top 1 accuracy\\
%          \hline
%          GPT4-fewshot & 0.272 & ? \\
%          Llama3-8B-fewshot & 0 & 0.38 \\
%          Mixtral 8x7B-fewshot & 0 & 0.46 \\
%          SFM-Seq (8B)-fewshot & 0.163 & ? \\
%          \hline

%     \end{tabular}
%     \label{tab:RNA:RNA generation tasks}
% \end{table}

\subsection{Protein binding RNA design}\label{sec:protein2rna}
RNA-binding proteins (RBPs) represent a large and diverse class of over 2,000 proteins that play a crucial role in regulating gene expression by interacting with RNA. Designing RNA decoys offers a powerful strategy to manipulate these interactions. Such decoys can sequester RBPs away from their natural RNA targets, act as competitors to displace natural RNA molecules from RBPs, or serve as scaffolds to recruit RBPs to specific RNA molecules or cellular locations.

% RNA binding proteins (RBPs) comprise a large class of over 2000 proteins that involved in regulating gene expression by interacting with RNA. Design RNA decoys to sequester RBPs from their natural targets, as competitors to displace natural RNA molecules from the RBP, or as scaffolds to bring RBPs to specific RNA molecules or locations within the cell. We evaluate our models on the task of protein binding RNA design by generating RNAs that bind to specific RBPs. The example is shown below:

\begin{example}
\noindent\textbf{Instructions}:  \texttt{Given} \pro{}MSEY$\cdots$SSGWGM\epro{}, \texttt{create an RNA molecule that binds to it}.	\\
\noindent\textbf{Response}:
\rna{}AAACAGG$\cdots$CGTACGACA\erna{}
\end{example}

We selected 200 targets and generated binding RNA for them. To evaluate the generation ability of \ourM{}, following \cite{xu2023prismnet}, we trained a predictor for each protein to predict the binding affinity between the RNA and the protein. Specifically, the final layer of the classifier is a sigmoid function, which produces an output value ranging from 0 to 1, indicating the strength of the RNA-protein binding. If the score is greater than 0.5, we consider the generated RNA to have successfully bound to the protein.

We compared the RNA sequences generated by \ourM{} 1B, 8B and 8x7B. Additionally, we used the predictors to evaluate the binding and non-binding RNA sequences from the test set. We also randomly selected RNA sequences of the same sizes from the unconditional generation setting for prediction (Section \ref{sec:unconditional_rna_generation}). %The results are reported in Table~\ref{tab:RNA:rbp2rna}. 

The results are summarized in Table~\ref{tab:RNA:rbp2rna}, which reports the average and median prediction scores, as well as the success rateâ€”the proportion of sequences with a prediction score above 0.5. We have the following observations: 
\begin{enumerate}
\item As expected, binding RNA sequences achieved the highest average prediction score of 0.714 and a success rate of 74.5\%, while the non-binding RNA sequences had the lowest average score of 0.274 and a success rate of 24.4\%. This confirms the reliability of the classifiers and serving as a benchmark for optimal performance.
\item For unconditioned RNA Sequences, with an average score of 0.391 and a success rate of 36.3\%, these sequences performed better than non-binding sequences but significantly worse than the binding sequences. This suggests that random RNA sequences have a moderate chance of being predicted as binders due to the intrinsic properties of RNA but lack the specificity achieved through conditioning.
\item For \ourM{} generated sequences, as we increase the model sizes, there is a clear trend that larger models perform better. The results also demonstrated that \ourM{} is more likely to generate RNA sequences that are likely to bind to the specified proteins when explicitly conditioned on them.
\end{enumerate}



% Following the approach in~\cite{xu2023prismnet}, we train a separate classification model for each protein to evaluate its interaction with RNA. For the generated RNA sequences, we input them into the model trained specifically for the conditioned proteins, and the prediction scores are then aggregated. To compare the model's performance, we sampled both binding and non-binding RNA sequences from the original test data, using the ground-truth labels for validation. Additionally, we evaluate RNA sequences generated without conditioning on specific proteins to assess the model's ability to follow instructions.
%Table~\ref{tab:RNA:rbp2rna} presents the results. The average and median prediction scores are reported for each group of RNA sequences, and we apply a threshold of 0.5 to classify successful interactions. Specifically, an RNA sequence with a score greater than 0.5 is considered a successful binding prediction.
%
% As we can see in Table~\ref{tab:RNA:rbp2rna}, the highest average score (0.714) and success rate (0.745) are achieved by the binding RNA sequences, as expected, since they are derived from the ground-truth test data. Non-binding sequences have the lowest scores, with an average score of 0.274 and a success rate of 0.244. The unconditioned sequences perform slightly better than the non-binding ones but still significantly worse than binding sequences, with an average score of 0.391. For the SFM-Seq models, performance improves with model size: the SFM-Seq (1B) model achieves a success rate of 0.409, while the larger SFM-Seq (8B) and SFM-Seq (8x7B) models reach success rates of 0.442 and 0.448, respectively. This suggests that increasing model size enhances RNA-protein interaction prediction performance.

\begin{table}[!h]
\centering
\begin{tabular}{lccc}
\toprule
Source          & AVG Score & Success rate (\%)\\
\midrule
Binding         & 0.714     & 74.5 \\
Non-binding     & 0.274     & 24.4 \\
Unconditioned   & 0.391     & 36.3 \\
\ourM{} (1B)    & 0.415     & 40.9 \\
\ourM{} (8B)    & 0.434     & 44.2 \\
\ourM{} (8x7B)  & 0.438     & 44.8 \\
\bottomrule
\end{tabular}
%% original table; do not delete because the median level is in that
% \begin{tabular}{cccc}
%          \toprule
%          Source & AVG Score & Median Score & Success Rate\\
%          \midrule
%          Binding & 0.714 & 0.877 & 0.745 \\
%          Non-binding & 0.274 & 0.117 & 0.244 \\
%          Unconditioned & 0.391 & 0.291 & 0.363 \\
%          SFM-Seq (1B) & 0.415 & 0.311 & 0.409 \\
%          SFM-Seq (8B) & 0.434 & 0.367 & 0.442 \\
%          SFM-Seq (8x7B) & 0.438 & 0.408 & 0.448 \\
%          \bottomrule
% \end{tabular}
\caption{Performance of designing protein-binding RNA given proteins: ``AVG Scores" refers to the average prediction scores across all sequences. Success ratio refers to the percentage of scores that are greater than 0.5. See Figure \ref{fig:enter-label} for the detailed distribution of the predicted scores.}
\label{tab:RNA:rbp2rna}
\end{table}.
