\section{Text capabilities }\label{sec:text_ability}
We assess the text-related capabilities of \ourM{} (8x7B) in this section. We first present several examples to provide readers with an intuitive understanding (Section \ref{sec:case_study}), and then benchmark results from AlpacaEval (Section \ref{sec:alpahaeval}).

\subsection{Case study}\label{sec:case_study}
%\TODO{@Yuan-Jyue}, pls input several text cases here.
Here we demonstrate our model's language capabilities in scientific discovery research. Specifically, we assessed the model's understanding of CRISPR gene editing, cancer immunotherapy, and drug discovery. A few selected cases are presented in this section for brevity. Additional cases can be found in Fig. \ref{SI_fig:case_study:CRISPR}, \ref{SI_fig:case_study:cancer_immunotherapy} and \ref{SI_fig:case_study:drug_discovery}.


\subsubsection{CRISPR gene editing}
 CRISPR gene editing has rapidly emerged as a powerful tool in biotechnology. Our model demonstrates a good understanding of CRISPR and is able to effectively analyze the advantages and disadvantages of commonly used CRISPR tools. 

  
\begin{mdframed}[backgroundcolor=white, linecolor=black, linewidth=1pt]
    \noindent
    \textbf{Instruction:} Could you provide me the advantages and disadvantages of Cas9? \\
    \textbf{Response:} \\
    Advantages: 1. Highly efficient and specific gene editing with a low risk of off-target effects. 2. Can target a wide range of sequences.  3. Cam be used to make specific alterations to genes in virtually any organism.  Disadvantages: 1. Potential for off-target effects when using Cas9. 2. The delivery of Cas9 can be inefficient. 3. Lack of control over the repair mechanism. 4. May cause chromosomal abnormalities. \\

    \noindent
    \textbf{Instruction:} Could you provide me the advantages and disadvantages of Cas12a, another widely used CRISPR protein? \\
    \textbf{Response:} \\
    Advantages: \\
    1. High specificty: Cas12a has higher specificity compared to Cas 9, which reduces off-target effects. 
    2. Simpler PAM requirements: Cas 12a requires a simpler PAM sequence than Cas9, making it easier to target specific genomic loci. 
    3. Multiplexing: Cas12a can be used for multiplexing genome editing, meaning multiple genomic loci can be edited in a single reaction. 
    Disadvantages: \\
    4. Lower efficiency: Cas12a is less efficient in terms of indel formation than Cas9. 
    5. Larger size: Cas12a is a larger protein than Cas9, which can make it harder to deliver the Cas12a system into cells. \\
    ... 
\end{mdframed}  
\captionof{figure}{Cases about CRISPR gene editing.} \label{fig:case_study:CRISPR}

\subsubsection{Cancer immunotherapy}
Our model is knowledgeable about existing checkpoint inhibitors used in cancer therapy and provides accurate explanations of their mechanisms of action. Specifically, in our experiment, it generates a valid peptide sequence for inhibiting the overexpressed PD-L1 protein (see Figure \ref{SI_fig:case_study:cancer_immunotherapy}), although the length of the peptide is incorrect. It demonstrates a solid understanding of the inhibition cascade between drugs and proteins.

\begin{mdframed}[backgroundcolor=white, linecolor=black, linewidth=1pt]  
    \noindent
    \textbf{Instruction:} Tell me existing checkpoint inhibitors for PD-L1. \\
    \textbf{Response:} The checkpoint inhibitors for PD-L1 that are currently approved by the FDA include atezolizumab, avelumab, duravalumab, and perbrolizumab.\\
    
    \noindent
    \textbf{Instruction:} PD-L1 is a protein that inhibits PD-1, a recptor found on T cells. This new drug X, is designed to inhibit PD-L1. If we introduce drug X, will T cell function be restored? Explain it step by step. \\
    \textbf{Response:}
    PD-1 is inhibited by interaction with PD-L1. If drug X inhibits PD-L1, PD-1 will no longer be inhibited. Therefore, T cell function will beestored. Therefore, the answer is yes.

\end{mdframed}  
\captionof{figure}{Cases about cancer immunotherapy.} \label{fig:case_study:cancer_immunotherapy}

\subsubsection{Drug discovery}
Drug discovery is an evolving process that incorporates empirical rules, computational design, and experimental validation. The model demonstrates an understanding of the fundamental principles of drug discovery. Specifically, when being asked questions about Lipinski’s Rule of Five and strategies (see Figure \ref{SI_fig:case_study:drug_discovery}) to improve a drug’s ADME properties, the model provides mostly accurate answers.

\begin{mdframed}[backgroundcolor=white, linecolor=black, linewidth=1pt]  
    \textbf{Instruction:} What approaches can drug experts use to enhance a drug's ADME properties? \\
    \textbf{Response:} Structure-based drug design: Computational techniques can be used to modify existing drugs or design new ones with desired ADME properties. 2. Lipophilicity: Lipophilicity can affect the ADME profile of a drug. 3. Hepatic and extracellular clearance: Understanding the hepatic and extracellular clearance of a drug is important for assessing its ADME profile. 4. Microsomal stability: Microsomal stability is a measure of how quickly a drug is broken down by liver enzymes. 5. Oral absorption: Oral absorption is determined by the solubility and permeability of a drug. 6. Permeability: The ability of a drug to cross cell membranes affects its bioavailability. 
    ... %7. Plasma protein binding: Plasma protein binding affects the concentration of free drug available to interact with its target. 8. Efflux transporters: Efflux transporters can affect the absorption, distribution, and excretion of a drug. 

\end{mdframed}  
\captionof{figure}{Case about drug discovery.} \label{fig:case_study:drug
_discovery}

\subsection{Comparison on AlpacaEval evaluation}\label{sec:alpahaeval}
To assess the linguistic capabilities of \ourM{}, we compared \ourM{} (8x7B) with the original Mixtral  (8x7B) using the AlpacaEval \cite{dubois2024length} dataset and method. AlpacaEval is a benchmark for LLM designed to gauge a model's proficiency in executing general user instructions. \ourM{}'s responses are compared against reference responses generated by original Mixtral using GPT-4 as the annotator. We focused on two metrics from AlpacaEval: the general win rate and the LC win rate (length-controlled win rate), which controls the bias introduced by the length of the responses. Our NatureLM model achieved a LC win rate of 31.8\% and a win rate of 25.2\% when benchmarked against Mixtral, showcasing that while the model possesses a foundational level of language comprehension, it exhibits constraints in its linguistic capacity.   A potential contributing factor to this limited performance is the overrepresentation of scientific data within the training dataset, which may skew the model's proficiency towards technical language and away from a more generalized understanding required for diverse user instructions. Fig. \ref{fig:case_study:alpacaeval} presents two illustrative examples from the AlpacaEval comparison. We will improve the text capabilities of \ourM{} in our future work. 
