%% This is file `medima-template.tex',
%%
%% Copyright 2018 Elsevier Ltd
%%
%% This file is part of the 'Elsarticle Bundle'.
%% ---------------------------------------------
%%
%% It may be distributed under the conditions of the LaTeX Project Public
%% License, either version 1.2 of this license or (at your option) any
%% later version.  The latest version of this license is in
%%    http://www.latex-project.org/lppl.txt
%% and version 1.2 or later is part of all distributions of LaTeX
%% version 1999/12/01 or later.
%%
%% The list of all files belonging to the 'Elsarticle Bundle' is
%% given in the file `manifest.txt'.
%%
%% Template article for Elsevier's document class `elsarticle'
%% with harvard style bibliographic references
%%
%% $Id: medima-template.tex 153 2018-12-01 11:38:32Z rishi $
%% $URL: http://lenova.river-valley.com/svn/elsarticle/trunk/medima-template.tex $
%%
%% Use the option review to obtain double line spacing
%\documentclass[times,review,preprint,authoryear]{elsarticle}

%% Use the options `twocolumn,final' to obtain the final layout
%% Use longtitle option to break abstract to multiple pages if overfull.
%% For Review pdf (With double line spacing)
\documentclass[preprint, 3p]{elsarticle}

%\documentclass[times,twocolumn,review]{elsarticle}
%% For abstracts longer than one page.
%\documentclass[times,twocolumn,review,longtitle]{elsarticle}
%% For Review pdf without preprint line
%\documentclass[times,twocolumn,review,nopreprintline]{elsarticle}
%% Final pdf
%%\documentclass[times,twocolumn,final]{elsarticle}
%%
%\documentclass[times,twocolumn,final,longtitle]{elsarticle}
%%


%% Stylefile to load MEDIMA template
%\usepackage{medima}
%\usepackage{framed,multirow}

%% The amssymb package provides various useful mathematical symbols
\pdfoutput = 1
\usepackage{lipsum}
\makeatletter
\def\ps@pprintTitle{%
	\let\@oddhead\@empty
	\let\@evenhead\@empty
	\def\@oddfoot{}%
	\let\@evenfoot\@oddfoot}
\makeatother

\usepackage{amssymb}
\usepackage{latexsym}

% Following three lines are needed for this document.
% If you are not loading colors or url, thenthese are
% not required.
\usepackage{url}
\usepackage{xcolor}

\usepackage{hyperref}

\definecolor{newcolor}{rgb}{.8,.349,.1}
\journal{Medical Image Analysis}

\usepackage{graphicx}
\usepackage{tabularx}
\usepackage{longtable}
\usepackage{nicefrac}
\usepackage{placeins}
\usepackage{arydshln}

\usepackage{amsmath}
\usepackage{multirow}
\usepackage{tikz}
\usepackage{floatrow}
\usetikzlibrary{matrix,chains,positioning,decorations.pathreplacing,arrows}
\usepackage{amsfonts}
\usepackage{makeidx}  % allows for indexgeneration
%\usepackage[numbers]{natbib}
\usepackage{amsmath, mathtools}
\usepackage{graphicx}
\usepackage{enumerate}
\usepackage{array}
\usepackage{caption}
\usepackage{subcaption}
\renewcommand\thesubfigure{\Alph{subfigure}}
\usepackage{stfloats} %For double column floating figures
\usepackage{booktabs}
%\usepackage{pgfplots}
\usepackage{verbatim}
\usepackage{algorithm2e}
\graphicspath{{figures/}}
\usepackage[acronym,nomain]{glossaries}
\usepackage{float}
\usepackage{xcolor}
\usepackage{tabularx}
\usepackage{comment}

\usepackage{multicol}
%\usepackage{hyperref}

\usepackage{tikz}
\usetikzlibrary{calc}
\usetikzlibrary{spy}
\usetikzlibrary{decorations.pathreplacing}
\usetikzlibrary{shapes}
\usetikzlibrary{through}
\usetikzlibrary{intersections}
\usetikzlibrary{decorations.markings}
\usetikzlibrary{3d}
\usetikzlibrary{positioning}

\hyphenation{CARPentry}

\usepackage{fancyhdr}
\usepackage{datetime}

\usepackage{quoting,xparse}
\usetikzlibrary{arrows, shapes, trees}
\usepackage{geometry}
\usepackage{hyperref}
\usepackage{ulem}

%Text-degree
\usepackage{textcomp}
%Diagonal ning line in tabular environment
\usepackage{diagbox}

%Formatting of scientific notation floats copied from latex
\usepackage{siunitx}
\sisetup{round-mode=places,round-precision=2}
\sisetup{scientific-notation = fixed, fixed-exponent = 0}
% TODO Notes package
% Put % before of what you want disabled

% Select what to do with todonotes:
%\usepackage[disable]{todonotes} % notes not shown
\usepackage[draft]{todonotes}   % notes shown
%\newcommand{\todofloat}[1]{}
\newcommand{\todofloat}[1]{\color{blue} \textbf{#1}}

% Select what to do with command \comment:
% \newcommand{\comment}[1]{}  %comment not showed
%\newcommand{\comment}[1]
%{\par {\bfseries \color{blue} #1 \par}} %comment showed
\newcommand{\R}[0]{\ensuremath{\mathbb{R}}}
\newcommand{\norm}[1]{\left\lVert#1\right\rVert}
\newcommand{\abs}[1]{\left\lvert#1\right\rvert}
\newcommand{\absunscaled}[1]{|#1|}
\newcommand{\diff}[1]{\ensuremath{\operatorname{d}\!{#1}}}
\newcommand{\intd}[1]{\ensuremath{\operatorname{d}\!{#1}}}
\newcommand{\sign}[0]{\text{sgn}}
\newcommand{\prox}[0]{\text{prox}}
\newcommand{\proj}[0]{\text{proj}}
\newcommand{\diag}[0]{\text{diag}}
\newcommand{\rep}[0]{\text{RELF}^{+}}
\newcommand{\rem}[0]{\text{RELF}^{-}}
\newcommand{\dom}[0]{\text{dom}}
\DeclareMathOperator*{\argmin}{arg\,min}
\DeclareMathOperator*{\argmax}{arg\,max}
\newcommand{\caseif}[0]{\; \text{if} \;}
\newcommand{\caseelse}[0]{\; \text{else} \;}
\newcommand{\caseand}[0]{\; \text{and} \;}
\newcommand{\st}[0]{\enskip \text{s.t.:} \enskip}
\DeclareCaptionLabelFormat{andtable}{#1~#2  \&  \tablename~\thetable}

\renewcommand{\vec}[1]{\mathbf{#1}}
\newcommand{\mat}[1]{\mathbf{#1}}
\newcommand{\domlv}[0]{\Omega_{\rm lv}}   % LV domain
\newcommand{\domrv}[0]{\Omega_{\rm rv}}   % RV domain
\newcommand{\dombiv}[0]{\Omega_{\rm biv}} % BiV domain
\newcommand{\domtor}[0]{\Omega_{\rm tor}} % Torso domain

%\newcommand{\domutc}[0]{\mathrm{UTC}}    % UTC domain
%\newcommand{\domuvcl}[0]{\mathrm{UVC}_l}  % LV UVC domain
%\newcommand{\domuvcr}[0]{\mathrm{UVC}_r}  % RV UVC domain
%\newcommand{\domuvc}[0]{\mathrm{UVC}}  % UVC domain

\newcommand{\domutc}[0]{D_{_{\mathrm{UTC}}}}  % UVC domain
\newcommand{\domuvcl}[0]{D_{_{\mathrm{UVC}_{\rm lv}}}}  % UVC domain
\newcommand{\domuvcr}[0]{D_{_{\mathrm{UVC}_{\rm rv}}}}  % UVC domain
\newcommand{\domuvc}[0]{D_{_{\mathrm{UVC}}}}  % UVC domain

\newcommand{\maputc}[0]{\mathcal{B}_{_{\mathrm{UTC}}}}
\newcommand{\mapuvcl}[0]{\mathcal{B}_{_{\mathrm{UVC}_{\rm lv}}}}
\newcommand{\mapuvcr}[0]{\mathcal{B}_{_{\mathrm{UVC}_{\rm rv}}}}
\newcommand{\mapuvc}[0]{\mathcal{B}_{_{\mathrm{UVC}}}}

\newcommand{\figref}[1]{Fig.~\ref{#1}}
\newcommand{\eqreffull}[1]{Eq.~\eqref{#1}}
\newcommand{\algref}[1]{Alg.~\ref{#1}}
\newcommand{\secref}[1]{Sec.~\ref{#1}}
\newcommand{\tabref}[1]{Tab.~\ref{#1}}

\newcommand\setrow[1]{\gdef\rowmac{#1}#1\ignorespaces}
\newcommand\clearrow{\global\let\rowmac\relax}
%\renewcommand\thesubfigure{\roman{subfigure}}

\newcolumntype{?}{!{\vrule width 1pt}}

\usepackage[symbol]{footmisc}
\def\correspondingauthor{\footnote{Corresponding author: gernot.plank@medunigraz.at, Medical University of Graz, Neue
        Stiftingtalstra√üe 6(MC1.D.)/IV,
        8010, Graz, Austria}}

%Extra quotation environment
%https://tex.stackexchange.com/questions/391726/the-quotation-environment
\NewDocumentCommand{\bywhom}{m}{% the Bourbaki trick
    {\nobreak\hfill\penalty50\hskip1em\null\nobreak
        \hfill\mbox{\normalfont(#1)}%
        \parfillskip=0pt \finalhyphendemerits=0 \par}%
}

\NewDocumentEnvironment{pquotation}{m}
{\begin{quoting}[
        indentfirst=true,
        leftmargin=\parindent,
        rightmargin=\parindent]
    \bywhom{#1}
\end{quoting}}

\clearrow

\newcommand\mycommfont[1]{\footnotesize\ttfamily\textcolor{blue}{#1}}
\SetCommentSty{mycommfont}

\makeglossaries



\definecolor{ColRev1}{RGB}{192,0,0}
\definecolor{ColRev2}{RGB}{0,96,160}
\definecolor{ColRev3}{RGB}{229,122,0}
\definecolor{ColRev4}{RGB}{230,0,153}
%\definecolor{ColBoth}{RGB}{121,63,191}
\definecolor{ColChange}{RGB}{0,160,0}
\definecolor{forestgreen}{RGB}{20,160,80}

\usepackage[acronym]{glossaries}

% anatomy
\newacronym{lv}{LV}{left ventricle}
\newacronym{rv}{RV}{right ventricle}
\newacronym{la}{LA}{left atrium}
\newacronym{ra}{RA}{right atrium}
\newacronym{cs}{CS}{coronary sinus}
\newacronym{lpv}{LPV}{left pulmonary veins}
\newacronym{lipv}{LIPV}{left inferior pulmonary vein}
\newacronym{lspv}{LSPV}{left superior pulmonary vein}
\newacronym{rpv}{RPV}{right pulmonary veins}
\newacronym{ripv}{RIPV}{right inferior pulmonary vein}
\newacronym{rspv}{RSPV}{right superior pulmonary vein}
\newacronym{laa}{LAA}{left atrial appendage}
\newacronym{raa}{RAA}{right atrial appendage}
\newacronym{svc}{SVC}{superior vena cava}
\newacronym{ivc}{IVC}{inferior vena cava}
\newacronym{tv}{TV}{tricuspid valve}
\newacronym{mv}{MV}{mitral valve}

\newacronym{bb}{BB}{Bachmann's bundle}
\newacronym{fo}{FO}{fossa ovalis}
\newacronym{pm}{PM}{pectinate muscle}
\newacronym{san}{SAN}{sino-atrial node}
\newacronym{hps}{HPS}{His-Purkinje System}
\newacronym{ic}{IC}{Inter-atrial Conduction}

% diseases
\newacronym{af}{AF}{atrial fibrillation}

% electrophysiology
\newacronym{ep}{EP}{electrophysiology}
\newacronym{ecg}{ECG}{electro-cardiogram}
\newacronym{egm}{EGM}{electrogram}
\newacronym{eam}{EAM}{electro-anatomical mapping}
\newacronym{pwd}{PWD}{P-Wave Duration}


% modeling
\newacronym{uac}{UAC}{universal atrial coordinates}
\newacronym{rd}{R-D}{Reaction-Diffusion}
\newacronym{re}{R-E}{Reaction-Eikonal}
\newacronym{bem}{BEM}{Boundary Element Method}
\newacronym{fem}{FEM}{Finite Element Method}
\newacronym{relf}{RELF}{Reaction-Eikonal Lead Field}
\newacronym{pca}{PCA}{Principal Component Analysis}
\newacronym{dbc}{DBC}{Dirichlet Boundary Conditions}

% imaging and image analysis
\newacronym{scn}{SCN}{SpatialConfiguration-Net}
\newacronym{ct}{CT}{computed tomography}
\newacronym{mri}{MRI}{Magnetic Resonance Imaging}

% maths
\newacronym{rmse}{RMSE}{relative mean square error}
\newacronym{ld}{LD}{Laplace-Dirichlet}

\begin{document}
	
	\begin{frontmatter}
		
		%\title{An efficient computational framework for the generation of biatrial volumetric models}
		% suggest another title, to broaden the focus  
		\title{An efficient end-to-end computational framework for the generation of ECG calibrated volumetric models 
			of human atrial electrophysiology}
		\author {Elena Zappon$^{a,b}$}
		\author {Luca Azzolin$^{c}$}
		\author {Matthias A.F. Gsell$^{a}$}
		\author {Franz Thaler$^{d,a,e}$}
		\author {Anton J. Prassl$^{a}$}
		\author {Robert Arnold$^{a}$}
		\author {Karli Gillette$^{f,g,a,b}$}
		\author {Mohammadreza Kariman$^{a}$}
		\author {Martin Manninger-W\"{u}nscher$^{h}$}
		\author {Daniel Scherr$^{h}$}
		\author {Aurel Neic$^{c}$}
		\author {Martin Urschler $^{e,b}$}
		\author {Christoph M. Augustin$^{a,b}$}
		\author {Edward J. Vigmond$^{i}$}
		\author {Gernot Plank$^{a,b}$\correspondingauthor{} \\ \ \\% leave an empty line between authors and affiliation\\
			$^a$ Division of Biophysics, Medical University of Graz, Graz, Austria \\
			$^b$ BioTechMed-Graz, Graz, Austria \\
			$^c$ NumeriCor Gmbh, Graz, Austria\\
			$^d$ Institute of Visual Computing, Graz University of Technology, Graz, Austria \\
			$^e$ Institute for Medical Informatics, Statistics and Documentation, Medical University of Graz, Graz, Austria \\
			$^f$ Scientific Computing and Imaging Institute, University of Utah, USA\\
			$^g$ Department of Biomedical Engineering, University of Utah, USA\\
			$^h$ Clinical Department of Cardiology, Medical University of Graz, Graz, Austria\\
			$^i$ University of Bordeaux, CNRS, Bordeaux, France\\}
		
		\begin{abstract}
			Computational models of atrial \gls{ep} are increasingly utilized for applications such as the development of advanced mapping systems, personalized clinical therapy planning, and the generation of virtual cohorts and digital twins. These models have the potential to establish robust causal links between simulated \textit{in silico} behaviors and observed human atrial EP, enabling safer, cost-effective, and comprehensive exploration of atrial dynamics. However, current state-of-the-art approaches lack the fidelity and scalability required for regulatory-grade applications, particularly in creating high-quality virtual cohorts or patient-specific digital twins. Challenges include anatomically accurate model generation, calibration to sparse and uncertain clinical data, and computational efficiency within a streamlined workflow.
			
			This study addresses these limitations by introducing novel methodologies integrated into an automated end-to-end workflow for generating high-fidelity digital twin snapshots and virtual cohorts of atrial EP. These innovations include: (i) automated multi-scale generation of volumetric biatrial models with detailed anatomical structures and fiber architecture; (ii) a robust method for defining space-varying atrial parameter fields; (iii) a parametric approach for modeling inter-atrial conduction pathways; and (iv) an efficient forward EP model for high-fidelity \gls{ecg} computation.
			
			We evaluated this workflow on a cohort of 50 atrial fibrillation patients, producing high-quality meshes suitable for reaction-eikonal and reaction-diffusion models and demonstrating the ability to simulate atrial ECGs under parametrically controlled conditions. These advancements represent a critical step toward scalable, precise, and clinically applicable digital twin models and virtual cohorts, enabling enhanced patient-specific predictions and therapeutic planning.
		\end{abstract}
		
		\begin{keyword}
			Atrial Electrophysiology; Volumetric Atrial Models; Cardiac Digit Twins; Universal Atrial Coordinates; Cardiac Modeling
		\end{keyword}
		
	\end{frontmatter}
	
	%\tableofcontents
	
	%\pagebreak
	
	
	\section{Introduction}
	\label{sec:intro}
	% Context
	Computational models of atrial \gls{ep} are increasingly being considered in a variety of applications, ranging from the industrial development of devices such as \gls{eam} systems \citep{reddy2023electrographic,szili2023electrographic}
	to the stratification and planning of clinical therapy \citep{boyle2019computationally}.
	These applications are built upon the mechanistic nature of biophysical models of atrial \gls{ep}, 
	and are based on the tacit assumption 
	that simulated \textit{in silico} behaviors closely correspond to the real human atrial \gls{ep} observed in patients.
	If sufficient regulatory-strength evidence of such a close causal relation is provided,
	\emph{in silico} models can be used to explore human atrial \gls{ep} more comprehensively, 
	in a safer and more cost-effective manner compared to the current paradigm based on preclinical animal testing and clinical trials \citep{dossel2012computational,schotten2011pathophysiological}. 
	%Such trustworthy atrial models can be used then to provide clinically useful predictions of patient responses to treatment-related perturbations, 
	%or to generate high fidelity synthetic \gls{ep} data for guiding the development of enhanced \gls{eam} systems and algorithms.
	However, current state-of-the-art models supporting evidence of such a tight causal and quantitatively accurate relation are limited. 
	The vast majority of computational studies use simplified anatomical models \citep{azzolin2023:_augmenta,labarthe2014:_bilayer,roney2023:_bia_vol} with uncalibrated default parameters \citep{boyle2019computationally, sakata2024:_assessing},
	and any comparison to directly observable quantities, such as \gls{ecg}, is usually limited \citep{ferrer2015detailed,loewe2016influence} or even not accounted for \citep{roney2023:_bia_vol}.
	This can be largely attributed to the limited capabilities of current modeling technologies in terms of anatomical model generation as well as calibration and simulation technologies.
	A direct comparison to observations would reveal major discrepancies between physical and virtual spaces,
	thus undermining the credibility of the model.
	
	The degree of fidelity needed, as well as the metrics used for measuring it, is application-dependent \citep{bhagirath2024:_bits}.
	Beyond the minimum requirement of a mechanistic relation -- that is, the models of atrial \gls{ep} are able to qualitatively recapitulate all the mechanisms of atrial \gls{ep} at play --
	in silico models can be calibrated to be representative of a group of patients of interest, 
	also covering anatomical and functional variability \citep{niederer2020creation}.
	Such representation of a patient cohort,
	rather than an individual, may facilitate generic interpretation of observations in real physical atria, and prediction of \gls{ep} responses to therapeutic interventions.
	Such sets of functionally-similar models referred to as virtual cohorts, have recently started to be considered for in silico trials and for safety and efficacy testing of new devices or therapies \citep{viceconti2020:_in_silico}.
	Nevertheless, creating virtual cohorts of in silico models will necessitate a shift in cardiac modeling, moving from a limited set of custom models to efficient and scalable workflows capable of generating large volumes of models quickly \citep{niederer2020creation}.
	% Model calibration for building virtual cohorts relies on sampling from inferred distributions 
	% of model parameters and inputs, where models are accepted as a valid instance of the cohort 
	% if the distribution of model predictions falls within a statistically plausible range, 
	% or are rejected otherwise \cite{niederer2020creation}.
	
	Most demanding are clinical applications geared toward precision cardiology, that is, to tailor therapies to individual patients. 
	There, models are sought that replicate cardiac anatomy and structures and quantitatively calibrate to match functional observations from an individual patient's heart in a one-to-one manner.
	% this might be a too strong requirement, removed it for now
	% under a broad range of physiological and testable conditions. 
	Such functionally equivalent models, where a particular stimulus or perturbation leads to the same emergent response in virtual and real space at a single time point, are referred to as digital twin snapshots. When the models are continuously or periodically updated with measurements, they become true digital twins \citep{bhagirath2024:_bits,corral2020digital,hopman2023right}.
	%Such digital twins, can be employed to make rigorous and specific predictions for that patient‚Äôs treatment
	
	% Need
	However, 
	%despite the high promises of the potentially transformative nature of digital twin technologies, 
	the ability of current atrial \gls{ep} modeling pipelines 
	for creating high-fidelity digital twin snapshots or virtual cohorts at scale 
	in a sufficiently efficient and robust manner %to be broadly applicable 
	is severely limited. 
	The key challenge is to describe the electrical sources in the atria accurately enough to predict the electrical potential field in its surroundings where all observable measurements are recorded. 
	In humans \emph{in vivo}, electrical measurements are in the form of \glspl{egm} by devices or \gls{eam} systems, or as \gls{ecg} at the body surface.
	While conceptually simple, the implementation of such a pipeline is vastly demanding, posing a long list of formidable challenges many of which remain unaddressed.
	In general, modeling pipelines for creating cardiac digital twin models of atrial \gls{ep} 
	are separated into two distinct stages, an anatomical and a functional twinning stage \citep{azzolin2023:_augmenta,labarthe2014:_bilayer,roney2023:_bia_vol}.
	
	% anatomical twinning
	At the anatomical stage,
	multi-label segmentation of tomographic images is performed to identify all relevant domains \citep{payer2017multi}
	which are turned into multi-label computational meshes \citep{crozier2016image,prassl09:_tarantula}
	to accurately represent biatrial anatomy.
	Involved procedures are notoriously laborious, 
	requiring numerous manual operator interventions by trained experts and significant computational resources \citep{azzolin2023:_augmenta,roney2023:_bia_vol}, 
	to obtain anatomically accurate representations of sufficient mesh quality for a given type of cardiac EP simulation.
	Specifically for atrial anatomies, these are often simplified \citep{harrild2000:_atria} 
	and represented as manifolds only \citep{azzolin2023:_augmenta,labarthe2014:_bilayer,roney2023:_bia_vol},
	thus limiting the achievable quantitative accuracy in representing electrical sources and associated potential fields.  
	
	% functional twinning
	At the functional twinning stage, 
	the fundamental core challenge of calibrating an atrial \gls{ep} model to clinical data
	%and, thus, to establish a close link between \emph{in silico} predictions
	%and the physical reality in an individual patient out of a highly variable patient population has to be addressed. 
	is to infer high-dimensional space-varying parameter fields governing the \gls{ep} behavior 
	from limited sparse clinical recordings
	that are afflicted with substantial observational uncertainties \citep{coveney2022calibrating,whittaker2020calibration}.
	This poses a number of technological problems.
	Model functionalization refers to conceiving a framework for comprehensively describing 
	a sufficiently high-dimensional parameter space 
	defined over geometrically complex objects, such as the heart \citep{bayer2018universal,roney2019uac}, 
	that encapsulates all relevant factors governing atrial \gls{ep}, and the genesis of the associated extracellular potential field.
	%Appropriate abstract reference frames are needed to prescribe space-varying parameter fields 
	%to geometrically complex objects such as the atria\cite{roney2019uac},
	These fields must be exposed to unattended algorithmic manipulation to facilitate parameter sweeps in order to minimize the mismatch between simulated and observed data within an optimization procedure \citep{grandits2021:_geasi,grandits2023:_geodesic_bp}.
	A computationally efficient yet accurate forward \gls{ep} model for generating the observed electrical recordings, 
	i.e. \glspl{ecg}, \glspl{egm} or \glspl{eam},  is required
	to cope with the computational burden of a large number of model evaluations incurring during optimization \citep{gillette2021:_framework,neic17:_reaction_eikonal,pezzuto2017:_ecg}.
	Computational workflows meeting these criteria and reported so far in the literature \citep{azzolin2023:_augmenta,boyle2019computationally,roney2023:_bia_vol}, are primarily limited in terms of anatomical and structural representation, functional calibration capabilities, as well as automation, efficiency, robustness, reproducibility, and numerical accuracy.
	
	% Task
	In this study, we address these limitations by developing novel methodologies that are essential for the scalable generation of atrial digital twin snapshots and virtual cohorts, calibrated by \glspl{ecg} and, potentially, \glspl{egm} or \glspl{eam}. 
	These comprise: 
	\begin{enumerate}[i)]
		\item An automated flexible multi-scale approach for generating anatomically accurate volumetric biatrial models with comprehensive parametric incorporation of atrial structures and fiber architecture, and
		of sufficiently high mesh quality to be suitable for most widely used forward \gls{ep} models;
		\item A robust method for generating a volumetric atrial reference frame 
		for defining space-varying atrial parameter fields, and their unattended manipulation in optimization sweeps;
		\item A novel method for the flexible parametric representation of \gls{ic} pathways;
		\item  An efficient and clinically-compatible \gls{ecg} forward model 
		that computes high fidelity gold standard \glspl{ecg} and \glspl{egm} with close to real-time performance. 
	\end{enumerate}
	
	We report on the integration of these methodological developments into a highly automated and efficient end-to-end workflow 
	for building high-fidelity digital twin snapshots of atrial \gls{ep} from clinical data. 
	We elucidate in detail the methodological underpinnings of all processing stages
	-- from image analysis over model generation to \gls{ecg} and \gls{egm} prediction --
	and evaluate the performance of the end-to-end workflow on a clinical cohort of patients
	treated for \gls{af}, by processing 50 patient datasets randomly selected from the Graz \gls{af} registry.
	For each patient, both coarse and fine-resolution meshes were generated 
	at $\approx$ \SI{0.90}{\milli\meter} and \SI{0.25}{\milli  \meter} for reaction-eikonal \citep{neic17:_reaction_eikonal}
	and reaction-diffusion mono- or bidomain models, respectively, together with a set of automatically generated volumetric \gls{uac}.
	We demonstrate the ability to compute atrial \glspl{ecg} by performing sweeps on parametrically-controlled and easy-to-generate inter-atrial pathways, and we show the possibility of automatically varying the position and shape of the \gls{san}, allowing fast calibration of the model at hands.
	
	The structure of the paper is as follows: Section \ref{Sec:methods} presents a comprehensive overview of the proposed end-to-end workflow, encompassing the anatomical twinning stage (Section \ref{Subsec:anatomical_twinning}) and the functional modeling stage (Section \ref{Subsection:functional_modeling_stage}). Specifically, our approach for generating anatomically accurate volumetric biatrial models, including anatomical structures and fibers, is detailed in Sections \ref{Subsubsection:Imaging}-\ref{Subsubsec:anatomical_structure}. Sections \ref{Subsubsec:UAC} and \ref{Subsubsec:Torso} describe the computation of the \gls{uac} and torso model surrounding the biatrial mesh. Our novel method for computing interatrial connections is outlined in Section \ref{Subsubsection:cables}, while the parameter sweeping process for model calibration is detailed in Section \ref{Subsubsec:baseline_sim}. The numerical results are provided in Section \ref{Sec:results}, with the discussion, limitations, and conclusions presented in Sections \ref{Sec:discussion}, \ref{Sec:limitation}, and \ref{Sec:conclusion}, respectively.
	
	
	
	%There is a need for volumetric models with appropriate widths of walls and structures, 
	%to generate \glspl{egm} that are able to closely match magnitude and dynamics of \glspl{egm}
	%observed with \gls{eam} systems.
	
	\section{Methods} 
	\label{Sec:methods}
	We provide a schematic overview of an end-to-end workflow for the generation of anatomically accurate volumetric biatrial-torso model
	% based on \gls{ecg}, \gls{egm} or \gls{eam} data.
	in Figure \ref{Fig:scheme}. 
	
	\begin{figure}[!t]
		\centering
		\includegraphics[width=0.9\textwidth]{images/scheme_pipelin.png}
		\caption{Schematic outline of the end-to-end framework for the generation of \gls{ecg}-calibrated volumetric models of patient-specific human atria. After image acquisition, the workflow comprises nine steps: automatic multilabel segmentation, automatic label augmentation on the blood pools, extrusion of the volumetric bilayer walls, automatic selection of atrial orifices, automatic annotation of anatomical structures and fiber generation, generation of \gls{uac}, registration and generation of a torso volume conductor, generation of \gls{ic}s, and cardiac electrophysiological simulation and P-wave generation. We moreover provide the paper section index where each step is detailed.}
		\label{Fig:scheme}
	\end{figure}
	
	The workflow comprises two major processing stages, (i) an anatomical twinning stage comprising the generation of the atrial anatomical model with structural labels and fiber architecture, the generation of an atrial reference frame for spatial parameter swiping, and a torso anatomical model,
	and (ii) a functional modeling stage, for 
	the definition of inter-atrial pathways, and
	the setting up of a forward \gls{ep} model for representing and calibrating atrial action potential propagation, 
	and for generating the associated clinically observable signals, such as the \glspl{ecg} and the \glspl{egm}.
	%, required for the functional assimilation of simulated prediction and observed data. 
	Both stages are closely linked to the choice of forward \gls{ep} mathematical model, 
	which influences, for instance, the targeted spatial mesh resolution of the biatrial anatomical model.
	
	The anatomical twinning stage is organized following an almost automatic and sequential step approach:
	\begin{enumerate}[i)]
		\vspace{-0.2cm}\item image segmentation: by exploiting a convolutional neural network, tomographic scans, either \gls{ct} or \gls{mri}, are automatically segmented;
		% using a convolutional neural network
		% taking a tomographic scan, \gls{ct} or \gls{mri}, as an input labels a minimum set of anatomical structures 
		% including blood pools in the heart and attached large vessels, and the left ventricular (LV) myocardium;
		\vspace{-0.2cm}\item label augmentation on the atrial blood pool: 
		%by extracting the surface and volume of the blood pool from the image stack, an automatic labeling process is conducted to identify the 
		veins and appendages on the atrial blood pools are identified and labeled, 
		%and to select the portion of segmented vein tissue that is not relevant to the atrial geometry and will be subsequently removed;
		for defining the border between atrial myocardium and attached vasculature;
		\vspace{-0.2cm} \item extrusion of the volumetric atrial walls: 
		by accounting for the identified labels on the blood pools, a sequence of prescribed operations is employed on the segmentation to extrude the endo- and epicardial walls endowed with a selected thickness, automatically opening veins and valves, and generating a corresponding smooth volumetric mesh;
		\vspace{-0.2cm}\item selection of atrial orifices: a fully automated process is carried out to identify the complete set of vein ostia and valves on the atrial walls, separated in endocardial and epicardial tissue;    
		\vspace{-0.2cm}\item automatic labeling of anatomical structures and fiber generation:
		the remaining anatomical regions are automatically annotated, and the atrial fiber architecture is computed;
		\vspace{-0.2cm}\item \glspl{uac} generation: a system of normalized volumetric universal coordinates is computed;
		\vspace{-0.2cm}\item integration of a torso volumetric conductor: a reference torso anatomical model is automatically integrated with the biatrial anatomical model 
		and lead fields are computed for all considered electrode positions
		for computing all clinical \gls{ecg} signals of interest.
	\end{enumerate}
	
	The functional modeling stage comprises:
	\begin{enumerate}[i)]
		\setcounter{enumi}{7}
		%geared towards the calibration of the atrial anatomical model to \gls{ep} data comprises:
		% viii) selecting an \gls{ep} mathematical model: for representing atrial sources such as a \gls{rd} bi- or monodomain formulation 
		% or an \gls{re} formulation, along with a definition of parameter spaces to be used for calibration;
		\vspace{-0.2cm}\item generating \glspl{ic}: a flexible set of conduction pathways that cannot be delineated from images are computed to facilitate inter-atrial impulse conduction 
		% the following I'd deem to detailed for being brought up here, this is for later then.
		%are built employing the Purkinje technology defined in \cite{gillette2021automated}, and the \gls{uac} description to ensure reproducibility across different atrial anatomy;
		\vspace{-0.2cm}\item generate \gls{ep} clinical outputs: employing either \gls{rd} monodomain formulation, or a \gls{re} model to represent the transmembrane voltage, $V_m$, in the atrial model, a parameter spaces of \gls{ep} parameter is selected to be used for the model calibration. The P-wave of the 12-lead \glspl{ecg} on the torso are computed using the respective lead fields  \citep{gillette2022personalized,potse2018scalable}. The same approach can be potentially employed to compute \glspl{egm} in the blood pools.
	\end{enumerate}
	%\gls{ecg} and \glspl{egm} prediction by complementing the biatrial source model with a model of the generated potential field 
	% such as pseudo-bidomain \cite{bishop2011bidomain}, or pointwise evaluation of the potential field by techniques such as 
	% $\phi_{\rm e}$-recovery \cite{bishop2011bidomain} or the lead field approach \cite{potse2018scalable,gillette2022personalized}, 
	% to compute the P-wave of the 12-lead \gls{ecg} system or \glspl{egm} in the atrial blood pools. 
	
	% While in this work we present results obtained by solving either the monodomain or the \gls{re} model in the atria, the workflow allows for the alternative use of the \gls{rd} Bidomain model. Similarly, \glspl{ecg} computation can be carried out by coupling one of the aforementioned cardiac \gls{ep} models with a pseudo-bidomain model  \cite{bishop2011bidomain} in atrial-torso geometry, or a pointwise evaluation of the potential field by $\phi_{\rm e}$-recovery \cite{bishop2011bidomain}.    
	
	% moved this upwards to the anatomical stage
	%viii) torso volume conductor model: in the absence of a torso scan in the image stack, an automated registration algorithm is implemented to define a reference torso geometry around the atria;
	
	The entire workflow is implemented in a single user-friendly software 
	building on the package \texttt{meshtool} \citep{neic2020automating} and on \texttt{python} code. 
	For meshing operations the software TetGen (Weierstrass Institute, Berlin, Germany) 
	and NetGen (CerbSim GmbH, Vienna, Austria) have been licensed 
	and integrated with \texttt{meshtool}. 
	An interactive mode is moreover available, allowing for the verification of the labeling process, the visual control of all processing stages, and the correction of any potential software errors, through the use of the software
	\href{https://numericor.at/rlb/wordpress/products/}{CARPentry Studio} (Numericor GmbH, Graz, Austria). 
	All required interactive steps are readily supported by the freely available starter version of \href{https://numericor.at/rlb/wordpress/resources/}{CARPentry Studio for academia}.
	% \href{https://numericor.at/rlb/wordpress/resources/}{CARPentry Studio Core}.
	Each step of our workflow will be extensively described in the following sections.
	%The torso registration and mesh building process, as well as the definition of the interatrial connection and the \gls{ep} simulations have been implemented in a sequential and semi-automatic pipeline.
	
	\subsection{Anatomical twinning stage}
	\label{Subsec:anatomical_twinning}
	\subsubsection{Imaging data acquisition and segmentation}
	\label{Subsubsection:Imaging}
	For developing and testing the workflow, 50 patients 
	diagnosed with \gls{af} and scheduled for \gls{af} ablation therapy were selected.
	Iodanized contrast \gls{ct} scans were acquired at an isotropic resolution of \SI{0.4}{\milli \meter} 
	as part of the routine standard of care at the Medical University of Graz hospital 
	for patients included in the local \gls{af} ablation registry. 
	This registry was approved by the ethics committee of the Medical University of Graz (reference number 26-217 ex 13/14) 
	and all patients gave written informed consent. 
	
	Each \gls{ct} dataset was segmented using the \gls{scn} \citep{thaler2021efficient}, 
	an automated multi-label segmentation method based on convolutional neural networks. 
	Prior to input to the \gls{scn}, 3D image stacks were cropped around the center of the heart, 
	re-sampled to an isotropic resolution of \SI{1.2}{\milli \meter}, 
	and Gaussian filtered with $\sigma=$\SI{1}{\milli \meter}.
	The \gls{scn} was trained to recognize and label seven cardiac domains, 
	comprising all four major cardiac \gls{lv}, \gls{rv}, \gls{la}, \gls{ra} blood pools, 
	the \gls{lv} myocardium, and the vascular blood pools in the aorta and the pulmonary artery.
	Segmentation quality was assessed interactively, to ascertain topological soundness. 
	Topological errors, if present, such as connections between the blood pools of the \gls{la} appendage and the \gls{lspv}, 
	or between the \gls{cs} and the \gls{la} were identified, 
	and manually corrected using the \texttt{ITK-SNAP} software \citep{py06nimg}. 
	
	%Even if the presented auto-segmented method has been here primarily used on \gls{ct} scans, 
	%the \gls{scn} has been previously tested and shown to be highly functioning on \gls{mri} scans as well. 
	
	\subsubsection{Automated label augmentation on the blood pools}
	\label{Subsubsection:LabelAugmentation}
	Label augmentation is performed to identify the atrial appendages and veins, 
	which will later be used in the wall extrusion process. 
	Among all the volumetric labels defined through segmentation, 
	only the \gls{la} and \gls{ra} blood pools are targeted at this stage, while all other anatomical labels serve as auxiliary references.
	% Label augmentation and landmark selection on the atrial blood pools are performed targeting the \gls{la} and the \gls{ra} labels while all other anatomical labels are used as auxiliary references.
	%for label selection and atrial wall extrusion.
	
	The blood pool label augmentation is carried out in two stages: 
	the first stage automatically defines key landmarks on the surface of the blood pools; 
	the second stage effectively labels the entire blood pool volumes. 
	
	In the first stage, beginning with the segmentation, triangular surface meshes 
	of the \gls{la} and \gls{ra} blood pools are generated. 
	Key landmarks to identify include the \gls{lpv}, \gls{rpv}, and the \gls{laa} in the \gls{la}, 
	and the \gls{svc}, \gls{ivc}, \gls{cs}, and \gls{raa} in the \gls{ra}. 
	The automatic landmarking process is based on the computation of the curvature of the surface mesh, 
	as presented in \cite{azzolin2023:_augmenta} (see Figure \ref{Fig:image_process}(a)). 
	The curvature is defined point-wise, allowing for clustering of the mesh nodes based on curvature thresholding. 
	% The veins and the \gls{cs} are identified using the surface curvature 
	% computed on the surface mesh as in \cite{azzolin2023:_augmenta} (see Figure\ref{Fig:image_process}(a)). 
	Clusters of nodes with the highest curvature -- five on the \gls{la} and four in the \gls{ra} -- are considered for anatomical structure selection and marked using the following algorithm (see Figure \ref{Fig:image_process}(b)):
	\begin{itemize}
		\item \textbf{\gls{la}:} The first landmark selected is the vertex of the \gls{laa}, identified as the central point of the \gls{la} cluster of node closest to the ascending aorta, or, if the aorta is not labeled, 
		as the cluster with the highest curvature. 
		A point cloud algorithm based on \gls{pca} is then employed to divide the remaining clusters into two groups: 
		the two clusters nearest to the \gls{laa} are designated as \gls{lpv}, 
		while the others are identified as \gls{rpv}.
		\item \textbf{\gls{ra}:} Similarly to the \gls{la}, the vertex of the \gls{raa} is first selected as the central point of the \gls{ra} cluster closest to the ascending aorta or the one with the highest curvature. 
		\gls{pca} is then applied to the two largest clusters in terms of nodes to identify the vena cavae. 
		The \gls{svc} is designated as the cluster nearest to the \gls{raa}, while the other is marked as \gls{ivc}. 
		Among the remaining clusters with the highest curvature, the one nearest to the \gls{ivc} is labeled as \gls{cs}.
	\end{itemize}
	
	\begin{figure}[!t]
		\centering
		\includegraphics[width=0.9\textwidth]{images/image_process.png}
		\caption{(a) Point-wise curvature on the blood pool surface mesh. The figure also highlights the correlation between the surface region of high curvature and anatomical structures to mark. 
			(b) Minimal set of labels selected based on surface curvature, including \gls{lpv}, \gls{rpv} and \gls{laa} on the \gls{la}, 
			and \gls{ivc}, \gls{svc} and \gls{raa} on the \gls{ra}. 
			(c) Instance of \gls{ld} solution and isosurfaces on the blood pool volumetric mesh. 
			(d) Vein ostiae and the discardable tissue labeled on the blood pool volumetric mesh.}
		\label{Fig:image_process}
	\end{figure}  
	
	The second stage entails 
	%The subsequent phase of the workflow entails 
	the generation of a volumetric hexahedral mesh of \gls{la} and \gls{ra} blood pools 
	from the initial volumetric segmentation of the \gls{scn}.% and of the portion of vein tissue to be removed. 
	For each atrium, a series of \gls{ld} problems is solved 
	to define a set of measures of distance between the previously defined veins and appendages.
	Dirichlet boundary conditions are generated by projecting the landmarks defined on the blood pool surface mesh 
	onto the volumetric mesh (see Figure \ref{Fig:image_process}(c)). 
	Isosurfaces of the \gls{ld} solves are then employed to label the veins ostiae 
	according to prescribed thresholds. 
	The remaining portions of the veins are considered discardable tissue, as they will not be part of the final anatomical model. Therefore, they are marked for removal. 
	Hereon, we will define this landmark by employing the name of the vessel followed by a negative sign, 
	e.g.\ the portion of discardable tissue of the \gls{svc} will be denoted as \gls{svc}$^-$  (see Figure \ref{Fig:image_process}(d)). 
	The same procedure is also applied to the \gls{cs}. 
	The labels defined on the volumetric hexahedral mesh are finally projected back onto an image stack
	that will then contain all augmented fine-grained labeling information.
	%, which will subsequently used for the wall extrusion.
	
	The high variability in the atrial anatomy occasionally led the automatic algorithm 
	to mislabel some structures. 
	Specifically, errors in the positioning of the \gls{ivc} and \gls{cs} were witnessed, 
	or in the identification of a reasonable portion of discardable \gls{ivc} blood pool volume. 
	In such cases, the workflow visualization modality allows for a swift manual correction.
	% , through the use the \texttt{CARPentry studio} software (NumeriCor GmbH, Graz, Austria). 
	% by activating the workflow visualization modality. 
	
	As the atrial walls are thin, their accurate segmentation from current routine clinical imaging
	with \gls{ct} and \gls{mri} is, in general, not feasible \citep{bishop2016three,dewland2013use}. 
	%\emph{Ex vivo} measurements have, however, provide valuable information about the atrial walls.
	% and they cannot be reliably segmented from \gls{ct} or \gls{mri} datasets.
	In our workflow, we therefore use \emph{a priori} knowledge \citep{azzolin2020influence,beinart2011left,varela2017novel,whitaker2016role} 
	on the \gls{ra} and \gls{la} endo- and epicardial width to construct the atrial walls.
	A two steps process is applied. First, atrial walls are grown by image operations 
	applied to the image stack, followed by a mesh generation step 
	to create conformal unstructured meshes with smooth boundaries.
	
	Using the augmented volumetric segmentation, atrial walls are grown by imposing a series of rule-based extrusion operations at the interface of two different landmarks, or at the interface of marked tissue with the image background. 
	%slices, labeled as mentioned in the previous section, to ensure the extrusion of the atrial walls of a prescribed width,
	% Three major sets of rules are applied: 
	% an extrusion set at the interface of labeled domains and the background, 
	% where new atrial wall tissue labels are effectively grown into the background; 
	% an exclusion set of rules comprising domains labelled as not pertaining to the atrial blood pool, 
	% to impede extrusion along these interfaces, 
	% and interfaces between differently labelled domains, to impede any extrusion of the walls into segmented domains; 
	% an inclusion set, where interfaces between differently labeled domains are allowed to grow into each other.
	% % Rules for creating new labelled volumes representing anatomical structures such as the atrial walls from label interfaces
	% are defined based on two sets of labels 
	% where an extrusion set defines the labelled volumes to be extruded, 
	% and another inclusion set defines either permissible label volumes into which extruding labels grow into,
	% or, conversely, an exclusion set comprising non-permissible labels that pose a boundary that either excludes the label interface from extrusion, 
	% or impedes any further extrusion.
	Extrusion operations may comprise both growth in an inward direction, \emph{i.e.} erosion, 
	or outward, corresponding to negative and positive growth along the surface normal of the labeled domain.
	% Extrusion is defined as outwards if growth is positive along the surface normal of the volume defined by set of labels, or
	% inward dilation i.e.\ erosion, is also possible and always permissible.
	The spatial granularity of extrusion is prescribed in physical length units, i.e. mm, 
	which is converted to discrete steps corresponding to the voxel size of \SI{400}{\micro \meter} of the segmented image. 
	
	%The new atrial wall volume created by extrusion is labelled accordingly to form endo- and epicardial layers.
	%To new volumes created by extrusions are assigned a new label representing the recovered anatomical structure.
	
	Specifically, volumetric representations of the atrial walls are then generated by the following rules:
	i) an inclusion set of labels where interfaces between differently labeled domains are allowed to grow into each other, comprising the blood pool labels of \gls{ra}, and \gls{la},
	the ostiae of both caval veins, \gls{svc} and \gls{ivc}, and the pulmonary veins, \gls{rpv} and \gls{lpv}; 
	ii) an exclusion set comprising domains labeled as not pertaining to the atrial blood pool, 
	to impede extrusion along these interfaces. This set includes the \gls{lv} and \gls{rv} blood pools, 
	the \gls{cs}, the \gls{svc}$^-$, \gls{ivc}$^-$, as well as right and left pulmonary veins, \gls{rpv}$^-$, \gls{lpv}$^-$ (see Figure \ref{Fig:image_process}). 
	% of the excess blood pools of caval and pulmonary veins labeled separately as volume to be removed, \gls{svc}$^-$, \gls{ivc}$^-$, \gls{rpv}$^-$, \gls{lpv}$^-$ (see Fig.~\ref{Fig:image_process}).
	The extrusion operations are performed then on the inclusion set, with the exclusion set serving as boundaries
	that impede any further growth.
	First, an inward extrusion is performed to create a minimal endocardial surface layer, 
	followed by an outward dilation to thicken the endocardial layer to the prescribed width,
	and a further outward extrusion to create an epicardial layer. 
	Distinct values for the thickness of the of \gls{la} and \gls{ra} endocardial and epicardial layers,
	as well as the wall width of each vein can be specified independently.
	As the exclusion set impedes any inward or outward extrusion,
	growth is then blocked at the interfaces of \gls{ra} and \gls{rv}, and \gls{la} and \gls{lv} blood pools, respectively, 
	to create the orifices of \gls{tv} and \gls{mv}. Similarly, atrial walls are not grown at the interface of the \gls{ra} and \gls{la} blood pool with the tissue of the veins and \gls{cs} to be dismissed, thus effectively creating the openings of all in- and outflow anatomical structures. 
	% all anatomical in- and outflow openings of both atria are automatically created.
	Additional rules may be applied to mitigate effects due to limited segmentation accuracy.
	For instance, the growth of the epicardial wall may be constrained around the ostiae of the \gls{lspv} 
	in cases where \gls{laa} and the \gls{lspv} are in close proximity, to avoid merging. 
	%Similarly, the growth is limited around the ostia of the right superior pulmunary vein (\gls{rspv}) and the \gls{svc} when these two veins are nearly in contact. The entire procedure is fully automated, relying solely on the comparison of positions of the atrial walls and tissue based on the already identified labels.
	%Such labels are then projected from the blood pools to the atrial bi-layer walls. Two sets of corresponding marks are thus identified, one for the endocardial and one for the epicardial regions. 
	
	% Now we extract a surface mesh first
	After the extrusion phase, the walls of both atria are volumetrically defined on the augmented image stack. 
	%of a given isotropic image resolution.
	The generation of a volumetric anatomical mesh starts with extracting a surface mesh
	that encloses all atrial wall labels.
	This initial surface mesh, conforming to a jagged voxel representation,
	is subjected to several remeshing and smoothing steps to obtain a smooth representation of all atrial walls, with smooth transitions between wall segments of different widths. Topological corrections, mesh quality checks, and improvements are moreover implemented to ascertain topological correctness and sufficiently high mesh quality. Finally, the surface mesh is resampled to match a prescribed target resolution. 
	Typically, for \gls{re} models a resolution of \SI{0.9}{\milli\meter}, sufficient to resolve endo- and epicardial atrial wall layers, was selected.
	For \gls{rd} simulation, a higher resolution of \SI{0.25}{\milli \meter} was used 
	to resolve slowly propagating wavefronts without producing numerical artifacts \citep{bishop24:_devarp_trap}.
	As a final step, volumetric meshes of the atrial walls are created by the inward meshing of the labeled surfaces, 
	again followed by a topological correction, reindexing, refinement, 
	and mesh quality-enhancing procedures.
	Mesh quality of the final volumetric mesh is measured as in \cite{karabelas2018:_towards} 
	based on \cite{batdorf1997computational,kanchi20073d,knupp2022worst} (referred to Figure \ref{Fig:dbcs}-left).
	
	Thus, at the end of this stage, the biatrial anatomy is represented by an unstructured volumetric tetrahedral mesh endowed with basic anatomical labels, including endo- and epicardial domains of \gls{ra}, \gls{la}, \gls{rpv}, \gls{lpv}, \gls{ivc} and \gls{svc}. 
	
	% \todo[inline]{Maybe we show mesh quality as an inset in the figure}
	%If the interactive mode is active, mesh quality metrics and characteristic lengths of the geometry can be extracted and visualized.
	
	\subsubsection{Automated selection of the atrial orifices}
	\label{Subsubsection:orifices}
	% we are operating here already at a volumetric mesh, right?
	The atrial orifices comprising \gls{lpv}, \gls{rpv}, \gls{mv}, \gls{tv}, 
	\gls{svc}, \gls{ivc}, and \gls{cs} are individually labeled as separated anatomical entities.
	Specifically, following \cite{azzolin2023:_augmenta},    
	orifices are detected as mesh nodes of the biatrial surface 
	that are shared between adjacent elements with disjoint endo- or epicardial labels of the same anatomical structure. 
	% adjacent tetrahedral elements of the same anatomical structure,
	% but are disjointly labelled as endo- or epicardial.
	Each orifice is therefore represented as a cluster $\Gamma$ of mesh nodes 
	arranged along a ring, and
	%Cluster of nodes spanning the ring of an orifice 
	is readily identified based on anatomical labels, 
	except for the differentiation into superior and inferior of the \gls{rpv} and \gls{lpv}. 
	There, each entity, \gls{rpv} and \gls{lpv}, consists of two clusters which are discriminated based on the Euclidian distances between their centroid and those of the \gls{ivc}, \gls{svc}, \gls{mv}, \gls{mv} as well as the tips of \gls{raa} and \gls{laa}. 
	For instance, the centroid of the \gls{lspv} is closer to the centroid of the \gls{svc} and to the \gls{raa} tip. 
	Given the identified orifices, the anatomical regions of the \gls{la} labeled as \gls{rpv} and \gls{lpv} are updated 
	to separate the inferior and superior pulmonary veins.
	
	At this stage, the sets of nodes spanning the orifices $\Gamma_{ivc}$, $\Gamma_{svc}$, $\Gamma_{tv}$, $\Gamma_{cs}$ on the \gls{ra}, and
	$\Gamma_{rpv} = \Gamma_{ripv} \cup \Gamma_{rspv}$, $\Gamma_{lpv} = \Gamma_{lpv} \cup \Gamma_{lpv}$, and $\Gamma_{mv}$, are known 
	and can be used as boundary conditions 
	(see Figure \ref{Fig:dbcs} for a graphical representation 
	of the identified clusters), along with the corresponding anatomical regions, as represented in Figure \ref{Fig:wall_to_fibers}(a). 
	
	% this is not needed
	% For the \gls{la}, the remaining rings are clustered into two sets
	% where the set closest to the \gls{laa} is identified as \gls{lpv}, 
	% and the other set as \gls{rpv}. 
	% % this does not make any sense
	% Once the centroid of the \gls{lpv} cluster is computed, 
	% the \gls{lspv}is located to the right of the centroid, 
	% while the \gls{lipv} is positioned to the left. 
	% The same procedure is repeated to identify \gls{ripv} and \gls{rspv}.
	
	
	%For instance, \gls{mv} and \gls{tv} orifices are identified as the largest set of nodes on each atrium.
	
	% For the \gls{la}, the remaining rings are clustered into two sets
	% where the set closest to the \gls{laa} is identified as \gls{lpv}, 
	% and the other set as \gls{rpv}. 
	% Once the centroid of the \gls{lpv} cluster is computed, 
	% the \gls{lspv}is located to the right of the centroid, 
	% while the \gls{lipv} is positioned to the left. 
	% The same procedure is repeated to identify \gls{ripv} and \gls{rspv}.
	
	% Regarding the \gls{ra}, the cluster closest to the \gls{raa} is designated as the \gls{svc}. 
	% The largest of the two remaining orifices is marked as the \gls{ivc}, 
	% while the last one is defined as the \gls{cs}. 
	
	% \todo[inline]{Should be moved to next section, maybe we can show this in the figure, struggling to picture this for myself.}
	% Finally, a plane is constructed passing through the centers of \gls{svc}, \gls{ivc}, and \gls{tv}, 
	% dividing the \gls{ra} into septal and lateral regions 
	% to be subsequently used to define atrial fiber architecture.
	
	\subsubsection{Automated labeling of anatomical structures and fiber generation}
	\label{Subsubsec:anatomical_structure}
	Following \cite{azzolin2023:_augmenta,ZhengAzzolinSanchezDosselLoewe2021},
	a further refined sub-classification of anatomical structures is performed 
	on the \gls{ra} and \gls{la} walls to define anatomical structures 
	and tissue regions of known differences in \gls{ep} behaviors. 
	Such structures -- including the \gls{san}, \gls{ct} and \glspl{pm}, the \gls{bb}, 
	and the muscular rim of the \gls{fo}, as well as \gls{raa} and \gls{laa} -- 
	are automatically annotated on a per-rule basis defining 
	location and width of the sought-after anatomical structures (refer to Figure \ref{Fig:wall_to_fibers}b).
	% A further refined sub-classification of anatomical structures is performed.
	% Following \cite{ZhengAzzolinSanchezDosselLoewe2021,azzolin2023:_augmenta},
	% anatomical structures or tissues of different \gls{ep} behaviours comprising 
	% the \gls{san}, \gls{ct} and \glspl{pm}, \gls{bb}, the rim of the \gls{fo} as well as \gls{raa} and \gls{laa}  
	% are automatically annotated on a per-rule basis (refer to Figure\ref{Fig:wall_to_fibers}b).
	
	The implemented rules are based on solutions of multiple instances of \gls{ld} problems 
	on the domain $\Omega_{myo}$, i.e. the right $\Omega_{\mathrm{RA}}$ or left $\Omega_{\mathrm{LA}}$ atrium, with varying boundary conditions. 
	Specifically, given a generic unknown $\varphi$, this involves solving
	\[ \begin{cases}
	-\Delta \varphi = 0 &\text{in } \Omega_{myo},\\
	\varphi = \varphi_i &\text{on } \Gamma_i,~i=1,\dots,m\\
	\nabla \varphi \cdot \mathbf{n} = 0 &\text{on } \Gamma_n,
	\end{cases}
	\]
	for $m = 2,3,4$ suitable \gls{dbc} $\varphi_i \in \R$, 
	set on generic partitions of the atria boundaries $\Gamma_i$, and Neumann boundary conditions $\Gamma_n$, 
	such that $\cup_{i = 1}^{m}\Gamma_i \cup \Gamma_n = \partial \Omega_{myo}$. The \gls{ld} solutions are used as measures of distance between the orifices and/or, the appendages. 
	Moreover, a plane passing through the center of the \gls{ivc} and \gls{svc}, and the center of mass of the \gls{ra} is computed to separate the \gls{ra} into lateral and septal regions. Hereon, we will call the intersection of such plane with the \gls{ra} wall as \gls{ra} roof.
	Similarly, the \gls{tv} annulus is partitioned into 
	a lateral and a septal tricuspid valve annulus, 
	$\Gamma_{tv_{l}}$ and $\Gamma_{tv_{s}}$, respectively. 
	A representation of the defined set of boundaries is shown in Figure \ref{Fig:dbcs}. 
	
	\begin{figure}[!t]
		\centering
		\includegraphics[width=0.9\textwidth]{images/Figure_meshquality_boundaries.png}
		\caption{Left: Mesh quality computed on the biatrial model. 
			Right: Boundaries on the \gls{ra} and \gls{la} 15 \gls{ld} problems employed to determine atrial anatomical structures and tissues known to have different \gls{ep} properties.}
		\label{Fig:dbcs}
	\end{figure}
	
	%$$ -\Delta \varphi = 0 \quad \text{in }\Omega_H$$
	% with up to three \glspl{dbc} referred to as $\varphi_a$, $\varphi_b$, and $\varphi_c$, imposed on specific parts of the boundary.
	% might be true, but why we are discussing this here?
	%The apices of the \gls{laa} and the \gls{raa}, 
	%which were manually selected as landmarks previously \cite{azzolin2023:_augmenta}, are here automatically identified in the preceding steps of the workflow.
	Compared to previous work \citep{azzolin2023:_augmenta,ZhengAzzolinSanchezDosselLoewe2021}, 
	three additional \gls{ld} solutions were computed on the \gls{ra}: 
	i) $\varphi_{v3}$ with \glspl{dbc} $\varphi_1 = 0$ and $\varphi_2 = 1$ imposed at the rings of \gls{svc} and \gls{ivc}, respectively; 
	ii) $\varphi_{r2}$ with $\varphi_1 = 0$ at the rings of the \gls{svc} and \gls{ivc}, and on the \gls{ra} roof, 
	and $\varphi_2 = 1$ at the ring of the \gls{mv}; and 
	iii) $\varphi_{w2}$ with \glspl{dbc}s $\varphi_1 = 0$ on the roof, $\varphi_2 = 1$ and $\varphi_3 = -1$ on the septal and lateral parts, respectively, at the \gls{mv} annulus. 
	The complete set of \glspl{dbc} is summarized in Table \ref{Table:dbcs}.
	These additional solutions are designed to improve the definition of the \gls{ct} and \glspl{pm}.
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lccccccccc}
			\toprule
			Atrium &$\varphi$ &$\varphi_1$ &$\Gamma_1$ &$\varphi_2$ &$\Gamma_2$ &$\varphi_3$ &$\Gamma_3$ &$\varphi_4$ &$\Gamma_4$ \\
			\hline \\[-2ex]
			\multirow{9}{*}{\gls{ra}} &$\varphi_{trans}$ &0 &$\Gamma_{endo}$ &1 &$\Gamma_{epi}$ &&&&\\
			&$\varphi_{ab}$ &-1 &$\Gamma_{raa}$ &0 &$\Gamma_{svc}$ &1 &$\Gamma_{tv_{l}}\cup \Gamma_{tv_{s}}$ &2 &$\Gamma_{ivc}$\\
			&$\varphi_{v}$ &0 &$\Gamma_{svc}\cup\Gamma_{raa}$ &1 &$\Gamma_{ivc}$ &&&&\\
			&$\varphi_{v2}$ &0 &$\Gamma_{ivc}$ &1 &$\Gamma_{raa}$ &&&&\\
			&$\varphi_{v3}$ &0 &$\Gamma_{svc}$ &1 &$\Gamma_{ivc}$ &&&&\\
			&$\varphi_{r}$ &0 &$\Gamma_{roof}$ &1 &$\Gamma_{tv_{l}}\cup \Gamma_{tv_{s}}$ &&&&\\
			&$\varphi_{r2}$ &0 &$\Gamma_{svc}\cup\Gamma_{roof}\cup\Gamma_{ivc}$ &1 &$\Gamma_{tv}$ &&&&\\
			&$\Gamma_{w}$ &-1 &$\Gamma_{tv_{l}}$ &1 &$\Gamma_{tv_{s}}$ &&&&\\
			&$\Gamma_{w2}$ &-1 &$\Gamma_{tv_{l}}$ &0 &$\Gamma_{roof}$ &1 &$\Gamma_{tv_{s}}$ &&\\
			\hline \\[-2ex]
			\multirow{6}{*}{\gls{la}} &$\varphi_{trans}$ &0 &$\Gamma_{endo}$ &1 &$\Gamma_{epi}$ &&&&\\
			&$\varphi_{ab}$ &-1 &$\Gamma_{laa}$ &0 &$\Gamma_{lpv}$ &1 &$\Gamma_{mv}$ &2 &$\Gamma_{rpv}$\\
			&$\varphi_{ab2}$ &0 &$\Gamma_{rpv}$ &1 &$\Gamma_{laa}$ &&&&\\
			&$\varphi_{v}$ &0 &$\Gamma_{lpv}$ &1 &$\Gamma_{rpv}$ &&&&\\
			&$\varphi_{r}$ &0 &$\Gamma_{rpv}\cup\Gamma_{lpv}\cup\Gamma_{mv}$ &1 &$\Gamma_{mv}$ &&&&\\
			&$\varphi_{r2}$ &0 &$\Gamma_{rpv}\cup\Gamma_{lpv}$ &1 &$\Gamma_{mv}$ &&&&\\
			\bottomrule
		\end{tabular}
		\caption{\glspl{dbc} for the 15 \gls{ld} problems employed to determine atrial anatomical structures and tissues known to have different \gls{ep} properties.}
		\label{Table:dbcs}
	\end{table}
	
	\begin{figure}[!t]
		\centering
		\includegraphics[width=0.9\textwidth]{images/wall_to_fibers.png}
		\caption{(a) Identification and labeling of the atrial orifices. 
			Different labels are used for endocardial and epicardial tissue layers. 
			(b) Anatomical structures identified on the biatrial anatomy, 
			including \gls{ct}, \gls{san}, \gls{pm}s, \gls{fo}, and part of the \gls{bb}. 
			(c) Generated rule-based atrial fiber architecture. 
			(d) Final biatrial volumetric mesh, augmented with labeled anatomical structures.}
		\label{Fig:wall_to_fibers}
	\end{figure}
	
	% \todo[inline]{In which way, I assume we use ranges of these solutions, we should spell this out here as this is different from previous work.}
	% \todo[inline]{We do not have a supplementary material as per yet, I don't think we will have one. I'd prefer to add boundary conditions to the figures, something like $\partial_{\rm {svc}}$}
	
	
	The labeling of the \gls{bb} follows \cite{ZhengAzzolinSanchezDosselLoewe2021}. 
	Using the additional solutions, we marked the area of the exit site of the \gls{san}, and, optionally, the muscular rim of the \gls{fo}, 
	and the anterior-central band of the \gls{bb}. 
	% In addition, our workflow also labels the \gls{san}, the rim of the \gls{fo}, 
	% and the anterior-central band of the \gls{bb}. 
	The \gls{san} is defined as a sphere with a radius of \SI{2.5}{\milli\meter} and centered on the node closest to the \gls{svc} with minimum $\varphi_{v2}$. 
	The \gls{fo} is defined as an annulus with thickness of \SI{2}{\milli\meter}, 
	located in the center of the atrial septum.
	The start and end points of the anterior-central band of the \gls{bb} 
	are selected as two points in the middle of the \gls{bb} lateral bands 
	of both \gls{ra} and \gls{la}.
	Additionally, an intermediate point is chosen as the epicardial node of the \gls{la} nearest to the midpoint of the \gls{ra} lateral \gls{bb}. 
	A geodesic path is, moreover, created using Dijkstra's algorithm between the starting and intermediate points, and between the intermediate and end points. 
	Around this path, a region of \SI{2}{\milli\meter} width is marked as \gls{bb}. 
	Unlike in \cite{azzolin2023:_augmenta,ZhengAzzolinSanchezDosselLoewe2021}, 
	no additional volumetric mesh is added to define the central bundle of the \gls{bb}. 
	While our workflow allows for marking the complete \gls{bb}, 
	in this work, we replace the anterior-central band of the \gls{bb}, 
	as well as all other inter-atrial conduction pathways,
	by conducting cables (see Section \ref{Subsubsection:cables}),
	increasing the flexibility of the modeling approach.
	
	Finally, the fiber architecture is computed using the rule-based method 
	presented in \cite{Piersanti2021}, 
	including the enhancements previously reported in \cite{azzolin2023:_augmenta,ZhengAzzolinSanchezDosselLoewe2021}. 
	A graphical overview of all anatomical structures, tissues, and fibers, 
	is given in Figure \ref{Fig:wall_to_fibers}.
	
	\subsubsection{Universal atrial coordinates generation}
	\label{Subsubsec:UAC}
	To compute the \glspl{uac} directly on the volumetric models 
	we expand upon the method outlined in \cite{roney2023:_bia_vol,roney2021constructing}. For each atrium, we define four coordinates:
	i) $\alpha_{\mathrm{RA}}$, the \gls{ivc}-to-\gls{svc} coordinate on the \gls{ra},
	and $\alpha_{\mathrm{LA}}$, the lateral-to-septal coordinate on the \gls{la} (see illustration in Figure \ref{Fig:uac}(a));
	ii) $\beta_{\mathrm{RA}}$, the lateral-to-septal coordinate on the \gls{ra} 
	that starts at the lateral \gls{tv} annulus, runs through the roof, and ends at the septal \gls{tv} annulus,
	and $\beta_{\mathrm{LA}}$, the posterior-to-anterior coordinate on the \gls{la}
	that starts at the posterior \gls{mv} annulus, runs through the roof, and ends at the anterior \gls{mv} annulus  
	(see Figure \ref{Fig:uac}(b));
	iii) $\gamma_{\mathrm{RA}}$ and $\gamma_{\mathrm{LA}}$, the relative transmural distance 
	between endocardium and epicardium in \gls{ra} and \gls{la}, respectively (see Figure \ref{Fig:uac}(c));
	iv) a binary value encoding to which atrium the other coordinates belong, always assigning shared points at the \gls{la}--\gls{ra} interface to the \gls{la}.
	% The fourth component is a binary value and specifies to 
	% which atrium the coordinate belongs. The points at the \gls{la}--\gls{ra} interface  to 
	% the \gls{la} by definition. 
	To compute the four coordinates, for each atrium, we first solve three \gls{ld} problems,
	imposing \gls{dbc} at pre-defined boundary surfaces. Secondly, to ascertain equal locations of the \gls{cs}, \gls{lipv}, and \gls{ripv} in the \gls{uac} space across different atrial models, a linear elasticity problem is solved.
	
	\begin{figure}[!t]
		\centering
		\includegraphics[width=0.9\textwidth]{images/uac.png}
		\caption{(a) Distribution of the 
			coordinate $\alpha$, representing the $\gls{svc}$-to-\gls{ivc} coordinate for the $\gls{ra}$, and the lateral-to-septal 
			coordinate for the \gls{la}. (b) Distribution of the coordinate $\beta$, representing the lateral-to-septal 
			coordinate for the \gls{ra}, and the posterior-to-anterior coordinate for the \gls{la}. (c) Distribution 
			of the coordinate $\gamma$, representing the endocardial-to-epicardial coordinate for both atria. (d) Projection of the labels on the space generated by the \gls{uac} before (left) and after (right) the solution of the linear elasticity problem. }
		\label{Fig:uac}
	\end{figure}
	
	% and to ensure that the anatomical structures are located at the 
	% same \glspl{uac} throughout the models, a linear elasticity problem is solved subsequently.
	
	%The second \gls{uac} is the lateral-to-septal 
	%coordinate in the \gls{ra}, called $\beta_{RA}$, whereas it is chosen from the posterior \gls{mv} to the anterior 
	%\gls{mv}, passing through the roof of the left atrium, and is denoted as $\beta_{LA}$. This last coordinate is 
	%denoted as $\beta_{RA}$ (see Figure\ref{Fig:uac}b). A third coordinate $\gamma_{LA,RA}$ is used to describe the 
	%transmural distance between the endocardium and the epicardium of both atria, whereas the last coordinate simply 
	%indicates the \gls{ra} or \gls{la} (see Figure\ref{Fig:uac}c). For each atria, we therefore solve three 
	%Laplace-Dirichlet problem, imposing Dirichlet boundary conditions along three set of boundary nodes. Moreover, 
	%to ensure that the anatomical structures of the veins are located in the same \gls{uac}, first we compute the 
	%standard \gls{uac} system, and then we solve a linear elasticity problem to move the opening on the vessels 
	%in the same \gls{uac} position. 
	
	The selection of the boundary surfaces is fully automated and based on the anatomical labels.
	% in the atria models defining the anatomical structures. 
	Specifically, the marked atrial orifices are utilized to define a set of reproducible points on both the vessel openings and the atrial walls. 
	Such points are selected according to the general shape of the atria, to follow the roof of the \gls{ra}, between the \gls{svc} and the \gls{ivc}, 
	and of the \gls{la}, between the \gls{lspv} and \gls{rspv}, and to set lateral and septal boundaries in both atria. 
	The points are then connected with quadrilateral surfaces. 
	The intersections of such surfaces with the atrial volumetric walls define six auxiliary surface meshes, three for each atrium, 
	effectively representing the interfaces separating the
	% The labels are utilized to identify    
	% reproducible 
	% points on both the openings of the vessels and on the atrial body. These points are then used to construct 
	% two auxiliary surface meshes, one for each atrium, to separate the
	\gls{ra} into a septal 
	$\Omega_{\mathrm{RA},sept}$ and a lateral  $\Omega_{\mathrm{RA},lat}$ part, 
	and the \gls{la} in an anterior $\Omega_{\mathrm{LA},ant}$ and a posterior $\Omega_{\mathrm{LA},post}$ part. 
	Moreover, to avoid splitting of the appendages, the \gls{raa} is assigned to the lateral side of the \gls{ra}, 
	and the \gls{laa} is assigned to the anterior 
	side of the \gls{la}. 
	% From this partition, three transmural interfaces are then extracted for each atrium. 
	Therefore, for the \gls{ra}  we obtain two interfaces, $\mathcal{I}_{tv,ivc}$ and $\mathcal{I}_{tv,svc}$, 
	from the \gls{tv} to the \gls{ivc} and \gls{svc}, respectively, 
	and one interface $\mathcal{I}_{\mathrm{ivc,svc}}$ along the roof from \gls{ivc} to \gls{svc}. 
	Similarly, the two interfaces $\mathcal{I}_{mv,lspv}$ and $\mathcal{I}_{mv,rspv}$,  
	from the \gls{mv} to the \gls{lspv} and \gls{rspv}, respectively, are computed for the \gls{la}, 
	together with the interface $\mathcal{I}_{lspv,rspv}$ along the roof from \gls{lspv} to \gls{rspv}. 
	The defined interfaces are furthermore used to partition the \gls{tv}, \gls{ivc}, and \gls{svc} into septal and lateral parts, 
	and the \gls{mv}, \gls{lspv}, and \gls{rspv} into anterior and posterior (see Figure \ref{Fig:UCAparametrization} for a graphical representation). 
	Such partition will be used in the following when solving the linear elasticity problem.
	
	\begin{figure}[!t]
		\centering
		\includegraphics[width=0.7\textwidth]{images/parameterization.png}
		\caption{(a) Interfaces and boundary surfaces in the \gls{ra} and \gls{la} and their parametrization.
			(b) Dirichlet values at the boundaries and interfaces used to compute the preliminary 
			$\alpha$ and $\beta$ components for the \gls{ra} and \gls{la} are shown. (c) Selected landmarks and the parametrization of the \gls{cs}, the \gls{lipv}, and the \gls{ripv}. (d) Final $\alpha$ and $\beta$ components after moving the orifices. The color are coded from blue $\!=\!0$ to red $\!=\!1$. }
		\label{Fig:UCAparametrization}
	\end{figure}
	
	% The partitioning also separates the \gls{tv}, \gls{ivc}, and \gls{svc} into septal and lateral parts, and 
	% the \gls{mv}, \gls{lspv}, and \gls{rspv} into anterior and posterior (see Figure\ref{Fig:UCAparametrization}).
	Both the interfaces and the atrial openings are then parametrized with a value $s \in [0, 1]$ with a Bi-Eikonal approach \citep{schuler2021cobiveco}.
	% A parametrization value $s \in [0, 1]$ is computed for each interface using a Bi-Eikonal approach \cite{schuler2021cobiveco}.
	The $\alpha$ and $\beta$ coordinates are computed by solving \gls{ld} problems 
	on the partitioned $\Omega_{\mathrm{RA}}$ and $\Omega_{\mathrm{LA}}$, 
	with \gls{dbc} assigned based on $s$, as defined in Table \ref{Tab:UACdirichlet}.
	%(see Figure \ref{Fig:UCAparametrizationFlat}(a) and Table \ref{Tab:UACdirichlet}). 
	For the $\gamma$ component, only two Dirichlet boundary conditions are imposed, 
	$0$ at the endocardium and $1$ at the epicardium.
	
	\begin{table}
		\centering
		%\addtolength{\tabcolsep}{5.0pt}
		\begin{tabularx}{\textwidth}{X|cc|cc}
			\toprule
			& \multicolumn{2}{c|}{$\Omega_{\mathrm{RA},sept}\;\;$ and $\;\;\Omega_{\mathrm{LA},ant}$} & 
			\multicolumn{2}{c}{$\Omega_{\mathrm{RA},lat}\;\;$ and $\;\;\Omega_{\mathrm{LA},post}$} \\
			& $\alpha$ & $\beta$ & $\alpha$ & $\beta$ \\
			\midrule
			$\Gamma_{tv}$, $\Gamma_{mv}$
			& -- & 1 
			& -- & 0 \\[0.15cm]
			$\Gamma_{ivc}$, $\Gamma_{lspv}$
			& $\mathrm{R}\,\cos\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$ 
			& $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$
			& $\mathrm{R}\,\cos\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$ 
			& $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((3\!+\!s)\,\frac{\pi}{2}\Big)$ \\[0.15cm]
			$\Gamma_{svc}$, $\Gamma_{rspv}$
			& $1\!+\!\mathrm{R}\,\cos\Big((1\!+\!s)\,\frac{\pi}{2}\Big)$ 
			& $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$
			& $1\!+\!\mathrm{R}\,\cos\Big((1\!+\!s)\,\frac{\pi}{2}\Big)$ 
			& $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((3\!+\!s)\,\frac{\pi}{2}\Big)$ \\[0.15cm]
			$\mathcal{I}_{tv,ivc}$, $\mathcal{I}_{mv,lspv}$
			& $0$ & $\Big(\frac{1}{2}\!+\!\mathrm{R}\Big)+(1\!-\!s)\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ 
			& $0$ & $s\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ \\[0.15cm]
			$\mathcal{I}_{tv,svc}$, $\mathcal{I}_{mv,rspv}$
			& $1$ & $\Big(\frac{1}{2}\!+\!\mathrm{R}\Big)+(1\!-\!s)\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ 
			& $1$ & $s\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ \\[0.15cm]
			$\mathcal{I}_{ivc,svc}$, $\mathcal{I}_{lspv,rspv}$
			& $\mathrm{R}\!+\!(1\!-\!2\,\mathrm{R})\,s$ & $\frac{1}{2}$ 
			& $\mathrm{R}\!+\!(1\!-\!2\,\mathrm{R})\,s$ & $\frac{1}{2}$ \\[0.15cm]
			\bottomrule
		\end{tabularx}
		\caption{Dirichlet boundary conditions for the Laplace problems that are solved to obtain 
			the preliminary \gls{uac} space in the septal and lateral \gls{ra} and the posterior and 
			anterior \gls{la} where $s \in [0, 1]$ is a parametrization value of the corresponding
			boundary surface and interface, respectively. In our framework,  $\mathrm{R}=0.1$.}
		\label{Tab:UACdirichlet}
	\end{table}
	\begin{table}
		%\addtolength{\tabcolsep}{7.0pt}
		\begin{tabularx}{0.6\textwidth}{l|cc}
			\toprule
			& $\alpha$ & $\beta$ \\
			\midrule
			$\Gamma_{cs,roof}$
			& $\alpha_{cs}\!+\!\mathrm{r}\,\cos\Big((2\!-\!s)\,\pi\Big)$
			& $\beta_{cs}\!+\!\mathrm{r}\,\sin\Big((2\!-\!s)\,\pi\Big)$ \\[0.15cm] 
			$\Gamma_{cs,tv}$
			& $\alpha_{cs}\!+\!\mathrm{r}\,\cos\Big(s\,\pi\Big)$ 
			& $\beta_{cs}\!+\!\mathrm{r}\,\sin\Big(s\,\pi\Big)$ \\[0.15cm]
			$\Gamma_{lipv,roof}$
			& $\alpha_{ipv}\!+\!\mathrm{r}\,\cos\Big(s\,\pi\Big)$ 
			& $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big(s\,\pi\Big)$ \\[0.15cm]
			$\Gamma_{lipv,mv}$
			& $\alpha_{ipv}\!+\!\mathrm{r}\,\cos\Big((2\!-\!s)\,\pi\Big)$
			& $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big((2\!-\!s)\,\pi\Big)$ \\[0.15cm] 
			$\Gamma_{ripv,roof}$
			& $(1\!-\!\alpha_{ipv})\!+\!\mathrm{r}\,\cos\Big(s\,\pi\Big)$
			& $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big(s\,\pi\Big)$ \\[0.15cm]
			$\Gamma_{ripv,mv}$
			& $(1\!-\!\alpha_{ipv})\!+\!\mathrm{r}\,\cos\Big((2\!-\!s)\,\pi\Big)$
			& $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big((2\!-\!s)\,\pi\Big)$ \\[0.15cm]
			\bottomrule
		\end{tabularx}
		
		\caption{Final position of the \gls{cs}, \gls{lipv}, and \gls{ripv} orifices with parameterization value $s \in [0, 1]$.}
		\label{Tab:UACdirichletLinElas}
	\end{table}
	
	% To ensure the same position in the \gls{uac} space of the \gls{cs}, \gls{lipv}, and \gls{ripv}, a linear elasticity problem in solved in the preliminary \gls{uac} space. 
	The linear elasticity problem is solved in the \gls{uac} space, by projecting the first three coordinates onto two thin cuboids, one for each atrium, 
	where $\alpha$ and $\beta$ are described along the edges of the square face, and $\gamma$ represents the thickness of the cuboid. 
	The \gls{ivc}, \gls{svc}, \gls{rspv}, and \gls{lspv} are moreover recast in circumferential form, on the edges of the square face representing the $\beta$ coordinate.    
	The square faces of each cuboid are then representing the endocardial and epicardial surfaces of the atrial walls.
	An updated Lagrangian formulation is then used to solve a linear elasticity problem, 
	prescribing rule-based Dirichlet boundary conditions for the \gls{cs}, \gls{ripv}, and \gls{lipv}.    
	To achieve this, the parametrization of the \gls{cs}, \gls{ripv}, and \gls{lipv} is carried out as follows. 
	Beginning with the mesh nodes of each orifice, i.e., $\Gamma_{ripv}$, $\Gamma_{lipv}$, and $\Gamma_{cs}$ as defined in Section \ref{Subsubsection:orifices}, 
	a node selection is grown across the biatrial surface in all directions. 
	The outermost selected nodes form two rings, one on the endocardium and one on the epicardium. 
	For each ring, four points are identified, corresponding to the minimum and maximum values of the $\alpha$ and $\beta$ components. 
	%(as shown in Figure \ref{Fig:UCAparametrizationFlat}(b) of Appendix \ref{appendix_UAC}).
	and for each of the four endocardial-epicardial point pairs, the shortest path is extracted on the selected surface.
	The \gls{cs} node selection is then divided into a roof and a \gls{tv} section 
	and parameterized using a \gls{ld} solve, with Dirichlet boundary conditions applied to the eight previously extracted paths. 
	Similarly, the \gls{lipv} and \gls{ripv} node selections are divided into a roof and a \gls{mv} section and parameterized in the same manner. %(refer to Figure \ref{Fig:UCAparametrizationFlat}(b)). 
	The final position of the \gls{cs}, \gls{ripv}, and \gls{lipv} after convergence of the linear elasticity problem is defined in Table \ref{Tab:UACdirichletLinElas} 
	%referred to Figure \ref{Fig:UCAparametrizationFlat}(c), and Table \ref{Tab:UACdirichlet}, 
	with $\mathrm{R}=0.1$, $\mathrm{r}=0.04$, $\alpha_{cs}=0.2$, $\beta_{cs}=0.8$, $\alpha_{ipv}=0.25$, and 
	$\beta_{ipv}=0.25$) and in Figure \ref{Fig:uac}(d), where the anatomical labels are also reported.
	
	% For this purpose, 
	% four mesh nodes on the endocardium and four on the epicardium of all the orifices are selected, representing the points at the orifice intersection with the interfaces at which  $\alpha$ and $\beta$ components have minimum and maximum values (referred to Figure\ref{Fig:UCAparametrizationFlat}(b) of Appendix \ref{appendix_UAC}).
	% minimum and maximum values of the $\alpha$ and $\beta$ components at the intersection of the orifice 
	% and the corresponding surface (see Figure\ref{Fig:UCAparametrizationFlat}b). 
	% Moreover, 
	% The \gls{cs} boundary surface 
	% is subdivided into a roof and a \gls{tv} part where each part is again parameterized using an Bi-Eikonak approach. 
	% The \gls{lipv} and \gls{ripv} boundary surfaces are also subdivided into a roof and a \gls{mv} part and are 
	% parameterized in the same manner (see Figure\ref{Fig:UCAparametrizationFlat}b). An updated Lagrangian 
	% formulation is then used to solve a linear elasticity problem with Dirichlet boundary conditions for the 
	% three orifices. The final position of the surfaces is defined in Table\ref{Tab:UACdirichlet} with 
	% $\mathrm{r}=0.04$, $\alpha_{\mathrm{CS}}=0.2$, $\beta_{\mathrm{CS}}=0.8$, $\alpha_{\mathrm{IPV}}=0.25$, and 
	% $\beta_{\mathrm{IPV}}=0.25$
	
	
	%The concept involves creating a plane passing through the selected points, which 
	%intersects the atrial wall to form three sets of connected mesh nodes. Each of these interface sets is then 
	%used to assign boundary conditions for the Laplace-Dirichlet solves. In practical implementation, a small set 
	%of points is defined on the roof, left, and right sides of the atria, connected through a series of continuous 
	%quadrilaterals.
	
	%The selection of boundary nodes is automated and computed based on the available labels partition of anatomical 
	%structures. These labels are utilized to identify reproducible points on both the openings of the vessels and 
	%the walls of the atria. The concept involves creating a plane passing through the selected points, which 
	%intersects the atrial wall to form three sets of connected mesh nodes. Each of these interface sets is then 
	%used to assign boundary conditions for the Laplace-Dirichlet solves. In practical implementation, a small set 
	%of points is defined on the roof, left, and right sides of the atria, connected through a series of continuous 
	%quadrilaterals.
	
	%Specifically, for the \gls{la}, we define seven points on the roof connecting the right and left superior 
	%pulmonary veins, and five points on each side of the atria connecting the \gls{rspv} or \gls{lspv} with the 
	%\gls{mv}. The boundaries defined by these last set of points are called as the right and left interface boundaries, 
	%respectively. The selection of points is entirely based on the general shape of the atria, taking into consideration 
	%the direction of the \gls{rspv} and \gls{lspv} in relation to the shape of the atrial body, as well as the relative 
	%position of the \gls{rspv} and \gls{lspv} to the mitral valve (for a more complete description, see Appendix 
	%\ref{appendix_UAC}). Once the quadrilateral set have been built, the mesh nodes of the \gls{la} wall are divided 
	%into anterior and posterior nodes, and the relative three interfaces are computed. Additionally, the set of anterior 
	%nodes is adjusted to include the \gls{laa}. The Laplace solutions are then calculated across the entire mesh, 
	%prescribing Dirichlet boundary values between 0 and 1 on the related interfaces: $\alpha_{LA}$ is obtained by 
	%assigning 0 to the left interface and 1 to the right; while for $\beta_{LA}$, we prescribe the solution to be 0.5 
	%on the roof, 0 on the posterior portion of the \gls{mv}, and 1 on the anterior portion of the \gls{mv}. Finally, 
	%to the transmural coordinate $\gamma_{LA}$ we assign Dirichlet boundary equal to 0 on the endocardial surface and 
	%1 on the epicardial surface. These \gls{uac}s represent the standard system, which must then be normalized to be 
	%compatible with different \gls{la} geometries. To achieve this, we project the three \gls{uac}s into a reference 
	%system represented by a thin three-dimensional box while maintaining the mapping between the original mesh node 
	%locations and the corresponding \gls{uac} representation (see Figure\ref{Fig:uac}d). Subsequently, a linear 
	%elasticity problem is solved to adjust the superior pulmonary veins to the reference location and shape of 
	%the box (for further details, refer to Appendix \ref{appendix_UAC}). Finally, the \gls{uac}s are mapped back 
	%to the \gls{la} geometry.
	
	%Concerning the \gls{ra}, the roof is defined by six points between the \gls{svc} and \gls{ivc}, while five points 
	%delineate the right side between the \gls{ivc} and the \gls{tv}. Additionally, five points identify the left 
	%boundary between the \gls{svc} and the \gls{tv}. Similar to the approach used for the \gls{la}, these points are 
	%chosen to correspond with the relative positions of the \gls{svc} and \gls{ivc} to the atrial body, as well as 
	%their relative locations to the \gls{tv} (for a more detailed explanation, see Appendix \ref{appendix_UAC}).
	%The set of quadrilaterals passing through these points are used to partition the mesh nodes into septal and 
	%lateral nodes of the \gls{ra}, and the corresponding interface node are defined. As for the \gls{la}, the 
	%\gls{raa} is automatically assigned to the lateral side of the \gls{ra}. The \gls{uac} $\alpha_{RA}$ is obtained 
	%by solving a Laplace problem with Dirichlet boundary values set to 0 on the right interface and 1 on the left one. 
	%The septal-lateral metric $\beta_{RA}$ is determined by imposing Dirichlet conditions of 0.5 on the roof, 0 on the 
	%septal portion of the \gls{tv}, and 1 on the lateral portion of the \gls{tv}. Finally, $\gamma_{RA}$ is computed 
	%similarly to the \gls{la}. Upon projection of the \gls{uac}s onto a three-dimensional box (see Figure\ref{Fig:uac}d), 
	%a linear elasticity problem is solved to adjust the \gls{ivc} and \gls{svc} to their reference locations, thus 
	%normalizing the right atrial \gls{uac} for different geometries. The final step involves remapping the \gls{uac} 
	%onto the original \gls{ra} mesh.
	
	\subsubsection{Torso volume conductor model}
	\label{Subsubsec:Torso}
	Since most clinical observations of atrial \gls{ep} are obtained by sampling the potential fields generated within the volume conductor, such as \glspl{ecg} or \glspl{egm}, volumetric models must be embedded in a torso domain to enable the prediction of these observations.
	However, the field of view of the \gls{ct} scans of the patients within the Graz \gls{af} registry was restricted to the heart to minimize the patient's exposure to x-ray radiation.
	Thus a full torso view including all \gls{ecg} recording sites was not available.
	Thus, to predict atrial P-waves in the \gls{ecg}, an automated procedure was implemented for registering a template heart-torso model with known electrode locations around the generated atria. The template was selected from a previously generated virtual cohort \citep{gillette2021automated, gillette2022personalized}.
	%, compromising 9 male and 5 female models of variable height, weight, and body mass index \cite{gillette2021automated, gillette2022personalized}
	% this part of the workflow was conceived 
	% to register anatomical models of torso and whole heart from a previously generated virtual cohort,
	% comprising 9 male and 5 female models of variable height, weight and body mass index
	% \cite{gillette2021automated, gillette2022personalized}.
	% A suitable torso matching the patient under consideration was selected as a template model, 
	More precisely, the atrial surfaces of the template were registered to the generated patient models by employing a global optimal 3D iterative closest point (GO-ICP) algorithm \citep{yang2015go,yang2013go}.
	The transformation was then stored and applied to the surface of the template torso
	along with the known positions of the electrodes \citep{gillette2021automated, gillette2022personalized} (see Figure \ref{Fig:torso_and_ica}(a)).
	
	\begin{figure}[!t]
		\centering
		\includegraphics[width=0.9\textwidth]{images/ica.png}
		\caption{(a) Combined conformal atria-torso mesh generated by registering a patient atrial model, 
			generated with our workflow, to a selected template torso model with positioned \gls{ecg} electrodes. The ventricles for the template heart are also represented. 
			(b) Representation of the electrical connections between the \gls{ra} and \gls{la}, including four \gls{ic} and the \gls{fo}.}
		\label{Fig:torso_and_ica}
	\end{figure}
	
	A volumetric mesh was subsequently generated between the surfaces of the torso and the atria, then merged with the volumetric mesh of the atria.
	% and the generated volumetric biatrial was then integrated with the torso mesh to obtain a conformal volumetric mesh 
	% of both atria embedded in a torso.
	For simplicity, other organs represented in the template model, such as the lungs, were excluded to avoid potential meshing issues caused by intersections with the atrial geometry.
	
	The obtained atria-torso model is suitable for the accurate simulation of \glspl{ecg} and \glspl{egm},
	either based on a bidomain model for simulating the entire extracellular potential field, $\phi_{\rm e}(\mathbf{x},t)$  \citep{bishop2011bidomain,neic17:_reaction_eikonal,ZAPPON2024112815}, 
	or on a lead field approach \citep{geselowitz1989theory,multerer2021uncertainty,potse2018scalable},
	yielding time traces of the extracellular potential sampled at discrete points $\phi_{\rm e,\mathbf{x}}(t)$. 
	
	% either based on \gls{rd} simulation of the entire extracellular potential field  torso \cite{neic17:_reaction_eikonal},
	% or based on lead field computations for predicting point potentials \cite{potse2018scalable} (see Fig.\ref{Fig:torso_and_ica}a).
	
	\subsection{Functional modeling stage}
	\label{Subsection:functional_modeling_stage}
	\subsubsection{Parametric modeling of inter-atrial connections}
	\label{Subsubsection:cables}
	Action potential propagation between \gls{ra} and \gls{la} is limited to 
	a few electrically excitable muscular strands of tissue called \glspl{ic}, 
	that serve as conduction pathways between the otherwise electrically insulated atria.
	% These \glspl{ic} are electrically excitable muscular strands of tissue.
	Known pathways include \gls{bb}, 
	the sheath of the \gls{cs}, the muscular rim of the \gls{fo} 
	and, posteriorly, a superior-posterior and a middle-posterior bridge.
	Histologically, \glspl{ic} arise sub-epicardially and branch outside of the atrial walls \citep{Margo2013,platonov2002morphology}. 
	While the \gls{bb} \citep{Bachmann1916,Margo2013,lemery2003anatomic} and the \gls{cs} musculature \citep{chauvin2000anatomic,antz1998electrical} have been shown 
	to be the most relevant electrical \gls{ic} \citep{sakamoto2005interatrial,knol2019bachmann}, 
	the specific number and location of the complete set of \gls{ic} remains unclear \citep{platonov2002morphology,platonov2007interatrial,sakamoto2005interatrial}.    
	
	In this study, we considered all five aforementioned \glspl{ic} as the only electrical connections between the \gls{ra} and \gls{la}.
	Except for the rim of the \gls{fo}, which forms an electrically conductive bridge within the volumetric biatrial mesh,
	\gls{ra} and \gls{la} are electrically fully decoupled using a nodal splitting approach \citep{costa2014:_fibrotic}. 
	All other \glspl{ic} traversing the space outside the atrial walls 
	are modeled as discrete electrically conducting strands, 
	that are anchored within the \gls{ra} and the \gls{la} at given sites defined through \glspl{uac} 
	(see Figure \ref{Fig:torso_and_ica}(b)), and constrained to the sub-epicardium.
	% The \gls{bb}, which presents both intra-atrial and inter-atrial components, is partially modeled through volumetric landmarks on the atrial body, as described in Section \ref{Subsubsec:anatomical_structure}, to represent the intra-atrial component, and partially with a cable, to describe the inter-atrial component. 
	% For structures such as the \gls{bb} that have both intra-atrial as well as inter-atrial components 
	% only the inter-atrial component was modelled as cable 
	% whereas the intra-atrial components were represented by labelling domains within the atrial walls.
	These strands are defined as auto-generated cables as previously utilized in \cite{gillette2022personalized}, and their electrical behavior is modeled with the approach used to define the \gls{hps} in the ventricles in \cite{neic17:_reaction_eikonal,vigmond2006:_sawtooth}. The cables used to define the \glspl{ic} were generated along the shortest path %within the atrial wall,
	connecting the anchoring location in \gls{ra} and \gls{la}, respectively.
	% using the same \gls{ep} modeling approach 
	% as for the representation of electrical conduction in the \gls{hps} \cite{clements, neic17:_reaction_eikonal}.
	% Anchoring the termini of the \glspl{ic} is constrained to the sub-epicardium, 
	%and allowing their electrical coupling to the atrial wall only on such entry points while keeping them fully isolated elsewhere.
	% All entry points of conducting strands of the \glspl{ic} were constrained to the sub-epicardium
	% and were electrically coupled to the atrial walls to conduct only at these connection points, 
	% while being electrically fully isolated elsewhere.
	
	% Specifically, we incorporated slow conduction properties 
	% corresponding to the landmark of the ring of \gls{fo} as defined previously. 
	
	%A superior-posterior and a middle-posterior bridges, and the \gls{cs} muscular band were represented by autogenerated cables, 
	%while the \gls{bb} was partly defined through tissue labeling and partly through the auto-generated cables (see Fig.\ref{Fig:torso_and_ica}b). 
	%All cable fascicles of the \gls{ic}s were confined sub-epicardially at the entry and exit points and were constrained to conduct only to the atrial wall tissue at these connection points, ensuring the physiological electrical isolation of the bundles. Entry and exit points of the cables were represented through \gls{uac} to ensure reproducibility on all atrial model. Tissue and cable tuning were performed to achieve physiological activation, as described subsequently.
	Of the five represented \glspl{ic}, the \gls{bb} is composed of both an intra- and an inter-atrial component, 
	where the intra-atrial component is modeled as tissue embedded within the atrial epicardial wall,
	as described in Section \ref{Subsubsec:anatomical_structure}, 
	and only the inter-atrial central component is modeled as a cable. 
	The \gls{bb} emerges at the junction of the \gls{ra} body and the \gls{svc}, 
	in close proximity to the sino-atrial node (\gls{san}), 
	and runs in an anterior direction around the \gls{raa}, where it branches into two bands, 
	one towards the \gls{ra} vestibule, 
	and the inter-atrial central \gls{bb} running toward the anterior \gls{la} wall, anchoring in the \gls{la} epicardium.
	From there another intra-atrial band runs epicardially towards the \gls{laa}, encircles the \gls{laa}, 
	and terminates superiorly between the \gls{laa} and the \gls{lspv}, 
	and inferiorly, at the \gls{mv} annulus \citep{Margo2013}. 
	
	The two posterior \gls{ic}s were defined as in \cite{loewe2016influence,wachter2015mesh}. 
	The superior-posterior bridge emerges in proximity of the \gls{ivc} at the 90\% of the \gls{ct}, 
	and connects the \gls{ra} with the \gls{la} wall near the \gls{ripv}, on the septal side. 
	The middle-posterior \gls{ic} starts at the junction of the \gls{ivc} with the atrial body, 
	on the \gls{ct}, and ends at the junction of the \gls{ripv}, on the inferior side of the \gls{ripv}.
	
	The musculature of the \gls{cs} was represented by a single cable providing a pathway 
	from the \gls{ra} wall, near the \gls{cs} ostium, to the posterior \gls{la} at the annulus of the \gls{mv}. 
	The origin of the cable in the \gls{ra} was chosen superior to the \gls{cs}, 
	at a distance between 3 and \SI{8}{\milli\meter} \citep{chauvin2000anatomic} from the \gls{cs} ostium, 
	while the \gls{la} anchoring site was selected according to \cite{wachter2015mesh}.
	
	The muscular rim of the \gls{fo} was modeled as tissue using previously defined labels, 
	endowed with slow conduction properties \citep{kharbanda2019current,harrild2000computer,padala2021anatomy}. 
	
	%Finally, the electrical isolation between the epicardium of the left and right atria was ensure by means of the splitting algorithm implemented in the EP simulator and previously used in \cite{gillette2022personalized}.  
	
	\subsubsection{Baseline calibration of the atrial activation sequence and P-wave simulation}
	\label{Subsubsec:baseline_sim}
	The genesis of the atrial P-wave is governed by the spatio-temporal evolution 
	of electrical depolarization wavefronts traversing the atria.
	%This process is governed by a high-dimensional parameter vector. 
	At each point in the atria, the propagation of the wavefronts 
	is governed by (i) a conduction velocity tensor 
	that is determined by intra- and extracellular conductivities along the eigenaxes, (ii)
	the tissue composition condensed into a scalar bidomain surface-to-volume ratio, 
	and (iii) the tissue's \gls{ep} properties related to cellular dynamics.
	Beyond these intrinsic conduction properties, the site of initiation of a depolarization wavefront, 
	typically at an exit site of the \gls{san} or at a pacing site,
	and the location of anchoring sites of inter-atrial conduction pathways 
	and their associated conduction velocities influence the distribution of electrical sources during depolarization
	and, thus, the morphology of the P-wave.
	As such, the parameter vector controlling atrial activation is high dimensional.
	While all these parameters are exposed for unattended manipulation in parameter sweeps, their calibration as a whole is computationally not feasible with current technology,
	and, further, observable data are very sparse and insufficient to constrain the calibration procedure.
	As such, we only demonstrate the ability of our fully parametric 3D biatrial model 
	to sweep over selected key parameters to generate physiologically meaningful sets of atrial P-waves.
	% and \glspl{egm}.
	% For all 50 models generated, extensive \gls{eam} datasets 
	% acquired with Carto XP (Biosense Webster, Israel) were available 
	% and could be used for the evaluation of simulated P-waves.
	
	Our workflow universally supports all common \gls{ep} modeling approaches 
	that build on the volumetric representation of atrial electrical sources, 
	including all \gls{rd} mono- and bidomain models \citep{bishop2011bidomain,potse2006comparison}, 
	as well as variants of \gls{re} models \cite{neic17:_reaction_eikonal}, 
	including plain Eikonal models with $V_{\rm m}$-recovery \citep{pezzuto2017:_ecg}.
	To demonstrate the feasibility of a sampling-based model calibration using the P-wave,
	we chose two popular approaches, 
	the computationally expensive \gls{rd} monodomain model, 
	requiring a high average spatial resolution of $\Delta x\approx$ \SI{0.25}{\milli\meter},  
	and a lightweight \gls{re} model at $\Delta x\approx$ \SI{0.90}{\milli \meter}. Both models are combined with the lead field approach for the P-wave computation. 
	We refrain here from providing a mathematical model formulation and refer to previous studies where all methodological underpinnings of the forward \gls{ep} modeling have been described in detail \citep{bishop2011bidomain,gillette2021:_framework,neic17:_reaction_eikonal}. 
	All simulations are executed using CARPentry \citep{neic17:_reaction_eikonal} for \gls{rd} and \gls{re} simulation runs, respectively, as well as openCARP \citep{plank2021opencarp} only suitable for \gls{rd} simulations.
	
	% All electrophysiology simulations in our study are executed using openCARP \cite{plank2021opencarp} and CARPentry \cite{neic17:_reaction_eikonal} for \gls{rd} and \gls{re} simulation runs, respectively.
	% To demonstrate suitability for both \gls{rd} and \gls{re} modeling 
	% we benchmark the performance of our workflow and evaluate its robustness
	% by testing two target resolutions typically used in \gls{rd} and \gls{re} modeling studies,
	% which were chosen as \SI{250}{\micro \meter} and, due to the thin-walled nature of the atria, \SI{900}{\micro \meter}, respectively.
	
	\paragraph{Baseline parameter setting for electrophysiological simulations}
	The dimensionality of the parameter space governing the spatial variation in conduction velocity 
	is reduced as follows. 
	First, the overall fiber arrangement is kept constant,
	and target conduction velocities are prescribed in the \gls{ct}, \gls{pm}s, \gls{ic}, \gls{fo}, 
	and the wider \gls{la} and \gls{ra} walls,  
	on a per-region basis 
	using reported ranges \citep{deng2012simulation}, eventually scaled to ascertain that the total atrial activation times computed with the \gls{re} model
	fall within the physiological ranges \citep{lemery2007normal}.
	
	Anisotropy in conduction velocity between longitudinal and transverse to the fiber directions was set to a value of 1.3 within the reported range of 1.0 to 1.6 \citep{gray1996incomplete,hansson1998right}.
	Atrial cellular dynamics were represented by the Courtemanche model \citep{courtemanche1998ionic},
	parameterized for various \gls{ep} regions following \cite{azzolin2023:_augmenta}
	with increased sodium channel conductivity to obtain the prescribed conduction velocities.
	% removed this, this does not matter for the P-wave
	% The maximum conductivity $g_{Ca,L}$ was moreover adjusted by a factor of 1.67 in the \gls{ct} 
	% to account for regional heterogeneities as described in \cite{krueger2013silico,loewe2015left}.
	To identify the electrical conductivities for the \gls{rd} model at the targeted mesh resolution that match the \gls{re} conduction velocities, the ForCEPSS framework \citep{gsell2024:_forcepss} was employed, using reported values \citep{roberts1979influence} for initialization. 
	The chosen parameter settings are summarized in Table \ref{tab:calibration}.
	% The baseline \gls{ep} parameter settings were used on various atrial models to simulate sinus rhythm. 
	% Atrial activation was initiated from the \gls{san} node, 
	% as identified by the workflow, situated on the posterior-superior side of the \gls{ct}. 
	% The cables used to define the \glspl{ic} were generated along the shortest path
	%within the atrial wall,
	% connecting the anchoring location in \gls{ra} and \gls{la}, respectively, 
	% except for \gls{bb} which was represented by a straight line connecting the sites
	% $\mathbf{x}_{BB_{RA}}$ and $\mathbf{x}_{BB_{LA}}$ (see \ref{Fig:torso_and_ica}). 
	% in the atria between the entry and exit point,
	%As such, these bundles do not necessarily represent the actual anatomical connection, 
	%but serve to mechanistically represent inter-atrial conduction. 
	Conduction velocities in the \gls{ic} cables were considered parameters, 
	tuned to obtain physiological inter-atrial activation delays and, thus,
	the separation of the contributions of \gls{ra} and \gls{la} to the P-wave.
	%Due to this potential lack for anatomical correspondence 
	%the calibrated velocities within these cables may not reflect the true 
	%inter-atrial conduction velocity in theses tissues, 
	%and may even fall outside the physiological range.
	%Moreover, each band of muscular tissue is surrogated by a single cable,
	% one cable was employed to represent a single band of muscular tissue, 
	%endowed with isotropic conduction properties. 
	
	All simulations were conducted for a duration of \SI[parse-numbers = false]{150}{\milli\second} 
	to cover the total atrial activation time and, thus, the entire duration of the P-wave. 
	All \glspl{ecg} were filtered with a \SI{150}{Hz} low-pass and a \SI{0.5}{Hz} high-pass, 
	and scaled by a factor of 0.2 to obtain \gls{ecg} signals with the observed magnitudes \citep{gillette2022personalized}.
	
	% Not sure this fits here, we have subsections of all of this in the following
	% -------------------------
	% ** see test R-D vs R-E
	%We utilized the described parameter settings to evaluate the assumptions and simplifications underlying the \gls{re} model by comparing computed \gls{ecg} results with those from the gold-standard monodomain model.
	% **test impact of \gls{ic} cables
	%Additionally, we tested our workflow ability to automatically and efficiently vary geometry-related parameters, effectively investigating their impact on \gls{ecg} morphology. To this end, we ran several simulations on various biatrial geometries to assess the impact of the anterior branch location of the \gls{bb} on the P-wave, 
	% **check volumetric effect of PMs
	% explore the function of the \gls{ra} endocardial layer formed between the \glspl{pm}, and 
	% **effect of SAN and RA CV choices
	%investigate the effects of different modeling assumptions on the \gls{san} and \gls{ra} conduction velocities on the P-wave.
	
	% tuned RD values GP, see pars.RD.R79tuned.par
	\begin{table}[!t]
		\centering
		\begin{tabular}{lllllll}
			\toprule
			& $v_{\rm l}$ & $v_{\rm t}$ &$g_{\rm{i,l}}$ & $g_{\rm {e,l}}$ &$g_{\rm{i,t}}$ & $g_{\rm {e,t}}$\\
			&m/s &m/s &S/m &S/m &S/m &S/m \\ 
			\hline \\[-2ex]
			\gls{ra}           & 0.97 & 0.74 & 0.583  & 0.742  & 0.232  & 1.162 \\
			\gls{la}           & 0.98 & 0.76 & 0.570  & 0.726  & 0.220  & 1.102 \\
			\gls{ct}           & 1.21 & 0.92 & 0.904  & 1.112  & 0.416  & 0.899 \\
			\glspl{pm}         & 1.30 & 0.99 & 1.053  & 1.339  & 0.408  & 2.042 \\
			\gls{bb} (atrial)  & 1.40 & 1.08 & 1.229  & 1.564  & 0.479  & 2.399 \\
			Rim \gls{fo}       & 0.33 & 0.24 & 0.080  & 0.102  & 0.034  & 0.168 \\
			\bottomrule
		\end{tabular}
		\caption{Conduction velocities in longitudinal and transverse directions, $v_l$ and $v_t$, 
			prescribed in the \gls{re} model along with calibrated conductivities 
			yielding matching velocities in the \gls{rd} models at the target average mesh resolution of 
			$\Delta x\approx$ \SI{0.25}{\milli \meter}. 
			The bidomain surface-to-volume ration was chosen as $\beta=$\SI{1400}{\centi \meter}$^{-1}$. }
		\label{tab:calibration}
	\end{table}
	
	% Elena, old
	% \begin{table}[!t]
	%     \centering
	%     \begin{tabular}{lllllll}
	%         \toprule
	%         & $v_{\rm l}$ & $v_{\rm t}$ &$g_{\rm{i,l}}$ & $g_{\rm {e,l}}$ &$g_{\rm{i,t}}$ & $g_{\rm {e,t}}$\\
	%         &\SI{}{\meter / \second} & \SI{}{\meter / \second} &\SI{}{\meter / \second} & \SI{}{\meter / \second} & \SI{}{\meter / \second} & \SI{}{\meter / \second} \\ 
	%         \hline \\[-2ex]
	%         \gls{la}           &0.97 &0.74 & 0.6262 & 0.6969 & 0.3274 & 0.4171 \\
	%         \gls{ra}           &0.98 &0.76 & 0.8338 & 0.9324 & 0.3117 & 1.1113 \\
	%         \gls{ct}           &1.21 &0.92 & 0.9481 & 1.1667 & 0.4225 & 0.9130 \\
	%         \glspl{pm}         &1.30 &0.99 & 1.5196 & 2.1912 & 0.6987 & 1.7135 \\
	%         \gls{bb} (atrial)  &1.40 &1.08 &  &  &   &  \\
	%         Rim \gls{fo}       &0.33 &0.24 & 1.4448 & 0.2482 &0.0276  & 1.0481 \\
	%         \bottomrule
	%     \end{tabular}
	%     \caption{Conduction velocities in longitudinal and transverse directions,  $v_l$ and $v_t$, 
	%     prescribed in the \gls{re} model along with calibrated conductivities 
	%     yielding matching velocities in the \gls{rd} models at the target average mesh resolution of 
	%     $\Delta x\approx$\SI{250}{\micro \meter}. 
	%     The bidomain surface-to-volume ration was chosen as $\beta=$\SI{1400}{\centi \meter}$^{-1}$. }
	%     \label{tab:calibration}
	% \end{table}
	
	We assessed the accuracy of the \gls{re} model relative to a gold standard \gls{rd} monodomain model for simulating atrial activation and associated P-waves.
	Using the baseline setup (see Table \ref{tab:calibration}), 
	a normal sinus activation with associated P-wave was simulated on two models of the same biatrial geometry
	discretized at different spatial resolutions of 
	\SI{0.9}{\milli\meter} and \SI{0.25}{\milli\meter} for \gls{re} and \gls{rd}, respectively,
	according to their specific numerical requirements \citep{neic2017eikonal}.
	%Both \gls{re} and monodomain simulations 
	% Both types of model the coarser \gls{re} as well as the higher resolution \gls{rd} variant,
	% discretized in space at average resolutions of \SI[parse-numbers = false]{0.9}{\milli \meter} and \SI[parse-numbers = false]{0.25}{\milli \meter}, respectively, 
	%were configured according to the baseline setup to prescribe conduction velocity, anisotropy, and \gls{ep} properties in all labeled atrial regions.
	%To ensure simulation compatibility, the \gls{re} model was first initialized by directly assigning conduction velocity tensors to each anatomical domain individually. Subsequently, using a standardized workflow \cite{gsell2024:_forcepss}, the \gls{rd} model parameters were calibrated to identify conductivity tensors, ensuring that the resulting conduction velocities on the high-resolution grid matched those prescribed for the \gls{re} model.
	%A comprehensive description of the \gls{ep} setup is compactly encoded in a \texttt{.json} dictionary,
	%that is made accessible through the supplementary material
	%to facilitate the seamless reproduction of \gls{rd} simulations with openCARP \citep{plank2021opencarp}, 
	%All most important parameters are conduction velocity related and are summarized in Table~\ref{tab:calibration}.
	
	%In the following, P-wave generation and \gls{ep} source simulation will be computed by employing the efficient \gls{re} model.
	% Sinus rhythm was simulated on various atrial models using the baseline
	% % a uniform and common 
	% electrophysiological parameter setting. 
	
	\paragraph{Effect of the \gls{bb} insertion in the \gls{la} on the P-wave}
	
	Activation of the \gls{la} is mediated through discrete \glspl{ic}
	with \gls{bb} being a primary pathway \citep{knol2019bachmann,sakamoto2005interatrial}.
	A site of insertion, $\mathbf{x}$, of an \gls{ic} defines an earliest activation site, $\varepsilon(\mathbf{x},t)$, of the \gls{la}
	and, thus, constitutes a major factor in shaping the \gls{la} contribution to the P-wave morphology.
	With traditional modeling approaches, relying on discrete meshing of the \glspl{ic}, elucidating such effects is challenging, 
	as any change in the insertion site requires major remeshing of the model.
	%The \gls{bb} is a group of parallel muscle strands serving as the primary unidirectional electrical connection between the \gls{ra} and \gls{la}. While connecting both chambers, in a healthy subject, the \gls{bb} is a passive receiver of the electrical signal coming from the \gls{ra}, whereas it plays a fundamental role in the activation of the \gls{la}. 
	In this work, we aim at investigate the impact of the location of the central \gls{bb} insertion on the \gls{la} activation and P-wave, by sweeping over the spatial parameter $\mathbf{x}_{BB_{LA}}$ representing the \gls{bb} insertion point on the \gls{la} wall, that is $\varepsilon(\mathbf{x}_{BB_{LA}},t_{la})$. The temporal parameter $t_{la}=l_{bb}/v_{bb}$ represents the activation time of $\varepsilon$, and depends on the length of the cable $l_{bb}$, and the designated conduction velocity $v_{bb}$. Here, we chose to prescribed $v_{bb}$, while varying $l_{bb}$ dependently on the insertion point $\mathbf{x}_{BB_{LA}}$. Therefore, changes in $\mathbf{x}_{BB_{LA}}$ will impact both the \gls{la} activation pattern, that is, the P-wave shape, and the initial activation time of the \gls{la}, corresponding to a shift in the portion of the P-wave related to the \gls{la} activation.
	% The impact of the location of the central \gls{bb} insertion on the \gls{la} activation and the P-wave is elucidated 
	% by sweeping over the parameters governing earliest \gls{la} activation, $\varepsilon(\mathbf{x}_{BB_{LA}},t_{la})$,
	% where $\mathbf{x}_{BB_{LA}}$ alters the \gls{la} activation pattern 
	% and $t_{la}=l_{bb}/v_{bb}$ shifts the portion of the P-wave related to \gls{la} activation
	% where $l_{bb}$ and $v_{bb}$ are the physical length of the \gls{bb} cable 
	% and $v_{bb}$ is the designated conduction velocity.
	The importance of these parameters is investigated concerning the sensitivity of the P-wave shape, and their ability to yield a large enough physiological envelope of P-waves to cover the true measured P-wave.
	% In the former case, the \gls{bb} affect the activation pattern of the \gls{la} and thus the total P-wave shape, whereas in the latter case, given the same activation site, the activation patterned is unvaried but shifts in time, thus prolonging the P-wave duration. 
	%Here, we aim to investigate the potential of our workflow and modeling approach in determining different shapes of the P-wave and, at the same time, explore the sensitivity of the P-wave to the position of the early activation sites on the \gls{la}.
	
	% we may consider ommitting this as the effect does not appear to be dramatic
	% we may also omit the argument regarding field strength as we are not evaluating this
	However, using a single \gls{ic} cable to mediate \gls{bb} conduction limits the model ability to produce single point-like \gls{la} activation sites, and potentially the \gls{ecg} shape generation.    
	As the number of inter-atrial cables is not limited in our modeling approach, we investigate a more realistic volumetric-like approximation of the \gls{bb} band,
	by using multiple cables.
	% , thus increasing the contribution to the extracellular potential field 
	% and to allow for a more gradual fan-shaped insertion on the \gls{la}.    
	Specifically, we select three points, $\mathbf{x}{BB_{LA}}^i$ ($i=1,2,3$), on the anterior wall of the \gls{la} as locations for anchoring three reference cables representing strands of the \gls{bb}, all of which originate from the midpoint of the \gls{ra} lateral band of the \gls{bb}. 
	As per the single-cable case, each location $\mathbf{x}{BB_{LA}}^i$ was then varied within a circular area of $\approx$ \SI{5}{\milli\meter}, defining a set of alternative entry points ${ \mathbf{x}{BB_{LA_{j}}}^i}{j=1}^{N{i}}$, and corresponding automatically generated cables.
	% For each $\mathbf{x}{BB{LA}}^i$, a corresponding cable was automatically generated. 
	By running a \gls{re} simulation for each tuple of cables $(\mathbf{x}{BB_{LA_{j}}}^1,\mathbf{x}{BB_{LA_{k}}}^2,\mathbf{x}{BB_{LA_{\ell}}}^3)_{j=1,\dots,N_1,k=1,\dots,N_2,\ell=1,\dots,N_3}$, 
	the set of P-waves was finally computed.
	
	% We may omit this experiment, I haven't found any studies modeling the 
	% inter-pectinate muscle space as tissue.
	\paragraph{Effect of the discrete RA endocardium on the P-wave}
	% \todo[inline]{I'd suggest to omit this section, I have not found any studies 
	% where the inter-pectinate muscle space was modeled as tissue.}
	The endocardium of the \gls{ra} consists of discrete structures 
	in the form of muscular strands comprising \gls{ct}, \gls{bb} and \glspl{pm},
	that are partially attached to a thin and smooth epicardial layer \citep{corradi2011atria,ho2009importance}. 
	
	The effect of the added endocardial tissue in atrial \gls{ep} simulation, 
	especially the inter-pectinate tissue, on the P-wave has never been study before.  
	To investigate the role of the \gls{ra} endocardium on the P-wave, 
	\gls{re} model simulations were run using two different parameter settings for each biatrial model. 
	In one setting, the \gls{ra} endocardium was considered cardiac conductive tissue, 
	while in the other, the tissue labeled as \gls{ra} endocardium was treated as part of the bath surrounding the heart. In the former case, conduction velocity tensors, as presented in Table \ref{Tab.RE_vs_RD}, were assigned to the \gls{ra} endocardium tissue. In the latter case, this tissue was marked as torso tissue.
	
	% This title does not reflect what you were doing here.
	%\paragraph{Effect of the \gls{san} modeling approach and of the RA conduction velocity on the P-wave}
	
	\paragraph{Calibration of the atrial activation sequence based on the P-wave of the 12 lead \gls{ecg}}
	We investigate the feasibility of calibrating the atrial activation sequence to the corresponding patient P-wave by sampling key parameters to generate a P-wave envelope covering the observed P-wave data.
	% The feasibility of calibrating the activation sequence in our fully parametrized atrial model
	% to the P-wave is evaluated by sampling key parameters to produce a large family of P-waves
	% where the envelope of signals covers the observed P-wave.  
	Due to the high dimensionality of the parameter space, we restrict sampling to three key parameters that are expected to be the most influential in altering the different phases of the P-wave morphology \citep{loewe2015left}. These are:
	\begin{itemize}
		\item the location of the leading pacemaker site within the \gls{san}, responsible for the initial \gls{ra} activation. 
		%specifically the discrete \gls{SAN} exit site that initiates \gls{RA} activation. 
		This site is known to be variable, allowing modulation of the onset phase of the P-wave.
		The \gls{san} is an elongated and highly heterogeneous \gls{ra} region
		with action potential properties, cell size, and capacity as well as conductance 
		changing from the center to periphery \citep{honjo_correlation_1996}, 
		and along inferior-superior and caudal-cranial gradients \citep{munoz_onset_2011}. 
		% Heterogeneity associated differential sensitivity to neurotransmitters 
		% as well as uneven cholinergic and adrenergic innervation 
		% are assumed to shift leading pacemaker site within the \gls{san}~\cite{munoz_onset_2011}.
		For calibration purposes, we avoided modeling the physiological function of the \gls{san} with biophysical detail and instead used only the spatial location of the exit site $\mathbf{x}_{\text{SAN}}$ as a parameter.
		Specifically, the \gls{san} was modeled as a focal activation site, with a short, cigarette-shaped band of \SI{2.2}{\centi\meter} in length \citep{boyett2000sinoatrial,monfredi2010anatomy}, located in the inferior-central part of the \gls{ct}. 
		Moreover, the \gls{san} was electrically isolated by a non-conductive region along the boundary of the \gls{ct}, replicating the effective block observed in physiological studies \citep{dobrzynski2005computer}. 
		% \Gls{san} exit sites, $\mathbf{x}_{\text{SAN}}$, were modeled as focal activation sites.
		% \todo[inline]{the following details are redundant in my view.}
		% representing the \gls{san} as either a sphere with a \SI{2}{\milli\meter} radius or a cylinder with its principal axis aligned along the \gls{ct}, with a radius of \SI{1.5}{\milli\meter} and varying lengths between \SI{3}{\milli\meter} and \SI{2}{\centi\meter}. 
		
		\item Calibration to the overall duration of earlier phase of the P-wave 
		was based on varying conduction velocities globally within the \gls{ra}.
		Velocity in the longitudinal direction $v_l$ was uniformly sampled in steps of \SI{0.5}{\meter/\second} 
		between \SI{0.6}{} and \SI[parse-numbers = false]{1.1}{\meter/\second}.
		% Assuming rotational isotropy, velocity in the transverse direction $v_t$ was determined from the fixed ratio  of $v_l/v_t$ at 0.75 between the minimum and maximum values, 
		% in accordance with \cite{deng2012simulation}.
		% \todo[inline]{the next is redundant, if you keep the ratio constant you don't need a step for sampling.
		% Suggest to remove and simply say the transverse velocity was determined from the fixed ratio.}
		Assuming rotational isotropy, velocity in the transverse direction $v_t$
		was sampled in steps of \SI{0.0375}{\meter/\second} 
		between \SI[parse-numbers = false]{0.45}{} and \SI[parse-numbers = false]{0.7875}{\meter/\second}, but kept smaller to the longitudinal velocity $v_l$, in accordance with \cite{deng2012simulation}.
		
		\item Calibration of the shape and duration of the late phase of the P-wave, was based on manual variation of the anchoring sites of the \glspl{ic} and of the global conduction velocities in the \gls{la}. Specifically, the \gls{bb} was modeled as a three-cable band initiating at the same origin in the \gls{ra}, and anchoring in the \gls{la} at three independent insertion sites. The other three \glspl{ic} were also varied in the anchoring position on the \gls{la}, and conduction velocities in the cables were manually calibrated to control the timing of the initial \gls{la} activation. 
	\end{itemize} 
	
	% The location of the anchoring sites of \glspl{ic} shape the depolarization wavefronts 
	% in the \gls{la} and thus the mid to terminal phase of the P-wave.
	% I don't understand this, the RA site does not matter
	%both \gls{ra} and 
	% For calibration two \glspl{ic} were considered,
	% the \gls{bb}, modeled as three cables of the same origin in the \gls{ra} 
	% and a fan-like anchoring in the \gls{la} at three independent insertion sites,
	% and the \gls{cs}, where conduction velocities in the cables were adjusted
	% to control the timing of initial \gls{la} activation.
	
	
	
	%To investigate the possibility of capturing real data \gls{ecg} features with our workflow, we manually calibrated the \gls{ep} of one generated biatrial model against the corresponding \gls{ecg} sinus rhythm data.
	
	
	% This should be default
	%We obtained a good match of the overall traces of most \gls{ecg} leads 
	%by assuming the \gls{ra} endocardium to be represented only by the \glspl{pm} and the \gls{ct}. 
	% Moreover, to gain a good alignment of the first part of the P-wave \cite{loewe2015left}, 
	% we defined the \gls{san} as a short, cigarette-shaped band of tissue \cite{monfredi2010anatomy,boyett2000sinoatrial} of \SI{2.2}{\centi\meter} in length, 
	% located on the inferior-central part of the \gls{ct}, 
	% and a non-conductive region of tissue along the septal boundary of the \gls{ct}, 
	% simulating the effective block observed in physiological studies \cite{dobrzynski2005computer}.
	
	% The second part of the P-wave \cite{loewe2015left} was calibrated 
	% by selecting the entry points on both \gls{ra} and \gls{la} of three \gls{bb} 
	% and one \gls{cs} conductive cables, and adjusting their corresponding conduction velocities.    
	
	% all of the following is a discussion
	% While this initial calibration achieves a good match for most \gls{ecg} leads
	% \begin{itemize}
	% \item variations in shape and orientation of the torso surrounding the heart,
	% \item potential differences in electrode positions compared to those used for patient \gls{ecg} recordings, 
	% \item the absence of scar tissue in the generated biatrial geometry -- which is likely present in \gls{af} subjects    
	% \end{itemize}
	% may hinder the complete match of simulated and real \glspl{ecg}. 
	% For all the aforementioned motives, hereon we will address the process of matching real \gls{ecg} data as pseudo-calibration.
	
	% While our investigation so far has targeted changes in \gls{la} activation, 
	% and thus the second part of the P-wave, we have not yet examined the factor possibly affecting the first part of the P-wave the location and shape of the \gls{san}, and the conduction velocity of the \gls{ra} epicardium. In this final numerical experiment, we aim to study these spatial and physical variables, moreover showing our framework ability to explore high dimensional space of varying parameters.       
	% To address this issue, we investigate the potential to reproduce the original \gls{ecg} 
	% by varying \gls{ra} spatial and physical parameters, specifically targeting the location and shape of the \gls{san} $\mathbf{x}_{\text{SAN}}$, 
	% and the conduction velocity of the \gls{ra} epicardium.     
	% We modeled the focal activation site representing the \gls{san} as either a sphere with a \SI{2}{\milli\meter} radius or a cylinder with its principal axis aligned along the \gls{ct}, with a radius of \SI{1.5}{\milli\meter} and varying lengths between \SI{3}{\milli\meter} and \SI{2}{\centi\meter}. 
	
	% velocity
	% Changes in conduction velocity were accounted for by varying the velocity in the fiber direction $v_f$ between \SI{0.6}{} and \SI[parse-numbers = false]{1.1}{\meter/\second} with uniform sampling steps of 0.05, and the velocity in the sheet direction $v_s$ between \SI[parse-numbers = false]{0.45}{} and \SI[parse-numbers = false]{0.7875}{\meter/\second} with uniform sampling steps of 0.0375, while maintaining a consistent ratio of $v_s/v_f$ at 0.75 between the minimum and maximum values, in accordance with \cite{deng2012simulation}.
	
	\paragraph{Data analysis}
	% \todo[inline]{I am not sure whether we need this section.}
	We performed both qualitative and quantitative analyses of \gls{ep} and P-wave variations resulting from the previously described modeling assumptions. The overall tissue activation was qualitatively analyzed by recording the total activation time of both atria. Changes in P-waves were quantitatively investigated using three different metrics. 
	
	In the presence of a reference signal $\phi_r^j$, either simulated or given as clinical \gls{ecg} data, we analyzed amplitude variation with the \gls{rmse} in percentage, as described in \cite{keller2010ranking,ZAPPON2024112815}. For each lead $j$, the \gls{rmse} is expressed as:
	$$ \text{\gls{rmse}}(\varphi_v^j)[\%] = \sqrt{\frac{\sum_{n=1}^N(\phi_v^j(n)-\phi_r^j(n))^2}{\sum_{n=1}^N (\phi_r^j(n))^2}} \cdot 100.$$
	We computed the \gls{rmse} when comparing the \gls{re} and \gls{rd} models when exploring the effect of the variation of the \gls{san} and \gls{ra} conduction velocity on the P-wave.
	
	When analyzing the variations of the \gls{bb}, whether using one or three cables, none of the computed \gls{ecg} signals matched the real \gls{ecg} data and could not be used as a reference to calculate the \gls{rmse}. For these test cases, we first defined an average P-wave, $\bar{\varphi}^j$, for each lead $j$. We then evaluated the overall morphological and amplitude P-wave variation by computing the mean absolute distance (MAD) of each resulting P-wave $\varphi_v^j$ from the average one as:
	\begin{equation}
	\label{eq:MAD}
	\text{MAD}(\varphi_v^j) = \frac{\sum_{t=1}^N |\varphi_v^j(t) - \bar{\varphi}^j(t)|}{N}, 
	\end{equation}
	where $\varphi_v^j(t)$ and $\bar{\varphi}^j(t)$ are the varying computed signal and the average signal at time $t$,  and $N$ is the total number of \gls{ecg} time samples. The MAD is then averaged over all leads.
	
	Additionally, in all cases, we accounted for the variation of \gls{pwd}, obtained using a sloping approach \citep{tan2000detection}. The \gls{pwd} was measured for each lead. Moreover, the \gls{pwd} averaged over all leads was given.
	
	\section{Results}
	\label{Sec:results}
	\subsection{Model generation workflow performance evaluation}
	We evaluated our atrial model generation workflow with respect to the required processing time 
	and the achieved degree of automation.
	50 contrast \gls{ct} datasets of \gls{af} patients were processed to generate biatrial models of target resolution of $\approx$ \SI{0.25}{\milli \meter} and $\approx$ \SI{0.9}{\milli \meter}, and the execution times of individual stages of the automated workflow were measured.
	% Two target resolutions were chosen, \SI{0.25}{\milli \meter} and
	% %, due to the thin-walled nature of the atria, 
	% \SI{0.9}{\milli \meter}, to generate computational meshes 
	% suitable for fast \gls{re} simulations as typically used for model calibration, 
	% and for computationally more expensive high fidelity \gls{rd} simulations.
	% 50 contrast CT datasets of AF patients were processed 
	% and the execution times of individual stages were measured.
	After each stage, visual checks were performed to detect processing errors. 
	Errors that required manual correction were recorded for each stage and at both target resolutions, and are summarized in Table \tabref{Tab:timings}.
	Benchmark results are reported for the execution of the workflow on a compute workstation equipped with an AMD Ryzen Threadripper PRO5965wx processor,
	%and NVIDIA GPU GeForce RTX 3090, employing 
	using 16 CPU cores. 
	%Results for each phase of the workflow were visually analyzed for each patient. 
	% 5 patients were moreover selected to investigate the effect of the \gls{bb} and of the \gls{ra} endocardium on the P-wave.
	% Average performance data and the number of processing errors requiring manual resolution  were measured over all 50 models for both \gls{re} and \gls{rd} resolution 
	% and are summarized in \tabref{Tab:timings}.
	
	Overall, all 50 cases in both resolutions were processed automatically in the majority of cases (38), 
	with only minimal user intervention required in 22 cases.
	The workflow produced anatomically highly detailed computational meshes, 
	with fine-grained domain annotation, fiber arrangement, and an anatomical reference frame (referred to Figure \ref{fig:_processing_results}).
	For the lower resolution \gls{re} model, the overall processing per model lasted, on average, less than ten minutes. 
	Higher resolution \gls{rd} compatible meshes were more costly to generate, specifically, the \gls{uac} generation stage where more than $\approx$84\% of the costs incurred.
	Importantly, a fully automatic meshing processes was achieved in all cases, yielding meshes free of topological errors and of, overall, excellent mesh quality. Worst element quality was always below 0.99, according to the quality metric \citep{karabelas2018:_towards},  
	which is considered a critical threshold in simulations using \gls{rd} solvers such as openCARP. Lower quality elements close to $0.99$ clustered mostly in high curvature regions, around the orifices of the thin-walled atria (referred to Figure \ref{fig:_processing_results}). 
	
	% \todo[inline]{To be moved to the Discussion}
	\subsubsection{Automation failure}  
	A manual intervention was needed to separate the \gls{lspv} from the \gls{laa} in 19 cases, and the \gls{cs} from the \gls{la} in 2 cases. The incorrect segmentation was however not due to the \gls{scn}, but to the low contrast in the acquired images. 
	At the second stage, during the labeling of the blood pools, manual intervention was required to switch the \gls{cs} with the \gls{ivc} landmarks in 21 cases.
	The final marking of the veins and the corresponding tissue to discard from the blood pool, including the \gls{cs}, \gls{laa}, and \gls{raa}, was automatically done by the workflow in 28 cases. For the remaining 22 cases, manual intervention was only needed to define or adjust the tissue to discard from the \gls{ivc} to ensure a better opening.
	
	% \subsection{Segmenting CT data}
	% The \gls{scn} segmented and labeled the major cardiac anatomical structures from the \gls{ct} scans of all 50 patients. 
	% %{\color{red} Figure\ref{}} shows both the \gls{ct} image and the corresponding segmentation for three patients. 
	% A manual intervention was needed to separate the \gls{lspv} from the \gls{laa} in 19 cases, and the \gls{cs} from the \gls{la} in 2 cases. The incorrect segmentation was however not due to the \gls{scn}, but to the low contrast in the acquired images. 
	% The computational cost of segmentation on average amounted to \SI[parse-numbers = false]{15}{\second}.
	
	% \subsection{Labeling the blood pools and building the volumetric mesh of atrial walls}
	% The anatomical structures on the surface of the blood pool were automatically annotated in all 50 patients. In 21 cases, manual intervention was required to switch the \gls{cs} with the \gls{ivc} landmarks. The extraction of the volumetric mesh for the blood pool and the computation of the Laplace-Dirichlet solutions, given the previously defined labels, were automatically obtained for all 50 patients. The final marking of the veins and the corresponding tissue to be removed from the blood pool, including the \gls{cs}, \gls{laa}, and \gls{raa}, was automatically done by the workflow in 22 cases. For the remaining cases, manual intervention was only needed to define the tissue to be removed from the \gls{ivc} to ensure its opening. The interpolation of the obtained landmarks to the image stack was automatically performed for all 50 patients. The labeling process of the blood pools in its entirety required $5 \pm 3$ minutes (refer to Table \ref{Tab.Times} for a summary of the computational expenses of the workflow phases).  
	
	
	% \begin{figure}[!t]
	%     \centering        \includegraphics[width=0.7\textwidth]{images/Times.png}
	%     \caption{Computational times required for generating volumetric bilayer atrial anatomical models with landmarks -- including orifice, appendage, PM, \gls{ct}, SA, FO, BB-- fibers, and UAC identification, using mesh sizes of \SI{1.00}{\milli \meter} and \SI{0.25}{\milli\meter}. Times are grouped by workflow phases (top row) and mesh size (bottom row). The graph on the bottom row excludes the computational time for UAC computation to emphasize variations in other generation phases."}
	%     \label{Fig.times}
	% \end{figure}
	
	\begin{figure}[!t]
		\centering        \includegraphics[width=0.9\textwidth]{images/Fig_mesh_variability_quality.png}
		\caption{(a) Representation of 35 out of 50 generated biatrial models. Only a subset of the generated geometries is represented to allow for a better visualization. (b) Examples of computed meshes of resolution of $\approx$ \SI{0.90}{\milli\meter} and $\approx$ \SI{0.25}{\milli\meter} and corresponding obtained mesh quality.}
		\label{fig:_processing_results}
	\end{figure}
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lccc}
			\toprule
			& \multicolumn{2}{c}{\textbf{Mesh resolution/Model}} & \textbf{Manual Correction} \\
			&\textbf{\SI{0.9}{\milli\meter}}/\gls{re} & \textbf{\SI{0.25}{\milli\meter}}/\gls{rd} & \# Cases \\ 
			\hline \\[-2ex]
			Automatic multilabel segmentation &$\approx$ \SI[parse-numbers = false]{15}{\second} & $\approx$ \SI[parse-numbers = false]{15}{\second} & 19 \\
			Label augmentation on the blood pools  &\SI[parse-numbers = false]{297}{} $\pm$ \SI[parse-numbers = false]{180}{\second} &\SI[parse-numbers = false]{297}{} $\pm$ \SI[parse-numbers = false]{180}{\second} & 22 \\
			Walls extrusion and meshing  &\SI[parse-numbers = false]{187}{} $\pm$ \SI[parse-numbers = false]{172}{\second} &\SI[parse-numbers = false]{628}{} $\pm$ \SI[parse-numbers = false]{557}{\second} & 0\\
			Selection of the atrial orifices  &\SI[parse-numbers = false]{4}{} $\pm$ \SI[parse-numbers = false]{6}{\second} &\SI[parse-numbers = false]{39}{} $\pm$ \SI[parse-numbers = false]{36}{\second} & 0\\
			Anatomical structures and fibers  &\SI[parse-numbers = false]{30}{} $\pm$ \SI[parse-numbers = false]{35}{\second} &\SI[parse-numbers = false]{1084}{} $\pm$ \SI[parse-numbers = false]{628}{\second} & 0\\
			Universal atrial coordinates  &\SI[parse-numbers = false]{8}{} $\pm$ \SI[parse-numbers = false]{14}{\second} &\SI[parse-numbers = false]{175}{} $\pm$ \SI[parse-numbers = false]{37}{\minute} & 0 \\
			\bottomrule
			\\[-2ex]
			Total processing time & \SI[parse-numbers = false]{9}{} $\pm$ \SI[parse-numbers = false]{7}{\minute} & \SI[parse-numbers = false]{209}{} $\pm$ \SI[parse-numbers = false]{51}{\minute} & --  
		\end{tabular}
		\caption{Average timings for the generation of all the 50 biatrial geometries 
			at target resolutions of $\approx$ \SI{0.90}{\milli\meter} and $\approx$ \SI{0.25}{\milli\meter}
			and the number of models requiring manual corrections are given for each processing stage.}
		\label{Tab:timings}
	\end{table}
	
	% The extrusion of both the surface and the volumetric mesh of the biatrial walls, the opening of the veins, \gls{cs}, and the valves, along with the corresponding marking of the epicardial and endocardial tissue, was correctly achieved in all 50 patients. The total time for the surface and volume mesh generation varied depending on the required mesh resolution, as reported in Table \ref{Tab.Times}. We tested the pipeline to generate the 50 anatomical meshes with two mesh sizes of \SI{0.9}{\milli \meter}, suitable for solving the reaction-eikonal model, and of \SI{0.25}{\milli \meter}, to be used with reaction-diffusion EP models. 
	% \todo[inline]{Improve the table by writing mean +/- stddev, we can omit then the text below}
	% The generation of the surface mesh with a mesh size of \SI{0.9}{\milli \meter} took $3 \pm 2$ minutes, $\ldots$
	% and between 6 and \SI[parse-numbers = false]{38}{\second} for the generation of the volumetric mesh, with an average computational cost of \SI[parse-numbers = false]{16}{\second}. 
	% Regarding the finer mesh, the generation of the surface mesh required between 2 and 7 minutes, with an average cost of 3 minutes, while the volumetric mesh was generated in 4 to 13 minutes, with an average expense of 7 minutes (refer to Table \ref{Tab.Times} for a summary of the computational expenses of the workflow phases).
	
	% \todo[inline]{Summary of the text below, 0 manual interventions needed, average +/- std duration, 
	% all of this is in the table, no need to comment on this.}
	%\subsection{Selection of the atrial orifices, anatomical structure, fiber generations, and UACs} $\ldots$
	% The final three steps of the cardiac workflow -- selecting and labeling the atrial orifices, defining the remaining anatomical structures, and generating the fibers -- were completed automatically for all 50 models. 
	% For a mesh size of \SI{0.9}{\milli \meter}, marking the orifices took between 2 and \SI[parse-numbers = false]{11}{\second}, averaging \SI[parse-numbers = false]{5}{\second}, while labeling the remaining anatomical structures and generating the fibers required between 15 and \SI[parse-numbers = false]{65}{\second}, averaging \SI[parse-numbers = false]{30}{\second}. When considering the finer mesh, marking the orifices took between 25 and \SI[parse-numbers = false]{76}{\second}, with an average time of \SI[parse-numbers = false]{39}{\second}, while annotating the anatomical structures and generating the fibers required an average of $18 \pm 10$ minutes (refer to Table \ref{Tab.Times} for a summary of the computational expenses of the workflow phases).
	
	%\gls{uac}s were generated for all 50 models at a resolution of \SI{0.9}{\milli \meter}, 
	% taking between 3 and \SI[parse-numbers = false]{22}{\second}, with an average time of \SI[parse-numbers = false]{8}{\second}. However, when the mesh was much finer, the computation of \gls{uac}s became significantly more time-consuming, requiring an average time of 175 minutes (refer to Table \ref{Tab.Times} for a summary of the computational expenses of the workflow phases). 
	
	%\todo[inline]{Suggest to remove this to the Discussion}
	% The majority of the computational costs were attributed to solving the linear elasticity problem to normalize the \gls{uac} solutions. This was mainly due to the poorly deformed elements that could potentially result from the projection of the Laplace-Dirichlet solutions on the \gls{uac} subspace, and hinder the solver convergence.
	
	\subsection{EP simulation workflow performance evaluation}
	The efficiency of our workflow in the function twinning stage, that is of simulating atrial \gls{ep} for high fidelity \gls{ecg} generation, was tested for both \gls{re} and \gls{rd} models for all 50 patients.
	Execution times of individual processing stages -- comprising torso generation, setting up of \glspl{ic}, computation of the \gls{ecg} lead fields, and the simulation of an entire atrial activation sequence initiated at the \gls{san} -- were measured (refer to Table \tabref{Tab.Times_re_rd} and Figure \ref{Fig.RD_vs_RE}).
	As shown previously in detail \citep{gillette2021:_framework}, owing to its relaxed mesh resolution dependency, 
	the \gls{re} model is significantly more lightweight,
	facilitating the setup of an \gls{ep} model in $\approx$ \SI{3}{\minute} and the computation of a full biatrial activation sequence with a high fidelity \gls{ecg} in $\approx$ \SI{1}{\second}.
	All steps of the workflow are computationally more costly due to the stricter mesh resolution requirements, requiring $\approx$ \SI{32}{\minute} for setup, and $\approx$ \SI{19}{\minute} 
	for computing the activation sequence and the \glspl{ecg}.
	While formally the same torso surface was employed, extra costs were incurred due to the required mesh conformity at the atrial surfaces,
	leading to a markedly higher number of elements in the \gls{rd} case. 
	
	%Using either a \gls{re} or \gls{rd} model for cardiac \gls{ep} with a coarse grid of \SI{0.9}{\milli\meter} and a fine mesh resolution of \SI{0.25}{\milli\meter}, respectively, affects not only the performance of our workflow in generating the biatrial model but also the definition and generation of the structures needed to compute the \gls{ecg}.
	
	%Although the registration of the reference torso surface onto the biatrial geometry is independent of the cardiac mesh resolution, the generation of the volumetric mesh inside the torso is influenced by the cardiac domain due to the required grid conformity at the interface between the atrial and torso domains. A torso mesh conformed to an atrial grid with a resolution of \SI{0.9}{\milli\meter} was generated in an average time of \SI[parse-numbers = false]{2}{\minute} and \SI[parse-numbers = false]{25}{\second}, whereas the corresponding mesh for an atrial geometry with a resolution of \SI[parse-numbers = false]{0.25}{\milli\meter} took \SI[parse-numbers = false]{22}{\minute} and \SI[parse-numbers = false]{30}{\second}, as reported in Table \ref{Tab.Times_re_rd}.
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lll}
			\toprule
			& \multicolumn{2}{c}{\textbf{Mesh resolution / Model}} \\
			&\SI{0.9}{\milli\meter} / \gls{re} &\SI{0.25}{\milli\meter} / \gls{rd}\\ 
			\hline \\[-2ex]
			Volumetric torso &$\approx$ \SI[parse-numbers = false]{2}{\minute} \SI[parse-numbers = false]{25}{\second} & $\approx$ \SI[parse-numbers = false]{22}{\minute} \SI[parse-numbers = false]{30}{\second}\\
			Interatrial connections  &$\approx$  \SI[parse-numbers = false]{19}{\second} & $\approx$ \SI[parse-numbers = false]{39}{\second}\\
			Lead field &$\approx$ \SI[parse-numbers = false]{35}{\second} &$\approx$ \SI[parse-numbers = false]{9}{\minute} \\
			\gls{ep} simulation  &$\approx$  \SI[parse-numbers = false]{27}{\second} &$\approx$ \SI[parse-numbers = false]{19}{\minute}\\
			\bottomrule
		\end{tabular}
		\caption{Execution times of setup and simulation using \gls{re} or \gls{rd} models.}
		\label{Tab.Times_re_rd}
	\end{table}
	
	% setting up the cables is a negligible cost, slightly more expensive to compute the geodesics, but at the end
	% we should be using straight lines anyways.
	% \todo[inline]{Removed paragraph on cables, not sufficiently relevant in my view.}
	The computational costs of creating the \gls{ic} cables also showed slight differences between the coarse \SI[parse-numbers = false]{0.9}{\milli\meter} mesh and the fine \SI[parse-numbers = false]{0.25}{\milli\meter} mesh, averaging \SI[parse-numbers = false]{19}{\second} and \SI[parse-numbers = false]{39}{\second}, respectively. However, this difference is negligible compared to the total costs of generating the biatrial and torso geometry, as well as the CPU time required for the computation of the lead field solution, which is of \SI[parse-numbers = false]{35}{\second} for the \gls{re} mesh, and \SI[parse-numbers = false]{9}{\minute} for the \gls{rd} grid, and the  \gls{ep} simulation, corresponding to \SI[parse-numbers = false]{27}{\second} of computation for the \gls{re} model, and to \SI[parse-numbers = false]{19}{\minute} for the \gls{rd} gold-standard monodomain model.
	
	\subsection{Assessing discrepancy between \gls{re} and \gls{rd} model}
	To evaluate the computational fidelity of the \gls{re} model compared to a gold standard \gls{rd} monodomain model \citep{nagel2022comparison}, 
	we compare the spatiotemporal distribution of electrical sources, $V_{\rm m}(\boldsymbol{x},t)$, 
	the activation maps, $\tau(\boldsymbol{x})$, and the P-wave generated by both models,
	employing the baseline parameter settings in Section \ref{Subsubsec:baseline_sim}. 
	Results are illustrated for a representative test case in Figure \ref{Fig.RD_vs_RE}.   
	
	
	\begin{figure}[!t]
		\centering \includegraphics[width=0.95\textwidth]{images/Figure_re_rd.png}
		\caption{First/Second row: Transmembrane potential $V_m$ propagation computed with the \gls{re} and \gls{rd} models, at different time instants. Third row: Activation map obtained by solving the \gls{re} and the \gls{rd} model. Fourth row: \glspl{ecg} obtained by solving the \gls{re} (black) and \gls{rd} (blue) models coupled with the lead field. The background colors refer to the activation map.  By coloring the background with time-bands of colors corresponding to the activation map, the activation map is compared with the \gls{ecg}, highlighting the relation between regional atrial activation and P-wave.}
		\label{Fig.RD_vs_RE}
	\end{figure}
	
	Overall, the model parameters calibrated to match \gls{rd} and \gls{re} conduction velocities led to nearly identical activation patterns. 
	Minor differences emerged due to imperfect matching of conduction velocities in the \gls{ra} and \gls{la} tissue,
	as well as in the \gls{ic} cables,
	and differences in anterograde and retrograde activation by the \glspl{ic} cables mediated by an electrotonic source-sink mismatch at the interface between cable and tissue (refer to Figure \ref{Fig.RD_vs_RE}, top panels). 
	These combined effects led to a slightly longer total activation time of the atria,
	which manifested in minor differences in the \gls{pwd}, 
	with an average variation of \SI[parse-numbers = false]{2}{\milli\second}.
	
	Differences in P-wave morphology and magnitude were negligible,
	with a maximum amplitude difference in lead II of 5\%, and an average \gls{rmse} across all leads of 3.71\% (refer to Table \ref{Tab.RE_vs_RD}).
	Magnitude differences could be attributed to the difference in spatial resolution,
	as these disappeared when running the \gls{re} simulation on the higher resolution mesh (not shown).
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lccccccccccccc}
			\toprule
			Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
			\hline \\[-2ex]
			\gls{rmse} [\%] &3.56 &3.44 &3.54 &5.16 &4.87 &5.17 &4.64 &3.66 &3.09 &1.69 &2.39 &2.34 &3.71\\
			\gls{pwd} \gls{re} [ms] &128 &131 &131 &125 &129 &89 &126 &131 &123 &126 &131 &131 &125.1\\
			\gls{pwd} \gls{rd} [ms] &121 &124 &121 &124 &129 &90 &120 &131 &123 &126 &131 &131 &122.6\\
			\bottomrule
		\end{tabular}
		\caption{\gls{rmse} in percentage, between \gls{ecg} signals computed with the \gls{re} and \gls{rd} models coupled with the lead field. The \gls{pwd} of the \gls{ecg} traces obtained with the \gls{re} and the \gls{rd} models coupled with the lead field are also reported. }
		\label{Tab.RE_vs_RD}
	\end{table}
	
	% \todo[inline]{This text we might move to the Discussion}
	% Our analysis clearly demonstrates that a \gls{re} simulation yields a T-wave 
	% that matches the prediction of a highest fidelity \gls{rd} model with an accuracy 
	% well below the bounds of model uncertainty, and at a very small fraction of the compute costs,
	% i.e. the evaluation of the \gls{re} model was $> 1000 \times$ faster than the \gls{rd} model.
	
	%Our analysis demonstrated that the cardiac sources and \glspl{ecg} generated by the \gls{re} model are nearly indistinguishable from those produced by the gold standard monodomain model. The marginal amplitude variations are considered negligible, as they fall below the previously investigated observational and residual uncertainties of \gls{ecg} measurements \cite{zappon2024quantifying}.
	
	% \todo[inline]{This text is way too detailed, we don't need to explain the atrial activation sequence.}
	% Activation of the atria was initiated at the \gls{san}, spreading anteriorly along the \gls{bb} and toward the \gls{raa}, and posteriorly through the \gls{ct} and \glspl{pm}. 
	% A reduction in conduction speed was observed between the \gls{pm}s and between the \gls{pm}s and \gls{ct}. 
	% At \SI[parse-numbers = false]{11}{\milli \second}, the wave was activating the \gls{bb} cable and spreading toward the anterior wall of the \gls{la}, activating the subendocardium at \SI[parse-numbers = false]{42}{\milli \second}. The two posterior cables were activated at \SI[parse-numbers = false]{44}{\milli \second} and \SI[parse-numbers = false]{60}{\milli \second}, respectively, while the \gls{cs} cable was reached by the signal at \SI[parse-numbers = false]{83}{\milli \second}. The \gls{fo} was activated at \SI[parse-numbers = false]{48}{\milli \second}, causing the spread of a circular wave on the septum of the \gls{la}. From the \gls{la} entry site of the \gls{bb}, the signal spread toward the \gls{laa}, faster following the remaining bands of the \gls{bb}, and slower toward the roof and the \gls{mv}. The first wave collision was observed on the anterior-septal portion of the \gls{laa}, between the wave generated by the \gls{bb} and the \gls{fo}. 
	% The posterior-superior \gls{ic} contributed to the initial activation of the lateral part of the \gls{ripv} at \SI[parse-numbers = false]{63}{\milli \second}, equally in all directions. 
	% The collision between the wave generated by the \gls{bb} and the posterior-superior \gls{ic} was observed between the \gls{rpv} and on the roof between the inferior pulmonary veins. The activation of the \gls{la} from the posterior-inferior \gls{ic} and the \gls{cs} bundle was overshadowed by the other waves. The posterior-lateral wall of the \gls{la} and the tip of the \gls{laa} were the last tissues to be activated.
	
	
	% The computed transmembrane voltage, $V_m$, on a coarse grid of \SI{0.9}{\milli\meter} 
	% with the \gls{re} model and on a fine mesh of \SI{0.25}{\milli\meter} with the monodomain model, were very similar, 
	% as represented in Figure \ref{Fig.RD_vs_RE}. 
	% Major variations were observed during the activation of the \gls{bb} entry site 
	% on the anterior wall of the \gls{la} and in the late activation of the posterior \gls{la}. 
	% Similarly, the activation maps produced by both models were very similar, 
	% with differences seen in the late posterior activation of the \gls{la}.
	
	% Differences in cardiac sources were minimally reflected in the P-wave, as shown in Figure \ref{Fig.RD_vs_RE}. 
	% Amplitude variations were observed at the P-wave peak, corresponding to the activation of the anterior part of the \gls{la}, across almost all leads, as shown by the \gls{ecg} background colors, 
	% while a reduction in \gls{pwd} was registered especially in limb leads. 
	% No significant morphological changes or differences in \gls{pwd} were detected. 
	% These minimal effects on the \gls{ecg} were further quantified by calculating the \gls{rmse} between the two \glspl{ecg} and their corresponding \glspl{pwd}. 
	% The maximum amplitude variation was recorded in lead II, with a value slightly bigger than 5\%, while the average \gls{rmse} across all leads was 3.71\%. \gls{pwd} remained unchanged in almost all precordial leads, and registered variations in the limb leads, especially aVR, with an average variation of \SI[parse-numbers = false]{2}{\milli\second}.
	
	% \begin{table}[!t]
	%     \centering
	%     \begin{tabular}{lccccccccccccc}
	%         \toprule
	%         Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
	%         \hline \\[-2ex]
	%         \gls{rmse} [\%] &0.29 &0.44 &0.41 &0.42 &0.27 &0.25 &0.59 &0.46 &0.33 &0.31 &0.21 &0.25 &0.36\\
	%         \gls{pwd} \gls{re} [ms] &114 &119 &120 &121 &97 &89 &111 &86 &88 &99 &101 &121 &105.5\\
	%         \gls{pwd} \gls{rd} [ms] &115 &119 &120 &121 &97 &90 &111 &86 &88 &99 &101 &121 &105.7\\
	%         \bottomrule
	%     \end{tabular}
	%     \caption{\gls{rmse} in percentage, between \gls{ecg} signals computed with the \gls{re} and \gls{rd} models coupled with the lead field. The \gls{pwd} of the \gls{ecg} traces obtained with the \gls{re} and the \gls{rd} models coupled with the lead field are also reported. }
	%     \label{Tab.RE_vs_RD}
	% \end{table}
	
	\subsection{Impact of Bachmann's Bundle insertion on the P-wave}
	The role of \gls{bb} insertion site $\mathbf{x}_{BB_{LA}}$ on the \gls{la} activation and P-wave morphology was investigated.
	The use of a cable-based \gls{ic} formulation readily facilitated a parametric sweeping of $\mathbf{x}_{BB_{LA}}$,
	where $\mathbf{x}_{BB_{LA}}$ was varied within a radius of $\approx$ \SI[parse-numbers = false]{5}{\milli \meter} 
	around the reference location, $\hat{\mathbf{x}}_{BB_{LA}}$, as used above.
	% The effect on activation maps and P-waves is shown in \ref{Fig.Tests_1BB}. 
	Shifts in $\mathbf{x}_{BB_{LA}}$ altered the location and timing of wavefronts collisions with waves initiated through the other \glspl{ic}.
	These were predominantly the rim of the fossa ovalis, activating the \gls{la} almost synchronously with the \gls{bb},
	and the posterior-superior \gls{ic}, activating the carina of the \glspl{lpv} with a delay of \SI[parse-numbers = false]{10}{\milli \second}. 
	This entailed a change in the total activation time of the \gls{la}, and induced a spread in \gls{pwd} and magnitude of the P-wave, with very limited effect on the corresponding morphology (refer to Figure \ref{Fig.Tests_1BB}). 
	%Over a simulation of \SI[parse-numbers = false]{150}{\milli \second}, 
	%we recorded a variation in the activation map 
	
	Variation of $\mathbf{x}_{BB_{LA}}$ alone led to differences in total activation time of \SI[parse-numbers = false]{23}{\milli \second}, 
	%between the shortest activation time of \SI[parse-numbers = false]{109}{\milli \second} and the longest activation time.
	in \gls{pwd} between \SI{12}{\milli \second} to \SI{24}{\milli \second} in individual leads, with a minimum and maximum average \gls{pwd} of 104 and \SI[parse-numbers = false]{120}{\milli \second} across all leads, respectively,    
	and in magnitude to a mean absolute distance \eqref{eq:MAD} between 4.4$\times$10$^{-4}$ and  2.6$\times$10$^{-3}$ mV.
	A quantitative summary on \gls{pwd} variability is given in Table \ref{Tab.PWD}.
	%Major changes in activation times were observed along the axis connecting the \gls{laa} (fastest total activation) 
	%and the \gls{rspv} (slowest total activation). 
	
	% \todo[inline]{Cannot follow this here, are you sure about the anatomical references?}
	% While the wave generated by the posterior-superior \gls{ic} resulted in a similar activation of the roof \textcolor{red}{incorrect in my view}, 
	% the activation of the posterior-lateral wall of the \gls{la} was mostly influenced by the position of the \gls{bb} entry site 
	
	%P-waves resulting from variations of $\mathbf{x}_{BB_{LA}}$ are represented in the first and third rows of Figure \ref{Fig.Tests_1BB}. 
	% For all activation sequences P-waves were positive and monophasic, except for leads aVL and V1.
	% Correct but obvious
	% --
	%Variation of the P-wave was recorded from \SI[parse-numbers = false]{37}{\milli \second}, 
	%corresponding to the signal reaching the \gls{la} at the \gls{bb} entry site nearest to the \gls{ra}, 
	%i.e., where the \gls{bb} cable was shortest. 
	%As a measure of amplitude variation, a mean absolute distance \eqref{eq:MAD} over the average P-wave was recorded between 0.00044 and 0.0026. 
	%An increase in absolute distance was steadily observed moving outward from the investigation area, 
	%without a preferential direction. 
	% \todo[inline]{This is also not correct, before 37ms there are no differences.}
	% Major variations of the P-wave were also recorded in two-time intervals: between 23 and \SI[parse-numbers = false]{71}{\milli \second}, and between 88 and \SI[parse-numbers = false]{125}{\milli \second}, 
	% corresponding to the activation phases of the anterior and posterior-lateral walls of the \gls{laa}, 
	% in accordance with the activation maps observations.
	
	% Corresponding to changes in the total activation time, variations in \gls{pwd} were observed in all leads, 
	% as reported in Table \ref{Tab.PWD}, with minimum and maximum average \gls{pwd} of 104 and \SI[parse-numbers = false]{120}{\milli \second}, respectively. 
	% Can be seen from the table
	%Major variations in \gls{pwd} were witnessed in leads aVL and aVF, 
	%and minimum variations in precordial leads V2 and V3. 
	% \todo[inline]{Next statement is obvious, can be omitted.}
	% Moreover, the P-waves with maximum and minimum \gls{pwd} were observed in the simulations where the maximum and minimum total activation times were recorded. 
	% % \todo[inline]{Next statement not really supported by the data we are showing.}
	% Overall, a more physiological sinus rhythm activation and corresponding P-wave duration were obtained 
	% for \gls{bb} entry sites situated in the central anterior wall of the \gls{la} and towards the \gls{laa}.
	
	\begin{figure}[!t]
		\centering \includegraphics[width=0.95\textwidth]{images/Test_case_1BB_2.png}
		\caption{Influence of \gls{bb} \gls{la} insertion site.  First row: 
			Biatrial activation map obtained for \gls{bb} reference insertion site along with P-waves in leads I and -aVR. The background of P-wave traces is color coded according to the activation map 
			to highlight the relation between 
			the dipole layer of depolarization wave fronts 
			and the associated amplitude in the P-wave. 
			Second row: \gls{la} activation maps corresponding to the minimum and maximum total activation time. 
			Third row: P-wave variation due to varying the \gls{bb} insertion site in the \gls{la}. 
			Insertion site an \gls{ecg} are color coded corresponding to the absolute distance from the mean \gls{ecg} (black), 
			averaged over time and the 12 leads.}
		\label{Fig.Tests_1BB}
	\end{figure}
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lccccccccccccc}
			\toprule
			Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
			\hline \\[-2ex]
			Min \gls{pwd} [ms] &93 &93 &108 &110 &104 &107 &99 &106 &106 &106 &106 &106 &104,2\\
			Max \gls{pwd} [ms] &117 &115 &124 &126 &119 &120 &118 &118 &118 &124 &124 &123 &120,5\\
			(Max - Min) \gls{pwd} [ms] &22 &16 &13 &16 &24 &15 &19 &12 &12 &18 &15 &14 &16,3\\
			\bottomrule
		\end{tabular}
		\caption{Maximum and minimum \gls{pwd}, and their differences, for each lead and averaged over all leads.}
		\label{Tab.PWD}
	\end{table}
	
	\subsection{Effect of a fan-like \gls{la} insertion Bachmann's Bundle on the P-wave}
	Our modeling approach flexibly supports an arbitrary number of \glspl{ic} cables which can be bundled to increase source strength for generating \glspl{ecg} and \glspl{egm},
	or to model inter-atrial coupling with higher anatomical complexity.
	This ability is showcased for investigating the influence of a distributed fan-like insertion of \gls{bb} into the \gls{la} with three independent coupling locations.
	An exemplary activation map along with the resulting distribution of P-waves is shown in Figure \ref{Fig.Three_bb}.
	As the same velocity was assumed in all cables the initial activation at the three \gls{la} insertion sites 
	was not synchronous but led to an elongated area of early activation along the location of the wide \gls{bb}.
	
	% \todo[inline]{The variability is marginal, not sure whether we should include this.}
	%We depicted areas of search for the \gls{la} entry points of the three cables representing the \gls{bb} in Figure \ref{Fig.Three_bb}, with an example of obtained activation maps, and the computed \gls{ecg} varying such entry points. 
	The total activation time was observed between 118 and \SI[parse-numbers = false]{125}{\milli\second}, comparable smaller than the \gls{bb} single-cable case. 
	
	The highest variation in the P-wave was observed between 80 and \SI[parse-numbers = false]{132}{\milli \second}, corresponding to the activation of the posterior wall of the \gls{la}.
	% Small or no variation, especially in the precordial leads, was observed in the time interval of activation of the anterior \gls{la} wall, between 28 to \SI[parse-numbers = false]{80}{\milli \second}. 
	% A slight increase in the range of variation of the absolute distance index compared to the single-cable case was registered, between 0.00014 and 0.0028.
	The average range of variation of the \gls{pwd} was registered from 102 to \SI[parse-numbers = false]{110}{\milli \second}, 
	thus smaller than the single-cable case. 
	The maximum and minimum \gls{pwd} over the leads was also reduced compared to the single-cable case, with maximum variation in lead aVL and II.    
	
	\begin{figure}[!t]
		\centering \includegraphics[width=0.9\textwidth]{images/Fig_three_bb.png}
		\caption{First row-left: Investigated \gls{la} entry site areas of the three cables representing the \gls{bb}. The different colors identify the three different regions of search. First row-center and right: An example of activation map obtained with the three-cable \gls{bb}. Second row: P-wave variation obtained by moving the \gls{la} entry sites of the three cables representing the \gls{bb}. The signals coloration correspond to the absolute distance from the mean \gls{ecg}, averaged over time and the leads. The mean \gls{ecg} is shown in black. By coloring the background of the lead with time-bands of colors corresponding to the activation map, the activation map is compared with the \gls{ecg}, highlighting the relation between regional atrial activation and P-wave.}
		\label{Fig.Three_bb}
	\end{figure} 
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lccccccccccccc}
			\toprule
			Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
			\hline \\[-2ex]
			Min \gls{pwd} [ms] &80 &100 &107 &110 &95 &107 &94 &98 &110 &110 &109 &109 &102,4\\
			Max \gls{pwd} [ms] &89 &114 &115 &118 &99 &115 &104 &108 &113 &113 &115 &116 &109,9\\
			(Max - Min) \gls{pwd} [ms] &9 &14 &8 &8 &4 &8 &10 &10 &3 &3 &6 &7 &7,5\\
			\bottomrule
		\end{tabular}
		\caption{Maximum and minimum \gls{pwd}, and their differences, for each lead and averaged over all leads.}
		\label{Tab.PWD_threeBB}
	\end{table}
	
	% \todo[inline]{Elena, how is this modelled in Caroline Roney's paper? Please look this up, also the models
	% if they are available.}
	
	\subsection{Effect of the RA endocardium on the P-wave}
	\begin{figure}[!t]
		\centering \includegraphics[width=0.95\textwidth]{images/Fig_PM.png}
		\caption{First row: P-wave obtained when including (blue) and excluding (black) the \gls{ra} endocardial tissue outside of the \gls{pm}s and \gls{ct}. The P-wave of all leads is compared to the activation map by coloring the background. Second row: Activation maps obtained when including (left) and excluding (right) the \gls{ra} endocardial tissue. Third row: Absolute difference between the activation maps obtained when including and excluding the \gls{ra} endocardial tissue.}
		\label{Fig.Tests_PM}
	\end{figure}
	
	The \glspl{ecg} obtained by simulating the atria, both including and excluding the \glspl{ra} endocardial layer outside of the \gls{pm} and \gls{ct}, are shown in Figure \ref{Fig.Tests_PM}. A decrease in wave amplitude between 14 and \SI[parse-numbers = false]{71}{\milli \second}, corresponding to the activation of the \gls{ra}, was observed in all leads. The \gls{rmse} in percentage is reported in Table \ref{Tab.PWD_PM}. The major amplitude variation, exceeding 1\%, was observed in leads II, V1, and V2, which capture the signal propagation towards the \gls{raa} and the \gls{pm}. Variations between 0.49\% and 0.91\% were observed in all other leads, except for aVL, where the variation was almost null. The \gls{pwd}, as reported in Table \ref{Tab.PWD_PM}, remained unchanged in all leads.
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lccccccccccccc}
			\toprule
			Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
			\hline \\[-2ex]
			\gls{rmse} [\%] &0,08 &0,63 &0,91 &1,19 &0,88 &0,58 &1,31 &1,18 &0,85 &0,72 &0,58 &0,49 &0,78\\
			PWD [ms] &114 &119 &120 &121 &97 &89 &111 &86 &88 &99 &101 &121 &105,5\\
			\bottomrule
		\end{tabular}
		\caption{\gls{rmse} between the \gls{ecg}s obtained when excluding the \gls{ra} endocardial tissue outside of the \gls{pm}s and \gls{ct}, and the \gls{ecg}s signal obtained including the \gls{ra} endocardial tissue, and \gls{pwd}, for each lead and averaged over all leads.}
		\label{Tab.PWD_PM}
	\end{table}
	
	The \gls{ecg} variations were related to the computed activation maps and their absolute difference, depicted in Figure \ref{Fig.Tests_PM}. When the \gls{ra} endocardial wall was included, the signal propagated faster on the posterior part of the \gls{raa}, on the posterior-lateral wall of the \gls{ra} near the \glspl{pm}, and on the septum. A total activation delay of about \SI[parse-numbers = false]{2.1}{\milli \second} was observed at the tip of the \gls{raa}, and \SI[parse-numbers = false]{2.7}{\milli \second} on the epicardium between the \glspl{pm}. The \gls{bb}, \gls{ra} anterior wall, and the \gls{ct} were activated at the same speed. However, the removal of the \gls{ra} endocardial layer between the \gls{ct} and the septum caused a delay in the activation of the septum and the \gls{ra} wall near the \gls{cs}. Although this delay did not affect the overall \gls{ra} activation, it slightly impacted the activation of the posterior wall of the \gls{la}, causing a total delay of up to \SI[parse-numbers = false]{2}{\milli \second} in the posterior-inferior region. Nonetheless, the total activation map of the \gls{la} remained almost unchanged, consistent with the computed \gls{ecg}s.
	
	\subsection{Calibration of the atrial activation sequence based on the P-wave of the 12 lead \gls{ecg}}
	
	One anatomical model of a patient was selected along with the recorded P-wave under normal sinus rhythm. 
	To keep the calibration procedure tractable, the parameter space to be explored was restricted to the initial exit site at the \gls{san}, the conduction velocity in the \gls{ra},
	and the location and timing of three insertion sites of \gls{bb} in the \gls{la}. 
	In a first pseudo-calibration step for a fixed \gls{san} exit site and baseline velocities in the \gls{ra}, 
	the location of \gls{bb} was varied in interactive simulation runs to obtain a close approximation of the terminal half of the P-wave. 
	One configuration yielding a good morphological fit in all leads, assessed only by visual inspection, was selected (see Figure \ref{Fig.Fig_SAN_sampling}, top row).
	
	
	% Figure \ref{Fig.Fig_SAN_sampling} (first row, right) shows the \glspl{ic} computed after manual pseudo-calibration of the \gls{la} activation, as well as the conduction block defined between the \gls{ct} and the septum.
	
	Keeping the \gls{bb} fixed, the physiological envelope of the P-wave signals was computed by sampling over $\mathbf{x}_{SAN}$, $v_f$, and $v_s$ (see Figure \ref{Fig.Fig_SAN_sampling}, mid row). 
	% P-wave traces are color-coded according to their distance to the clinical P-wave measured as \gls{rmse}.
	The P-wave envelope under this sampling covered the observed clinical signal in most leads quite well, 
	with the exceptions of leads aVL, -aVR and V1. 
	While the discrepancy in the low amplitude P-waves in leads aVL and -aVR were rather small,
	this was not the case in lead V1 where the biphasic P-wave could not be approximated.
	The terminal negative deflection of the measured P-wave was likely mediated by fibrotic tissue in this AF patients with impaired electrophysiologicalgls{ep} excitability and repolarization properties.
	%As the atria were modeled as healthy this specific feature
	% could not be replicated.
	
	Quantitatively, the P-waves were well approximated, 
	with a \gls{rmse} in the range of $[1.72\%,2.79\%]$.
	The minimum \gls{rmse} was achieved 
	when the \gls{san} was positioned at 62\% along the \gls{ivc}-\gls{svc} axis and 35\% along the lateral-to-septal axis (see Figure \ref{Fig.Fig_SAN_sampling}, top row), 
	with conduction velocities of $v_f = \SI[parse-numbers = false]{1.1}{\meter/\second}$ and $v_s = \SI[parse-numbers = false]{0.45}{\meter/\second}$. 
	The \gls{rmse} values, as well as the \gls{pwd} for each lead, 
	obtained with the optimal parameter set that minimized \gls{rmse} are presented in Table \ref{Tab.PWD_SAN}.
	On average, simulated \gls{pwd} was longer than clinical \gls{pwd}, with \SI[parse-numbers = false]{139}{\milli\second} versus \SI[parse-numbers = false]{130}{\milli\second}, suggesting that calibration could be further improved 
	by including \gls{ic} and \gls{la} velocities in the sampled parameter space.  
	
	% We moreover measured the \gls{pwd} of the \gls{ecg} traces that minimized the \gls{rmse} and compared these values with the \gls{pwd} from the clinical data, as shown in Table \ref{Tab.PWD_SAN}. The largest differences in \gls{pwd} were observed in leads II, V2, and V6, while the smallest differences were found in leads III, aVL, V1, and V4. Overall, all \gls{ecg} traces showed a \gls{pwd} greater than \SI[parse-numbers = false]{120}{\milli\second}, consistent with the clinical data.
	
	
	%Variations in activation sites and conduction velocities affect both the amplitude and shape of the entire P-wave. 
	%Higher variability in amplitude was observed in limb leads II, aVF, and III, and in precordial lead V1. 
	%The envelope covered the clinical \gls{ecg} in most of the leads. 
	% However, sufficient positive P-wave amplitude was not achieved in aVL and -aVR, while an excessively positive P-wave was obtained in V1, independent of the parameters used. 
	% Moreover, similar qualitative P-wave variability was observed regardless of the shape of the activation impulse. 
	%This result was confirmed quantitatively, with a \gls{rmse} range of $[1.72\%,2.79\%]$ for spherical activation impulses and $[1.76\%,2.63\%]$ for cylindrical activation impulses.
	
	% For spherical activation impulses, the minimum \gls{rmse} was achieved when the \gls{san} was positioned at 62\% along the \gls{ivc}-\gls{svc} axis and 35\% along the lateral-to-septal axis (refer to Figure \ref{Fig.Fig_SAN_sampling}, first row, right, for a graphical representation), with conduction velocities of $v_f = \SI[parse-numbers = false]{1.1}{\meter/\second}$ and $v_s = \SI[parse-numbers = false]{0.45}{\meter/\second}$. 
	
	% The cylindrical activation impulse that minimized the \gls{rmse} had its principal axis located between 32\% and 37\% along the \gls{ivc}-\gls{svc} axis, and at 43\% along the lateral-to-septal axis, with a total length of \SI{4.75}{\milli\meter} (refer to Figure \ref{Fig.Fig_SAN_sampling}, first row, right, for a graphical representation). The conduction velocities for this case were $v_f = \SI[parse-numbers = false]{0.6}{\meter/\second}$ and $v_s = \SI[parse-numbers = false]{0.45}{\meter/\second}$. 
	
	% The \gls{rmse} values for each lead corresponding to the parameter set that minimized \gls{rmse} are presented in Table \ref{Tab.PWD_SAN}. The lowest \gls{rmse} was observed in leads III and aVL, while the highest \gls{rmse} was noted in leads -aVR and V1.
	
	\begin{figure}[!t]
		\centering \includegraphics[width=0.95\textwidth]{images/Figure_SAN_sampling.png}
		\caption{P-wave based calibration of the atrial activation sequence. 
			Top row, left: Constant \glspl{ic} configuration 
			obtained by interactive pseudo-calibration. 
			Top row - right: Spatial sampling ranges of \gls{san} exit sites 
			with optimal location that minimized the mismatched as measured by the \gls{rmse}. 
			Bottom row: Physiological envelope of P-waves (colored traces) obtained by varying 
			\gls{san} exit site, and the velocities $v_f$ and $v_s$ in the \gls{ra} 
			together the clinical P-wave (black).
			%, and the cylindrical \gls{san} activation site, $v_f$ and $v_s$. 
			Traces are color-coded according to their distance to the clinical P-wave
			measured as \gls{rmse}.
			%The color referred to the \gls{rmse}, in percentage, computed between each signal and clinical data. The clinical \gls{ecg} is represented in black.
		} \label{Fig.Fig_SAN_sampling}
	\end{figure}
	
	\begin{table}[!t]
		\centering
		\begin{tabular}{lccccccccccccc}
			\toprule
			Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
			\hline \\[-2ex]
			min \gls{rmse} [\%] &1.16 &1.68 &2.27 &2.06 &1.33 &1.02 &2.60 &1.51 &2.22 &1.87 &1.58 &1.37 &1.72 \\[1ex]
			\gls{pwd} [ms]  &&&&&&&&&&&&& \\
			Clinical data &139 & 147 & 123& 124 &127 &146 &134 & 125 &118 &132 &120 &129 &130.33\\
			Simulated \gls{ecg} &136 &137 &135 &150 &133 &151 &145 &142 &131 &129 &131 &151 &139.25 \\
			Diff with data &3 & 10 &12 &26 &6 &5 &11 &17 &13 &3 &11 &22 &8.92\\[1ex]
			\bottomrule
		\end{tabular}
		\caption{Minimum \gls{rmse}, in percentage, \gls{pwd} of the \gls{ecg} data, of the computed \gls{ecg} with minimum \gls{rmse}, and their difference, obtained by varying the position of the spherical or cylindrical activation impulse representing the \gls{san}, and the conduction velocity $v_f$ and $v_s$ of the \gls{ra}.}
		\label{Tab.PWD_SAN}
	\end{table}
	
	% \begin{table}[!t]
	%     \centering
	%     \begin{tabular}{lccccccccccccc}
	%         \toprule
	%         Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
	%         \hline \\[-2ex]
	%         \textbf{Spherical min(\gls{rmse})} &&&&&&&&&&&&&\\
	%         \gls{rmse} [\%] &1.16 &1.68 &2.27 &2.06 &1.33 &1.02 &2.60 &1.51 &2.22 &1.87 &1.58 &1.37 &1.72 \\[1ex]
	%         \textbf{Cylindrical min(\gls{rmse})} &&&&&&&&&&&&&\\
	%         \gls{rmse} [\%] &1.06 &1.47 &2.29 &1.92 &1.27 &0.94 &3.58 &1.64 &2.17 &1.87 &1.57 &1.31 &1.76\\
	%         \bottomrule
	%     \end{tabular}
	%     \subcaption{Minimum \gls{rmse}, in percentage, obtained by varying the position of the spherical or cylindrical activation impulse representing the \gls{san}, and the conduction velocity $v_f$ and $v_s$ of the \gls{ra}.}
	%     \bigskip
	%     \begin{tabular}{lccccccccccccc}
	%         \toprule
	%         Lead &aVL &I &-aVR &II &aVF &III &V1 &V2 &V3 &V4 &V5 &V6 &Average\\ 
	%         \hline \\[-2ex]
	%         \textbf{Clinical data} &&&&&&&&&&&&& \\
	%         \gls{pwd} [ms] &139 & 147 & 123& 124 &127 &146 &134 & 125 &118 &132 &120 &129 &130.33\\[1ex]
	%         \textbf{Spherical min(\gls{rmse})} &&&&&&&&&&&& \\
	%         \gls{pwd} [ms] &136 &137 &135 &150 &133 &151 &145 &142 &131 &129 &131 &151 &139.25 \\
	%         Diff with data &3 & 10 &12 &26 &6 &5 &11 &17 &13 &3 &11 &22 &8.92\\[1ex]
	%         \textbf{Cylindrical min(\gls{rmse})} &&&&&&&&&&&&&\\
	%         \gls{pwd} [ms] & 131 &132 &130 &150 &133 &137 &135 &147 &131 &129 &131 &151 &134.42\\
	%         Diff with data &8 &15 &7 &26 &6 &9 &1 &22 &13 & 3 &11 &22 &6.08\\
	%         \bottomrule
	%     \end{tabular}
	%     \subcaption{\gls{pwd} of the \gls{ecg} data, and of the computed \gls{ecg} signals computed varying the site of the spherical or cylindrical activation impulse, and \gls{ra} $v_f$ and $v_s$, and their absolute difference with data \gls{pwd}. The \gls{pwd} of the computed \gls{ecg} referred to the \gls{ecg} traces that minimize the \gls{rmse} to the clinical data.}
	%     \label{Tab.PWD_SAN}
	% \end{table}
	
	
	
	\section{Discussion}
	\label{Sec:discussion}
	In this study, we present a highly automated scalable end-to-end workflow for generating patient-specific anatomical digital twins of human atria for their efficient calibration based on clinical data, e.g. the \glspl{ecg}.
	% patient-specific anatomically accurate volumetric biatrial models directly from medical images.
	Our novel workflow is comprehensive, 
	including the generation of an anatomical reference frame, that facilitates parametric encoding of all space-varying \gls{ep} model properties,
	as required for model calibration, 
	a detailed annotation of all relevant anatomical landmarks and structures,
	including a fiber architecture based on a detailed anatomy-informed set of rules, and a new method for lightweight and flexible \glspl{ic}.
	
	Our workflow, implemented in a single software building on meshtool \citep{neic2020automating},
	generates smooth biatrial anatomical representations of sufficient mesh quality 
	for a range of resolutions suitable for the entire spectrum of atrial \gls{ep} simulations.
	Both coarser meshing at $\approx$ \SI{0.9}{\milli \meter} for real-time calibration using \gls{re} type models
	% for real-time parameter sweeps suitable for model calibration 
	% using \gls{re} type models 
	\citep{neic17:_reaction_eikonal,pezzuto2017:_ecg},
	as well as finer meshing with sufficient resolution for fully mechanistic \gls{rd} modeling studies, is supported. 
	The workflow efficiently generates simulation-ready models within less than 10 minutes
	at a reference resolution of approximately \SI[parse-numbers = false]{0.90}{\milli \meter}.
	In contrast to previous approaches for building volumetric biatrial models,
	that used statistical shape models to generate virtual anatomies \citep{nagel2021:_biatrial_ssm,roney2023:_bia_vol},
	our approach generates accurate representations of an individual patient's atrial anatomy, 
	with fidelity being limited only by image quality and segmentation accuracy.
	A key advantage of our approach is its robustness, 
	as the description of anatomy is built on a volumetric image stack.
	This avoids the more common manifold extrusion procedure prone to topological mesh errors.
	
	Building on this anatomical model generation pipeline, the workflow is moreover extended for \gls{ecg} based model calibration, employing either  
	% following a dual strategy that combines 
	\gls{re}-based real-time \gls{ep} modeling, suitable for \gls{ep} calibration,
	% for calibration
	% with 
	or a full fidelity computationally more demanding \gls{rd} model, suitable for predictive simulations.
	Inter-atrial conduction is modeled by initially imposing electrical insulation between the \gls{ra} and \gls{la} \citep{costa2014:_fibrotic}, and then discretely reconnecting them using a novel, highly flexible, and physiologically constrained method. This approach employs auto-generated cables to establish inter-atrial conduction pathways, accurately representing all anatomically significant \glspl{ic} between the \gls{ra} and \gls{la}. Our approach facilitate the generation of pathways with prescribed        
	% This lightweight approach is flexible, facilitating the generation of pathways with prescribed 
	conduction velocities to connect arbitrary parametrically steerable locations within \gls{ra} and \gls{la} on the fly, 
	and, thus, avoids rigid and error-prone explicit meshing of inter-atrial bundles \citep{nagel2021:_biatrial_ssm}.
	
	Finally, we employ a forward \gls{ecg} generation framework based on the \gls{relf} model 
	that yields full fidelity \glspl{ecg} with real-time performance, 
	to support a clinically compatible model calibration procedure.
	To this end, volumetric atrial models are swiftly registered with reference torso models
	with corresponding electrodes, and integrated into volumetric conformal atria-torso meshes. 
	All combinations of cardiac \gls{ep} source and field models are supported.
	As mesh conformity is preserved, high fidelity \glspl{ecg} and \glspl{egm} can be computed
	over the entire solution domain based on either bidomain or lead field formulation, 
	without being restricted to lower fidelity methods such as potential recovery \citep{bishop2011bidomain},
	or boundary element approaches, that require smooth low pass pass-filtered coarse representations 
	of the atrial sources \citep{schuler2019:_spatial}. 
	
	%In this proof-of-concept study we demonstrate the ability of this end-to-end workflow
	%to generate multi-resolution anatomical models in a highly automated and scalable manner.
	The ability to compute full fidelity \glspl{ecg} with close to real-time performance is demonstrated by exploring important parameters governing the genesis of the P-wave, which were impossible or notoriously difficult to explore with previous approaches, e.g., the anatomy of \glspl{ic}.
	% that are impossible or notoriously difficult to explore with previous approaches.
	We show the equivalence in activation patterns and \glspl{ecg} between the lower resolution \gls{re} model, and the high resolution \gls{rd} model, with discrepancies well below the overall model uncertainty.
	These combined features make our end-to-end workflow suitable for large-scale atrial \gls{ep} modeling studies.
	The achieved model efficiency supports a fast exploration of the high-dimensional parameter space 
	spanned by atrial anatomy, structure and \gls{ep}, 
	by facilitating unattended sweeps over important space-varying parameters
	which is needed for the automation of optimization loops 
	using \gls{ecg} or \gls{egm} observations as target for calibration.
	
	
	%Our workflow allowed for the use of both \gls{rd} and \gls{re} model to simulate the cardiac \gls{ep}, with no exception, and for their coupling with either the Pseudo-bidomain model, the $\phi_e$-recovery of the Lead Field method to generate \gls{ecg} traces. Between the available options, in this work, we elect the \gls{relf} method to compute the \glspl{ecg} with close to real-time performance. We prove the ability of the \gls{relf} to accurately approximate the cardiac sources and to generate reliable \glspl{ecg} in sinus rhythm. 
	
	
	%of sufficiently high mesh quality to be suitable for most widely used forward \gls{ep} models. An automated multi-label segmentation of atrial anatomy is performed employing a \gls{scn}. 
	%A first set of principal atrial anatomical structures are automatically identified and labeled on the segmented blood pool. Endocardial and epicardial atrial walls are generated on the segmentation by extrusion of the blood pool on the background space based on a prescribed set of rules, allowing for an automatic and robust opening of the veins, \gls{cs}, and valves. Volumetric meshes are generated directly from the segmentation, yielding smooth anatomical representations of excellent mesh quality over a range of resolutions suitable for the entire spectrum of atrial cardiac electrophysiology simulations.
	%Detailed right atrial endocardial anatomical and heterogeneous \gls{ep} functional structures are explicitly computed, including a fiber architecture based on a detailed anatomy-informed set of rules.
	%A functional reference frame for defining space-varying atrial parameter fields is presented in the form of a new and robust set of volumetric \glspl{uac}.
	
	%We also proposed a new method to represent \glspl{ic} using auto-generated cables. Our method is flexible, allowing for the rapid generation of any number of \glspl{ic} and the calibration of conduction properties for each one, and functionally realistic, ensuring electrical isolation of the \glspl{ic} from the atrial body except at the entry and exit points. 
	
	%Finally, we defined an efficient and clinically compatible \gls{ecg} forward generation framework to compute high fidelity \glspl{ecg}. Our framework allowed for the rapid generation of a volumetric atria-torso mesh by registering a reference torso geometry --and a corresponding set of electrodes -- to fit the volumetric atrial model.  Our workflow allowed for the use of both \gls{rd} and \gls{re} model to simulate the cardiac \gls{ep}, with no exception, and for their coupling with either the Pseudo-bidomain model, the $\phi_e$-recovery of the Lead Field method to generate \gls{ecg} traces. 
	%Between the available options, in this work, we elect the \gls{relf} method to compute the \glspl{ecg} with close to real-time performance. We prove the ability of the \gls{relf} to accurately approximate the cardiac sources and to generate reliable \glspl{ecg} in sinus rhythm.  
	
	%The entire workflow, implemented in a single software building on meshtool \cite{neic2020automating}, completes in less than 10 minutes when generating models at a reference resolution of approximately \SI[parse-numbers = false]{0.90}{\milli \meter}. 
	
	%Additionally, it allows for easy and automatic variation of anatomical and physical \gls{ep} parameters crucial for \glspl{ecg} generation, such as \glspl{ic} position and number, \gls{san} position and size, and the conduction velocity of the atrial body.
	
	%Therefore, our workflow is suitable for exploring the high-dimensional \gls{ep} parameter space, automating and optimizing the \gls{ecg} calibration procedure.
	
	% We demonstrate the applicability of the model for modeling the influence of inter-atrial conduction paths on the P-wave in the ECG,
	% using a detailed torso model and a bidomain fidelity ECG model,
	% and validate the ability of the volumetric model to produce physiological \glspl{ecg}.
	% Combined with the \gls{relf} model of electrophysiology we show biatrial forward simulations with full biophysical fidelity
	% at real-time performance.
	% The presented modeling approach allows the generation of high fidelity bi-atrial anatomy models at scale, 
	% and real-time ECG simulations for P-wave based model calibration.       
	
	\subsection{Scalable generation of volumetric biatrial anatomy models}
	Robust computational workflows capable of generating comprehensive volumetric biatrial anatomical models, with fiber architecture and anatomical reference frames at scale remain challenging.  
	Only a few approaches have been reported to date, exhibiting significant variation in the degree of automation, anatomical and structural fidelity, flexibility in prescribing fiber architecture, and their ability to generate clinically observable signals such as \glspl{egm} and \gls{ecg} with sufficient accuracy.
	
	The majority of atrial modeling studies using larger cohorts relied upon a bilayer formulation \citep{labarthe2014:_bilayer}
	which greatly simplifies the meshing procedure and reduces computational expenses
	at the cost of decreased biophysical fidelity. 
	For instance, it has been demonstrated that the behavior of bilayer models 
	may differ strikingly from full 3D models
	with respect to propagation patterns and arrhythmia dynamics \citep{roney2021constructing}.
	Discrepancies may stem from various factors such as e.g.\ altered source-sink relations,      
	the impact of mediating transmural conduction through resistive coupling between the endocardial and epicardial layers using finite elements of incompatible dimensionality,
	or differences in inter-atrial conduction pathways.
	Another important limitation of bilayer models is their reduced accuracy in representing the atrial sources, and a lack of suitable approaches for computing higher-fidelity extracellular potential fields \citep{bishop2011bidomain}, due to incompatibility with standard \gls{fem}.
	%, 
	% as implemented in cardiac \gls{ep} simulators such as openCARP \citep{plank2021opencarp}.
	% bilayer models -- due to their manifold nature -- are not compatible with standard \gls{fem} 
	% as implemented in cardiac \gls{ep} simulators such as openCARP \cite{plank2021opencarp}. 
	As such, modeling clinical data, such as the \glspl{egm} or \gls{ecg}, with high fidelity is challenging. Moreover, modeling extracellular potential field-related physiological events, such as bath loading effects, cannot be represented at all \citep{bishop2011bathloading}.
	These factors combined limit their trustworthiness and application scope.
	
	Volumetric biatrial meshes generated in early pioneering studies used artisanal handcrafted non-scalable methods that were tractable only for constructing small-size cohorts, 
	often consisting of a single model only \citep{harrild2000:_atria,seemann2006:_atria}.  
	Later, model-generation pipelines were proposed, gradually refined and increasingly automated
	towards a more industrialized process for producing models \emph{en masse}.
	A first mostly complete workflow was reported in a study by \cite{krueger2012personalization} 
	which was applied to generate a cohort of nine biatrial models.
	However, as numerous manual interventions were required in segmentation and landmark selection, for fiber mapping, and anatomical region definition, scalability remained limited.
	
	A further automated approach using a segmentation-derived endocardial surface mesh as input
	was reported and tested in 29 patients in \cite{azzolin2023:_augmenta}.
	The approach bears similarities to ours in that 
	e.g.\ surface curvature measures are used to identify structures such as the pulmonary veins, 
	but also differs markedly in other processing steps 
	which are more challenging to perform when using manifold meshes only. 
	For instance, the identification of the mitral valve orifice,
	which is trivially identified as a label interface in our volumetric approach,
	required a rigid registration with a mean statistical shape model.
	Moreover, important anatomical landmarks such as the \gls{fo} and its rim 
	or the \gls{cs} were not identified,
	and additional manual procedures were required to identify anatomical regions such as the \gls{laa}.
	% \todo[inline]{Are there more differences we should highlight?}
	
	The largest atrial modeling cohort study comprising 1000 biatrial models 
	has been reported in \cite{roney2023:_bia_vol}. 
	While this is indicative of robustness and scalability, 
	the numerous interactive manual interventions required suggest a labor intense procedure.
	Moreover, the workflow was primarily tailored for generating patient-specific bilayer models,
	as only these were derived from clinical imaging data.
	Support for generating volumetric models was limited to volumetric remeshing of pre-existing manifold meshes built from a statistical shape model \citep{rodero2021linking}, 
	and auxiliary data such as fiber architecture and \glspl{uac} were transferred over from an atlas.
	A direct generation of anatomically accurate patient-specific 3D models from images,
	as per our approach, is currently not supported.
	Similarly, in the study by \cite{nagel2021:_biatrial_ssm}, 
	volumetric biatrial models were also generated using statistical shape model manifolds as input
	to a heterogeneous workflow that integrated five different mesh generation and manipulation tools.
	
	%In \cite{krueger2012personalization}, the authors presented a set of tools for personalizing computational models of \gls{af} subjects and tested them on nine patients. An automatic segmentation procedure was used for five of the subjects. For each patient, a volumetric biatrial model was computed from the obtained segmentation, with prescribed wall thicknesses. A manual procedure was then implemented to select anatomical regions, including the \gls{ct}, \glspl{pm}, \gls{bb}, atrioventricular rings, and appendages. A set of atlas fiber architectures was mapped onto the obtained geometry using an automatic registration technique.  As the final step, the pulmonary veins of the \gls{la} were manually selected. While this represents the first complete workflow for atrial model personalization, it requires manual segmentation and definition of anatomical regions. 
	
	% Automated generation
	
	%A more automated approach was implemented by Azzolin et al. in \cite{azzolin2023:_augmenta}. Their workflow, tested on 29 patients, takes as input the endocardial surface of the \gls{la}, either already segmented or obtained from mapping data. If a closed surface is provided, the pipeline identifies the pulmonary veins as regions with the highest curvature, excluding the manually indicated \gls{laa}. The \gls{mv} is identified in an average position according to a rigid alignment with the \gls{la} surface and an average statistical shape atrial model. Although the atrial orifices and anatomical regions are selected in a procedure very similar to ours, the pipeline still requires the manual definition of the appendages. Furthermore, the \gls{fo} and its surrounding ring are not identified, and no method is offered to aid in the segmentation of the atria from \gls{ct} or \gls{mri} scans.
	
	%In the study by Roney et al. \cite{roney2023:_bia_vol}, left and right atrial blood pools were semi-automatically segmented using CemrgApp \cite{RAZEGHI2020100570}. The workflow then progresses by working on a mesh surface representing the endocardial layer of the atria. Each atrial orifice (valves and veins) is manually opened using sphere clipping. Manual intervention is also required to identify the atrial landmarks -- six in the left atrium and six in the right atrium -- to subsequently compute the major anatomical structures. The workflow allows for the generation of either surface or volumetric bilayer models, depending on user preference. For surface models, the epicardial surface is defined using statistical shape models, and the endocardial and epicardial surfaces are then extruded to a prescribed thickness to define the bilayer model. For volumetric models, the space between the two surface walls is meshed, prescribing a suitable mesh resolution for \gls{rd} models. Laplace-Dirichlet solvers are used to define major endocardial anatomical structures, including \gls{pm}, \gls{ct}, and \gls{bb}, however lacking the representation of the \gls{fo}. Depending on user choice, either ruled-based or atlas fibers are implemented based on the previously computed \glspl{uac}. The entire workflow was tested on 100 instances of a \gls{ct}-derived statistical shape model. Although the authors investigated the potential of the workflow on a larger cohort than presented in this paper, the 100 atrial models produced were computed using statistical shape models, which may present less variability than a real cohort of subjects/patients -- e.g., all models presented four pulmonary veins and nicely shaped atrial appendages. Furthermore, the workflow does not present a fully automated segmentation method and requires a certain degree of manual intervention from the user, impairing reproducibility.
	
	%Similarly, in the study by Nagel et al. \cite{nagel2021:_biatrial_ssm}, bi-atrial models were also generated using a statistical shape model. In their approach, five different mesh generation and manipulation tools were integrated into a workflow for generating biatrial meshes. 
	%Their approach, in its current form, cannot be used for biophysical \gls{ep} simulations as all models have a lower resolution of approximately \SI{1.5}{\milli\meter} and suffer from severe mesh quality issues. The explicit representation of inter-atrial connections led to degenerated elements, which impede the use of such models for any \gls{rd} simulations.
	
	
	% Pros of our approach -> reads more like a conclusion, 
	%and most of this we say in the intro to the Discussion
	\subsection{Performance considerations}
	The scalability of model generation methods depends on two major factors,
	the degree of automation to limit time-consuming interactive processing,
	and the robustness and speed of the processing steps, from medical image to final mesh.
	Our workflow achieves the highest degree of automation reported so far, 
	with minimal to no interactive processing at all. 
	Using a \gls{scn} \citep{thaler2021efficient}
	all \gls{ct} data sets were automatically segmented within $<$\SI{10}{\second}.
	As all required labels were detected robustly and accurately from all contrast CT datasets used in this study, 
	with no need for manual label correction.
	The additional landmarking of veins and appendages can be fully automated, requiring minimal user input to verify or, if necessary, correct the auto-generated labels. This process includes an interactive step supported by a custom tool designed for quick label correction, reducing the overall operation time to less than \SI{1}{\minute}.
	For all subsequent steps, including the generation of the volumetric atrial walls,
	the assignment of a rule-based fiber architecture and the computation of \glspl{uac},
	full automation has been achieved. 
	Performance measured over 50 models yielded execution times of only $<$\SI{10}{\minute} and $<$\SI{30}{\minute}
	for generating an atrial model at mesh resolutions suitable for \gls{re} and \gls{rd} simulations, respectively.
	%Additionally, the workflow is flexible, allowing for the generation of just one of the two chambers instead of a complete bi-atrial model.
	
	%ensuring the computation of a high-quality mesh for each prescribed resolution. This makes our model suitable for \gls{ep} simulations using either \gls{re} or \gls{rd} models, as well as for mechanical and electromechanical simulations \cite{}. 
	%the landmarking process to identify atrial orifices and major anatomical structures is completely automated and does not require any manual intervention. Moreover, we provide a very detailed representation of the major anatomical structures of the atria, 
	%including both the fossa ovalis and its surrounding ring. The workflow is tested on a cohort of 50 \gls{ct} scans obtained from AF patients, encompassing a large cohort of highly variable anatomies.
	
	
	\subsection{Mesh quality}
	% building volume on the image stack is a discriminating factor
	A common limitation of most previous approaches is the mesh quality of the generated models. To evaluate this, we randomly sampled publicly available datasets of biatrial anatomies and found degenerated elements in all the sampled meshes \citep{nagel2021:_biatrial_ssm,roney2023:_bia_vol}.
	These deficiencies went unnoticed as all studies employed \gls{rd} monodomain models
	which cope well with poor mesh quality.
	However, employing these meshes for higher fidelity bidomain-based \gls{egm} generation is often challenging, 
	since it requires the solution of sensitive elliptic problems, whose solution may be associated with slow convergence or even solver divergence in the presence of degenerated elements when employing classical solvers \citep{conley2016overcoming,schneider2018decoupling}. 
	% as solving elliptic problems is sensitive, with degenerated elements leading to solver divergence. 
	Mesh quality issues and topological errors may arise from volumetric mesh construction methods that depend on the direct extrusion of the atrial walls from the mesh manifolds, \citep{azzolin2023:_augmenta,nagel2021:_biatrial_ssm}. In contrast, our approach is more robust, as the atrial wall volume is generated directly from the image stack in a volumetric manner. This ensures that the initial volumetric representation of the atrial myocardium is topologically sound.
	Although our approach is less prone to topological issues, mesh topology errors may still occur at later stages, such as during surface smoothing operations aimed at removing jaggedness from the voxel-based grid or during remeshing to align mesh resolution with a prescribed target. To enhance robustness, a set of topological clean-up operations is repeatedly applied after each critical meshing step. All 50 models in this study were meshed fully automatically, achieving element quality exceeding a threshold of 0.99, which has been empirically identified as critical for cardiac bidomain simulations.
	
	\subsection{Universal Atrial Coordinates} 
	
	Anatomical reference frames are key for scalable modeling studies,
	as they provide a parametric encoding of all spatial model properties
	and, thus, facilitate an unattended parameter manipulation as required for parameter sweeps \citep{bayer2018universal,roney2019uac}.
	Accurate and robust mapping based on such reference frames relies on features such as
	geometric linearity, that is, isolines are evenly distributed in space,
	and uniqueness to ensure that any parameter set encodes a unique location in space. 
	For atrial models, the first universal spatial reference system has been reported 
	for manifold models by \cite{RONEY201965}.
	This approach was modified through anatomical normalization \citep{roney2021constructing} and later extended to volumetric models \citep{roney2023:_bia_vol}. The volumetric extension relies on \glspl{uac} defined on endocardial and epicardial manifold surfaces, which serve as boundary conditions for solving a \gls{ld} problem on the volumetric mesh to achieve transmural interpolation. Additionally, by solving another \gls{ld} problem, a normalized transmural distance field is computed.
	
	While this approach is straightforward to implement, 
	the generated volumetric \glspl{uac} were of moderate quality.
	Shortcomings of this approach include manual selection of initial landmark on the \gls{la} and \gls{ra} surface, which impedes a fully automated computation of the \gls{uac}, and the construction of geodesic path to define interfaces between lateral and septal, and anterior and posterior regions of the atria, to be used as boundary conditions for \gls{ld} solves, which does not guarantee the consistent identifications of anatomical structures and reference point locations across different geometries. Moreover, \glspl{uac} are always initially built and normalized over surface biatrial models and later extended to volumetric tissue by linear projection on the transmural mesh nodes, thus impairing a sound definition of transmural interfaces and \gls{uac} distribution. Lastly, the volumetric \glspl{uac} were not tested across meshes with different resolutions.       
	%In \cite{RONEY201965}, the authors introduced a set of two \glspl{uac}, $\alpha$ and $\beta$ for surface biatrial models, representing a lateral-to-septal description and a posterior-to-anterior coordinate, respectively, for the \gls{la}, and a \gls{ivc}-to-\gls{svc} description, and a lateral-to-septal one, respectively, for the \gls{ra}. 
	% Do not explain what Caroline did, this is in her study, focus only on
	% what were the strengths (e.g. ease of implementation) and the weaknesses ()
	% --------------
	% 
	% The following text is an explanation of the method as described in the paper,
	% there is no need for recapitulating this here.
	% -----
	%The definition of such coordinates was based on the manual selection of three landmark points on the \gls{la} surface and of two points on the \gls{ra} surface. Landmark points on the \gls{la} are located at the junction of the \gls{lspv} with the atrial body, the junction of the \gls{rspv} with the atrial body, and near the \gls{fo}, while the two on the \gls{ra} at the junctions of the \gls{svc} and \gls{ivc} with the atrial body. To divide the \gls{la} into anterior and posterior regions, three boundaries were identified: the geodesic path between the \gls{lspv} and \gls{rspv} landmarks, the geodesic path between the \gls{rspv} and \gls{mv} passing through the \gls{fo}, and the geodesic path between the \gls{lspv} and the \gls{mv}. The \gls{laa} was subsequently assigned to the anterior portion of the \gls{la}. Similarly, the \gls{ra} was divided into lateral and septal parts by computing geodesic paths between the \gls{ivc} and \gls{svc} landmarks, and between these landmarks and the \gls{tv}. Two sets of \glspl{uac} coordinates were then calculated using Laplace-Dirichlet solvers with prescribed \glspl{dbc} along the defined paths. Additionally, an initial attempt was made to normalize the \glspl{uac} across different geometries by selecting a single geodesic path for each atrium, running posterior-anterior on the \gls{la} wall and lateral-septal on the \gls{ra}. This geodesic path was utilized as a map between the \glspl{uac} solutions and the normalized geodesic distance. The map was then applied to the entire \glspl{uac} system. The proposed method was, however, designed only over biatrial surfaces. Furthermore, the manual selection of landmark points and the definition of geodesic paths between these landmarks did not guarantee the consistent identifications of anatomical structures and reference point locations across different geometries. Lastly, the choice of paths for coordinate rescaling was not well defined.
	% Again, the following text is a recapitulation of Caroline's paper
	% ---
	%A more detailed description of the scaling process was provided in \cite{roney2021constructing}. In this follow-up work, the authors extended the scaling of the \glspl{uac} to include additional anatomical structures. Four points were uniquely identified for each opening, including pulmonary veins, \gls{laa}, \gls{svc}, \gls{ivc}, \gls{cs}, and \gls{raa}. Such points were mesh nodes for which the $\alpha$ and $\beta$ \glspl{uac} had minimum and maximum values. A series of geodesic paths connecting the selected points were computed, along with \gls{uac} isolines whenever the $\alpha$ and $\beta$ coordinates of a selected node were equal to 0 or 1. To ensure rescaling, these paths were projected onto horizontal and vertical lines. The \gls{laa} and \gls{raa} were normalized by solving Laplace-Dirichlet models. 
	% This is some text you may use as this describes a weakness 
	% ----
	% \todo[inline]{May be used, describes a weakness}
	% While this new normalization method allowed for a more consistent selection of corresponding anatomical structures across different geometries, the initial selection of landmark points remained manual, and the \glspl{uac} method was still applicable only to surface meshes.
	% Already described above now
	% ---
	% The \glspl{uac} system was finally extended to volumetric meshes in \cite{roney2023:_bia_vol}. In this work, the authors first utilized the method presented in \cite{roney2021constructing} to compute the \gls{ra} and \gls{la} $\alpha$ and $\beta$ coordinates, and scale them, on the endo- and epicardial surfaces. The corresponding quantities are then projected onto the transmural mesh nodes by solving Laplace-Dirichlet problems.
	%Additionally, a third \gls{uac}, $\gamma$, was introduced as a measure of wall transmurality. 
	% \todo[inline]{The following text you may use as this describes a shortcoming}
	% However, it is unclear how the geodesic boundary paths defined on the endo- and epicardial surfaces to divide the \gls{la} in posterior-anterior parts, and the \gls{ra} in lateral-septal parts,         
	% were projected into the volumetric mesh. Furthermore, the scaling process was conducted only on the endo- and epicardial surfaces, and thus failed to account for variations in \glspl{uac} within transmural nodes. Lastly, the volumetric \glspl{uac} were not tested across meshes with different resolutions.
	% \todo[inline]{Text may be used as you describe here how we overcome limitations, it is too long though.}
	
	Our approach overcomes these limitations by utilizing automatically selected anatomical landmarks and paths, and        
	directly implementing \glspl{uac} on volumetric meshes. We replace the anatomical paths previously used to divide the \gls{la} into posterior-anterior parts and the \gls{ra} into lateral-septal parts with well-defined volumetric interfaces that separate not only the endo- and epicardial surface regions, but also the transmural mesh nodes. The computation of these interfaces is fully automated, based on pre-labeled anatomical regions and a set of rules to define reproducible points across anatomies. The resulting interfaces therefore ensure consistent identification of anatomical structures across different anatomical models. Moreover, they represent a well-defined transmural boundary for \gls{ld} computation, thus avoiding previous ambiguity in the distribution of the transmural coordinate.
	
	% The interfaces are defined as the intersections between the atrial volumes and the quadrilateral surfaces connecting the selected points.
	% The \gls{uac} generation method included in our workflow is directly implemented on volumetric biatrial grids and utilizes anatomical landmarks that are automatically selected through a rule-based approach within the workflow itself. 
	% This ensures consistency across different atrial geometries while accounting for transmural anatomical variations, thereby overcoming issues presented in previous work. 
	% Our \gls{uac} system comprises four coordinates: the $\alpha$ and $\beta$ coordinates, as presented in \cite{RONEY201965}; the transmural coordinate $\gamma$ defined in \cite{roney2023:_bia_vol}; and an additional binary coordinate to distinguish the \gls{ra} from the \gls{la}. 
	% We replace the anatomical paths previously used to divide the \gls{la} into posterior-anterior parts and the \gls{ra} into lateral-septal parts with well-defined interfaces that separate not only the endo- and epicardial surface regions but also the transmural mesh nodes. The computation of these interfaces is fully automated, based on pre-labeled anatomical regions and a set of rules to define reproducible points across anatomies. The interfaces are defined as the intersections between the atrial volumes and the quadrilateral surfaces connecting the selected points. The scaling of our \gls{uac} system is based on solving a linear elasticity problem to align the \gls{ripv}, \gls{lipv}, and \gls{cs} into common reference positions. For this, we utilized the method proposed in \cite{roney2021constructing} to define four reference points on the \gls{ripv}, \gls{lipv}, and \gls{cs}. The four points are then used to parameterize the vein and \gls{cs} openings. The linear elasticity model is solved in a 3D space, accounting for atrial transmurality and potential anatomical variations in the biatrial volumes.
	
	We demonstrate the robustness of our method for computing \glspl{uac} by testing it on 50 biatrial geometries from real \gls{af} patients. Additionally, we show the applicability of our method across different mesh resolutions. The computation of our \glspl{uac} is efficient on meshes suited for the \gls{re} model, though it requires longer computational times when applied to finer grids appropriate for \gls{rd} models. The majority of the computational costs were attributed to solving the linear elasticity problem to normalize the \gls{uac} solutions. This was mainly due to the poorly deformed elements that could potentially result from the projection of the Laplace-Dirichlet solutions on the \gls{uac} subspace, and hinder the solver convergence. Enhancing the computational efficiency of \glspl{uac} for finer meshes will be a focus of future work.
	% \textcolor{red}{Moved over from the Results, could also go into the limitations: \\
	% .}
	
	%\todo[inline]{We should say a word on our transmural solution, do we preserve a bi-layer endo-epi structure?}
	
	% -------------------------
	\subsection{Modeling of inter-atrial connections}
	% {\color{red}Highlight that more interatrial connections can be easily added to the one here considered. Highlight that easy mapping of the interatrial connections between different models.}
	%The issue of representing \glspl{ic} across the atria has been addressed previously in a number of works, with different levels of detail.
	Electrical conduction between the atria is mediated by discrete \glspl{ic}
	which play an important \gls{ep} role.
	During normal sinus rhythm, the activation of the \gls{la} is mediated by \glspl{ic}, 
	% where the insertion points of the \glspl{ic} on the \gls{la} 
	determining the points of \gls{la} earliest activation which critically shape the mid- to terminal part of the P-wave.
	Further, \glspl{ic} provide pathways for inter-atrial reentrant circuits,  
	which represents a substrate for maintaining atrial flutter tachycardia.
	
	% \cite{sakamoto2005interatrial}, histological study of the interatrial connection system, considering 4 connections: the BB, the CS bandle, an inferior-posterior and a superior-anterior bundle (other than the ring of the fossa ovalis). The inferior-posterior and superior-posterior bundle were positioned in corrspondence of the fossa ovalis. 
	Detailed histological studies identified five electrical main connections 
	between the otherwise electrically isolated atria \citep{sakamoto2005interatrial}:
	the \gls{bb}, the coronary sinus bundle, an inferior-posterior bundle, a superior-anterior bundle, and the muscular ring around the fossa ovalis. 
	These histo-anatomical insights are helpful for providing fundamental anatomical constraints to a model,
	but individual anatomy and \gls{ep} properties may be highly variable.
	For instance, not all \glspl{ic} may be present or electrically active in a given individual,
	the locations of \gls{ra} origin and \gls{la} insertion,
	as well as the conduction properties of the \glspl{ic} in between, may also vary to a significant extent.
	Thus, this leads to different \glspl{ic} dominating inter-atrial activation and, thus, 
	govern the site of early activation of the \gls{la}, or the \gls{ra} for a retrograde atrial activation.
	As such, all factors defining the \glspl{ic} must be considered variable,
	and, thus, flexibility represented to support their inference. 
	
	Three different approaches for incorporating the \glspl{ic} in a biatrial model have been considered previously. These were: 
	i) timed electrical stimuli on the \gls{ra} to mediate inter-atrial conduction, without including any \glspl{ic} physical representation
	% inter-atrial conduction mediated by delivering appropriately timed electrical stimuli \cite{Piersanti2021}, 
	ii) anatomical bridges,  explicitly represented by meshing geometries of assumed \glspl{ic} \citep{krueger2012personalization,loewe2016influence,wachter2015mesh}, 
	and, 
	iii) single ohmic resistor connecting \gls{ra} and \gls{la}        
	% \gls{ra} and \gls{la} were connected by a single ohmic resistor 
	\citep{roney2023:_bia_vol,roney2021constructing}. 
	%Independently of the used approach, previous studies considered the \glspl{ic} anatomy as fixed, 
	%with a constant number of \glspl{ic} and immutable sites of \gls{ic} origin and insertion.
	All these approaches are limited in terms of flexibility and generalizability. 
	Timed stimulation is flexible, but does not represent the mechanism mediating inter-atrial conduction,
	and as such does not generalize to physiological representations such as the atrial flutter.
	Explicit anatomical meshing is of limited flexibility and robustness, 
	as any change in the origin or insertion site requires remeshing.
	This approach has been employed previously in \cite{krueger2012personalization}, 
	where, in absence of \glspl{uac}, complex \emph{ad-hoc} algorithms have been used 
	to parameterize and explicitly \glspl{ic} mesh \citep{loewe2016influence,wachter2015mesh}. 
	While feasible in principle, these approaches are complex and challenging to implement robustly, as the integration of thin inter-atrial strands with 3D atrial walls is highly prone to topological mesh errors.
	To provide an example, the explicit meshing of \gls{bb} used in the study by  \cite{nagel2021:_biatrial_ssm} 
	led to degenerated elements in all models throughout the cohort.
	Finally, fixed resistive 0D coupling, as used in various studies, using bilayer manifold models \citep{roney2023:_bia_vol,roney2021constructing}
	is flexible, but the approach is not suitable for connecting volumetric atria 
	due to dimensionality mismatch between 0D and 3D elements, which may cause source-sink mismatches. 
	In principle, this problem may arise also in manifold models where 0D resistors are used 
	to couple the two 2D surfaces representing \gls{ra} and \gls{la}.
	In the absence of a physiological interpretation of the coupling resistor, 
	viable ranges for which transduction between the atria is feasible have to be found by trial and error \citep{roney2023:_bia_vol}.
	Moreover, a further limitation is the inability to impose physiological constraints on the inter-atrial conduction delays,
	as transduction is rigidly dictated by the membrane time constant and the choice of coupling resistance.
	
	% -- I think that this is fudged together, but do not necessarily want to discuss this
	% -- removed this for now.
	%A mechanistic interpretation justifying the modeling approach and given supporting methodological details 
	%are very sparse, rendering a replication of simulation results challenging.
	
	%The inferior-posterior and superior-posterior bundles were positioned in correspondence with the fossa ovalis. Although this histological approach provides valuable anatomical details, it does not offer a practical method for integrating these structures into computational models for \gls{ep} simulations.
	
	% \cite{krueger2012personalization}The \gls{bb} was electrically isolated from the myocardial tissue, and four additional bridges were manually added using strands of volumetric tissue. Moreover, although the \gls{ic} are anatomically detailed, they are not flexible enough to ensure rapid \gls{ep} model calibration.
	%In \cite{krueger2012personalization}, a high number of \glspl{ic} is added to a detailed biatrial volumetric model, including the \gls{bb}, and a lower anterior, upper, middle and lower posterior \glspl{ic}, as a bundle of volumetric tissue. Although this approach provides detailed anatomical description, it lacks the flexibility for rapid \gls{ep} model calibration, which is crucial for efficient simulations.
	
	% \cite{wachter2015mesh}, report a very complex landmarks system to parametrized the atria model. Basic landmarks around the principal vessels, valves and structures were selected and connected along the atrial walls. More points were then identify a given \% of such root to identify the \gls{ic}s. Iteratrial connections included a superior-posterior bridge, a middle-posterior bridge, the coronary sinus bundles, a middle-anterior bridge (which corresponds to our BB and the BB general description), and an inferior anterior bridge. The entry and exit points were connected with 3D fascicular tissue.
	% \cite{loewe2016influence}, atrial connections follows \cite{wachter2015mesh}, without the inferior-anterior connections. Same structure as above.
	%A complex landmark system to parametrize the biatrial model and facilitate the mapping of major \glspl{ic} has been developed by Wachter et al. in \cite{wachter2015mesh}. They selected basic landmarks around the principal vessels, valves, and structures, which were connected along the atrial walls. More points were identified at given percentages of such roots to pinpoint the interatrial connections (\gls{ic}s). These connections included a superior-posterior bridge, a middle-posterior bridge, the coronary sinus bundles, a middle-anterior bridge, and an inferior anterior bridge. The entry and exit points were connected with 3D fascicular tissue. While comprehensive, this method is highly complex and may be difficult to implement in practice. Similarly, in \cite{loewe2016influence}, the atrial connections followed the methodology of \cite{wachter2015mesh}, excluding the inferior-anterior connections. While this maintains consistency in structural representation, it retains the complexity and potential implementation challenges of the original method.
	
	Our novel approach to incorporating \glspl{ic} in atrial models overcomes these limitations,
	as it is highly flexible, versatile, robust, and computationally efficient,
	and adds \gls{ic} anatomy as a parameter that can be probed during model calibration. 
	%It involves the definition of 3D cables that connect two preselected points across the atria. Our representation is physiologically accurate since the cables remain completely electrically isolated from the myocardial tissue except at their entry and exit sites.
	An arbitrary number of \glspl{ic} can be included to connect arbitrary sites of origin and insertion 
	on \gls{ra} and \gls{la}, respectively, 
	with prescribed conduction velocities governing the activation delay across the \gls{ic}.
	The use of \glspl{uac} to define origin and insertion sites, facilitates a seamless mapping of  \glspl{ic} 
	between different anatomies.
	Further, modeling the \glspl{ic} using a cable formulation computed with the same approach of generating the \gls{hps}, overcomes 
	dimension mismatch problems between 1D and 3D model components. 
	%facilitating also the control of anterograde and retrograde conduction delays when entering or leaving an \gls{ic} \cite{boyle2010purkinje}.
	%Finally, integrating a cable is of negligible computational cost.
	
	
	% \subsection{Dual multi-resolution multi-fidelity strategy}
	% Patient-specific modeling of atrial \gls{ep} comprises three distinct stages
	% which are 
	% i) an anatomical twinning stage where an anatomical model is reconstructed from imaging data, 
	% ii) a functional modeling stage where model parameters are calibrated to minimize the mismatch 
	% to observations, and 
	% iii) a prediction stage at which the calibrated model is used to predict the model's response to a perturbation
	% such as a pacing or an arrhythmia induction protocol \cite{azzolin2021:_vulnerability}, 
	% to a substrate modification by ablation therapy \cite{dasi2024:_ablation_vs_drugs},
	% or the administration of a drug \cite{dasi2022:_in_silico}.
	% Most atrial modeling studies focus on the anatomical twinning and prediction stages,
	% largely using uncalibrated models, skipping the functional twinning stage.
	% A major reason is the vast costs involved in model calibration.
	% Owing to the large parameter space, either a vast number of forward simulations is required 
	% with sampling-based methods \cite{gillette2021:_framework}, 
	% or expensive tailored optimization methods must be employed \cite{grandits2021:_geasi,grandits2023:_geodesic_bp}.
	% In both approaches, the use of high fidelity \gls{ep} forward models, e.g. as \gls{rd} models, is inadequate 
	% as it requires high spatiotemporal resolution and, inevitably, a significant computational cost.
	% % As such, for a comprehensive exploration of a high dimensional parameter space, \gls{rd} models are untractable.
	
	% To resolve these discrepancies in terms of model requirements
	% -- fast forward simulations for functional calibration and physiologically accurate simulations for predictions -- 
	% we propose a dual multi-resolution multi-fidelity strategy.
	% For model calibration, we use an extremely lightweight \gls{re} model \cite{neic17:_reaction_eikonal}
	% where spatio-temporal resolution constraints are relaxed 
	% such that models can be evaluated much faster than high fidelity \gls{rd} models, 
	% up to a factor of $\times$1000 faster, achieving close to real-time performance.
	% This speed combined with an easier calibration of most critical properties 
	% such as conduction velocities, which are directly exposed as model parameters,
	% facilitate a more comprehensive exploration of the parameter space 
	% which can lead to higher fidelity in model calibration \cite{gillette2021:_framework}.
	% For predictive simulations calibrated parameters are seamlessly transferred to a higher resolution \gls{rd} model
	% using an automated framework for matching activation patterns \cite{gsell2024:_forcepss}
	% to ascertain that low and high fidelity models produce the same calibrated outputs.
	% This relies on the ability of the anatomical stage to produce geometrically closely matched meshes 
	% at lower and higher spatial resolutions of about $\approx$ \SI{1}{\milli \meter} and $\approx$ \SI{0.3}{\milli \meter}, respectively.
	
	% It is important to note that the low fidelity \gls{re} model used in here 
	% under conditions such as normal sinus rhythm or pacing can be essentially considered equivalent 
	% to the high fidelity model. As we show (see Fig.~\ref{Fig.RD_vs_RE}), discrepancies between the models are marginal.
	% However, high fidelity of \gls{rd} models may offer better predictive capabilities 
	% for investigating phenomena such as inducibility of re-entrant arrhythmias \cite{azzolin2021:_vulnerability}
	% as current \gls{re} models do not account for all mechanisms at play.
	
	\subsection{Calibration of atrial \gls{ep} models - using the P-wave as objective}
	%Comparison to surrogate markers such as dominant frequency of local activation rates 
	%have been considered as \cite{},
	Atrial \gls{ep} modeling studies aiming to achieve a patient-specific calibration 
	by 1:1 matching with direct observations
	-- \glspl{eam}, \glspl{egm} or the \gls{ecg} -- are rare.
	Rather, adjustments of conductive and cellular dynamics properties informed by literature data
	are implemented to match global metrics such as the total atrial activation time.
	For a more accurate spatial calibration, \gls{eam} activation maps have been used 
	to infer intrinsic tissue conductivities or conduction velocities \citep{lubrecht2021:_piemap}. 
	High density \gls{eam} datasets covering both the \gls{ra} and the \gls{la} 
	appear ideal for calibrating atrial \gls{ep} models, 
	as they may provide a detailed view of the overall atrial activation sequence.
	However, while such datasets can be acquired, in principle,
	they are not widely available, as their acquisition tends to prolong procedures.
	Moreover, the spatiotemporal registration uncertainty between measurements of an \gls{eam} manifold 
	%and the atrial endocardia in the patient 
	and the image-derived model is significant, 
	posing major challenges for calibration procedures. 
	%where observations on two distinct manifolds
	% of uncertain spatial relations must be compared and matched        
	Using \Glspl{egm} from which \gls{eam} maps are derived is even more challenging, 
	as \glspl{egm} provide only a very local view on tissue activation. 
	These may be helpful for inferring tissue characteristics from fibrotic patches,
	e.g.\ \gls{egm} magnitude and temporal separation of fractionated complexes in \glspl{egm},
	correlate with the severity and structure of fibrosis, but do not offer benefits 
	over using \gls{eam} maps directly. 
	
	In principle, the P-wave in the \gls{ecg} appears a most natural choice for calibrating atrial \gls{ep} models,
	as it is abundantly available and can be non-invasively recorded.
	However, the P-wave provides only a global view of the atrial activation sequence, and
	inferring space-varying model parameters from such limited data is challenging.
	Owing to the ill-posed nature of this inverse problem, this may not even be feasible.
	Also challenging but, in principle, feasible, as shown recently for the ventricular activation sequence \citep{gillette2021:_framework,grandits2023:_geodesic_bp},
	is the identification of model parameters that produce atrial activation sequences  
	and replicate, with high fidelity, the P-wave in a clinical standard \gls{ecg}.
	However, there may be more than one, potentially many, parameter sets yielding the same P-wave
	which raises questions of identifiability and uniqueness \citep{grandits2024:_identify}.
	Due to these challenges, only a few attempts have been made to calibrate an atrial \gls{ep} model to the P-wave \citep{loewe2015left}, mostly limited to qualitative visual comparisons between simulated and recorded P-waves, without a quantitative assessment of the differences, as shown in this work. 
	
	Further obstacle to P-wave-based model calibration can stem from incomplete parameterization that fails to accurately produce the actual activation sequence, and insufficient computational performance that hinders a feasible calibration process, especially when relying on traditional \gls{rd} models \citep{deng2012simulation,fedele2023comprehensive,ferrer2015detailed,loewe2015left,nagel2022comparison,roney2023:_bia_vol}
	combined with high fidelity pseudo-bidomain or lower fidelity extracellular potential recovery models \citep{labarthe2014:_bilayer,nagel2021:_biatrial_ssm}, or limited fidelity in the predicted P-wave.       
	% A further obstacle impeding P-wave based model calibration are 
	% the computational cost of simulating the activation sequence and the fidelity of the predicted P-waves
	% associated to it.
	% approaches relying on \gls{rd} models \cite{deng2012simulation,loewe2015left,ferrer2015detailed,fedele2023comprehensive,roney2023:_bia_vol}
	% combined with high fidelity pseudo-bidomain or lower fidelity extracellular potential recovery models \cite{labarthe2014:_bilayer,nagel2021:_biatrial_ssm} 
	% a P-wave based calibration is challenging to implement.
	% Reasons may comprise an incomplete parameterization that is not able to produce the actual activation sequence,
	% insufficient computational performance impeding a calibration in a tractable manner, 
	% or limited fidelity of the predicted P-wave.
	Bilayer models are popular as they mitigate performance issues by reducing the overall problem size, 
	but are still orders of magnitude slower compared to volumetric \gls{re} models as used in this study,
	and the fidelity of predicted P-waves is limited.
	Volumetric biatrial \gls{rd} models immersed in a torso model using a full- or pseudo-bidomain 
	are able to produce high-fidelity P-waves \citep{loewe2015left}.
	They offer the advantage of providing the entire potential field $\phi{\rm e}$ throughout the torso, 
	in which \gls{ecg} or more extensive body surface potential maps are embedded \citep{ferrer2015detailed, ZAPPON2024112815}. 
	However, the approach is computationally by far too expensive for calibration studies, 
	and the additional information on the torso potential field is not readily exploited 
	in the absence of body surface potential mapping data.
	
	Our analysis clearly demonstrates 
	that the \gls{relf} model, combined with an extensive parameter space -- including also the anatomy of the \glspl{ic} -- is able to produce activation sequences
	and associated P-waves with full bidomain fidelity, and at real-time performance.
	As shown previously for a ventricular \gls{ecg} model \citep{gillette2021:_framework}, and in line with atrial \gls{ecg} model \citep{nagel2022comparison},
	% and in this work for an atrial \gls{ecg} model, 
	the \gls{relf} model yields \glspl{egm} and \glspl{ecg} that are not discernible from those 
	produced with a full fidelity \gls{rd} bidomain model, as anticipated on theoretical grounds \citep{geselowitz1989theory}. 
	As shown in Figure \ref{Fig.RD_vs_RE}, differences in activation sequence and P-wave are minor,
	and can be deemed negligible in view of the overall model uncertainties. 
	
	In our study we focused on demonstrating the flexibility and efficiency of our framework 
	in supporting comprehensive automated exploration of physical and geometrical parameter spaces,
	and refrained from attempting to accurately calibrate the atrial \gls{ep} model to the P-wave for various reasons.
	As the atrial model was integrated with torso geometry and electrode positions taken from a different subject,
	and none of the pathologies of the patient treated by AF ablation were considered,
	discrepancies can be anticipated, and the feasibility of achieving a good fit is not even guaranteed. 
	Nonetheless, we quantified differences between simulated and real \glspl{ecg} using the \gls{rmse}, 
	showing that the P-wave can be closely captured in most leads, 
	except aVL, aVR, and V1. 
	P-wave features such as positivity in lead aVL and -aVR, as well as the biphasic characteristic of lead V1 were replicated by the model. 
	
	%Using the fast \gls{relf} model enabled us to run a large set of simulations varying single or multiple parameters automatically.
	Importantly, the fidelity and efficiency of our framework support, in principle,
	a comprehensive exploration of the parameter space spanned by atrial anatomy and \gls{ep},  
	and, thus, provides a basis for building future applications geared towards creating digital twins 
	based on P-wave calibration.
	A full single forward simulation lasted $\approx$ \SI{27}{\second} only 
	where the actual evaluation of the \gls{ep} model amounted only to $\approx$ \SI{3}{\second}.  
	%  \todo[inline]{Suggest to replace the following with data on a single sweep.}
	% By conducting a brute force sweep of \gls{bb} entry sites in the \gls{la}, we obtained more than 50k simulation runs in less than two weeks, including the computation of \gls{bb} cables for each simulation and the calculation of cardiac sources. Furthermore, over 8k simulations were completed in less than three days. As this sampling process is a parallel computing problem, the parameter search could be further accelerated by utilizing HPC clusters.           
	
	
	%Our workflow enabled the use of all \gls{rd} and \gls{re} models for computing cardiac sources, as well as all the aforementioned approaches for \glspl{ecg} computation, without exception. In this work, we chose to highlight the potential of the \gls{relf} model for representing cardiac activation and generating the corresponding \glspl{ecg}, as it is the most efficient cardiac-torso combination among those available and previously discussed. By comparing the transmembrane potential and activation maps computed with the \gls{re} and monodomain models, we demonstrate the ability of the \gls{re} model to maintain the accuracy of the \gls{ep} solution in biatrial simulations. Compared to other studies that employ a similar approach for atrial \gls{ep} \cite{gillette2022personalized}, our framework allows for a more detailed and complex description of atrial anatomical and functional structures, including a higher number of physiologically realistic \glspl{ic}. Our numerical results highlight the \gls{re} model capability to provide a stable, and reliable solution even when such complexity is introduced into the \gls{ep} problem. 
	
	%Additionally, the generation of 12-lead \glspl{ecg} using both the coupled monodomain-lead field and \gls{relf} approaches demonstrates that \gls{relf} reliably reconstructs \gls{ecg} signals, with negligible amplitude variability compared to the typical observational and residual uncertainties associated with \gls{ecg} recordings.   
	
	
	%Moreover, in \cite{gillette2021:_framework,gillette2022personalized}, the lead filed has been demonstrated to be suitable for model calibration of biventricular and whole heart geometries when coupled with the \gls{re} model, while proving beneficial in terms of computational costs compared to more traditional approaches, such as the Pseudo-bidomain \cite{loewe2016influence,ZAPPON2024112815} or the $\phi{\rm e}$-recovery \cite{ferrer2015detailed}. 
	
	
	
	%It is important to note that the low fidelity \gls{re} model used in here 
	%under conditions such as normal sinus rhythm should be considered equivalent to the high fidelity model.
	%As we show (see \ref{}, discrepancy between the models is marginal.
	
	
	
	
	
	%\subsection{Assessing the approximation error of reaction-eikonal model}
	%Various modeling approaches are available for representing atrial electrophysiology and generating P-waves, each differing in biophysical accuracy and computational cost. The most common approach involves using a \gls{rd} monodomain model to represent cardiac sources \cite{deng2012simulation,loewe2015left,ferrer2015detailed,fedele2023comprehensive,roney2023:_bia_vol}, combined with a $\phi{\rm e}$-recovery technique \cite{labarthe2014:_bilayer,ferrer2015detailed} or a pseudo-bidomain model \cite{loewe2015left,loewe2016influence,ZAPPON2024112815} for \gls{ecg} computation.
	%However, a significant drawback of the \gls{rd} modeling approach is its dependence on mesh resolution, which typically must be on the order of \SI{0.4}{\milli \meter} or smaller. This requirement leads to highly detailed mesh grids, resulting in vast computational costs, that are further increased by using biophysically detailed models of cellular dynamics as reaction terms \cite{plank2008:_ecme}. As such, calibration of \gls{rd}-based models is notoriously difficult since a vast number of expensive forward simulations is required to minimize the discrepancy between model prediction and observed data. 
	
	%Therefore, atrial personalization for digital twinning could be more effectively achieved by employing lightweight modeling approaches, such as the \gls{re} formulation \cite{neic17:_reaction_eikonal}. Due to the independence of solution accuracy from spatial resolution, \gls{re} models offer a significant advantage in computational speed, with performance gains ranging from approximately 100 to over 1000 times faster, depending on the specific implementation. Additionally, \gls{re} models allow for easier calibration of cardiac potential propagation properties, such as conduction velocity, as these are directly exposed as model parameters. Since \gls{re} models can be implemented to achieve real-time performance, evaluating millions of forward simulations becomes feasible, enabling a much more detailed exploration of the parameter space. This can lead to higher fidelity in model calibration \cite{gillette2021:_framework}. Differences between the \gls{re} and \gls{rd} mono- and bidomain solutions for biventricular simulations have been previously analyzed in \cite{neic17:_reaction_eikonal,gillette2021:_framework}. Both studies demonstrate that the activation sequences generated by the \gls{re} and \gls{rd} models are closely matched and that the action potential morphology and repolarization patterns computed with the \gls{re} and \gls{rd} models are in good agreement.
	
	%We carried out the \gls{ecg} computation by solving the lead field approach, which has been shown in \cite{geselowitz1989theory} to yield the same extracellular electrograms as a full \gls{rd} bidomain model. Moreover, in \cite{gillette2021:_framework,gillette2022personalized}, the lead filed has been demonstrated to be suitable for model calibration of biventricular and whole heart geometries when coupled with the \gls{re} model, while proving beneficial in terms of computational costs compared to more traditional approaches, such as the Pseudo-bidomain \cite{loewe2016influence,ZAPPON2024112815} or the $\phi{\rm e}$-recovery \cite{ferrer2015detailed}. 
	
	%        Other \gls{ecg} studies for atrial \gls{ep} have been carried out using a pseudo-bidomain formulation \cite{loewe2015left,ferrer2015detailed}. The pseudo-bidomain approach offers the advantage of computing the entire $\phi{\rm e}$ field in the torso domain, thus allowing for the direct computation of other torso-related outputs beyond the \gls{ecg}, such as body surface potential maps \cite{ferrer2015detailed, ZAPPON2024112815}. 
	%While the lead field method requires solving a Neumann-Laplace problem for each electrode to compute the \gls{ecg}, the pseudo-bidomain approach involves solving a single Dirichlet-Laplace problem. The \gls{ecg} is then obtained by sampling the pseudo-bidomain solution at torso mesh nodes corresponding to the desired electrode locations. However, this Dirichlet-Laplace problem must be solved at each time step of the \gls{ecg} generation.
	%Although the computational cost of the lead field solution $Z_i$ is non-negligible, taking seconds or minutes depending on the mesh resolution, it is calculated only once during the initialization of the \gls{ep} simulation and only for a limited number of electrodes $i$, based on the selected \gls{ecg} configuration. Moreover, our implementation \cite{gillette2021:_framework} allows for storing the $Z_i$ solution and reusing it whenever the same torso mesh and cardiac and torso conduction properties are employed. The lead field approach is therefore particularly efficient in multi-simulation contexts, such as exploring \gls{ecg} variability based on variations in cardiac \gls{ep} parameters.
	
	%       Our workflow enabled the use of all \gls{rd} and \gls{re} models for computing cardiac sources, as well as all the aforementioned approaches for \glspl{ecg} computation, without exception. In this work, we chose to highlight the potential of the \gls{relf} model for representing cardiac activation and generating the corresponding \glspl{ecg}, as it is the most efficient cardiac-torso combination among those available and previously discussed. By comparing the transmembrane potential and activation maps computed with the \gls{re} and monodomain models, we demonstrate the ability of the \gls{re} model to maintain the accuracy of the \gls{ep} solution in biatrial simulations. Compared to other studies that employ a similar approach for atrial \gls{ep} \cite{gillette2022personalized}, our framework allows for a more detailed and complex description of atrial anatomical and functional structures, including a higher number of physiologically realistic \glspl{ic}. Our numerical results highlight the \gls{re} model capability to provide a stable, and reliable solution even when such complexity is introduced into the \gls{ep} problem.         
	%      Additionally, the generation of 12-lead \glspl{ecg} using both the coupled monodomain-lead field and \gls{relf} approaches demonstrates that \gls{relf} reliably reconstructs \gls{ecg} signals, with negligible amplitude variability compared to the typical observational and residual uncertainties associated with \gls{ecg} recordings.   
	
	%      Additionally, we demonstrate the flexibility and efficiency of our framework in exploring physical and geometrical parameter spaces. The fast \gls{relf} model, combined with a versatile implementation of the \gls{ep} model and \glspl{ic}, enabled us to run a large set of simulations varying single or multiple parameters automatically. By conducting a brute force sweep of \gls{bb} entry sites in the \gls{la}, we obtained more than 50k simulation runs in less than two weeks, including the computation of \gls{bb} cables for each simulation and the calculation of cardiac sources. Furthermore, over 8k simulations were completed in less than three days. As this sampling process is a parallel computing problem, the parameter search could be further accelerated by utilizing HPC clusters.           
	
	
	\subsection{Role of left and right atrial electrophysiology in the genesis of the P-wave}
	Despite the vast acceleration in speed achieved with high fidelity \gls{ep} forward models, such as the \gls{relf},
	% that facilitates an efficient comprehensive exploration of the high dimensional parameter space,
	the use \emph{a priori} knowledge on the genesis of the P-wave is key to constrain the inference 
	by limiting the admissible parameter space and the potentially large number of different atrial activation sequences
	that produce the same P-wave.
	
	From a macroscopic perspective, the genesis of the P-wave is well understood.
	However, at a mesoscopic size scale, that is  
	the relation between the physics of depolarization wavefronts traversing individual atrial structures,
	and their relative contribution to the global P-wave, is still unclear.
	% this is the case to a much lesser extent. 
	These aspects have been investigated only in a limited number of studies  \citep{ferrer2015detailed,loewe2016influence,loewe2015left}.
	In \cite{loewe2015left}, the separate contribution of \gls{ra} and \gls{la} depolarization 
	to the P-wave was investigated in two atrial and torso models derived from healthy subjects. 
	%For both subjects, cardiac sources were predefined to ensure that the resulting \gls{ecg} was qualitatively comparable to corresponding clinical data. The role of the \gls{ra} and \gls{la} in activation was then examined by recording the \gls{ecg} with activation confined to either the right or left atria. 
	Using a pre-defined activation sequence, the study showed 
	that the \gls{ra} predominantly influences the P-wave in precordial leads V1 and V2, 
	as well as limb leads II, aVF, and III, while the \gls{la} governs the central to terminal portion of the P-wave. %However, no variations in the cardiac sources were accounted for in their study.
	The effect of varying the atrial activation sequence on P-wave features such as positivity or negativity in different leads was investigated in \cite{loewe2016influence} 
	in eight biatrial models of healthy subjects by initiating atrial activation at a fixed set of locations 
	representing the \gls{san} exit sites.
	%from an anterior-superior to a posterior-inferior position within the \gls{ra} across eight predefined locations. 
	%The \gls{ra} was then activated at the \gls{san} with the same electrical impulse for each selected location. 
	%and qualitatively analyzing its impact on P-wave positivity  or negativity in different leads.
	%of leads I and II, the amplitude of lead II, and the negativity of lead aVR. 
	%The study also provided a qualitative and quantitative assessment of the contributions of \gls{ra} and \gls{la} activation to the P-wave. 
	%However, the analysis was limited by the fixed number of \gls{san} locations, and the lack of consideration for other influential parameters, such as atrial conduction velocities. 
	%Furthermore, variation of shape and dimension of the electrical impulse activating the \gls{san} was not mentioned. 
	The relative contributions of different atrial regions to the P-wave was moreover analyzed in \cite{ferrer2015detailed},
	for a fixed atrial activation sequence, with prescribed \gls{san} exit sites and \gls{ic} locations.
	
	%. These effects were investigated by fixing the cardiac activation setting and activating individual anatomical regions in the biatrial model. However, no variations in the locations of the \glspl{ic} were explored, nor were the major factors affecting \gls{ra} activation, such as conduction velocities in the atrial body and the shape and location of the \gls{san}.
	
	In our study, we investigate the envelope of P-waves associated with changes in both 
	\gls{ra} and \gls{la} anatomical and physiological factors, including the location and number of early activation sites in the \gls{la}, the position and shape of the \gls{san}, and the conduction velocity in the fiber and sheet directions of the \gls{ra} body. 
	By separately analyzing the \gls{ecg} variations due to these factors, we also highlight the roles of \gls{ra} and \gls{la} activation in shaping the P-wave.
	
	The generated P-wave envelop did not fully cover the P-wave observed in the modeled patient,
	most evident in the leads aVL and aVR (refer to Figure \ref{Fig.Fig_SAN_sampling}). 
	The terminal phase of the P-wave is not well covered, most prominently visible in the more lateral precordial leads V3-V6,
	and in the leads aVR and II.
	This discrepancy may stem from various factors. Firstly, the model was built from a patient treated by AF ablation,
	with a significant fibrotic burden, which remained unaccounted for in our model, and was instead assumed to be structurally healthy. 
	As such, the dipoles at the interface between healthy atrial tissue and fibrotic patches are missing,
	leading to a nearly zero dipole once both atria were fully activated.
	Finally, torso and lead positions were taken from another subject and were not specific to the given subject.
	
	\subsubsection{Discrete interatrial conduction}
	The site of earliest activation of the \gls{la} is governed by the \gls{ra} activation sequence
	and the location and conductive properties of the \glspl{ic}.
	In this study, we only investigated the effect of varying location and size of the insertion site
	of \gls{bb} on the anterior wall of the \gls{la} upon the P-wave. 
	The use of bundles for modeling \glspl{ic} facilitates automated sweeps over these important parameters
	that govern the coupling spatio-temporal coupling of \gls{ra} and \gls{la}.
	% This is notoriously difficult to achieve with explicit mesh-based representations of \glspl{ic} \cite{nagel2021:_biatrial_ssm}
	% or with simple timed stimuli which require the detection of an activation at the entrance of an \gls{ic} in the \gls{ra}, 
	% the calculation of a conduction time through an \gls{ic} of an assumed length and conduction velocity \cite{Piersanti2021}, 
	% to mimic the effect of an \gls{ic}.
	Consistent with \cite{loewe2015left}, our simulations show 
	that \gls{la} activation influences only the mid to terminal portion of the P-wave (see Figure \ref{Fig.Tests_1BB}).
	%but in contrast to previous work, we demonstrate that the location of \gls{bb} entry sites affects the entire second half of the P-wave. 
	Major differences due to varying \gls{bb} insertion in the \gls{la} were witnessed in leads aVL, I, aVR, and III, 
	while effects in other limb leads and precordial leads were minor.
	Using three \glspl{ic} to model a fan-like insertion of the \gls{bb} into the anterior \gls{la} 
	had led only to a minor spread in the P-wave envelope (see Figure \ref{Fig.Three_bb}).
	%Greater variability in the P-wave is observed when varying the location of the \gls{san} along the \gls{ct}, its shape and dimensions, and the conduction velocities in the \gls{ra} body. Our analysis suggests that the early activation site and shape in the \gls{ra}, as well as its conduction velocities, have a major influence on the entire P-wave. This is expected, as these factors determine not only the depolarization of the \gls{ra} but also the activation of the \gls{la}, which in a healthy subject is completely driven by conduction through the \glspl{ic}, which are themselves activated depending on the \gls{ep} signal traveling through the \gls{ra} tissue. These findings are in agreement with \cite{loewe2016influence,ferrer2015detailed}.  
	
	%\subsection{Baseline simulation setup for sinus rhythm}
	%As part of the atrial modeling framework, we provide a baseline parameter setting for simulating sinus rhythm cardiac \gls{ep} and generating physiological 12-lead \glspl{ecg}. The proposed activation begins at the \gls{san}, which is automatically located by the workflow at the junction of the \gls{svc} and the \gls{ct}. The activation then propagates rapidly posteriorly through the \gls{ct} and \glspl{pm}, and anteriorly through the lateral bands of the \gls{bb}. As the cardiac potential travels through these cables, it initiates the activation of the \gls{la} anteriorly, and posteriorly at the superior and inferior junctions of the \gls{ripv} with the \gls{la} body via the posterior \glspl{ic}, and near the \gls{mv} through the \gls{cs} \gls{ic}. Fiber orientations and conduction velocities in various anatomical regions are computed and selected based on previous studies \cite{Piersanti2021,azzolin2023:_augmenta} and physiological findings \cite{gray1996incomplete,lemery2007normal,deng2012simulation}. The resulting biatrial \gls{ep} activation and the generated \gls{ecg} align with several previous publications \cite{lemery2007normal,deng2012simulation,labarthe2014:_bilayer,loewe2015left}.
	
	
	\subsubsection{Sino-atrial node}
	The \gls{san} was originally described as a crescent-shaped area 
	at the junction of the \gls{svc} and the \gls{ra} \citep{anderson1983surgical,keith1907form}, and later found to extend along the \gls{ct}, and toward the \gls{ivc} \citep{akima1978method,anderson1978development,boyett2000sinoatrial,csepe_human_2016, fedorov_conduction_2012,monfredi2010anatomy}. 
	Normal \gls{san} automaticity depends on its depolarized resting potential 
	of $\approx$ \SI{-60}{\milli\volt} relative to neighboring atrial tissue,
	with a more hyperpolarized resting potential of $\approx$ \SI{-85}{\milli\volt} \citep{csepe_human_2016}.
	To prevent hyperpolarization of \gls{san} and, thus, inhibition of automaticity,
	the \gls{san} is electrically mostly insulated to safeguard pacemaker function \citep{fedorov_conduction_2012,joyner_propagation_1986}.
	Electrical coupling of the \gls{san} with surrounding atrial myocardium 
	is limited to specific discrete sites,referred to as \gls{san} exit pathways, 
	which can number up to five \citep{fedorov_optical_2010,li_redundant_2017}.
	As such, the location of exit pathways and their activity 
	determine the earliest atrial activation site and the shape of the initial depolarization wavefront, 
	which not only affects the activation sequence of the \gls{ra}, 
	but may also influence the order of activation of the \glspl{ic}, and, in turn, alter the earliest activation site on the \gls{la} \citep{antz_electrical_1998,li_redundant_2017}.
	Consequently, \gls{san} anatomy and exit pathways are crucial parameters to consider 
	when modeling atrial activation and its reflection in the P-wave.
	
	In modeling studies, the \gls{san} is often represented as a spherical region at a fixed location  \citep{fedele2023comprehensive,ferrer2015detailed,loewe2016influence}, 
	with only a few studies accounting for the anatomical shape of the \gls{san}         \citep{gillette2022personalized,labarthe2014:_bilayer}.
	% However, the activation of the \gls{san} 
	% and the spatial spread in the potential locations of \gls{san} exit sites indicates 
	% a significant variability in the site initiating atrial activation \cite{loewe2016influence,ferrer2015detailed}.    
	In line with previous work \citep{ferrer2015detailed,loewe2016influence}, here we demonstrate 
	a marked dependency of the P-wave upon exit sites (refer to Figure \ref{Fig.Fig_SAN_sampling}). By easy variation of \gls{san} size and location, we show that our framework may easily allow for inferring both \gls{san} anatomy and exit sites based on \gls{ecg} calibration.
	% In our framework high fidelity P-waves can be efficiently predicted 
	% to support, in principle, the inference of both \gls{san} anatomy and exit site.
	
	
	%As such, the \gls{san} anatomy and exit sites can be considered key parameters 
	%influencing the overall shape of the P-wave. 
	
	%While the cell composition of the \gls{san}, its connection with the \gls{ra} myocardium, 
	%and the mechanisms by which the electrical impulse exits the \gls{san} are still not fully %understood %\cite{monfredi2010anatomy,fedorov2009structural,sanchez2005sinus,dobrzynski2005computer}.
	%It has been shown that the \gls{san} maintains electrical coupling 
	%with the surrounding atrial myocardium through up to five specific sites 
	%referred to as \gls{san} exit pathways \cite{li_redundant_2017, fedorov_optical_2010}.
	%it has been demonstrated that several exit sites at discrete locations exist.
	%\todo[inline]{Moh, please help us out here with your wisdom!}
	
	%In modeling studies the \gls{san} is often represented by as a spherical region at a fixed location  \cite{loewe2016influence,ferrer2015detailed,fedele2023comprehensive}, 
	%with only a few studies accounting for \gls{san} shape as an elongated cylindrical region
	% \cite{labarthe2014:_bilayer,gillette2022personalized}.
	%However, the activation of the \gls{san} 
	%and the spatial spread in the potential locations of \gls{san} exit sites indicates 
	%a significant variability in the site initiating atrial activation \cite{loewe2016influence,ferrer2015detailed}. 
	%This does not only alter the activation of the \gls{ra}, 
	%but may also influence the order of activation of the \glspl{ic},
	%which, in turn, will alter the earliest activation site on the \gls{la}.
	%As such, the \gls{san} anatomy and exit sites can be considered key parameters 
	%influencing the overall shape of the P-wave. 
	
	%In computational \gls{ep} modeling, this complexity poses the \gls{san} shape and position as variables within a space-varying parameter set, with the potential to significantly alter the overall shape of the P-wave. 
	
	% MOH's text, integrated in the text above
	%\todo[inline]{Finish integrating Moh's stuff from below}
	%The human SAN is a three-dimensional crescent-shaped structure located at the junction of the superior vena cava, extending along the crista terminalis toward the inferior vena cava \cite{csepe_human_2016, fedorov_conduction_2012}.
	
	%Normal SAN automaticity depends on its relatively depolarized resting potential of approximately -60 mV, in contrast to the neighboring atrial tissue, which rests at a more hyperpolarized potential of around -85 mV \cite{csepe_human_2016}. While this difference in resting potential could theoretically cause hyperpolarization of SAN pacemaker cells and inhibit their automaticity, insulating structures surrounding the SAN prevent this, thereby safeguarding its pacemaking function \cite{joyner_propagation_1986,fedorov_conduction_2012}.
	
	%The SAN maintains electrical coupling with the surrounding atrial myocardium through specific sites called SAN exit pathways which can number up to five \cite{li_redundant_2017, fedorov_optical_2010}.
	
	%The exit pathways allow variability in the earliest atrial activation site, which not only affects the activation sequence of the RA but also influences the activation of the LA, which depends on RA activation \cite{li_redundant_2017, antz_electrical_1998}. Consequently, the SAN's anatomy and its exit pathways are crucial parameters to consider in understanding atrial activation dynamics.
	
	
	\subsubsection{Effect of the \gls{ra} endocardium on the P-wave}
	The anterior and lateral endocardial walls of the \gls{ra} are primarily composed of \glspl{pm} and \gls{ct} tissue and are attached to a thin-walled \gls{la} epicardium, consisting of only a few layers of myocytes \citep{lang2022imaging,matsuyama2004anatomical}. 
	While explicitly accounted for in bilayer models, this is often not the case in volumetric models \citep{roney2023:_bia_vol}.
	There, two different approaches have been used. Either, \gls{ct}, \glspl{pm} and the \gls{ra} portion of \gls{bb}
	were explicitly meshed, or a continuous right endocardial layer of prescribed thickness was generated
	to represent the \gls{ra} endocardium, 
	onto which the locations of the \gls{ct} and \glspl{pm} are projected either by using an anatomical atlas, or on a per-rule basis.
	Atrial \gls{ep} in the \gls{ra} endocardium is then simulated 
	by assigning faster conduction velocities in the \gls{ct} and \glspl{pm}, 
	and slower conduction velocities to the remaining endocardial tissue between the protruding structures \citep{azzolin2023:_augmenta,ferrer2015detailed,loewe2015left,roney2023:_bia_vol}. 
	As such, the absence of tissue between the \gls{ct} and \glspl{pm} is typically not accounted for 
	which can be interpreted as an increase in the effective epicardial wall width and the associated source strength. 
	
	We investigate the effect of modeling the \gls{ra} endocardial tissue as an electrically active layer 
	upon atrial sources and the P-wave. 
	Treating the \gls{ra} endocardial layer in between \gls{ct} and \glspl{pm} as blood pool 
	slowed down the activation of the posterior-lateral wall of the \gls{ra} and the \gls{raa}. 
	While this had a minor impact on the total activation time of the \gls{ra}, 
	a noticeable reduction in peak P-wave amplitude by $\approx$1\% of was witnessed, with no change in P-wave morphology. 
	The observed deviations in P-wave can be deemed as minor relative to the overall uncertainty.
	This may be not the case when simulating endocardial recorded \glspl{egm},
	when electrodes are located near the atrial sources.
	
	% Explicit meshing of the \gls{ra} endocardium is challenging due to limitations in resolution and contrast of 
	% current clinical tomographic imaging modalities. 
	% A detailed segmentation of \gls{ct} and \glspl{pm} that could be used for an explicit representation in a mesh 
	% is therefore currently not feasible in a scalable manner, 
	% but is likely to become tractable with the advent of latest generation
	% photon counting detector CT scanners \cite{} which will facilitate the accurate identification of these structures.
	
	
	% The section below I see largely as a limitation
	
	%\subsection{ECG pseudo-calibration}
	
	%Calibration of atrial sources based on \gls{ecg} data has been seldom addressed in previous work \cite{loewe2015left,gillette2022personalized}. In \cite{loewe2015left}, the authors presented a qualitative comparison between simulated and real P-waves by visually depicting both \gls{ecg} signals in different images, though no quantitative results were provided. In \cite{gillette2022personalized}, the \gls{ep} activation of a whole heart model was calibrated using \glspl{ecg}, with differences between simulated and real \glspl{ecg} quantified through the L2 norm. However, the atrial modeling approach in this study lacked some anatomical and functional structures, including the posterior and \gls{cs} \glspl{ic}.
	
	%The set of atrial geometries developed in our work does not include corresponding torso geometries. Therefore, each atrium was fitted with a reference realistic torso and the associated electrode sites. While this approach is useful for investigating atrial cardiac activation and the role of cardiac \gls{ep} parameters in P-wave generation, it hinders the precise definition of a cardiac digital twin. The shape and orientation of the torso surrounding the heart, as well as potential differences in the location and distance of electrodes relative to the cardiac source, can significantly influence \gls{ecg} generation \cite{zappon2024quantifying}. Furthermore, the dataset used comprises only \gls{ct} scans of patients with \gls{af}, yet potential scar tissue and non-conductive regions in the biatrial models were not considered.
	
	% This stuff I moved up to the P-wave calibration section
	%While some discrepancies between the P-wave generated by our biatrial-torso model and the corresponding clinical data were anticipated, our pseudo-calibration process demonstrates that a good match for the P-wave can be achieved in most leads using our framework. By quantifying differences between simulated and real \glspl{ecg} using the \gls{rmse}, we show that the P-wave can be closely captured in all leads except aVL, aVR, and V1, where amplitude differences were observed between the pseudo-calibrated \gls{ecg} and the clinical data. However, our P-wave model successfully preserves the positivity in lead aVL and -aVR, as well as the biphasic characteristic of lead V1. By automatically defining the biatrial-torso geometry and systematically varying cardiac \gls{ep} parameters, we demonstrate that our framework is well-suited for an efficient pseudo-calibration of atrial \gls{ep} based on \gls{ecg} matching.
	
	\section{Limitations}
	\label{Sec:limitation}
	While our comprehensive end-to-end workflow for generating biatrial anatomies combined with a fast forward \gls{relf} model 
	is able to deliver anatomically accurate models for predicting \gls{ecg} and \glspl{egm} with full physical fidelity,
	numerous notable limitations exist.
	
	% this is not a P-wave calibration study
	Our study introduces a technological framework for the calibration of atrial \gls{ep} models 
	using observable extracellular recordings such as the \glspl{egm} and the \gls{ecg} -- specifically, the P-wave -- 
	as objective to infer the high dimensional parameter space of such models. 
	However, in its current form, an immediate inference to match a P-wave is not readily feasible. 
	Despite the computational efficiency, a brute force sampling approach may still be too expensive \citep{gillette2021:_framework}.
	Similar to previous work on identifying the ventricular activation sequence, a dedicated approach 
	combining \emph{a priori} knowledge to constrain the search space in combination with an optimization approach 
	that facilitates a gradient-based search for parameters is likely to be more accurate and efficient \citep{grandits2021:_geasi,grandits2023:_geodesic_bp}.
	% Moreover, while sampling some narrow  physiologically constrained parameter corridors
	% as performed in here 
	% may yield a better approximation than current standard methods 
	% that rely on using literature data only for model components 
	% without an evaluation of the achieved agreement of the overall model with observations, 
	% this cannot be guaranteed for two major reasons.
	Sampling narrow, physiologically constrained parameter corridors, as done here, may provide a better approximation than current standard methods. However, this improvement cannot be guaranteed for two key reasons.        
	First, as the atrial activation sequence cannot be measured accurately, not even invasively with most advanced \gls{eam} methodology,
	a solid ground truth reference is missing.
	Secondly, it is highly likely that a large number of parameter choices exist 
	which yield identical perfect matches of the \glspl{ecg}.
	Thus, uniqueness and identifiability is a serious concern that must be adequately addressed 
	to narrow down the number of different sampled parameters.
	The development of suitable strategies and their evaluation was beyond the scope of this study,  and substantial further research efforts are required.
	
	An important factor limiting the anatomical accuracy of models generated by our end-to-end workflow 
	is the assumption of uniform wall thicknesses. 
	As the accurate segmentation of the atrial walls from current clinical images is challenging,
	in our volumetric wall reconstruction, the atrial wall widths were prescribed and held constant over individual anatomical regions. 
	However, atrial wall widths in a patient's heart may be different and more heterogeneous than represented in the model.
	As a consequence, the local variation in local conduction velocity mediated by altered source-sink relation may not be accurately represented.
	
	A further limitation is the reduced computational efficiency in generating finer resolutions for \gls{rd} models.
	This is largely caused by the computation of the \glspl{uac}, and not due to the extra cost of generating finer meshes per se. Specifically, projecting the biatrial mesh onto the \gls{uac} space to solve the linear elasticity problem 
	can induce mesh deformations and, consequently, iterative solver convergence issues.
	% with the 
	% as a relatively simple preconditioner is used. 
	More advanced preconditioners will be incorporated \citep{augustin2016anatomically} in the future to enhance computational efficiency. 
	Due to the nearly identical anatomical meshes, a faster alternative may be 
	the Euclidian interpolation of \glspl{uac} from the lower resolution \gls{re} to the higher resolution \gls{rd} mesh, but this has not been investigated here.
	
	While our novel flexible and rapid approach to modeling \glspl{ic} is advantageous for model calibration, 
	the used of single cables only approximates the effects of conduction between the atria, 
	but not the actual sources contributing to the extracellular potential fields.
	%In our framework, each \gls{ic} is represented by a single cable connecting the \gls{ra} and \gls{la}, based solely on their entry sites. 
	The actual anatomical structures underlying the \glspl{ic} are rather muscular bands or 3D strands. 
	As such, the volumetric source density of the cable approach underestimates the contribution to the potential field
	in their vicinity.
	While this does not constitute a problem for simulating activation sequences and the \gls{ecg}, 
	simulating the sensing of \gls{egm} will be inaccurate and also methodologically challenging to implement, 
	owing to the non-conformity of \gls{ic} grid with the atrial or the volume conductor mesh.
	This problem is readily mitigated by using bundles instead of single cables,
	which may be straightforwardly incorporated, as shown for modeling a fan-like insertion of the \gls{bb},
	where a bundle of three \glspl{ic} was used (see Figure \ref{Fig.Three_bb}). 
	Alternatively, the cable-based representation can be replaced with an explicit mesh of the \glspl{ic}
	once the model has been calibrated.
	
	%Another drawback of our current approach is the limitation of imposing only unidirectional conduction properties on each cable, neglecting the potential effects of tissue anisotropy. This is, however, an implementation issue that we will address in future work.
	Further, the cables representing the \glspl{ic} are automatically generated, 
	following the geodesic paths connecting two mesh nodes on \gls{ra} and \gls{ra}, respectively. 
	This might result in a mismatch in trajectory and length between modeled and real \gls{ic} and, thus, physiologically non-feasible conduction velocities may be required to meet prescribed inter-atrial conduction delays.
	Specifically, this may complicate the transfer of parameters from \gls{re} to \gls{rd} models 
	where conduction velocity ranges are limited.
	
	%However, this issue can be mitigated by tuning the conduction velocities within the cables to achieve physiological activation times at the anatomical regions at the end of the cables, such as for the \gls{bb} entry sites in the \gls{la}.
	% Both CT and MRI tomographic imaging of patients treated for atrial conditions is becoming abundantly available, 
	% including their use for modeling applications. However, the field of view is often limited to the heart itself,
	% excluding the extracardiac space with the thorax were observations are recorded from.
	% %%%
	% \todo[inline]{Continue from here}
	Further, we have not included the segmentation of the torso along with the corresponding atrial geometries. 
	Large clinical data cohorts often lack \gls{ct} or \gls{mri} scans of the torso. 
	Nevertheless, our workflow supports the meshing of a generic torso surface around the generated atrial geometries, 
	making the integration of automatic torso segmentation and corresponding meshing into our workflow a straightforward enhancement.
	
	Finally, we have not yet incorporated considerations for atrial disorders and pathologies, 
	which would be essential for clinical applications. While idealized damage and non-conductive substrates can be easily modeled using our \gls{ep} simulator‚Äîalong with pacing protocols to simulate conditions such as \gls{af}, our current \gls{relf} model does not support \gls{ecg} simulations for rhythm disorders with reentrant phenotypes, as the underlying Eikonal model driving activation does not accommodate reentrant patterns. 
	Advancements in altering atrial depolarization profiles may be necessary. Furthermore, real damaged tissue data from clinical sources is not yet integrated into our workflow. However, our reference framework is designed to easily accommodate new functionalities and define complex parameter variations, paving the way for the automatic and near-real-time creation of atrial digital twins.
	
	\section{Conclusions}    
	\label{Sec:conclusion}
	In this study, we developed an automated, efficient, and flexible multi-scale workflow for creating anatomically accurate biatrial models with high-quality meshes, suitable for various forward \gls{ep} models, and extended it to a function twinning stage, where efficient model calibration based on \gls{ecg} data is possible. The workflow utilizes a segmented blood pool and a deep learning-based segmentation network to automatically identify and label principal atrial structures. It moreover generates endocardial and epicardial walls based on predefined rules, ensuring robust model openings and high mesh quality. Detailed anatomical and functional structures, including fiber architecture and space-varying parameters, are incorporated into the models. 
	We further introduced a flexible method for representing \glspl{ic} using auto-generated cables, allowing for realistic and customizable conduction properties. Our framework finally includes a clinically compatible \gls{ecg} forward generation system, integrating both \gls{rd} and \gls{re} models with Lead Field methods for accurate \gls{ecg} trace generation, and possibly allows for the integration of the Pseudo-bidomain approach or $\phi_e$-recovery. Among these, we show that the \gls{relf} method proved effective for real-time \gls{ecg} computation and reliable source approximation in sinus rhythm. 
	
	Overall, our workflow facilitates efficient exploration of the \gls{ep} parameter space for calibration of biatrial models based on \gls{ecg} data, making it a valuable tool for advancing cardiac \gls{ep} research.
	
	\FloatBarrier
	
	\section*{Acknowledgments}
	
	This research received support from the Austrian Science Fund (FWF) grant no. 10.55776/I6476, and from the European Union‚Äôs Horizon 2020 research and innovation program under the Marie Sk\l{}odowska-Curie grant TwinCare-AF (grant agreement no. 101148636). The National Institutes of Health supported this work under grant 1R01HL158667 and by the Austrian Science Fund (FWF) under grant 10.55776/P37063.
	E. Zappon acknowledges her membership to INdAM GNCS - Gruppo Nazionale per il Calcolo Scientifico (National Group for Scientific Computing, Italy).
	\begin{figure}[!h]
		\centering
		\includegraphics[width = 0.4\textwidth]{images/eu_funded_en.jpg}
	\end{figure}
	
	
	
	%ACRONYMS
	
	% \newglossaryentry{ep}{name=EP, description={Electrophysiology}}
	% \newglossaryentry{cdt}{name=CDT, description={Cardiac Digital Twin}}
	% \newglossaryentry{eam}{name=EAM, description={Electro-anatomical map}}
	% \newglossaryentry{ecg}{name=ECG, description={Electrocardiogram}}
	% \newglossaryentry{egm}{name=EGM, description={Electrogram}}
	% \newglossaryentry{rd}{name=R-D, description={reaction-riffusion}}
	% \newglossaryentry{re}{name=R-E, description={reaction-eikonal}}
	% \newglossaryentry{bem}{name=BEM, description={Boundary Element Method}}
	% \newglossaryentry{fem}{name=FEM, description={Finite Element Method}}
	% \newglossaryentry{hps}{name=HPS, description={His-purkinje System}}
	% \newglossaryentry{relf}{name=RELF, description={reaction-eikonal Lead Field}}
	% \newglossaryentry{tnnp}{name=TNNP, description={Ten-Tusscher Ionic Model}}
	% \newglossaryentry{cv}{name=CV, description={Conduction Velocity}}
	% \newglossaryentry{se}{name=SE, description={Subendocardial}}
	% \newglossaryentry{3dwh}{name=3DWH, description={3D Whole Heart}}
	% \newglossaryentry{mri}{name=\gls{mri}, description={Magnetic Resonance Imaging}}
	% \newglossaryentry{ms}{name=MS, description={Mitchell Schaeffer}}
	% \newglossaryentry{apd}{name=APD, description={Action Potential Duration}}
	% \newglossaryentry{lv}{name=LV, description={Left Ventricle}}
	% \newglossaryentry{rv}{name=RV, description={Right Ventricle}}
	% \newglossaryentry{la}{name=LA, description={Left Atrium}}
	% \newglossaryentry{ra}{name=RA, description={Right Atrium}}
	% \newglossaryentry{uac}{name=UAC, description={Universal Atrial Coordinates}}
	% \newglossaryentry{uvc}{name=UVC, description={Universal Ventricular Coordinates}}
	% \newglossaryentry{utc}{name=UTC, description={Universal Torso Coordinates}}
	% \newglossaryentry{bn}{name=BN, description={Bi-Eikonal Normalization}}
	% \newglossaryentry{af}{name=AF, description={Atrial Fibrillation}}
	% \newglossaryentry{scn}{name=SCN, description={SpatialConfiguration-Net}}
	% \newglossaryentry{lspv}{name=LSPV, description={Left Superior Pulmonary Vein}}
	% \newglossaryentry{rspv}{name=RSPV, description={Right Superior Pulmonary Vein}}
	% \newglossaryentry{ripv}{name=RSPV, description={Right Inferior Pulmonary Vein}}
	% \newglossaryentry{cs}{name=CS, description={Coronary Sinus}}
	% \newglossaryentry{laa}{name=LAA, description={Left Atrial Appendage}}
	% \newglossaryentry{raa}{name=RAA, description={Right Atrial Appendage}}
	% \newglossaryentry{svc}{name=SVC, description={Superior Vena Cava}}
	% \newglossaryentry{ivc}{name=IVC, description={Inferior Vena Cava}}
	% \newglossaryentry{pca}{name=PCA, description={Principal Component Analysis}}
	% \newglossaryentry{lpv}{name=LPV, description={Left Pulmunary Veins}}
	% \newglossaryentry{rpv}{name=RPV, description={Right Pulmunary Veins}}
	% \newglossaryentry{tv}{name=TV, description={Tricuspid Valve}}
	% \newglossaryentry{mv}{name=MV, description={Mitral Valve}}
	% \newglossaryentry{ct}{name=\gls{ct}, description={Computed Tomography Images}}
	% \newglossaryentry{ic}{name=IC, description={Interatrial Connection}}
	% \newglossaryentry{bb}{name=BB, description={Bachmann's Bundle}}
	% \newglossaryentry{sa}{name=SA, description={Sinoatrial Node}}
	% \newglossaryentry{fo}{name=FO, description={Fossa Ovalis}}
	% \newglossaryentry{pm}{name=PM, description={Pettinate Muscle}}
	% \newglossaryentry{pwd}{name=PWD, description={P-wave Duration}}
	% \newglossaryentry{dbc}{name=BC, description={Dirichlet Boundary Conditions}}
	% \printglossary[style=long,nonumberlist]
	
	
	% LateX generates the ``References'' ing automatically and switches
	% to 9 point type for the bibliography.  If you use BibTeX (recommended),
	% follow the examples in the sample 'refs.bib' file to enter your references,
	% and leave the following line unchanged.
	
	%%Harvard
	\bibliographystyle{model2-names}
	
	%\bibliographystyle{ieeetr}
	
	\bibliography{References}
	\FloatBarrier
	
	% \appendix
	% \section{UAC computation}
	% \label{appendix_UAC}
	
	% \todo[inline]{The figures in this appendix are very nice, but without a text this is not a real appendix. 
	% Either we move the text over to here, or we don't have an appendix and move these figures into the main text.}
	% % \begin{figure}
	% % \centering
	% % \includegraphics[width=0.9\textwidth]{images/parameterization_v2.png}
	% % \caption{Interfaces and boundary surfaces in the \gls{ra} and \gls{la} and their parametrization.}
	% % \label{Fig:UCAparametrization}
	% % \end{figure}
	
	% % \begin{figure}
	% % \centering
	% % \includegraphics[width=0.9\textwidth]{images/parameterization_flat_v3.png}
	% % \caption{Overview of the Dirichlet values at the boundary surfaces and interfaces including 
	% % their parametrization $s \in [0, 1]$ (color-coded from blue $\!=\!0$ to red $\!=\!1$) used to compute 
	% % the \glspl{uac}. 
	% % In \textbf{(a)} the Dirichlet values at the boundaries and interfaces used to compute the preliminary 
	% % $\alpha$ and $\beta$ components for the \gls{ra} and \gls{la} are shown. Row \textbf{(b)} depicts the 
	% % selected landmarks and the parametrization of the \gls{cs}, the \gls{lipv}, and the \gls{ripv}. The 
	% % final $\alpha$ and $\beta$ components are illustrated in row \textbf{(c)} after moving the orifices.}
	% % \label{Fig:UCAparametrizationFlat}
	% % \end{figure}
	
	% \begin{table}
	% \centering
	% \addtolength{\tabcolsep}{5.0pt}
	% \begin{tabularx}{\textwidth}{X|cc|cc}
	% \toprule
	% & \multicolumn{2}{c|}{$\Omega_{\mathrm{RA},sept}\;\;$ and $\;\;\Omega_{\mathrm{LA},ant}$} & 
	% \multicolumn{2}{c}{$\Omega_{\mathrm{RA},lat}\;\;$ and $\;\;\Omega_{\mathrm{LA},post}$} \\
	% & $\alpha$ & $\beta$ & $\alpha$ & $\beta$ \\
	% \midrule
	% $\Gamma_{tv}$, $\Gamma_{mv}$
	%     & -- & 1 
	%     & -- & 0 \\[0.15cm]
	% $\Gamma_{ivc}$, $\Gamma_{lspv}$
	%     & $\mathrm{R}\,\cos\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$ 
	%     & $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$
	%     & $\mathrm{R}\,\cos\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$ 
	%     & $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((3\!+\!s)\,\frac{\pi}{2}\Big)$ \\[0.15cm]
	% $\Gamma_{svc}$, $\Gamma_{rspv}$
	%     & $1\!+\!\mathrm{R}\,\cos\Big((1\!+\!s)\,\frac{\pi}{2}\Big)$ 
	%     & $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((1\!-\!s)\,\frac{\pi}{2}\Big)$
	%     & $1\!+\!\mathrm{R}\,\cos\Big((1\!+\!s)\,\frac{\pi}{2}\Big)$ 
	%     & $\frac{1}{2}\!+\!\mathrm{R}\,\sin\Big((3\!+\!s)\,\frac{\pi}{2}\Big)$ \\[0.15cm]
	% $\mathcal{I}_{tv,ivc}$, $\mathcal{I}_{mv,lspv}$
	%     & $0$ & $\Big(\frac{1}{2}\!+\!\mathrm{R}\Big)+(1\!-\!s)\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ 
	%     & $0$ & $s\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ \\[0.15cm]
	% $\mathcal{I}_{tv,svc}$, $\mathcal{I}_{mv,rspv}$
	%     & $1$ & $\Big(\frac{1}{2}\!+\!\mathrm{R}\Big)+(1\!-\!s)\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ 
	%     & $1$ & $s\,\Big(\frac{1}{2}\!-\!\mathrm{R}\Big)$ \\[0.15cm]
	% $\mathcal{I}_{ivc,svc}$, $\mathcal{I}_{lspv,rspv}$
	%     & $\mathrm{R}\!+\!(1\!-\!2\,\mathrm{R})\,s$ & $\frac{1}{2}$ 
	%     & $\mathrm{R}\!+\!(1\!-\!2\,\mathrm{R})\,s$ & $\frac{1}{2}$ \\[0.15cm]
	% \bottomrule
	% \end{tabularx}
	
	% \caption{Dirichlet boundary conditions for the Laplace problems that are solved to obtain 
	% the preliminary \gls{uac} space in the septal and lateral \gls{ra} and the posterior and 
	% anterior \gls{la} where $s \in [0, 1]$ is a parametrization value of the corresponding
	% boundary surface and interface, respectively. In our framework,  $\mathrm{R}=0.1$.}
	% \label{Tab:UACdirichlet}
	% \end{table}
	
	% \begin{table}
	% \centering
	% \addtolength{\tabcolsep}{5.0pt}
	% \begin{tabularx}{0.6\textwidth}{X|cc}
	% \toprule
	% & $\alpha$ & $\beta$ \\
	% \midrule
	% $\Gamma_{cs,roof}$
	%     & $\alpha_{cs}\!+\!\mathrm{r}\,\cos\Big((2\!-\!s)\,\pi\Big)$
	%     & $\beta_{cs}\!+\!\mathrm{r}\,\sin\Big((2\!-\!s)\,\pi\Big)$ \\[0.15cm] 
	% $\Gamma_{cs,tv}$
	%     & $\alpha_{cs}\!+\!\mathrm{r}\,\cos\Big(s\,\pi\Big)$ 
	%     & $\beta_{cs}\!+\!\mathrm{r}\,\sin\Big(s\,\pi\Big)$ \\[0.15cm]
	% $\Gamma_{lipv,roof}$
	%     & $\alpha_{ipv}\!+\!\mathrm{r}\,\cos\Big(s\,\pi\Big)$ 
	%     & $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big(s\,\pi\Big)$ \\[0.15cm]
	% $\Gamma_{lipv,mv}$
	%     & $\alpha_{ipv}\!+\!\mathrm{r}\,\cos\Big((2\!-\!s)\,\pi\Big)$
	%     & $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big((2\!-\!s)\,\pi\Big)$ \\[0.15cm] 
	% $\Gamma_{ripv,roof}$
	%     & $(1\!-\!\alpha_{ipv})\!+\!\mathrm{r}\,\cos\Big(s\,\pi\Big)$
	%     & $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big(s\,\pi\Big)$ \\[0.15cm]
	% $\Gamma_{ripv,mv}$
	%     & $(1\!-\!\alpha_{ipv})\!+\!\mathrm{r}\,\cos\Big((2\!-\!s)\,\pi\Big)$
	%     & $\beta_{ipv}\!+\!\mathrm{r}\,\sin\Big((2\!-\!s)\,\pi\Big)$ \\[0.15cm]
	% \bottomrule
	% \end{tabularx}
	
	% \caption{Final position of the \gls{cs}, \gls{lipv}, and \gls{ripv} orifices with p
	% arameterization value $s \in [0, 1]$.}
	% \label{Tab:UACdirichletLinElas}
	% \end{table}
	
\end{document}


