\section{Related Work}
\label{sec:related-work}

{\bf Protein Structure Retrieval.} With the growth of the volume of protein structures, structure retrieval has become a critical task in protein data management. 
AlphaFind \citep{prochazka2024alphafind} is a web tool designed to identify structurally similar proteins in AlphaFold Database \citep{varadi2022alphafolddb} by compressing data from $\sim$23 TB to $\sim$20 GB using vector embeddings, narrowing down candidates with a neural network. The similarity of the search result is evaluated by US-align~\citep{zhang2022usalign}. Another state-of-the-art method, FoldSeek \citep{van2024foldseek}, accelerates protein structure searches by representing tertiary amino acid interactions as sequences over a 3D interaction structural alphabet, which derives from vector quantization by VQ-VAE \citep{oord2017vqvae}. 
However, the representation only models the structure of two contiguous residues in a chain.  

\input{section/tab/tab-3-settings}

{\bf Protein Inverse Folding.} Protein inverse folding aims to predict diverse sequences that can fold into a given protein structure.
ProteinMPNN \citep{dauparas2022mpnn} is a deep learningâ€“based method for protein sequence design that excels in both in silico and experimental evaluations. By leveraging a message-passing neural network with enhanced input features and edge updates, ProteinMPNN is capable of designing monomers, cyclic oligomers, protein nanoparticles, and protein-protein interfaces, rescuing previously failed designs generated by Rosetta \citep{adolf2018rosetta,baek2021rosetta} or AlphaFold \citep{jumper2021alphafold}.
ESM-IF1 \citep{hsu2022esmif1} employs a sequence-to-sequence Transformer to predict protein sequences from backbone atom coordinates. %As end-to-end deep learning approaches, these approaches suffer from hallucinations of generating non-functional sequences. 

{\bf Antibody Inverse Folding.} AbMPNN \citep{dreyer2023abmpnn} inherits the model architecture of ProteinMPNN, and trains an antibody-specific variant for antibody design. It outperforms generic protein design models in sequence recovery and structure robustness, especially for hyper-variable CDR-H3 loops. AntiFold \citep{hoie2024antifold} is an antibody-specific inverse folding model, which is fine-tuned on ESM-IF1, with both solved and predicted antibody structures. However, it should be emphasized that Antifold infers CDR sequences based on the structure of the variable domain and the sequence of the framework regions. Consequently, the accuracy of CDR sequence inference is influenced not only by the structure of the CDRs but also by the sequence and structural information of the framework regions. Previous studies utilizing antibody sequence language models without structural information have demonstrated that the sequence of the framework regions can partially predict the CDR sequences, particularly for relatively conserved residues. As a result, the requirement for the framework sequence as input complicates the inference of CDR sequences that can bind to different antigens while maintaining an identical framework.



{\bf Antibody Co-Design.} In recent years, deep learning models have emerged as powerful data-driven approaches for antibody design. RefineGNN \citep{jin2022refinegnn} is the first structure sequence co-design method that alternatively predicts the atom coordinates and residue types in CDRs by auto-regression. DiffAb \citep{luo2022diffab} and IgGM \citep{wang2024iggm} utilize diffusion models to generate the structure and sequence of CDRs based on the framework regions and the target antigen, with DiffAb oriented for specific antigens. MEAN \citep{kong2023mean} and dyMEAN \citep{kong2023dymean} employ graph neural networks to predict the structure and sequence of CDRs. Table \ref{tab:settings} presents a comparative analysis of various antibody design task configurations.

