\section{Results} 
\label{sec:res}


\subsection{IgSeek Approach}
\label{sec:subsec-res-setting}
The gist of IgSeek for structure-to-sequence generation is isomorphic structure retrieval, which allows for the exploration of a large and diverse antibody CDR structure database. Fig.~\ref{fig:framework} illustrates the framework of IgSeek. Given an antibody CDR database where both structures and sequences are available, IgSeek first constructs a CDR vector database, where vector embeddings index the structural proximity of the CDRs. In this offline stage, we pre-train a {\em Multi-channel Equivariant Graph Neural Network (MEGNN)} to encode the structure of CDR loops into fixed-length vectors within the CDR database. Specifically, MEGNN aligns the spatial structure distance between pairs of CDRs with equal lengths and similar conformations. Subsequently, for a CDR structure $G$ whose sequence is to be predicted, we first deploy the pre-trained MEGNN to generate an embedding $\vect{h}_G$ for $G$. $\vect{h}_G$ then serves as the search key to query the {\em $K$-nearest neighbors ($K$-NN)} structurally similar CDR loops in the vector database. Finally, the $K$-NN results, associated with their corresponding residue sequences, are collected to predict the sequence of $G$ by ensemble and Bernoulli sampling.A detailed description of our methodology can be found in Appendix~\ref{sec:method}.


{\bf Datasets.} 
We evaluate our {\igseek} and other baselines using both solved and predicted antibody structures. The training set consists of CDR pairs sampled from $11,023$ solved CDR loops in the {\em Structural Antibody Database (SAbDab)} \citep{dunbar2013sabdab,schneider2021sabdab}. To construct the CDR vector database, we utilize $24,479$ solved CDR loops from SAbDab before January 1, 2024 (SAbDab-before-2024). In addition, $4,449$ solved CDR loops released between January 3, 2024 and May 29, 2024 from SAbDab (SAbDab-2024) serve as the test set to evaluate the performance of {\igseek} and its competitors.
In addition to the solved antibody structures from SAbDab, we also conduct experiments on $5,111$ CDR loops from the {\em Structural T-Cell Receptor Database (STCRDab)} \citep{leem2018stcrdab} to evaluate the model generalization ability. Furthermore, we evaluate the model efficiency using $5,000$ predicted CDR-H3 loops from the {\em Observed Antibody Space (OAS-H3)} \citep{kovaltsuk2018oas,olsen2022oas}. IMGT numbering scheme \cite{lefranc2003imgt} is utilized for antibody datasets. More details of datasets and experiment settings can be found in Appendix \ref{sec:appendix-data} and Appendix~\ref{sec:appendix-param}. 


\begin{figure*}[t]
\centering
    \includegraphics[width=0.95\columnwidth]{section/fig/bar_aar_line_time.pdf}
\vspace{-3mm}
\caption{The comparison of average AAR and inference speed. (a) AAR in SAbDab-2024 dataset. (b) AAR in STCRDab. (c) Inference speed. }
\vspace{-5mm}
\label{fig:exp-seq-design}
\end{figure*}

\subsection{IgSeek for CDR Structure Retrieval}
\label{sec:subsec-res-search}


\begin{wrapfigure}{r}{0.6\textwidth}
  \centering
  \vspace{-5mm}
  \hspace{-2mm}
  \includegraphics[width=\linewidth]{section/fig/bar_ret_auc.pdf}
  \vspace{-3mm}
  \caption{IgSeek vs. FoldSeek in CDR retrieval.}
  \label{fig:exp-search}
  \vspace{-3mm}
\end{wrapfigure}


In this set of experiments, we compare {\igseek} with the state-of-the-art structure searching model, FoldSeek~\cite{van2024foldseek}, by examining the quality of the retrieved isomorphic structures. Introduction to competitors is deferred to Appendix~\ref{sec:appendix-baseline}. Specifically, for a given query CDR $q$, the retrieved CDR $r$ is considered a positive instance if their RMSD is less than 1 \AA. To ensure the robustness of our evaluation, we omit any query CDR for which there are no candidates in the CDR database with a distance of less than 1 \AA\ from the query. This strategy allows us to focus on instances where meaningful comparisons can be made, thereby enhancing the result reliability.



Fig. \ref{fig:exp-search} presents the experimental results of {\igseek} and FoldSeek, illustrating the model performance on the retrieved sequences using the AUROC metric. As we can observe, {\igseek} outperforms FoldSeek on four types of CDR loops while maintaining comparable performance on CDR-H3 and CDR-L1, indicating its capability of identifying structurally similar CDRs across diverse CDR loops. It is worth noting that {\igseek} achieves a 2.6x speed-up in structure retrieval time compared to FoldSeek. Since this improvement in speed does not come at the cost of accuracy, it demonstrates that {\igseek} strikes a superior trade-off between efficiency and accuracy. The ability to quickly retrieve high-quality structural matches can greatly enhance workflows in antibody design, as shown in Sec.~\ref{sec:subsec-res-gen}.


\begin{figure*}[t]
  \centering
  \includegraphics[width=0.95\textwidth]{section/fig/scatter_cdr_123.pdf}
  \vspace{-2mm}
  \caption{Embeddings of CDRs in the SAbDab-before-2024 datasets projected onto 2D Space.}
   \label{fig:exp-viz}
  \vspace{-4mm}
\end{figure*}


\subsection{IgSeek for CDR Sequence Design}
\label{sec:subsec-res-gen}
In this set of experiments, we compare {\igseek} with the state-of-the-art models for protein and antibody sequence design, including ProteinMPNN~\citep{dauparas2022mpnn}, ESM-IF1~\citep{hsu2022esmif1}, AbMPNN~\citep{dreyer2023abmpnn}, and AntiFold~\citep{hoie2024antifold}. Specifically, the MEGNN in {\igseek} is trained on the SAbDab-before-2024 dataset to construct the CDR vector database. Subsequently, the trained MEGNN is utilized to generate embeddings for the CDRs in the SAbDab-2024 dataset. For each query CDR in the SAbDab-2024 dataset, we retrieve the top-$10$ nearest neighbors from the SAbDab-before-2024 dataset in the CDR vector database, ensuring that the lengths of the retrieved sequences match that of the query. Finally, we proceed to sample the amino acids for each position in the CDR sequences to generate the predicted result for the query CDR. Note that existing protein and antibody inverse folding methodologies such as ProteinMPNN and AntiFold typically generate at least two samples for evaluation. In our experiments, we follow the settings of ProteinMPNN and present the best results of all other methods for evaluation. Average {\em amino acid recovery (AAR)} is utilized to evaluate model performance, which quantifies the accuracy of the predicted sequences. For a query CDR $q$, the AAR is defined as the ratio of overlapping positions between the predicted sequence $\hat{\vect{s}}_q$ and ground-truth sequence $\vect{s}_q$:
$
    \text{AAR}\left(\hat{\vect{s}}_q,  \vect{s}_q\right) = \frac{1}{L} \sum_{l = 1}^{L} \mathbb{I}(\hat{\vect{s}}_q(l), \vect{s}_q(l)).
$


Fig. \ref{fig:exp-seq-design} (a) illustrates the average AAR for each model on the SAbDab-2024 dataset. As we can observe, Antifold and AbMPNN achieve much better results compared to ProteinMPNN and ESM-IF1, highlighting the advantages of fine-tuning pre-trained protein design models specifically on the antibody dataset. Additionally, {\igseek} outperforms its competitors by at least 2.9\% on light chain CDR loops (CDR-L) and achieves results comparable to state-of-the-art methods on heavy chain CDR loops (CDR-H). We incorporate an additional variant of {\igseek} that uses RMSD as a secondary sorting metric, denoted as {\igseek}+Kabsch. Notice that we do not deploy the Kabsch algorithm to search the entire database. Instead, we validated the RMSD of the top-ranked CDRs identified by IgSeek until we identified the top 10 CDRs with RMSD less than 1 \AA. Notably, {\igseek}+Kabsch consistently outperforms all baselines across six types of CDR loops, highlighting the effectiveness of our retrieval-based strategy. The marked advantage of {\igseek}+Kabsch on CDR-H3 loops is particularly noteworthy, as CDR-H3 is often considered one of the most hypervariable regions.

\noindent
{\bf Remark.} We observe a performance degradation in AntiFold and AbMPNN on the SAbDab-2024 dataset compared to the results reported by \cite{hoie2024antifold}. One possible reason for this discrepancy is that these two models heavily depend on antibody backbone structures as auxiliary information, while only the structures of CDRs are given in our settings.


{\bf Generalization Performance.}
Next, we evaluate the model inference performance on the STCRDab dataset without any further model training. To conduct this evaluation, we randomly draw around $80\%$ of the CDR loops to generate selection templates, while the remaining $20\%$ are used as queries. Fig. \ref{fig:exp-seq-design} (b) displays the average AAR of each model on the STCRDab dataset. As we can see, {\igseek} takes the lead by at least $30\%$ on CDR loops from chain A and chain B, respectively. These impressive results further underscore the potential of structure retrieval approaches in mitigating hallucinations during sequence inference, demonstrating that {\igseek} can effectively generalize to unseen data while maintaining high accuracy in sequence recovery.


{\bf Efficiency Evaluation.}
We evaluate the model efficiency using the OAS-H3 dataset. Fig. \ref{fig:exp-seq-design} (c) reports the inference time of {\igseek} compared with other baseline models, all without any model retraining. As we can observe, {\igseek} achieves at least $20$x speed-up compared to baseline methods, which demonstrates that our {\igseek} achieves a better trade-off between effectiveness and efficiency. This enhanced inference speed is particularly beneficial in practical applications like high-throughput antibody design where rapid sequence generation is crucial. 


{\bf Visualization. }
To investigate the representation generated by MEGNN, we conduct a visualization analysis on the SAbDab-before-2024 dataset by T-SNE \citep{van2008tsne}. Fig. \ref{fig:exp-viz} presents the visualization results of top-$60$ CDR representations in each cluster, where PyIgClassify cluster labels \citep{adolf2015fccc} (refer to Appendix \ref{sec:appendix-data}) are utilized in this set of experiments. As Fig.~\ref{fig:exp-viz} illustrates, {\igseek} produces a high-quality visualization that clearly organizes the embeddings of CDR loops from distinct clusters into separate groups with minimal overlaps. 
Furthermore, the visualization not only demonstrates the effectiveness of {\igseek} in distinguishing CDRs among different clusters but also highlights its ability to capture structural information inherent in CDR loops. This visual clarity and distinct grouping underscore the robustness and discriminative capability of {\igseek} in embedding isomorphic CDR structures closer together while ensuring distinct clusters remain well-separated, which facilitates the identification and retrieval of CDR loops based on their structural characteristics.

\begin{figure*}[t]
  \centering
  \includegraphics[width=0.95\textwidth]{section/fig/case_CDRL1.pdf}
  \vspace{-3mm}
  \caption{A Case study using 8W8R CDR-L1 as an example.}
  \label{fig:exp-case}
  \vspace{-3mm}
\end{figure*}


\input{section/tab/tab-5-dymean-comparison}


\subsection{Case Study}
\label{sec:subsec-case-study}

{\bf Example.}
We first use the 8W8R CDR-L1 as an example to illustrate the query and generation process of {\igseek}. Step 1: given the backbone structure of the 8W8R CDR-L1 loop, we employ the pre-trained MEGNN to generate its embeddings. Step 2: we retrieve the top-10 nearest neighbors of the 8W8R CDR-L1 loop from the CDR vector database $\mathcal{Z}$. Step 3: we utilize the aligned sequences from the retrieved records to generate the residue probability distribution at each position. Step 4: Finally, we sample the output result from this distribution. In this example, we observe that the AAR of the sequence generated by {\igseek} outperforms other competitors by at least 0.27, demonstrating the effectiveness of our approach.



{\bf Incorporation with structure generation models.}
Next, we evaluate the sequence prediction capabilities of Igseek using dyMEAN \cite{kong2023dymean} and RFdiffusion \cite{watson2023rfdiffusion} as structure generators for unseen antibody structures, focusing on 8R1C CDR-L1 loop. The pipeline first employs dyMEAN or RFdiffusion to generate missing CDR structures, followed by sequence predictions by Igseek for this loop, and the experimental results are presented in Table~\ref{tab:dymean-comparison}.The analysis reveals significant improvement in sequence recovery rates for 8SGN CDR-L1 when using Igseek, demonstrating the effectiveness of Igseek as a key component in real-world antibody design pipelines. In particular, our experiments indicate that the quality of structural prediction substantially influences the accuracy of sequence generation, suggesting that the development of precise CDR structure prediction models would further enhance the overall antibody design pipeline.