\newpage
\appendix
\section{Appendix}\label{app}
All the experiments %can be
are 
conducted with 4 RTX-4090 GPUs.
% \subsection{Ablation Design}\label{app:ablatondesign}
% We conduct four ablation experiments on the model design. In this section, we present their specific designs. Here we still use the notation as in the main paper $\mathbf{A}, \mathbf{B}, \mathbf{g}, \mathbf{h}$
% \subsubsection{Downsample}
% We maintain the gate mechanism and the 2-way input in the design. The whole process can be described in Eq.\ref{app:alg:downsample}


\subsection{Pretraining}\label{app:pretrain}
We adhere to most of the pretraining hyperparameters and settings used for HyenaDNA, including the dataset selection. However, we implemented some small differences. We pretrained on sequences with a maximum length of 16K, while other pretraining sequence lengths were 1K or 2K, aligning with HyenaDNA's recommendation to use sequences 2-4 times longer than those required for downstream tasks. We set the global batch size to 512 and trained for 400 epochs with a learning rate of 1e-3. For MLM pretraining, selected nucleotides were "masked" by replacing them with "N" to predict the original nucleotide type. Table \ref{table:pretrain} and Table \ref{table:pretraintime} illustrate the parameter settings for models of various sizes and the time needed to pretrain. The pretraining dataset is HG38\footnote{\url{https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.26}}.
\begin{table}[h]
\centering
\caption{\textbf{Pretraining hyperparameters.} Values used during pretraining are reported.}
\label{table:pretrain}
\begin{tabular}{@{}ll@{}}
\toprule
\textbf{Hyperparameter}      & \textbf{Value}                         \\ \midrule
Learning Rate                & $1 \times 10^{-3}$                     \\
Batch Size                   & 256                                    \\
Weight Decay                 & 0.1                                    \\
Dropout                      & 0.0                                    \\ \midrule
Optimizer                    & AdamW                                  \\
Optimizer Momentum           & $\beta_1 = 0.9$, $\beta_2 = 0.999$     \\
Learning Rate Scheduler      & Cosine Decay                           \\
Training Epochs              & 200                                    \\ \bottomrule
\end{tabular}
\end{table}

\begin{table}[h]
\centering
\caption{\textbf{Pretraining time.} The time required for pretraining different model sizes with varying sequence lengths on 4 RTX-4090 GPUs is reported.}
\label{table:pretraintime}
\begin{tabular}{@{}lcccc@{}}
\toprule
\textbf{Parameters}          & \textbf{386K} & \textbf{1.7M} & \textbf{7.4M} & \textbf{27.4M} \\ \midrule
Sequence Length              & 1K            & 1K            & 16K           & 2K             \\
Time                         & 50 mins       & 80 mins       & 14 hours      & 4 hours        \\ \bottomrule
\end{tabular}
\end{table}

\subsection{Downstream Tasks}
All hyperparameters for the ConvNova model on downstream tasks can be found in Table \ref{table:hp}.

\begin{table}[H]
\caption{\textbf{Hyperparameters for ConvNova model on all downstream tasks.} Specific settings for each task include sequence length, dilation rate, hidden dimension, number of GCBs, and model size.} 
\label{table:hp}
\resizebox{\textwidth}{!}{
\begin{tabular}{@{}lccccc@{}}
\toprule  
Tasks&   Sequence Length   &   Dilation Rate  &  Hidden Dim & Num GCB &  Size\\ \midrule 
NT   &      1K        &       4/1        &      128    &   5     &  1.7M\\
Deepstarr & 1K        &        4         &      128    &   5     &  1.7M\\
Genomic Benchmark& 1K &        4         &       64  &   5     &    386K\\
Chromatin Profile Prediction &   2K        &        4         &    512/256  &   5     &  27.4M/6.8M\\
Gene Finding& 16K     &        4         &      196    &   10    &  7.4M\\
\bottomrule
\end{tabular}
}               
\end{table}

\subsubsection{Nucleotide Transformer Benchmark}\label{app:nt}
\textbf{Objective:} Sequence classification.

\textbf{Models and Setup:} We follow the HyenaDNA setup~\citep{nguyen2024hyenadna}, splitting the dataset into 90/10 training and test sets and fine-tuning DNABERT-2, HyenaDNA, NT-v2, and Caduceus-Ph for 20 epochs using pre-trained weights from Hugging Face.

\textbf{Experiment Configuration:} All models are tested with 32-bit floating point precision, and ConvNova uses consistent hyperparameters across tasks as detailed in Table \ref{table:hp}, with ten random seeds to report mean values, shown in Tables \ref{table:ntall1} and \ref{table:ntall}. Optimizer we use AdamW and share the same setting as Table \ref{table:pretrain}. Learning rate we use 1e-3 for ConvNova and official learning rate for the baselines.

\subsubsection{Chromatin Profile Prediction}\label{app:deepsea}
\textbf{Objective:} Perform 919 binary classification tasks on DNA fragments using 4.4 million training data points. Even small improvements (<1\%) are significant for this challenging task.

\textbf{Models and Setup:} The baseline models include HyenaDNA (0.4M, 3.3M, 6.5M), NT-v2 (50M, 100M, 250M, 500M), DNABERT-2, and DeepATT (previous state-of-the-art CNN for this task), trained for 10 epochs. All models are tested using 32-bit floating point precision.

\textbf{Experiment Configuration:} The train/test split follows the DeepSEA paper methodology. The optimizer used is AdamW with settings from Table \ref{table:pretrain}, and a learning rate of 1e-3 for ConvNova, with official learning rates used for the baselines.

We provide more detailed comparisons of different models in Table \ref{table:alldeepsea}. Besides the pretrained ConvNova model, we also train ConvNova from scratch. Remarkably, even without pretraining, the model outperformed all versions of HyenaDNA at similar parameter sizes, further demonstrating ConvNova's superior ability to encode long-sequence information. Hyperparameters can be found in Table \ref{table:hp}.

\begin{table}[H]
\caption{\textbf{Chromatin Profile Prediction detailed comparison.} AUC Score (↑) is reported for ConvNova, HyenaDNA, NT v2, DNABERT-2, DeepATT performance on transcription factors (TF), DNase I hypersensitive sites (DHS), histone modifications (HM), and average score. The best values per task are bolded, and the second best are underlined.}
\label{table:alldeepsea}
\resizebox{\textwidth}{!}{
\begin{tabular}{@{}l>{\columncolor{low}}l>{\columncolor{low}}l|lll|llll|l|>{\columncolor{low}}l>{\columncolor{low}}l>{\columncolor{low}}l|l@{}}
\toprule
\multicolumn{1}{c}{\multirow{1}{*}{Tasks}} & \multicolumn{10}{c}{Pretrain}                                                                                                                   & \multicolumn{4}{c}{From Scratch} \\ 
\midrule
\rowcolor{titlebg}\multirow{2}{*} & \multicolumn{2}{c}{\textbf{ConvNova}} & \multicolumn{3}{c}{\textbf{Hyena-DNA}} & \multicolumn{4}{c}{\textbf{NTv2}}                      & \textbf{DNABERT-2}          & \multicolumn{3}{c}{\textbf{ConvNova}}      & \textbf{DeepATT} \\
                    & 6.8M               & 27.4M                & 0.4M                 & 3.3M                 & 6.5M                 & 50M              & 100M             & 250M             & 500M             & 117M                 & 1.7M               & 6.8M               & 27.4M               & 7.8M \\ \midrule
\textbf{TF}                     & 96.77  & \textbf{96.99} & 94.42 & 94.72 & 95.91 & 96.17 & 96.23 & 96.64 & \underline{96.76} & 96.16 & 96.62 & 96.63 & 96.77 & 96.35 \\
\textbf{HM}                     & 86.54  & 86.70  & 84.64  & 84.96  & 85.46  & 86.23 & 86.20 & \textbf{86.81}  & \underline{86.76}  & 86.18  & 85.74  & 85.74  & 85.76  & 86.16 \\
\textbf{DHS}                    & 93.42  & \textbf{93.71} & 90.58 & 90.90 & 92.33 & 92.50 & 92.65 & \underline{93.13} & 93.39 & 92.50 & 92.93 & 92.94 & 93.03 & 92.52 \\
\textbf{Avg.}                    & 92.25  & \textbf{92.47} & 89.89 & 90.19 & 91.23 & 91.63 & 91.69 & \underline{92.19} & 92.30 & 91.61 & 91.76 & 91.77 & 91.85 & 91.68 \\ \bottomrule
\end{tabular}
}
\end{table}

% Please add the following required packages to your document preamble:
% \usepackage{booktabs}
% \begin{table}[]
% \begin{tabular}{@{}l|lll|lll|llll|l@{}}
% \toprule
% \multicolumn{1}{c}{} & \multicolumn{11}{c}{Pretrain}                                                                                                \\ \midrule
%               & \multicolumn{3}{c|}{ConvNova}    & \multicolumn{3}{c|}{Hyena-DNA} & \multicolumn{4}{c|}{NT v2}                       & DNABERT-2 \\
%                      & 1.7M  & 6.8M  & 27.4M          & 0.4M     & 3.3M     & 6.5M    & 50M            & 100M  & 250M  & 500M           & 117M      \\ \midrule
% TF                   & 96.62 & 96.77 & \textbf{96.99} & 94.42    & 95.49    & 95.64   & 97.17          & 96.23 & 92.3  & 96.76          & 96.16     \\
% HM                   & 85.74 & 86.54 & 86.70          & 84.64    & 84.88    & 85.08   & 86.23 & 86.2  & 92.19 & \textbf{86.76} & 86.18     \\
% DHS                  & 92.93 & 93.42 & \textbf{93.71} & 90.58    & 91.73    & 92.07   & 92.5           & 92.65 & 91.69 & 93.39          & 92.50     \\
% avg                  & 91.76 & 92.25 & \textbf{92.47} & 89.89    & 90.7     & 90.93   & 91.63          & 92.5  & 91.63 & 92.30          & 91.61     \\ \bottomrule
% \end{tabular}
% \end{table}


\subsubsection{GeneFinding}\label{app:genefinding}
\textbf{Objective:} Classify each nucleotide based on its gene structure context, predicting exon-intron boundaries and capturing long-range dependencies for accurate annotation.

\textbf{Models and Setup:} The baseline models include NT-H, DNABERT2, HyenaDNA, and GENA\_LM. ConvNova is tuned for 10 epochs, while the results of the other baselines are taken from BEND. 

\textbf{Experiment Configuration:} The optimizer used is AdamW with settings from Table \ref{table:pretrain}, and a learning rate of 1e-3 for ConvNova.
\paragraph{Dilation and Layer impact on Gene Finding}


We find that the gene-finding task requires long-range dependency. Both decreasing dilation rate and the number of layers can result in performance decline in Figure \ref{fig:mccdilation}. Models are trained from scratch in all experiments. For Dilation impact, the model dimension is 256 and contains 5 GCBs. For GCB counts impact, the model dimension is set to 256, and the dilation rate is set to 4. For additional hyperparameter settings in this task, readers can refer to Table \ref{table:hp}.

\begin{figure}[hbtp!]
\begin{minipage}[]{\textwidth}
 \centering
 \caption{\textbf{The impact of dilation rate and GCB counts on the MCC in ConvNova model.}}
 \subfigure[Impact of dilation on GeneFinding. ]{\includegraphics[width=0.45\textwidth]{imgs/mcc_by_dilation.png}}
 \subfigure[Impact of GCB counts on GeneFinding. ]{\includegraphics[width=0.45\textwidth]{imgs/mcc_by_layers.png}}
 \label{fig:mccdilation}
\end{minipage}
\end{figure}

\subsubsection{Genomic Benchmark}\label{app:genomicbenchmark}
\textbf{Objective:} Sequence classification task.

\textbf{Models and Setup:} All models are fine-tuned for ten epochs. ConvNova is used as the backbone with a width of 64, 5 GCBs, and a parameter size of 386K, pretrained on sequences of length 1000. The baselines used in this task are CaduceusPh, HyenaDNA, and the CNN model provided in the benchmark. We follow the train-valid split (90/10) as provided by HyenaDNA~\citep{nguyen2024hyenadna}.

\textbf{Experiment Configuration:} The models are trained with 16-precision float points. We use five random seeds and report the mean result across all models. Hyperparameters for this task can be found in Table \ref{table:hp}.The optimizer used is AdamW with settings from Table \ref{table:pretrain}, and a learning rate of 1e-3 for ConvNova, with official learning rates used for the baselines.

\subsection{Ablations experiments}
\subsubsection{ Gating Mechanism and Dual Branch}\label{app:ablation}

In this section, we provide more details about the ablation experiments. There are two variants of the ConvNova model: "Single Gate" and "Dual Branch", and one variant without any gating mechanism.


\paragraph{Single Gate}

Our "Single Gate" design follows a sequential process as shown in Eq.\ref{puregate}. First, a single \texttt{Conv1D} layer processes the input feature $\mathbf{A} \in \mathbb{R}^{l \times d}$ with weight matrix $\mathbf{W}_d \in \mathbb{R}^{k \times d \times d}$ and bias term $\mathbf{b}_d \in \mathbb{R}^d$ to produce $\mathbf{z} \in \mathbb{R}^{l \times d}$. This output is then processed through two parallel activation paths - \texttt{GELU} activation and \texttt{sigmoid} activation ($\sigma$). The final output is obtained through residual addition of element-wise multiplication ($\odot$) of activated $\mathbf{h}$ and gated $\mathbf{g}$. To maintain parameter count parity with ConvNova, we increase the dimension of the \texttt{Conv1D} layer.


\begin{equation}\label{puregate}
\textcolor{black}{
\centering
\begin{aligned}
    &\mathbf{z} = \mathbf{W}_d * \mathbf{A} + \mathbf{b}_d \\
    &\mathbf{h} = \operatorname{GELU}(\mathbf{z}) \\
    &\mathbf{g} = \sigma(\mathbf{z}) \\
    &\mathbf{A} = \mathbf{A} + \mathbf{h} \odot \mathbf{g}
\end{aligned}
}
\end{equation}

\paragraph{Dual Gate}

Our "Dual Branch" design processes input features through two parallel convolutional paths with gating mechanisms, as shown in Eq.\ref{dualgate}. The inputs $\mathbf{A}, \mathbf{B} \in \mathbb{R}^{l \times d}$ are processed through two separate \texttt{Conv1D} layers with weights $\mathbf{W}_{1, d/\sqrt{2}}, \mathbf{W}_{2, d/\sqrt{2}} \in \mathbb{R}^{k \times d \times d/\sqrt{2}}$ and biases $\mathbf{b}_{1, d/\sqrt{2}}, \mathbf{b}_{2, d/\sqrt{2}} \in \mathbb{R}^{d/\sqrt{2}}$ to produce intermediate features $\mathbf{z}_1, \mathbf{z}_2 \in \mathbb{R}^{l \times d/\sqrt{2}}$. The first path applies \texttt{GELU} activation to $\mathbf{z}_1$, while the second applies \texttt{sigmoid} activation to $\mathbf{z}_2$. The final output $\mathbf{A}$ is obtained through residual addition of element-wise multiplication ($\odot$) of activated $\mathbf{h}$ and gated $\mathbf{g}$, and $\mathbf{B}$ is obtained through residual addition of gated $\mathbf{g}$.


\begin{equation}\label{dualgate}
\centering
\textcolor{black}{
\begin{aligned}
    &\mathbf{z}_1 = \mathbf{W}_{1, d/\sqrt{2}} * \mathbf{A} + \mathbf{b}_{1, d/\sqrt{2}} \\
    &\mathbf{z}_2 = \mathbf{W}_{2, d/\sqrt{2}} * \mathbf{B} + \mathbf{b}_{2, d/\sqrt{2}} \\
    &\mathbf{h} = \operatorname{GELU}(\mathbf{z}_1) \\
    &\mathbf{g} = \sigma(\mathbf{z}_2) \\
    &\mathbf{A} = \mathbf{A} + \mathbf{h} \odot \mathbf{g} \\
    &\mathbf{B} = \mathbf{B} + \mathbf{g}
\end{aligned}
}
\end{equation}

\paragraph{Addition w/o Gate}

This variant uses simple addition operations for feature aggregation without any gating mechanism. As shown in Eq.\ref{Add}, the input features $\mathbf{A}, \mathbf{B} \in \mathbb{R}^{l \times d}$ are processed through convolutional layers and GELU activation to produce intermediate features $\mathbf{h}, \mathbf{g} \in \mathbb{R}^{l \times d/\sqrt{2}}$. These are then combined through addition to produce the final outputs $\mathbf{A}$ and $\mathbf{B}$.


\begin{equation}\label{Add}
\centering
\textcolor{black}{
\begin{aligned}
    \mathbf{h} &= \operatorname{GELU}(\mathbf{W}_{1, d/\sqrt{2}} * \mathbf{A} + \mathbf{b}_{1, d/\sqrt{2}}) \\
    \mathbf{g} &= \operatorname{GELU}(\mathbf{W}_{2, d/\sqrt{2}} * \mathbf{B} + \mathbf{b}_{2, d/\sqrt{2}}) \\
    \mathbf{A} &= \mathbf{A} + \mathbf{h} + \mathbf{g} \\
    \mathbf{B} &= \mathbf{B} + \mathbf{g}
\end{aligned}
}
\end{equation}

\subsubsection{ Dilation Mechanism, Kernel Size and local dependency analysis}\label{app:dilation_kernel_ablation}
In this section, we present the results of the ablation experiments on the dilation rate and kernel size. We perform experiments on the NT benchmark, testing kernel sizes of 3, 5, 7, 9, 11 and dilation rates of 1, 2, 3, 4. We report the average results across 18 tasks from the entire benchmark.

As shown in Table \ref{tab:dilation_kernel}, the performance generally improves with increasing dilation rate and kernel size. However, to balance compatibility with long-range tasks and maintain relatively small model parameters, we ultimately select a kernel size of 9 and a dilation rate of 4.

\begin{table}[h]
\centering
\caption{Performance of different kernel sizes and dilation rates across 18 tasks in the NT benchmark. The values represent the average performance for all tasks.}
\adjustbox{max width=\textwidth}{
\begin{tabular}{lrrrrrrrr}
\toprule
\textbf{Kernel Size \textbackslash Dilation Rate} & 1 & 2 & 3 & 4 \\ \midrule
3 & 63.710 & 67.440 & 69.630 & 71.705 \\
5 & 63.939 & 70.324 & 73.049 & 73.688 \\
7 & 68.353 & 71.141 & 74.049 & 74.598 \\
9 & 68.959 & 72.257 & 74.241 & 74.530 \\
11 & 69.408 & 73.437 & 74.835 & 74.549 \\
\bottomrule
\end{tabular}
}
\label{tab:dilation_kernel}
\end{table}

Furthermore, we observe that even for short-range tasks (all having a sequence length of 500 bp), the intensity of local dependencies varies significantly. For instance, the tasks H3K4me2, H3K4me3, and H3K14ac, as discussed in %Section
\S 
\ref{discussion}, exhibit strong local dependencies, with better performance observed when the dilation rate is small. On the other hand, tasks like splice site prediction show a much stronger global dependency, which benefits from a larger dilation rate. See Table \ref{nt_dilation_kernel} for more details.

\begin{table}[htbp]
    \centering
    \caption{Performance of different dilation rates and kernel sizes on the NT benchmark. For example, \texttt{k3\_d1} represents a kernel size of 3 and a dilation rate of 1, with similar notation for other configurations.}
    \begin{minipage}{0.9\textwidth}
        \centering
        \adjustbox{max width=\textwidth}{
        \begin{tabular}{lcccccccc}
            \toprule
            \multicolumn{1}{l}{\multirow{2}{*}{\textbf{Task}}} & \multicolumn{8}{c}{\textbf{Histone Modifications}} \\
                               & \textbf{H3} & \textbf{H3K4me1} & \textbf{H3K4me2} & \textbf{H3K4me3} & \textbf{H3K9ac} & \textbf{H3K14ac} & \textbf{H3K36me3} & \textbf{H3K79me3} \\
            \midrule
            k3\_d1 & 70.14 & 43.14 & 41.26 & 49.12 & 54.00 & 54.80 & 54.48 & 62.38 \\
            k3\_d2 & 74.66 & 50.54 & 47.52 & 56.50 & 58.92 & 60.16 & 63.18 & 67.50 \\
            k3\_d3 & 79.24 & 52.44 & 49.50 & 55.18 & 61.60 & 62.72 & 63.66 & 68.50 \\
            k3\_d4 & 78.72 & 51.94 & 47.54 & 54.12 & 62.36 & 62.30 & 63.74 & 68.48 \\
            k5\_d1 & 70.76 & 50.26 & 48.28 & 56.86 & 58.02 & 61.62 & 60.44 & 65.38 \\
            k5\_d2 & 76.80 & 55.88 & 53.50 & 62.30 & 63.58 & 66.24 & 65.04 & 70.30 \\
            k5\_d3 & 80.88 & 55.58 & 52.16 & 58.86 & 65.62 & 65.22 & 66.02 & 70.84 \\
            k5\_d4 & 80.48 & 55.12 & 50.46 & 56.56 & 65.60 & 63.58 & 65.58 & 69.70 \\
            k7\_d1 & 71.76 & 53.12 & 50.72 & 59.78 & 61.06 & 65.16 & 62.80 & 67.32 \\
            k7\_d2 & 77.72 & 56.62 & 53.66 & 62.54 & 65.96 & 66.92 & 66.06 & 70.78 \\
            k7\_d3 & 81.20 & 56.42 & 53.38 & 58.04 & 66.56 & 65.30 & 66.54 & 70.96 \\
            k7\_d4 & 81.62 & 56.46 & 51.02 & 57.60 & 66.38 & 64.24 & 65.96 & 69.10 \\
            k9\_d1 & 72.52 & 54.70 & 52.90 & 60.50 & 62.26 & 66.02 & 63.94 & 69.14 \\
            k9\_d2 & 78.02 & 56.76 & 55.38 & 62.04 & 66.80 & 67.46 & 67.62 & 71.52 \\
            k9\_d3 & 81.58 & 56.16 & 52.20 & 58.64 & 66.70 & 65.00 & 66.28 & 71.84 \\
            k9\_d4 & 81.18 & 55.36 & 48.54 & 56.60 & 65.84 & 64.40 & 66.08 & 70.02 \\
            k11\_d1 & 72.60 & 55.90 & 53.78 & 62.80 & 63.02 & 66.40 & 66.14 & 69.58 \\
            k11\_d2 & 79.62 & 56.74 & 55.20 & 63.84 & 67.08 & 67.24 & 68.20 & 71.54 \\
            k11\_d3 & 81.66 & 56.12 & 53.98 & 58.66 & 66.88 & 64.62 & 66.84 & 71.22 \\
            k11\_d4 & 80.66 & 55.16 & 47.84 & 55.76 & 66.82 & 63.20 & 66.28 & 70.18 \\
            \bottomrule
        \end{tabular}
        }
    \end{minipage}
    \vspace{1cm} % 这里控制表格之间的空白
    \begin{minipage}{0.9\textwidth}
        \centering
        \adjustbox{max width=\textwidth}{
        \begin{tabular}{lcccccccc}
            \toprule
            \multicolumn{1}{l}{\multirow{2}{*}{\textbf{Task}}} & \multicolumn{3}{c}{\textbf{promoter}} & \multicolumn{3}{c}{\textbf{splice\_sites}} & \multicolumn{2}{c}{\textbf{enhancer}} \\
            & \textbf{all} & \textbf{non\_tata} & \textbf{tata} & \textbf{all} & \textbf{acceptor} & \textbf{donor} &  & \textbf{types} \\
            \midrule
            k3\_d1 & 94.68 & 94.82 & 94.54 & 47.08 & 79.72 & 78.16 & 54.04 & 48.96 \\
            k3\_d2 & 96.06 & 96.04 & 95.74 & 49.26 & 80.62 & 79.84 & 52.36 & 49.76 \\
            k3\_d3 & 96.50 & 96.44 & 96.20 & 60.36 & 84.10 & 83.10 & 53.84 & 50.66 \\
            k3\_d4 & 96.62 & 96.64 & 96.38 & 91.42 & 87.54 & 87.12 & 56.46 & 48.96 \\
            k5\_d1 & 95.10 & 95.30 & 94.88 & 48.20 & 79.84 & 78.20 & 52.44 & 51.30 \\
            k5\_d2 & 96.36 & 96.48 & 96.32 & 53.84 & 82.22 & 81.28 & 52.96 & 50.50 \\
            k5\_d3 & 96.68 & 96.80 & 96.54 & 88.70 & 86.96 & 85.74 & 55.68 & 49.62 \\
            k5\_d4 & 96.76 & 96.74 & 96.18 & 96.38 & 96.08 & 95.62 & 56.14 & 51.84 \\
            k7\_d1 & 95.36 & 95.46 & 95.46 & 48.50 & 79.72 & 78.92 & 55.70 & 51.48 \\
            k7\_d2 & 96.54 & 96.48 & 96.32 & 59.54 & 83.76 & 83.12 & 53.26 & 49.00 \\
            k7\_d3 & 96.76 & 96.70 & 96.28 & 95.38 & 88.86 & 88.82 & 57.42 & 49.88 \\
            k7\_d4 & 96.76 & 96.68 & 96.48 & 96.24 & 97.02 & 96.40 & 58.02 & 50.02 \\
            k9\_d1 & 95.56 & 95.66 & 95.28 & 49.16 & 80.36 & 79.40 & 53.58 & 49.60 \\
            k9\_d2 & 96.54 & 96.54 & 96.34 & 70.96 & 84.78 & 84.16 & 54.06 & 48.12 \\
            k9\_d3 & 96.70 & 96.66 & 96.40 & 95.94 & 91.98 & 90.98 & 56.82 & 48.94 \\
            k9\_d4 & 96.74 & 96.82 & 96.92 & 96.54 & 97.08 & 96.48 & 58.60 & 50.00 \\
            k11\_d1 & 95.68 & 95.84 & 95.06 & 49.08 & 80.78 & 79.04 & 54.56 & 49.18 \\
            k11\_d2 & 96.40 & 96.46 & 96.40 & 84.60 & 85.62 & 84.84 & 55.92 & 48.20 \\
            k11\_d3 & 96.64 & 96.58 & 96.22 & 95.84 & 96.86 & 95.62 & 57.16 & 48.58 \\
            k11\_d4 & 96.82 & 96.66 & 96.60 & 96.76 & 97.16 & 86.46 & 61.68 & 50.80 \\
            \bottomrule
        \end{tabular}
        }
    \end{minipage}
    \label{nt_dilation_kernel}
\end{table}



\subsection{Supervised Methods against Foundation Models}\label{app:supvsfoundation}
We conduct a comparison between previous supervised models and DNA foundation models on both the NT benchmark and genomic benchmark. For our comparison, we select the popular Basenji and the recently established state-of-the-art model, LegNet, which excels in short DNA regulatory regions.  Originally, these supervised models are designed for specific tasks. We apply them to the benchmarks used for foundation models. As observed in Table \ref{table:traditional_vs_foundation_nt} and Table \ref{table:traditional_vs_foundation_genomic_benchmark}, although they show decent performance on specific tasks, they do not achieve the same level of consistency across various datasets as foundation models do. This inconsistency may arise from the enhanced modeling capabilities afforded by pretraining in foundation models.
\begin{table}[h]
  \caption{\textbf{Supervised Methods against Foundation Models on NT Benchmark.} MCC/F1-score is reported for pretrained NTv2, HyenaDNA, DNABERT-2, Caduceus-Ph, ConvNova, Basenji, LegNet ConvNova*(* means trained from scratch)}. The best values per task are bold, and the second-best are underlined. ± indicates the error range across five random seeds.
  \label{table:traditional_vs_foundation_nt}
  \centering
  \adjustbox{max width=\textwidth}{
  \begin{tabular}{lrrrr>{\columncolor{low}}rrrr}
    \toprule
    & \textbf{NTv2} & \textbf{HyenaDNA} & \textbf{DNABERT-2} & \textbf{Caduceus-Ph} & \textbf{ConvNova} & \textbf{Basenji} & \textbf{LegNet} & \textbf{ConvNova*}\\ 
    & (500M)     & (1.6M)      & (117M)    & (1.9M)    & (1.7M)   & (7.4M) & (2.1M) & (1.7M) \\
    \midrule
    \rowcolor{titlebg} \textit{\textbf{Histone}} & & & & & & & & \\
    \midrule
    H3              & 78.17 \small{±2.54} &  78.14 \small{±1.70} &  79.31 \small{±0.68} &  80.48 \small{±1.04} &  \textbf{81.50} \small{±0.80} & 78.05 \small{±1.83} & 76.24 \small{±0.83} & \underline{81.18} \small{±1.68}\\
    H3K4me1         & 51.64 \small{±1.12}  &  44.52 \small{±2.59} &  48.34 \small{±4.63} &  52.83 \small{±0.96} &  \textbf{56.60} \small{±1.01} & 42.70 \small{±3.16} & 47.47 \small{±1.36} & \underline{55.36} \small{±2.26}  \\
    H3K4me2         & 37.24 \small{±2.25}  &  42.68 \small{±2.66} &  43.02 \small{±2.92} &  \underline{49.88} \small{±2.65} &  \textbf{57.45} \small{±2.27} & 34.73 \small{±1.66} & 45.75 \small{±0.56} & 48.54 \small{±3.24} \\
    H3K4me3         & 50.30 \small{±1.77}  &  50.41 \small{±3.15} &  45.43 \small{±3.33} &  \underline{56.72} \small{±2.58} &  \textbf{67.15} \small{±0.93} & 38.85 \small{±1.49} & 52.67 \small{±2.54} & 56.60 \small{±2.70} \\
    H3K9ac          & 61.05 \small{±1.40}  &  58.50 \small{±1.75} &  60.04 \small{±1.27} &  63.27 \small{±2.29} &  \textbf{68.10} \small{±1.91} & 56.34 \small{±2.86} & 58.21 \small{±0.92} & \underline{65.84} \small{±1.14}\\
    H3K14ac         & 57.22 \small{±2.19}  &  56.71 \small{±2.40} &  54.49 \small{±4.99} &  60.84 \small{±2.94} &  \textbf{70.71} \small{±2.32} & 49.88 \small{±1.60} & 56.64 \small{±2.12} & \underline{64.40} \small{±0.90}\\
    H3K36me3        & 60.50 \small{±1.75}  &  59.92 \small{±1.06} &  57.58 \small{±2.38} &  61.12 \small{±1.44} &  \textbf{68.31} \small{±1.19} & 52.73 \small{±1.84} & 57.54 \small{±1.03} & \underline{66.08} \small{±1.88} \\
    H3K79me3        & 65.78 \small{±2.34}  &  66.25 \small{±3.65} &  64.38 \small{±0.48} &  67.17 \small{±2.03} &  \textbf{72.08} \small{±1.23} & 62.43 \small{±1.29} & 63.78 \small{±1.72} & \underline{70.02} \small{±0.72} \\
    H4              & 79.87 \small{±1.34}  &  78.15 \small{±1.58} &  78.18 \small{±0.98} &  80.10 \small{±1.00} &  \textbf{81.12} \small{±0.93} & 77.60 \small{±2.02} & 78.73 \small{±1.93} & \underline{80.78} \small{±0.82} \\
    H4ac            & 55.22 \small{±2.20}  &  54.15 \small{±2.96} &  51.80 \small{±0.10} &  59.26 \small{±3.67} &  \textbf{66.10} \small{±1.20} & 46.05 \small{±1.77} & 55.29 \small{±2.60} & \underline{63.56} \small{±1.14} \\
    \midrule
    \rowcolor{titlebg} \textit{\textbf{Regulatory}} & & & & & & & & \\
    \midrule
    Enhancer     & 54.51 \small{±1.94}  &  53.13 \small{±4.52} &  52.50 \small{±1.44} &  55.20 \small{±2.56} &  \textbf{57.60} \small{±2.52} & 57.54 \small{±2.75} & 55.25 \small{±3.23} & \underline{58.60} \small{±3.80} \\
    Enhancer Types & 43.36 \small{±1.75}  &  48.16 \small{±2.48} &  44.32 \small{±1.18} &  47.17 \small{±2.85} &  49.75 \small{±2.82} & \textbf{52.67} \small{±2.94} & 44.72 \small{±2.04} & \underline{50.00} \small{±1.80} \\
    Promoter All   & \textbf{96.82} \small{±0.47}  &  95.57 \small{±0.18} &  96.23 \small{±0.17} &  96.65 \small{±0.16} &  \textbf{96.82} \small{±0.22} & 95.79 \small{±0.24} & 96.52 \small{±0.33} & 96.74 \small{±0.06}\\
    Promoter non-TATA & \textbf{97.45} \small{±0.69}  &  95.86 \small{±0.37} &  \underline{97.17} \small{±0.17} &  96.31 \small{±0.50} &  96.76 \small{±0.21} & 95.99 \small{±0.18} & 96.45 \small{±0.27} & 96.82 \small{±0.28}\\
    Promoter TATA  & 96.53 \small{±0.81}  &  95.88 \small{±0.53} &  \textbf{96.99} \small{±0.49} &  96.21 \small{±0.81} &  96.34 \small{±0.38} & 95.89 \small{±0.78} & 96.25 \small{±0.45} & \underline{96.92} \small{±0.32} \\
    \midrule
    \rowcolor{titlebg} \textit{\textbf{Splice sites}} & & & & & & & \\
    \midrule
    Splice Acceptor & \textbf{97.99} \small{±0.66}  &  96.98 \small{±0.49} &  97.49 \small{±0.36} &  94.21 \small{±0.37} &  96.23 \small{±0.41} & \underline{97.65} \small{±0.34} & 85.80 \small{±0.58} & 97.08 \small{±0.32} \\
    Splice Donor    & \textbf{98.50} \small{±0.43}  &  95.27 \small{±1.07} &  94.33 \small{±0.27} &  94.69 \small{±0.67} &  96.62 \small{±0.61} & \underline{97.82} \small{±0.66} & 84.72 \small{±0.95} & 96.48 \small{±0.32} \\
    Splice All      & \textbf{98.15} \small{±1.01}  &  94.05 \small{±1.08} &  93.75 \small{±1.25} &  92.87 \small{±1.73} &  96.33 \small{±0.31} & \underline{98.07} \small{±0.45} & 54.71 \small{±2.38} & 96.54 \small{±0.36} \\
    \bottomrule
  \end{tabular}
  }
\end{table}

\begin{table}[ht]
  \caption{\textbf{Supervised Methods against Foundation Models on Genomic Benchmark.} Top-1 accuracy (↑) is reported for pretrained HyenaDNA, Caduceus-Ph, ConvNova, Basenji, LegNet, and the original CNN baseline (trained from scratch). The best values are in bold, and the second-best is underlined. ± indicates the error range across five random seeds.}
  \label{table:traditional_vs_foundation_genomic_benchmark}
  \centering
  \adjustbox{max width=\textwidth}{
  \begin{tabular}{lrrr>{\columncolor{low}}rrr}
    \toprule
    Task    & \textbf{CNN}        & \textbf{HyenaDNA}   & \textbf{Caduceus-Ph} & \textbf{ConvNova}  & \textbf{Basenji}   & \textbf{LegNet} \\
             & (264K)     & (436K)     & (470K)      & (386K)   & (7.4N)              & (2.1M)            \\
    \midrule
    \rowcolor{titlebg} \textit{\textbf{Enhancers}} & & & & & & \\
    \midrule
    Mouse Enhancers         &  0.730 \small{±0.032} &  \underline{0.779} \small{±0.013} &  0.754 \small{±0.074} &  \textbf{0.784} \small{±0.009} &  0.659 \small{±0.155} &  0.504 \small{±0.000} \\
    Human Enhancers Cohn    &  0.702 \small{±0.021} &  0.718 \small{±0.008} &  \textbf{0.747} \small{±0.004} &  \underline{0.743} \small{±0.005} &  0.712 \small{±0.030} &  0.739 \small{±0.004} \\
    Human Enhancer Ensembl  &  0.744 \small{±0.122} &  0.832 \small{±0.006} &  0.893 \small{±0.008} &  \underline{0.900} \small{±0.004} &  \textbf{0.905} \small{±0.007} &  0.879 \small{±0.002} \\
    \midrule
    \rowcolor{titlebg} \textit{\textbf{Species Classification}} & & & & & & \\
    \midrule
    Coding vs. Intergenomic &  0.892 \small{±0.008} &  0.904 \small{±0.008} &  0.915 \small{±0.003} &  \textbf{0.943} \small{±0.001} &  0.905 \small{±0.004} &  \underline{0.939} \small{±0.002} \\
    Human vs. Worm          &  0.942 \small{±0.002} &  0.961 \small{±0.002} &  \textbf{0.973} \small{±0.001} &  \underline{0.967} \small{±0.002} &  0.957 \small{±0.003} &  0.965 \small{±0.001} \\
    \midrule
    \rowcolor{titlebg} \textit{\textbf{Regulatory Elements}} & & & & & & \\
    \midrule
    Human Regulatory        &  0.872 \small{±0.005} &  0.862 \small{±0.004} &  \underline{0.872} \small{±0.011} &  \textbf{0.873} \small{±0.002} &  0.764 \small{±0.005} &  0.764 \small{±0.006} \\
    Human Non-TATA Promoters &  0.861 \small{±0.009} &  0.887 \small{±0.005} &  \underline{0.946} \small{±0.007} &  \textbf{0.951} \small{±0.003} &  0.919 \small{±0.006} &  0.942 \small{±0.007} \\
    Human OCR Ensembl       &  0.698 \small{±0.013} &  0.744 \small{±0.019} &  \textbf{0.828} \small{±0.006} &  0.793 \small{±0.004} &  0.766 \small{±0.009} &  \underline{0.802} \small{±0.004} \\
    \bottomrule
  \end{tabular}
  }
\end{table}

\subsection{Neighborhood Importance in DNA Modeling}

To support our hypothesis that the inductive bias of CNNs—emphasizing neighboring nucleotides—is beneficial for DNA modeling, we make modifications to the Rotary Position Embedding (RoPE) mechanism. Specifically, we adjust the original $\theta$ values, enabling each attention head to have distinct $\theta$ values.  

Additionally, we modified the initialization of the bias term in the linear layers of the $K$ and $Q$ projections. The bias was initialized to $[0, 0, 0, 0, 0, \ldots, 1]$, while all other parameters retained the initialization strategy of NTv2 (standard deviation $\sigma = 0.02$, mean $\mu = 0$). This adjustment ensures that the initialization process places greater emphasis on neighboring nucleotides. Refer to Figure  \ref{fig:local_weight_more} for an illustration of the improved attention map, which places greater emphasis on neighboring tokens.

To validate our hypothesis, we train NTv2 from scratch on three tasks where ConvNova with small dilation values performs better (H3K4me2, H3K4me3, and H3K14ac). The results, presented in Table \ref{table:local_weight_more}, show that the adjusted NTv2 consistently outperform its original version. However, despite these improvements, our naive modifications to the transformer architecture does not surpass ConvNova trained from scratch(See Table \ref{table:traditional_vs_foundation_nt}). To enhance transformer and state space model (SSM) designs for DNA modeling further dedicated efforts is needed. 

Our primary goal here is to validate the hypothesis and provide an explanation for why CNNs might outperform transformers in this context.
\begin{table}[h!]
\caption{Performance comparison of NTv2 and NTv2*, where NTv2* represents our modified version. Results demonstrate that NTv2* achieves significant improvements.}
\centering
\renewcommand{\arraystretch}{1.3}
\setlength{\tabcolsep}{8pt}
\begin{tabular}{ccc}
\toprule
\multicolumn{1}{c}{\textbf{Task}} & \multicolumn{1}{c}{\textbf{NTv2*}} & \multicolumn{1}{c}{\textbf{NTv2}} \\
\midrule
H3K14ac & 46.65 & 34.42 \\
H3K4me2 & 33.10 & 25.79 \\
H3K4me3 & 34.62 & 21.40 \\
\bottomrule
\end{tabular}
\label{table:local_weight_more}
\end{table}

\begin{figure}[h!]
\centering
\begin{minipage}{0.45\textwidth}
    \centering
    \includegraphics[width=\textwidth]{imgs/modified_ntv2.png}  % Replace with your image path
    \label{fig:attention_map_ntv2_star}
\end{minipage}%
\hspace{0.02\textwidth}  % Adjust this value to control the gap between images
\begin{minipage}{0.45\textwidth}
    \centering
    \includegraphics[width=\textwidth]{imgs/original_ntv2.png}  % Replace with your image path
    \label{fig:attention_map_ntv2}
\end{minipage}
\caption{Visualization of the attention maps for NTv2* (Left) and NTv2 (Right). The asterisk (*) denotes the modified version. The modified initialization places more emphasis on neighboring tokens in the attention map.}
\end{figure}\label{fig:local_weight_more}



\begin{table}[htbp]
  \caption{\textbf{Results for all models on NT benchmark across genomic regions.} Ten random seeds were used for each model. The best values per task are bolded.}
  \label{table:ntall1}
  \centering
  \resizebox{\textwidth}{!}{
    \begin{tabular}{lllllllllll}
    \toprule
    \textbf{Seed}  &  2222  & 42 & 43 & 44 & 45 & 46 & 47 & 48 & 49 & 50             \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Promoter All}}} \\
    \rowcolor{low} ConvNova (4 dilation) & 96.99 & 96.84 & 96.81 & 96.82 & 96.75 & 96.82 & 96.64 & 96.60  & 96.96 & 96.96 \\
    Caduceus-Ph  & 96.66 & 96.61 & 96.64 & 96.61 & 96.51 & 96.79 & 96.77 & 96.49 & 96.73 & 96.68 \\
    HyenaDNA & 95.55 & 95.62 & 95.44 & 95.58 & 95.62 & 95.58 & 95.56 & 95.65 & 95.75 & 95.39 \\
    DNABERT-2 & 96.38  & 96.27  & 96.15  & 96.21  & 96.05  & 96.19  & 96.31  & 96.22  & 96.29  & 96.11 \\
    NTv2 & 96.47 & 97.19 & 96.95 & 96.35 & 97.12 & 96.63 & 97.21 & 96.85 & 96.40 & 97.00 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Promoter Non-TATA}}} \\
    \rowcolor{low} ConvNova (4 dilation)  & 96.55 & 96.81 & 96.73 & 96.83 & 96.88 & 96.61 & 96.77 & 96.83 & 96.87 & 96.72 \\
    Caduceus-Ph & 96.14 & 95.81 & 96.14 & 96.52 & 96.79 & 95.98 & 96.47 & 96.36 & 96.39 & 96.50 \\
    HyenaDNA & 95.81 & 96.23 & 95.66 & 96.00    & 95.96 & 95.60  & 95.63 & 96.04 & 95.87 & 95.83 \\
    DNABERT-2 & 97.18  & 97.02  & 97.31  & 97.22  & 97.04  & 97.29  & 97.15  & 97.03  & 97.26  & 97.12 \\
    NTv2 & 96.89 & 97.44 & 98.01 & 96.81 & 97.29 & 98.14 & 97.52 & 96.78 & 97.93 & 97.66 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Promoter TATA}}} \\
    \rowcolor{low}ConvNova (4 dilation)& 96.57 & 96.08 & 95.92 & 96.57 & 96.07 & 96.40  & 96.57 & 96.09 & 96.41 & 96.72 \\
    Caduceus-Ph & 95.40  & 96.72 & 96.40  & 95.73 & 96.23 & 96.74 & 96.24 & 95.71 & 96.56 & 96.41 \\
    HyenaDNA  & 95.92 & 96.06 & 96.23 & 95.89 & 96.41 & 95.55 & 95.58 & 96.16 & 95.58 & 95.42 \\
    DNABERT-2 & 97.38  & 97.12  & 96.52  & 97.41  & 97.18  & 96.53  & 97.40  & 97.15  & 96.54  & 96.87 \\
    NTv2 & 95.75 & 96.89 & 97.13 & 95.92 & 96.48 & 96.12 & 97.05 & 96.73 & 95.88 & 97.34 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Splice Sites All}}} \\
    \rowcolor{low}ConvNova (4 dilation) & 96.46 & 96.18 & 96.6  & 96.41 & 96.28 & 96.50  & 95.82 & 96.30  & 95.93 & 95.85 \\
    Caduceus-Ph & 96.29 & 96.33 & 96.21 & 96.36 & 96.58 & 96.04 & 96.19 & 95.84 & 96.03 & 96.27 \\
    HyenaDNA & 96.97 & 96.99 & 96.93 & 96.74 & 96.96 & 97.22 & 97.05 & 96.89 & 97.47 & 96.61 \\
    DNABERT-2  & 94.32  & 94.18  & 93.71  & 94.29  & 94.21  & 93.72  & 94.34  & 94.23  & 93.74  & 94.03 \\
    NTv2 & 97.14 & 98.47 & 98.97 & 97.88 & 98.63 & 98.09 & 97.57 & 98.42 & 98.28 & 98.05 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Splice Sites Acceptor}}} \\
    \rowcolor{low}ConvNova (4 dilation) & 96.65 & 96.01 & 96.87 & 96.70  & 96.86 & 96.59 & 96.37 & 97.06 & 96.50  & 96.59 \\
    Caduceus-Ph & 97.36 & 96.41 & 96.74 & 96.61 & 96.24 & 96.83 & 96.74 & 96.87 & 96.50  & 96.56 \\
    HyenaDNA & 94.20  & 95.37 & 95.21 & 94.97 & 96.06 & 95.94 & 95.2  & 94.48 & 95.73 & 95.56 \\
    DNABERT-2 & 94.57  & 94.21  & 94.23  & 94.55  & 94.19  & 94.24  & 94.59  & 94.18  & 94.25  & 94.28 \\
    NTv2 & 98.57 & 97.89 & 98.14 & 97.45 & 98.64 & 97.72 & 98.09 & 97.33 & 98.38 & 97.69 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Splice Sites Donor}}} \\
    \rowcolor{low}ConvNova (4 dilation) & 96.57 & 96.34 & 96.52 & 96.51 & \underline{96.44} & 96.02 & 96.04 & 96.38 & 96.12 & 96.33 \\
    Caduceus-Ph & 92.47 & 93.21 & 94.74 & 92.61 & 92.78 & 93.56 & 91.14 & 94.36 & 92.57 & 91.22 \\
    HyenaDNA & 94.38 & 94.59 & 95.04 & 93.09 & 94.63 & 93.38 & 94.52 & 93.67 & 92.97 & 94.22 \\
    DNABERT-2 & 92.51  & 94.76  & 93.72  & 92.48  & 94.81  & 93.74  & 92.53  & 94.79  & 93.71  & \textbf{94.52} \\
    NTv2 & 98.93 & 98.67 & 98.43 & 98.12 & 98.54 & 98.79 & 98.26 & 98.68 & 98.37 & 98.21 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Enhancer}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 59.00    & 55.08 & 57.09 & 58.54 & 55.33 & 58.26 & 58.77 & 58.42 & 58.10  & 57.39 \\
    Caduceus-Ph & 53.72 & 53.64 & 54.04 & 53.50  & 57.76 & 56.75 & 55.94 & 55.35 & 54.66 & 56.63 \\
    HyenaDNA & 55.34 & 53.88 & 54.43 & 51.44 & 51.56 & 54.28 & 55.53 & 48.61 & 53.18 & 53.05 \\
    DNABERT-2  & 52.11 & 51.06 & 52.15 & 53.67 & 53.23 & 53.24 & 51.56 & 52.08 & 52.88 & 53.06 \\
    NTv2 & 53.25 & 55.80 & 54.10 & 52.99 & 56.45 & 54.75 & 53.40 & 55.50 & 54.99 & 53.85 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{Enhancer Types}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 49.08 & 48.32 & 49.24 & 52.56 & 48.72 & 48.90  & 51.24 & 49.45 & 49.96 & 49.98 \\
    Caduceus-Ph  & 47.78 & 44.58 & 45.93 & 46.01 & 48.04 & 50.02 & 47.98 & 47.09 & 46.93 & 47.35 \\
    HyenaDNA & 47.71 & 47.77  & 48.06 & 48.49 & 50.34 & 49.67 & 49.45 & 48.43 & 45.68 & 46.09 \\
    DNABERT-2 & 43.68 & 44.17 & 43.29 & 45.50  & 43.23 & 45.29 & 44.22 & 45.18 & 45.40  & 43.28 \\
    NTv2 & 42.15 & 44.90 & 43.20 & 41.85 & 44.25 & 43.80 & 42.50 & 45.10 & 43.05 & 42.75 \\
\bottomrule
    \end{tabular}}
\end{table}

\begin{table}[htbp]
  \caption{\textbf{(Cont.) Results for all models on NT benchmark across genomic regions.}}
  \label{table:ntall}
  \centering
  \resizebox{\textwidth}{!}{
    \begin{tabular}{lllllllllll}
    \toprule
    \textbf{Task/Model}  &  2222  & 42 & 43 & 44 & 45 & 46 & 47 & 48 & 49 & 50             \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3}}} \\
    \rowcolor{low}ConvNova (4 dilation) & 81.62 & 81.78 & 81.04 & 82.24 & 82.12 & 82.06 & 81.36 & 81.12 & 80.91 & 80.69 \\
    \rowcolor{low}ConvNova (1 dilation) & 76.90  & 77.14 & 77.28 & 77.73 & 76.80  & 76.59 & 77.24 & 77.26 & 77.68 & 76.96 \\
    Caduceus-Ph& 80.29 & 79.85 & 78.63 & 81.47 & 80.24 & 81.50  & 81.35 & 80.35 & 79.95 & 81.02 \\
    HyenaDNA  & 76.88 & 77.01 & 78.23 & 77.24 & 79.84 & 78.39 & 77.56 & 79.27 & 79.11 & 77.90 \\
    DNABERT-2  & 79.61 & 79.34 & 79.12 & 78.90  & 79.02 & 79.61 & 78.68 & 79.42 & 79.99 & 79.41 \\
    NTv2 & 76.55 & 80.32 & 78.30 & 77.52 & 80.71 & 79.25 & 76.89 & 77.82 & 76.94 & 77.38 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3K4me1}}} \\
    \rowcolor{low}ConvNova (4 dilation) & 58.88 & 56.99 & 56.37 & 56.20  & 56.70  & 55.59 & 55.71 & 55.71 & 56.48 & 57.41 \\
    \rowcolor{low}ConvNova (1 dilation) & 55.67 & 57.8  & 56.62 & 57.64 & 55.28 & 55.65 & 55.80  & 56.32 & 57.44 & 55.82 \\
    Caduceus-Ph & 53.31 & 53.00    & 54.54 & 51.87 & 53.40  & 53.42 & 52.34 & 51.91 & 52.01 & 52.45 \\
    HyenaDNA & 43.04 & 44.54 & 44.04 & 44.55 & 46.02 & 44.9  & 41.93 & 45.26 & 46.13 & 44.77 \\
    DNABERT-2 & 50.32 & 47.21 & 48.09 & 49.38 & 49.22 & 43.71 & 46.91 & 50.22 & 47.82 & 50.51 \\
    NTv2 & 51.56 & 50.96 & 52.35 & 51.81 & 50.52 & 52.03 & 52.31 & 52.72 & 51.25 & 50.93 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3K4me2}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 55.80  & 54.68 & 51.81 & 52.52 & 54.51 & 52.20  & 51.30  & 55.01 & 54.91 & 54.45 \\
    \rowcolor{low}ConvNova (1 dilation) & 56.53 & 59.72 & 57.26 & 58.60  & 56.13 & 56.35 & 58.72 & 56.25 & 58.67 & 56.24 \\
    Caduceus-Ph& 47.96 & 47.23 & 52.47 & 50.14 & 51.35 & 50.14 & 52.18 & 50.01 & 48.72 & 48.56 \\
    HyenaDNA  & 40.5  & 40.02 & 42.43 & 43.32 & 41.80  & 44.07 & 43.78 & 42.39 & 44.82 & 43.65 \\
    DNABERT-2  & 43.10  & 42.10  & 45.20  & 41.90  & 43.01 & 40.10  & 43.80  & 44.40  & 43.40  & 43.20 \\
    NTv2 & 37.76 & 36.32 & 37.59 & 37.57 & 35.56 & 37.87 & 39.49 & 37.04 & 36.35 & 36.86 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3K4me3}}} \\
    \rowcolor{low}ConvNova (4 dilation)   & 59.23 & 62.11 & 60.42 & 59.98 & 59.82 & 60.10  & 60.58 & 58.73 & 61.45 & 59.60 \\
    \rowcolor{low}ConvNova (1 dilation)   & 67.13 & 66.8  & 67.97 & 67.79 & 66.45 & 67.01 & 66.75 & 68.00    & 67.42 & 66.22 \\
    Caduceus-Ph& 56.93 & 56.09 & 58.64 & 54.56 & 57.95 & 57.84 & 58.54 & 56.49 & 54.14 & 56.03 \\
    HyenaDNA  & 47.65 & 52.13 & 50.12 & 52.10  & 52.32 & 51.43 & 49.46 & 49.72 & 47.26 & 51.90 \\
    DNABERT-2 & 46.30  & 46.40  & 46.60  & 46.00  & 45.80  & 45.60  & 46.40  & 44.80  & 44.30  & 42.10 \\
    NTv2 & 51.70 & 48.94 & 50.52 & 49.17 & 52.03 & 50.15 & 49.87 & 51.34 & 50.75 & 48.53 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3K9ac}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 68.98 & 69.98 & 66.18 & 67.54 & 68.39 & 66.18 & 67.54 & 68.39 & 68.67 & 69.10 \\
    \rowcolor{low}ConvNova (1 dilation) & 63.85 & 65.33 & 67.12 & 65.15 & 66.24 & 63.66 & 65.38 & 66.82 & 65.17 & 66.20 \\
    Caduceus-Ph& 63.78 & 62.28 & 64.20  & 65.04 & 61.06 & 60.98 & 65.24 & 63.59 & 63.55 & 62.95 \\
    HyenaDNA  & 57.57 & 57.51 & 58.37 & 60.25 & 58.57 & 59.78 & 59.19 & 59.31 & 57.53 & 56.88 \\
    DNABERT-2  & 58.80  & 60.40  & 61.00  & 58.82  & 60.41  & 60.98  & 58.77  & 60.78  & 61.23  & 59.20 \\
    NTv2 & 60.45 & 61.90 & 62.22 & 59.98 & 60.85 & 61.57 & 62.10 & 60.02 & 61.78 & 59.65 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3K14ac}}} \\
    \rowcolor{low}ConvNova (4 dilation) & 66.24 & 66.61 & 65.91 & 68.03 & 65.98 & 65.83 & 65.6  & 66.58 & 64.93 & 66.22 \\
    \rowcolor{low}ConvNova (1 dilation) & 69.76 & 73.03 & 69.26 & 72.57 & 69.95 & 69.88 & 71.03 & 69.46 & 72.23 & 69.92 \\
    Caduceus-Ph& 61    & 59.36 & 62.4  & 60.7  & 62.14 & 57.9  & 62.83 & 61.31 & 59.65 & 61.08 \\
    HyenaDNA  & 56.08 & 57.59 & 54.5  & 54.31 & 57.31 & 55.19 & 58.26 & 57.09 & 57.95 & 58.85 \\
    DNABERT-2  & 56.31  & 55.72  & 53.42  & 55.14  & 55.61  & 55.36  & 49.52  & 54.54  & 53.81  & 55.72 \\
    NTv2 & 55.03 & 58.65 & 56.74 & 59.18 & 57.01 & 55.82 & 57.33 & 58.40 & 56.10 & 57.92 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3K36me3}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 67.12 & 67.58 & 68.44 & 69.43 & 68.07 & 68.54 & 69.13 & 67.75 & 67.85 & 69.22 \\
    \rowcolor{low}ConvNova (1 dilation)  & 69.31 & 67.01 & 66.1  & 66.43 & 67.17 & 68.31 & 66.66 & 66.24 & 66.59 & 67.01 \\
    Caduceus-Ph & 62.53 & 61.53 & 61.16 & 62.3  & 59.68 & 61.42 & 61.46 & 60.33 & 60.81 & 59.93 \\
    HyenaDNA & 60.27 & 60.52 & 59.8  & 59.01 & 58.88 & 59.66 & 60.73 & 60.09 & 60.98 & 59.3 \\
    DNABERT-2  & 59.12  & 58.11  & 58.11  & 57.16  & 56.14  & 57.15  & 58.25  & 58.21  & 58.65  & 55.27 \\
    NTv2 & 58.75 & 61.42 & 60.15 & 62.10 & 59.81 & 61.68 & 60.25 & 58.90 & 60.77 & 61.12 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H3K79me3}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 72.24 & 71.37 & 71.05 & 70.85 & 72.41 & 72.73 & 72.67 & 72.19 & 72.35 & 72.91 \\
    \rowcolor{low}ConvNova (1 dilation)  & 70.99 & 71.83 & 69.34 & 72.57 & 71.17 & 70.66 & 71.85 & 70.23 & 71.24 & 71.67 \\
    Caduceus-Ph& 68.78 & 68.28 & 67.77 & 68.02 & 65.14 & 66.89 & 66.37 & 66.49 & 66.58 & 67.42 \\
    HyenaDNA & 62.6  & 65.71 & 67.33 & 66.82 & 66.97 & 65.52 & 65.23 & 67.68 & 67.63 & 66.99 \\
    DNABERT-2  & 64.10  & 64.10  & 65.10  & 64.10  & 64.12  & 65.13  & 64.32  & 64.21  & 64.82  & 63.91 \\
    NTv2 & 64.23 & 66.14 & 67.22 & 65.99 & 63.87 & 65.43 & 68.12 & 64.68 & 66.79 & 65.31 \\
    \bottomrule
    \end{tabular}}
\end{table}


\begin{table}[htbp]
  \caption{\textbf{(Cont.) Results for all models on NT benchmark across genomic regions.}}
  \label{table:ntall3}
  \centering
  \resizebox{\textwidth}{!}{
    \begin{tabular}{lllllllllll}
    \toprule
    \textbf{Task/Model}  &  2222  & 42 & 43 & 44 & 45 & 46 & 47 & 48 & 49 & 50             \\ 
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H4}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 80.44 & 80.97 & 82.36 & 80.76 & 81.37 & 81.17 & 81.62 & 80.19 & 80.74 & 81.54 \\
    \rowcolor{low}ConvNova (1 dilation)  & 77.29 & 76.05 & 77.06 & 76.83 & 78.82 & 76.29 & 76.88 & 77.42 & 76.81 & 79.06 \\
    Caduceus-Ph & 80.51  & 80.02 & 80.21 & 79.94 & 81.12 & 79.72 & 79.36 & 79.82 & 80.25 & 80.33 \\
    HyenaDNA  & 78.33 & 77.61 & 77.60  & 77.84 & 77.85 & 77.96 & 79.73 & 78.58 & 79.24 & 76.79 \\
    DNABERT-2 & 78.82 & 78.42 & 77.83 & 78.83 & 78.42 & 77.26 & 78.62 & 78.23 & 77.54 & 78.14 \\
    NTv2 & 78.53 & 80.21 & 79.65 & 79.11 & 81.20 & 78.92 & 79.76 & 81.00 & 80.45 & 79.83 \\
    \midrule
    \rowcolor{titlebg}\multicolumn{11}{l}{\textit{\textbf{H4ac}}} \\
    \rowcolor{low}ConvNova (4 dilation)  & 64.78 & 65.04 & 62.44 & 64.47 & \underline{66.65} & 63.18 & 64.95 & 66.23 & 65.63 & 64.14 \\
    \rowcolor{low}ConvNova (1 dilation)  & 64.9  & 66.89 & 66.77 & 66.9  & 65.68 & 65.32 & 66.14 & 65.77 & 66.82 & 65.77 \\
    Caduceus-Ph& 58.21 & 60.68 & 60.68 & 60.69 & 58.96 & 57.55 & 59.52 & 60.47 & 60.27 & 55.59 \\
    HyenaDNA & 53.77 & 53.99 & 51.30  & 57.10  & 55.75 & 54.30  & 51.32 & 55.36 & 55.34 & 53.22 \\
    DNABERT-2 & 51.82 & 51.92 & 51.73 & 51.83 & 51.92 & 51.73 & 51.83 & 51.92 & 51.73 & 51.83 \\
    NTv2 & 53.68 & 56.94 & 54.72 & 57.41 & 54.20 & 55.84 & 56.10 & 54.95 & 53.02 & 55.35 \\
\bottomrule
\end{tabular}}
\end{table}

\begin{figure}[hbt!]
\subsection{Relative occupancy of Yeast histone marks}
    \centering
    \renewcommand{\thesubfigure}{\Alph{subfigure}}
    \setcounter{subfigure}{0}
    \subfigtopskip=2pt
    \subfigbottomskip=2pt
    \subfigcapskip=-2pt

    \subfigure{
        \label{fig:H3K14ac}
        \includegraphics[width=\linewidth]{imgs/H3K14ac.png}}
    \quad
    \subfigure{
        \label{fig:H3K4me2}
        \includegraphics[width=\linewidth]{imgs/H3K4me2.png}}

    \vspace{0.5em}

    \subfigure{
        \label{fig:H3K4me3}
        \includegraphics[width=\linewidth]{imgs/H3K4me3.png}}
    \quad
    \subfigure{
        \label{fig:H3K9ac}
        \includegraphics[width=\linewidth]{imgs/H3K9ac.png}}

    \caption{\textbf{Relative occupancy of histone marks on chromosome IV.}
        \textbf{A)}  Histone H3K14ac vs H3 YPD ChIP chip ratios
        \textbf{B)}  Histone H3K4me2 vs H3 YPD ChIP chip ratios
        \textbf{C)}  Histone H3K4me3 vs H3 YPD ChIP chip ratios
        \textbf{D)}  Histone H3K9ac vs H3 YPD ChIP chip ratios. 
        Relative occupancy of histone marks on chromosome IV. Data were obtained from the Yeast Genome Database. For complete datasets and additional information, please refer to \url{https://www.yeastgenome.org/dataset/E-WMIT-3\#resources}.
    }
    \label{fig:histone_marks}
\end{figure}

% \begin{table}[h!]
% \centering
% \resizebox{\textwidth}{!}{
% \begin{tabular}{c c c c c c c c c c c}
% \hline
% \multicolumn{1}{c}{} & \multicolumn{10}{c}{\textbf{Model Performance Across Tasks}} \\ 
% \cline{2-11}
% \multicolumn{1}{c}{Model} & \multicolumn{10}{c}{\textbf{Epigenetic Marks (Part 1)}} \\
% \cline{2-11}
% & \textbf{H3} & \textbf{H3K14ac} & \textbf{H3K36me3} & \textbf{H3K4me1} & \textbf{H3K4me2} & \textbf{H3K4me3} & \textbf{H3K79me3} & \textbf{H3K9ac} & \textbf{H4} & \textbf{H4ac} \\
% \hline
% DNABERT-2 & 79.01 & \underline{56.23} & \underline{60.92} & 48.55 & 34.35 & 42.22 & 62.69 & \underline{57.60} & 77.73 & 45.66 \\
% DNABERT-S & \textbf{79.42} & 52.00 & 55.10 & 48.53 & 37.19 & 36.43 & 63.90 & 53.90 & 80.30 & 47.75 \\
% GPN & \underline{79.30} & 52.38 & 60.77 & \underline{51.66} & \underline{40.86} & \underline{45.38} & \underline{66.15} & 57.01 & \textbf{81.56} & \underline{51.66} \\
% \rowcolor{low} ConvNova & 76.94 & \textbf{58.43} & \textbf{61.04} & \textbf{53.11} & \textbf{42.86} & \textbf{52.21} & \textbf{67.67} & \textbf{61.75} & \underline{80.30} & \textbf{54.89} \\
% \hline
% \end{tabular}}

% \vspace{1em}

% \resizebox{\textwidth}{!}{
% \begin{tabular}{c c c c c c c c c c}
% \hline
% \multicolumn{1}{c}{Model} & \multicolumn{2}{c}{\textbf{Promoter}} & \multicolumn{5}{c}{\textbf{TF (Human)}} & \multicolumn{1}{c}{\textbf{Virus}} & \multicolumn{1}{c}{\textbf{Splice}} \\
% \cline{2-3} \cline{4-8} \cline{9-9} \cline{10-10}
% & \textbf{notata} & \textbf{tata} & \textbf{0} & \textbf{1} & \textbf{2} & \textbf{3} & \textbf{4} & \textbf{Covid} & \textbf{Reconstruct} \\
% \hline
% DNABERT-2 & \underline{93.33} & 64.14 & 68.16 & 71.39 & 66.83 & \textbf{62.05} & 75.55 & 70.29 & 86.19 \\
% DNABERT-S & \textbf{93.37} & \underline{64.45} & \textbf{69.85} & \textbf{74.65} & 65.04 & \underline{56.28} & 74.94 & \underline{70.78} & 85.43 \\
% GPN & 92.02 & 53.78 & 66.37 & \underline{72.31} & \underline{73.31} & 49.73 & \underline{76.21} & 70.46 & \underline{88.03} \\
% \rowcolor{low} ConvNova & 92.16 & \textbf{66.06} & \underline{69.82} & 71.83 & \textbf{74.52} & 50.58 & \textbf{76.65} & \textbf{73.86} & \textbf{88.29} \\
% \hline
% \end{tabular}}
% \caption{\textbf{Comprehensive performance comparison across all tasks.} Results show percentage accuracy for each model across epigenetic marks prediction, promoter detection, transcription factor (TF) prediction, virus detection and splice site reconstruction tasks. Best results are in \textbf{bold}, second best are \underline{underlined}.}
% \label{table:results_merged}
% \end{table}

% eval/accuracy 0.66689
% eval/auroc 0.7242

% 0 \& \text0& \text0& \text0& \text0& \text0& \text0& \text