\section{Experiments}
\label{experiments}
In this section, we present the results of two parts of experiments: pocket-ligand docking and pocket-aware 3D drug design. Through pre-training, the 3DMolFormer model is theoretically capable of being applied to the conformation generation of small molecules. However, as \cite{ConformationWrong} pointed out, the existing benchmarks for conformation generation are wrong; therefore, this experiment is not conducted.

Following the configuration of the GPT-2 small model~\citep{GPT-2}, the 3DMolFormer model with a total of 92M parameters has 12 transformer layers, each containing 12 self-attention heads, and the embedding dimension is 768. The maximum length for the parallel sequences is set to 2048, which covers over 99\% of the samples in the training set as well as all samples in the test set for protein-ligand docking.

For pre-training, all samples with a coordinate range larger than 40 are screened out. Then, we replicate each protein pocket five times and each pocket-ligand complex twenty times, mixing them with small molecule conformations, resulting in a total of 228M training data samples. 3DMolFormer is pre-trained on this dataset for only one epoch, using a batch size of 10K implemented by gradient accumulation. The maximal learning rate is set to $5\times10^{-4}$ with a warmup period of 1\% steps followed by cosine decay. An AdamW optimizer~\cite{AdamW} with a weight decay factor of $0.1$ is employed, and the coefficient $\alpha$ in the loss function of Eq. (\ref{pretrainloss}) is set to $1.0$. The pre-training process takes less than 48 hours with 4 A100 80G GPUs. For further details on the selection of hyper-parameters for pre-training, please refer to Appendix~\ref{app2}.

\subsection{Protein-ligand Binding Pose Prediction}
\label{exp-docking}
Experiments of protein-ligand binding pose prediction are conducted in the targeted and semi-flexible docking scenario, where the protein pocket for binding is specified and fixed, while the ligand conformation is entirely flexible.

\paragraph{Data} Following Uni-Mol~\citep{Uni-Mol}, we use PDBbind v2020~\citep{PDBbind} as the training set for supervised fine-tuning on protein-ligand docking and CASF-2016~\citep{CASF-2016} as the test set, which includes 285 test samples. In addition, we apply the same data filtering process as Uni-Mol to remove training samples with high similarity to the protein sequences or molecular structures of the complexes in the test set, which results in a training set comprising 18,404 ground-truth complexes. 

\paragraph{Baselines} We select four search-based methods: AutoDock4~\citep{AutoDock4}, AutoDock Vina~\citep{AutoDockVina,AutoDockVina2}, Vinardo~\citep{Vinardo}, and Smina~\citep{Smina}, along with Uni-Mol~\citep{Uni-Mol}, which is currently the state-of-the-art deep learning method for targeted docking, as our baselines.

\paragraph{Ablation Studies} Two variants of 3DMolFormer are established: (1) training a 3DMolFormer model from scratch on the fine-tuning set for protein-ligand docking without pre-training (w/o PT), and (2) fine-tuning the pre-trained 3DMolFormer model without data augmentation (w/o DA).

\paragraph{Evaluation} The root mean square deviation (RMSD) between the predicted ligand pose and the ground truth is used to assess binding pose accuracy. Specifically, two metrics are employed: (1) the percentage of RMSD results that fall below predefined thresholds, with higher percentages indicating better performance, and (2) the average RMSD, where lower values are preferred.

\paragraph{Fine-tuning} For supervised fine-tuning for pocket-ligand binding pose prediction, we train the model for 2000 epochs with a batch size of 128. The maximum learning rate is set to $1\times10^{-4}$, with a warmup period of 1\% of the steps and cosine decay applied thereafter. The training process takes less than 24 hours with 4 A100 80G GPUs.

\begin{table}[t]
\caption{Experimental results of 3DMolFormer, its variants, and other baselines on protein-ligand binding pose prediction, following the results reported in Uni-Mol~\citep{Uni-Mol}. ($\uparrow$) / ($\downarrow$) denotes that a higher / lower value is better. The best result in each column is \textbf{bolded}.}
\label{docking-results}
\begin{center}
\begin{tabular}{cccccc}
\hline
Methods & \%$<$1.0\AA~($\uparrow$) & \%$<$2.0\AA~($\uparrow$) & \%$<$3.0\AA~($\uparrow$) & \%$<$5.0\AA~($\uparrow$) & Avg.~($\downarrow$)
\\ \hline 
AutoDock4 & 21.8 & 35.4 & 47.0 & 64.6 & 3.53\\
AutoDock Vina & 44.2 & 64.6 & 73.7 & 84.6 & 2.37 \\
Vinardo & 41.8 & 62.8 & 69.8 & 76.8 & 2.49 \\
Smina & \textbf{47.4} & 65.3 & 74.4 & 82.1 & 1.84\\
Uni-Mol & 43.2 & 80.4 & 87.0 & 94.0 & 1.62 \\ \hline
3DMolFormer w/o PT & 15.5 & 57.8 & 78.1 & 92.4 & 2.25\\
3DMolFormer w/o DA & 10.3 & 51.0 & 74.9 & 91.6 & 2.45 \\
3DMolFormer & 43.8 & \textbf{84.9} & \textbf{96.4} & \textbf{98.8} & \textbf{1.29} \\\hline
\end{tabular}
\end{center}
\vspace{-0.2cm}
\end{table}

\paragraph{Results}
As shown in Table~\ref{docking-results}, 3DMolFormer outperforms all baselines in both average RMSD and the percentage of predictions with RMSD less than 2.0, 3.0, and 5.0 \AA. Notably, it significantly surpasses other methods in the percentages for RMSD below 3.0 and 5.0 \AA. This indicates that 3DMolFormer is less prone to making "large errors" compared to the baselines, reflecting its robustness. However, for the percentage of predictions with RMSD below 1.0 \AA, the search-based method Smina outperforms the deep learning approaches, suggesting that there is still room for improvement in the ability of deep learning methods to capture the intricate interactions between protein pockets and ligands. Moreover, the ablation studies demonstrate that the pre-training and data augmentation both play a crucial role in the training of the 3DMolFormer docking model.

It is worth noting that, unlike all baseline methods, 3DMolFormer does not require an initialized 3D conformation of the ligand as input, indicating that the model has acquired the capability to predict small molecule 3D conformations through pre-training. This feature enhances the usability of 3DMolFormer compared to previous docking approaches.

Additionally, the average time taken by 3DMolFormer to predict a binding pose is 0.8 seconds using 1 A100 80G GPU, and this can be significantly accelerated through parallel inference. This suggests that 3DMolFormer has great potential for applications in large-scale virtual screening. For further details and results of experiments on protein-ligand binding pose prediction, please refer to Appendix~\ref{app3}.

\subsection{Pocket-aware 3D Drug Design}
\label{exp-drug-design}
In the experiments for pocket-aware 3D drug design, small molecule ligands and their 3D conformations are designed to bind well with a specified pocket on a protein whose structure remains fixed.

\paragraph{Data} Following previous works~\citep{Pocket2Mol,TargetDiff,DecompDiff}, we select 100 protein pockets from the CrossDocked2020~\citep{CrossDocked} dataset that exhibit low similarity ($<30\%$) to the protein sequences of pocket-ligand complexes used in pre-training, thereby establishing our targets for 3D drug design.

\paragraph{Baselines} We compare 3DMolFormer against various baselines for pocket-aware 3D molecular generation, including AR~\citep{AR}, liGAN~\citep{liGAN}, GraphBP~\citep{GraphBP}, Pocket2Mol~\citep{Pocket2Mol}, TargetDiff~\citep{TargetDiff}, and DecompDiff~\citep{DecompDiff}. Additionally, we report the results of the ligands corresponding to the 100 protein pockets in the CrossDocked2020 dataset for reference.

\paragraph{Evaluation} In alignment with previous works, we evaluate 100 3D molecules generated for each protein pocket. Four metrics are selected to comprehensively assess the potential of generated molecules in practical drug design: (1) \textbf{Vina Score}, which directly estimates the binding affinity based on the generated 3D molecules; (2) \textbf{Vina Dock}, representing the best possible binding affinity of the molecules estimated by re-docking; (3) \textbf{QED} (Quantitative Estimate of Drug-likeness)~\citep{QED}; and (4) \textbf{SA} (Synthetic Accessibility)~\citep{SA}\footnote{Here the original SA score has been linearly transformed to $[0,1]$, as illustrated in Appendix~\ref{app4}.}. We employ Quick Vina 2~\citep{QuickVina2} to estimate the binding affinity, which is an efficient alternative to AutoDock Vina. For all metrics, we report their average values across designed drug molecules for all protein pockets. Following \cite{DESERT} and \cite{DecompDiff}, we also report the percentage of designed drug molecules meeting specific criteria: Vina Dock$<-8.18$, QED $>0.25$, and SA$>0.59$. This percentage, referred to as the \textbf{Success Rate}, reflects the performance of different methods in multi-objective drug design, which is a common scenario in practical drug discovery.

\paragraph{Reward Function} For the aforementioned drug design objectives, we formulate a composite reward function for the RL fine-tuning process ($R(m)$ in Eq.~(\ref{RLloss})). First, a reverse sigmoid function~\citep{MolRL-MGPT} is applied to transform the Vina Dock score into a range of $[0,1]$, where higher values are preferable:
\begin{equation}
R_{\mathrm{Dock}}(m)=1/(1+10^{0.625\cdot(\mathrm{VinaDock}(m)+10)}),
\end{equation}
where $m$ refers to a small molecule.

Next, we utilize a step function for QED and SA, as these properties are auxiliary to the docking score; thus, they only need to exceed certain thresholds rather than aiming for higher values.
\begin{equation}
R_{\mathrm{QED}}(m)=\mathbb{I}(\mathrm{QED}(m)>0.25), \quad R_{\mathrm{SA}}(m)=\mathbb{I}(\mathrm{SA}(m)>0.59),
\end{equation}
where $\mathbb{I}(\cdot)$ represents the indicator function.

Finally, the mean of these three scores is employed as the RL reward function:
\begin{equation}
R(m)=\frac{1}{3}\big[R_{\mathrm{Dock}}(m)+R_{\mathrm{QED}}(m)+R_{\mathrm{SA}}(m)\big].
\end{equation}
This composite reward is also used as the multi-objective criteria for selecting drug candidates from all generated molecules.

\begin{table}[t]
\caption{Experimental results of 3DMolFormer and other baselines on pocket-aware 3D drug design, following the results reported in DecompDiff~\citep{DecompDiff}.  ($\uparrow$) / ($\downarrow$) denotes that a higher / lower value is better. The best result in each column is \textbf{bolded}.} 
\label{drugdesign-results}
\begin{center}
\begin{tabular}{cccccc}
\hline
Methods & Vina Score~($\downarrow$) & Vina Dock~($\downarrow$) & QED~($\uparrow$) & SA~($\uparrow$) & Success Rate~($\uparrow$)
\\ \hline 
Reference & -6.36 & -7.45 & 0.48 & 0.73 & 25.0\% \\ \hline
AR & -5.75 & -6.75 & 0.51 & 0.63 & 7.1\% \\
liGAN & - & -6.33 & 0.39 & 0.59 & 3.9\% \\
GraphBP & - & -4.80 & 0.43 & 0.49 & 0.1\% \\
Pocket2Mol & -5.14 & -7.15 & \textbf{0.56} & 0.74 & 24.4\% \\
TargetDiff & -5.47 & -7.80 & 0.48 & 0.58 & 10.5\% \\
DecompDiff & -5.67 & -8.39 & 0.45 & 0.61 & 24.5\% \\\hline
3DMolFormer & \textbf{-6.02} & \textbf{-9.48} & 0.49 & \textbf{0.78} & \textbf{85.3\%} \\ \hline
\end{tabular}
\end{center}
\vspace{-0.2cm}
\end{table}

\paragraph{Fine-tuning} For the reinforcement learning fine-tuning aimed at pocket-aware 3D drug design, we execute 500 RL steps for each protein pocket, with a batch size of 128 and a constant learning rate of $1\times10^{-4}$. The parameter $\sigma$ in Eq.~(\ref{RLloss}) is set to 100. The RL process for each protein pocket takes less than 8 hours using 1 A100 80G GPU and 128 CPU cores, with the computation of the Vina Dock reward running in parallel on the CPU cores.

\paragraph{Results}
As shown in Table~\ref{drugdesign-results}, the molecules designed by 3DMolFormer outperform those of all baselines across four metrics: Vina Score, Vina Dock, SA, and Success Rate. Notably, it exhibits a significant advantage in Success Rate, becoming the first method to exceed the reference values provided in the dataset for this key metric. Additionally, the result of QED also significantly surpasses the predefined threshold. This indicates that 3DMolFormer demonstrates superior performance in binding affinity optimization and multi-objective joint optimization compared to existing 3D drug design methods, highlighting its strong potential for real-world applications in drug discovery.

For further details, results, and a case study of experiments on pocket-aware 3D drug design, please refer to Appendix~\ref{app4}.