\documentclass{midl} %
%
%
\usepackage[textwidth=2cm]{todonotes}
\usepackage{adjustbox}
\usepackage{float}
\usepackage{booktabs}
\usepackage{bold-extra}
\usepackage{xcolor}



%
%
%
%


\usepackage{mwe} %
%
%
%
%

\jmlrproceedings{}{}

\title[Sequence models for continuous cell cycle stage prediction from brightfield images]{Sequence models for continuous cell cycle stage prediction from brightfield images}

\midlauthor{\Name{Louis-Alexandre Leger\nametag{$^{1}$}\midljointauthortext{Contributed equally}} \orcid{0009-0007-9183-173X} \Email{louis-alexandre.leger@epfl.ch}\\
\Name{Maxine Leonardi\nametag{$^{1}$}\midlotherjointauthor} \orcid{0000-0002-9780-9126} \Email{maxine.leonardi@epfl.ch}\\
\Name{Andrea Salati\nametag{$^{1}$}\midlotherjointauthor} \orcid{0009-0006-6373-3668} \Email{andrea.salati@epfl.ch}\\
\setsharedfootnotesymbol
\Name{Felix Naef\nametag{\thinspace$^{1}$}\midljointsharedauthortext{Shared supervision}} \orcid{0000-0001-9786-3037} \Email{felix.naef@epfl.ch}\\
\Name{Martin Weigert\nametag{$^{1,2}$}\midlothersharedauthor} \orcid{0000-0002-7780-9057} \Email{martin.weigert@tu-dresden.de}\\
\addr $^{1}$ Institute of Bioengineering, School of Life Sciences, Ecole Polytechnique Fédérale de Lausanne (EPFL), Lausanne, Switzerland\\ 
\addr $^{2}$ Center for Scalable Data Analytics and Artificial Intelligence (ScaDS.AI), Dresden/Leipzig, Germany
}

          
%
%
\usepackage{amsmath}
\DeclareMathOperator{\R}{\mathbb{R}}
\DeclareMathOperator{\Z}{\mathbb{Z}}
\DeclareMathOperator{\N}{\mathbb{N}}
\DeclareMathOperator{\Q}{\mathbb{Q}}
\DeclareMathOperator{\C}{\mathbb{C}}


\DeclareMathOperator{\Tr}{\mathrm{Tr}}
\DeclareMathOperator{\sgn}{\mathrm{sgn}}
\DeclareMathOperator{\Det}{\mathrm{det}}


\DeclareMathOperator{\erf}{\mathrm{erf}\,}
\DeclareMathOperator{\eps}{\varepsilon}
\DeclareMathOperator{\const}{\text{const}}
\DeclareMathOperator{\nequiv}{\not \equiv}
\DeclareMathOperator{\bessel}{J}
\DeclareMathOperator{\F}{\mathcal{F}}
\DeclareMathOperator{\J}{\mathcal{J}}

\DeclareMathOperator{\BigO}{\mathcal{O}}

\DeclareMathOperator{\?}{\negthickspace}

\DeclareMathOperator*{\argmax}{arg\,max}
\DeclareMathOperator*{\argmin}{arg\,min}

\newcommand{\cmmnt}[1]{\ignorespaces}


\newcommand{\Id}{Id}
\newcommand{\dpartial}[2]{\ensuremath{\frac{\partial #1}{\partial #2}}}
\newcommand{\ddiff}[2]{\ensuremath{\frac{d #1}{d #2}}}
\newcommand{\dvector}[2]{\ensuremath{\begin{pmatrix} #1 \\ #2\end{pmatrix}}}
\newcommand{\colvector}[2]{\ensuremath{\begin{pmatrix}#1\\#2\end{pmatrix}}}
\newcommand{\scalar}[1]{\ensuremath{\langle #1 \rangle}}
\newcommand{\legendre}[2]{\ensuremath{\big(\tfrac{#1}{#2}\big)}}
\newcommand{\mean}[1]{\ensuremath{\langle #1 \rangle}}
\newcommand{\norm}[1]{\ensuremath{\left\lVert #1 \right\rVert}}

\newcommand{\Var}{\ensuremath{\text{Var}}}
\newcommand{\Cov}{\ensuremath{\text{Cov}}}



\newcommand{\sizetwo}[2]{\ensuremath{#1\!\times\!#2\xspace}}
\newcommand{\sizethree}[3]{\ensuremath{#1\mkern-2mu\times\mkern-2mu#2\mkern-2mu\times\mkern-2mu#3\xspace}}

\newcommand{\negspace}{\ensuremath{\mkern-10mu}}%
\newcommand\abs[1]{\lvert #1 \rvert}
\newcommand\ceil[1]{\lceil #1 \rceil}
\newcommand\floor[1]{\lfloor #1 \rfloor}


\makeatletter
\DeclareRobustCommand\onedot{\futurelet\@let@token\@onedot}
\def\@onedot{\ifx\@let@token.\else.\null\fi\xspace}

\def\eg{\emph{e.g}\onedot} \def\Eg{\emph{E.g}\onedot}
\def\ie{\emph{i.e}\onedot} \def\Ie{\emph{I.e}\onedot}
\def\cf{\emph{cf.}\xspace} \def\Cf{\emph{Cf.}\xspace}

\def\etc{\emph{etc}\onedot} \def\vs{\emph{vs}\onedot}
\def\wrt{w.r.t\onedot} \def\dof{d.o.f\onedot}
\def\etal{\emph{et~al}\onedot}
\makeatother

\newcommand{\um}{\ensuremath{\mu m}\xspace}
\newcommand{\nm}{\ensuremath{nm}\xspace}
\newcommand{\uM}{\ensuremath{\mu M}\xspace}
\newcommand{\ug}{\ensuremath{\mu g}\xspace}


\newcommand{\simi}{\mathord{\sim}}
\newcommand{\NA}{\mathit{NA}}
\newcommand{\MB}{\mathrm{MB}}
\newcommand{\figpart}[1]{#1)}


\newcommand{\Fucci}{\textsc{Fucci}\xspace}
\newcommand{\Mamba}{\textsc{Mamba}\xspace}
\newcommand{\LSTM}{\textsc{LSTM}\xspace}

\newcommand{\dataregular}{\textsc{Regular}\xspace}
\newcommand{\datadrug}{\textsc{Drug}\xspace}

\newcommand{\DTW}{\ensuremath{\Delta_{DTW}}\xspace} 
%
\begin{document}


\maketitle

\begin{abstract}
Understanding cell cycle dynamics is crucial for studying biological processes such as growth, development and disease progression. While fluorescent protein reporters like the \Fucci system allow live monitoring of cell cycle phases, they require genetic engineering and occupy additional fluorescence channels, limiting broader applicability in complex experiments.
In this study, we conduct a comprehensive evaluation of deep learning methods for predicting continuous \Fucci signals using non-fluorescence brightfield imaging, a widely available label-free modality. 
To that end, we generated a large dataset of 1.3 M images of dividing RPE1 cells with full cell cycle trajectories to quantitatively compare the predictive performance of distinct model categories including single time-frame models, causal state space models and bidirectional transformer models.
We show that both causal and transformer-based models significantly outperform single- and fixed frame approaches, enabling the prediction of visually imperceptible transitions like 
G1/S within 1h resolution. Our findings underscore the importance of sequence models for 
accurate predictions of cell cycle dynamics and highlight their potential for label-free imaging.
\end{abstract}

\begin{keywords}
Cell cycle prediction, label-free microscopy, sequence-models
\end{keywords}

\section{Introduction}

The cell cycle is the driving force behind the growth and development of all living organisms. This well-studied sequence of cellular events is tightly regulated, as aberrations in such mechanisms can lead to genomic instability, a key driver of various diseases including cancer \cite{kastan_cell-cycle_2004}. 
Live cell fluorescence microscopy has become a powerful tool for studying cell cycle progression, particularly through the genetic engineering of fluorescent reporters like the \Fucci system~\cite{sakaue-sawano_visualizing_2008,stallaert_structure_2022}. This system enables the distinction of cell cycle phases from single images by fluorescently tagging the two proteins Cdt1 and Geminin whose expression changes distinctively with the cell cycle~(\figureref{fig:Overview_of_data}). 
Despite its utility, the classic \Fucci system and recent variants~\cite{sakaue-sawano_genetically_2017,grant_accurate_2018} are limiting in practice as they occupy two of the few available microscopy channels, reducing the ability to study other cellular processes simultaneously and requiring genetic modification that might interfere with the endogenous cell cycle regulation. 
\begin{figure}[t!]
    \floatconts
      {fig:Overview_of_data}
      {\caption{\textbf{Multi-modal imaging of \Fucci-reporter cells reveals a continuous representation of cell cycle states.} 
      \textbf{a)} Time-lapse imaging of Fucci-reporting cells allows for precise quantification of cell cycle
    staging through the characteristic oscillations of fluorescent reporter intensities. 
    \textbf{b)} Representative time-lapse images of brightfield, H2B, and 
    \Fucci channels across one full (M-M) cell cycle.
    \textbf{c)} Quantification of integrated logarithmic fluorescence intensities from \Fucci reporters in a representative full M-M (mitosis to mitosis) track. Vertical lines mark the time points corresponding to the images shown in d. 
    \textbf{d)} Log-transformed \Fucci manifold for continuous inference of cell cycle states. 
    }}
    {\includegraphics[width=1\linewidth]{Figures/Overview_of_data.png}}
\end{figure}
In contrast, brightfield microscopy is an easily accessible and label-free imaging modality that does not require genetic engineering, providing limited specificity and imaging contrast. Although some cell cycle transitions, such as nuclear envelope breakdown, are marked by distinct morphological changes that are easily detectable, most cell cycle transitions are visually indiscernible in individual brightfield images of cells.
In this paper, we ask whether leveraging the \emph{temporal information} in time-lapse brightfield microscopy images of cells would allow to predict \emph{continuous} cell cycle states without the need for fluorescent reporters such as \Fucci. 
To address this, we study several sequence-based deep learning models, including transformers \cite{vaswani_attention_nodate} and recently proposed state-space models (\eg Mamba~\cite{gu_mamba_2024}). 
In particular, we will investigate both \emph{causal} sequence models that only use information from previous time points, as well as \emph{non-causal} models that may ingest  the entire sequence. 
Providing a new dataset of over 1~M images of segmented and tracked RPE-1 cells with accompanying ground truth \Fucci signals, we show that both causal and transformer-based non-causal models significantly outperform single-frame approaches, enabling the prediction 
of morphologically important cell state transitions 
like G1/S within 1h resolution from live-cell brightfield imaging alone. 
\vspace{-.2cm}

%
%
%

\subsection{Related work}

The prediction of individual cell states from single microscopy images has found considerable interest in the literature. For instance~\cite{rappez_deepcycle_2020} and \cite{narotamo_machine_2021, li_predicting_2024} showed that deep learning-based models allow to classify \emph{discrete} cell cycle states such as the G1, S, or G2 phase from static multichannel images of cells. In particular, \cite{eulenberg_reconstructing_2017,blasi_label-free_2016,jin_imbalanced_2021,he_cell_2022} explored the use of brightfield and phase-imaging data for cell cycle classification without fluorescence labeling. However, accurately annotating discrete cell cycle stages highly depends on the imaging modality with nuclear stains like DAPI or Hoechst providing clear features, while phase or brightfield imaging require substantial more manual annotation expertise. 
Beyond static images, several studies have investigated how the \emph{temporal information} of images sequences can be leveraged to capture dynamic cell behaviors~\cite{wang_live-cell_2020, zhao_insights_2024,chu_prediction_2020}. 
However, most approaches have so far been focused on well-defined cell phases such as mitosis, where large morphological changes facilitate annotation and prediction~\cite{held_cellcognition_2010,moreno-andres_livecellminer_2022,jose_automatic_2024, su_spatiotemporal_2017}. A notable exception is \cite{ulicna_learning_2023} which applies dynamic time warping (DTW) to features from an unsupervised autoencoder, enabling to temporally align cell trajectories and to continuously predict cycle states across interphase. 
\emph{Sequence models} such as recurrent neural networks and transformers~\cite{vaswani_attention_nodate} have been shown to be effective for modeling temporal data~\cite{hewamalage2021recurrent,wen2022transformers}, with \emph{state-space models} such as Mamba~\cite{gu_efficiently_2022,gu_mamba_2024} having recently gained considerable interest due to their training and inference efficiency. 
As the cell cycle is a continuous causal process, sequence models appear to be a natural fit for our task. However, studies that address this question systematically remain missing, with the closest work being~\cite{jose_automatic_2024} which uses recurrent neural networks for cell cycle prediction, yet for the case of a small number of discrete cell states. 



\section{Method}

\subsection{Dataset}


We generated a large dataset of dividing human \Fucci RPE1 cells using joint brightfield and fluorescence time lapse microscopy. Movies spanning 72 hours were acquired at a 5 minute time resolution, capturing multiple cell cycles. In addition to the brightfield modality, we acquired a nuclear marker channel (Histone H2B) and the two \Fucci channels ($\Fucci_{1/2}$). Based on the H2B channel, we segmented the cell nuclei with a custom StarDist model~\cite{schmidt_cell_2018} and tracked them across frames using TrackMate~\cite{tinevez_trackmate_2017}.
Note that since the amount of H2B histones needed by cells to pack DNA doubles during S (DNA replication), the H2B channel does contain information on cell cycle progression. Below, we leverage this as a control to assess predictive performance of brightfield vs H2B.   
Full cell cycle tracks (from one mitosis to the next, M-M) were identified using K-Means clustering and ground-truth \Fucci signals were computed by normalizing the average \Fucci intensities measured across the segmented nuclear mask~(\figureref{fig:Overview_of_data}c,d).
The training dataset comprises 5,188 full (M-M) cell cycle tracks with an average track length of 230 frames. Each track contains paired brightfield and H2B images of size $64\times64$ centered on the nucleus and the corresponding integrated \Fucci signals. To evaluate model performance, we created two additional test datasets: \dataregular, which contains 358 additional full tracks from RPE1 cells acquired at similar conditions as the training set, and \datadrug, which comprises 73 complete tracks of cells treated with the cell cycle inhibitor \emph{Palbociclib} that heavily distorts the cell cycle and which is used in the clinic to treat breast cancer. In total, this training and testing dataset consists of approximately 1.3 M images and \Fucci signals, all of which we make publicly available alongside this paper\footnote{\url{https://zenodo.org/records/14774038}}.




\begin{figure}[t!]
    \floatconts
      {fig:architechture}
      {\caption{\textbf{Overview of prediction approach.} \textbf{a)} We use a ResNet-18~\cite{he2016deep} to extract single frame embeddings from a brightfield sequence which are fed into a sequence model that predicts both \Fucci channels. \textbf{b)} Sequence models explored in this paper: Single Frame MLP, Fixed-frame CNN, causal state-space models \eg Mamba~\cite{gu_mamba_2024}, bidirectional models \eg transformers~\cite{vaswani_attention_nodate}.}}
    {\includegraphics[width=1\linewidth]{Figures/arch.pdf}}
\end{figure}

\subsection{Method}

The cell cycle prediction task can be formalized as follows: Given a temporal sequence of $N$ brightfield images $X \in \mathbb{R}^{N \times 64 \times 64}$, each associated with a corresponding two-dimensional \Fucci signal $Y \in \mathbb{R}^{N \times 2}$, the goal is to train a model $f$  that predicts the normalized \Fucci intensities across the entire sequence in a supervised manner, \ie. $f: \mathbb{R}^{N \times 64 \times 64} \to \mathbb{R}^{N \times 2}~$. Note that the length $N$ of the input is not fixed and the sequences is not required to span the entire cell cycle, allowing us to analyze the impact of temporal context on prediction performance.
For this cell cycle prediction task, we evaluated three conceptually distinct model classes: single-frame models, causal models and non-causal models~(\cf~\figureref{fig:architechture}). 
Each model uses first a ResNet-18~\cite{he2016deep} as feature extractor which for every input image in the sequence independently creates a 512-dimensional embedding. The sequence of embeddings is then fed into the proper sequence model head that differs between the three classes: \emph{Single-frame models} serve as a baseline, predicting the \Fucci signal from each image embedding independently without leveraging temporal information. We use a simple 4 layer MLP with hidden dimension 512. \emph{Fixed-frames models} that use a fixed history of past frames for prediction and for which we use a causal convolutional neural networks~\cite{van2016wavenet} with a fixed causal temporal receptive field.  \emph{Causal models} that incorporate past and present temporal context in more flexible way and which are able to potentially capture arbitrary long temporal dependencies. In particular, we compare LSTMs~\cite{hochreiter_long_1997}, Mamba~\cite{gu_mamba_2024}, and causal transformers with masked attention. \emph{Bidirectional models} process the entire sequence bidirectionally, using both past and future frames for inference (\ie non-causal information). We use a standard transformer (4 layers) as representative architecture.
To ensure a fair comparison, we choose all sequence heads to have the same number of parameters ($\approx 1M$). All transformer variants additionally use rotary positional embeddings~\cite{su2024roformer} to encode the relative temporal position of each frame.
We train each model for 150 epochs while randomly sampling subtracks of variable lengths, using a learning rate of $10^{-4}$ and $L_1$ loss. We use standard data augmentation such as random rotations and flips. 



\begin{figure}[b!]
    \floatconts
      {fig:Prediction_bf}
      {\caption{\textbf{Predictions on unperturbed RPE cells \dataregular.} \textbf{a)} Distribution of $L1$ errors across the different  models.
    \textbf{b)} Predictions of \Fucci signals on two example tracks: one with accurate and one with poor predictions. The ground truth signal is shown in black.
    \textbf{c)} Average prediction error, and \textbf{d)} variability of ground truth \Fucci signals as a function of normalized cell cycle time $\tau$.
      }} 
      {\includegraphics[width=.9\linewidth]{Figures/Predictions_bf.pdf}\vspace{-.4cm}} 
\end{figure}

\subsubsection{Performance metrics}

We evaluated model prediction by computing the mean $L_1$ error for each \Fucci channel across a given track. Further, we use the \emph{dynamic time warping distance} \DTW between the predicted and ground truth \Fucci signal that takes into account both the signal prediction error as well as the temporal misalignment between the two signals. We use the default \DTW distance implementation from the \texttt{dtaidistance} package~\cite{meert_dtaidistance_2020} with a penalty of 0.1. 
Additionally, we introduced two biologically meaningful cell cycle checkpoints (\figureref{fig:Overview_of_data}) and measured the time difference between our predicted and observed checkpoints in minutes. The first checkpoint $t_{G1/S}$ is the onset of the Geminin (\Fucci2) signal, marking the G1/S phase transition \cite{sakaue-sawano_genetically_2017}. 
While the classic \Fucci reporter does not provide an exact molecular landmark of the S/G2 transition, we use the disappearance of the Cdt1 (\Fucci1) signal as an approximate landmark for evaluating S/G2 transition predictions $t_{S/G2}$. We use these two landmarks to categorize cells into discrete G1, S, and G2 phase classes for which we compute the F1-score of predictions. 


\section{Results}



\begin{table}[t] 
    \centering
    \floatconts
      {tab:table1_bf} 
      {\caption{\textbf{Cell cycle prediction accuracy for brightfield on \dataregular}. Shown are mean and standard deviation of the $L_1$ error per \Fucci channel and \DTW for full tracks across \dataregular.  \DTW when using H2B images as input modality for comparison.} 
      }
    
    \footnotesize
    \begin{tabular}{lllllll}
        \toprule
        & \multicolumn{3}{c}{Brightfield} & \multicolumn{1}{c}{Histone H2B} \\
        \cmidrule(lr){2-4} \cmidrule(lr){5-5}
        \textbf{Models} & $L_{1, FUCCI_1}$ & $L_{1, FUCCI_2}$ & \DTW & \DTW \\
    \midrule
    Single Frame & 0.193 ± 0.066 & 0.146 ± 0.045 &  3.735 ± 0.863 &   2.595 ± 1.201 \\
    Causal CNN & 0.157 ± 0.078 & 0.122 ± 0.049 &  2.468 ± 0.917 &  2.165 ± 1.210 \\
    Mamba & 0.112 ± 0.072 & 0.091 ± 0.049 & 1.444 ± 0.898 &   1.426 ± 0.949 \\
    Transformer & \textbf{0.066} \textbf{± 0.038} & \textbf{0.062} \textbf{± 0.037} & \textbf{1.285} \textbf{± 0.553} & \textbf{1.155} \textbf{± 0.612} \\
    \bottomrule
      \end{tabular}
      
    \end{table}
    
\subsection{Comparison of prediction accuracy across sequence models}


\begin{table}[b] 
    \centering
\floatconts
  {tab:Table2} 
  {\caption{\textbf{Prediction accuracy of biological checkpoints and selected cell cycle states from brightfield images on \dataregular.}}}
  {
\footnotesize
\begin{tabular}{llllll}
\toprule
 \textbf{Models}& $\Delta t_{G1/S}[\text{min}]$ & $\Delta t_{S/G2}[\text{min}]$ & $G1$ & $S$ & $G2$ \\
\midrule
Single Frame & 191.3 & 117.9 & 0.71 & 0.64 & 0.87 \\
Causal CNN & 146.8 & 113.3 & 0.78 & 0.64 & 0.86 \\
Mamba & 111.4 & 102.4 & 0.83 & 0.72 & 0.89 \\
Transformer & \textbf{60.1} & \textbf{57.2} & \textbf{0.90} & \textbf{0.85} & \textbf{0.93} \\
\bottomrule
\end{tabular}
}
\end{table}



We first compare the performance of the different sequence models on predicting the \Fucci signals from brightfield images for full (M-M) cell cycle tracks.
As seen in \figureref{fig:Prediction_bf}a, the single frame model predicts extremely noisy signals that deviate substantially from the ground truth \Fucci signal, whereas both causal and bidirectional models achieve qualitatively much better predictions on both \Fucci channels and generally aligns with the expected trends (\cf~Supp.~\figureref{fig:histograms}a,b, Supp.~\figureref{fig:umap}).
To quantitatively assess the performance of the different models, we show the mean $L_1$ error for each \Fucci channel across all tracks as well as the average \DTW distance between the predicted and ground truth \Fucci signals for the different models in \tableref{tab:table1_bf}.
As expected, the single frame model which operates without integrating temporal information performs the worst across all metrics (\DTW = 3.735), while integrating the full bidirectional (non-causal) sequence information via a transformer achieves the best prediction (\DTW = 1.285). Surprisingly, there is a notable difference between the performance of the fixed-frames model (causal CNN) and the state-space model (Mamba), with the former performing substantially worse than the latter (\DTW = 2.468 vs. 1.444). This suggests that models that allow information propagation across the entire sequence can be more effective than models that only use a fixed-size temporal context. 
We additionally computed \DTW when training with the H2B channel as input modality, which a priori should be a substantially easier task as the H2B signal is biologically correlated with the cell cycle. Indeed, this is corroborated by the performance of the single frame model that vastly improves in this case (\DTW = 2.595). Interestingly, both state-space models as well as the transformer only marginally improve, suggesting that these models are able to extract temporal cues from the brightfield images comparable to the easier H2B modality. 
Similar observation can be made for the prediction of the biological checkpoints, as shown in \tableref{tab:Table2} and \figureref{fig:visual_landmarks}.
Again, the bidirectional transformer significantly outperforms other methods for all metrics, including the prediction of biological checkpoints (Table \ref{tab:Table2}, Supp. \figureref{fig:visual_landmarks}). As expected, the majority of prediction errors happen near the $t_{G1/S}$ and $t_{S/G2}$ landmarks and are proportional to the standard deviation of the data (\figureref{fig:Prediction_bf} b, c).


%


%
%
%
%
%
%
%
%
%
%
%
%
%
%
%
%
%
%
%



    






%

\vspace{-.2cm}
\subsection{Prediction on partial tracks}


So far we focused on full (M-M) tracks of non-perturbed cells, all of which exhibit fairly stereotypical cell cycle trajectories and for which sequence models are able to base their predictions on a well defined starting points (\ie the cell division event).
We now evaluate the performance of the different models on partial tracks, where the starting point is not known a priori, which is a more challenging task.
\begin{figure}[t]
\floatconts
  {fig:partial_track_bf}
  {\caption{\textbf{Comparative performance of temporal encoders on partial cell cycle tracks (brightfield).}
     Shown is the average $L_1$ error of both \Fucci signals when using partial tracks as input, parametrized by their relative start and end time $\tau_1 \leq \tau_2 \in [0,1]$. 
 }}
{\includegraphics[width=1\linewidth]{Figures/triang.pdf}\vspace{-.2cm}}
\end{figure}
In \figureref{fig:partial_track_bf}, we show the average $L_1$ error with cropped partial tracks as input, indicated by their relative start and end time $\tau_1 \leq \tau_2 \in [0,1]$. These partial tracks ranged from single-frame portions (along the diagonal) to entire tracks (lower-right element). Causal models still achieved better accuracy in predicting \Fucci values compared to non-temporal MLPs or fixed history CNNs. Surprisingly, the performance advantage of transformers over causal methods observed in full tracks diminishes on partial tracks. For all models, the error is maximal for segments that end near $t_{S/G2}$. Errors in $\Fucci_1$ show the same pattern while the errors in the $\Fucci_2$ arise mostly when taking segments from the beginning of the cell cycle (Supp. \figureref{fig:partial_track_fucci}).


\subsection{Prediction on out-of-distribution perturbations}

Finally, we evaluate the model performance on \datadrug, \ie biologically strongly perturbed cells that can be considered out-of-distribution. 
Specifically, these cells were treated with the drug \emph{Palbociclib}, a CDK4-6 inhibitor, that increases the cell cycle duration almost two-fold from $\sim$20h to 40 h (\figureref{fig:drug_predictions} a) and specifically the G1 phase duration, leading to a strongly distorted cell-cycle (\figureref{fig:drug_predictions} b). 
As expected, almost all models demonstrated a significant drop in accuracy when predicting \Fucci signal in these unseen drug-treated cells, as indicated by all evaluation metrics (Table \ref{tab:Table3}). The notable exception is the bidirectional transformer, that provides reasonable predictions and correctly captures distortions in the G1 phase (\figureref{fig:drug_predictions} c) that all other models significantly underestimated. Interestingly, the MLP outperformed the other causal models on this distorted data, potentially as the later overfitted on the training data.
When performing the same analysis on H2B, we found slightly better predictive performance (Supp. \tableref{tab:sup_table_WT,tab:sup_table_drug}, Supp. \figureref{fig:predictions_h2b},  Supp. \figureref{fig:partial_track_h2b}), which is expected due to the stronger correlation with the cell cycle. 


\begin{figure}[t!]
 %
 %

\floatconts
  {fig:drug_predictions}
  {\caption{\textbf{Results on perturbed RPE cells \datadrug.}
  \textbf{a,b)} Effect of CDK4/6 inhibition on cell cycle durations and onset of G1, S, and G2/M phases. The inhibitor extends G1 duration while leaving S and G2/M unchanged.
  \textbf{c)} Example \Fucci predictions on \datadrug. 
 }}
{\includegraphics[width=1\linewidth]{Figures/Drug_predictions.pdf}\vspace{-.2cm}}
\end{figure}


\begin{table}[b!] 
    \centering
    \floatconts
      {tab:Table3} 
      {\caption{\textbf{$L_1$ error and \DTW for \Fucci channels on \datadrug.}}}
      {
    \footnotesize
    \begin{tabular}{llll}
    \toprule
     \textbf{Models} & $L_{1, FUCCI_1}$ & $L_{1, FUCCI_2}$ & \DTW \\
    \midrule
    Single Frame & 0.239 ± 0.082 & 0.182 ± 0.056 & 5.329 ± 1.147 \\
    Causal CNN & 0.252 ± 0.113 & 0.161 ± 0.059 & 4.323 ± 1.302 \\
    Mamba & 0.485 ± 0.090 & 0.259 ± 0.045 & 3.563 ± 1.918 \\
    Transformer & \textbf{0.147} \textbf{± 0.056} & \textbf{0.139} \textbf{± 0.048} & \textbf{3.022} \textbf{± 0.985} \\
    \bottomrule
    \end{tabular}
    }
    \end{table}
    
\vspace{-.1cm}


\section{Discussion and Conclusions}
In this study, we generated and released a large dataset of cycling RPE1 cells under both normal and drug-treated conditions and used it to investigate the utility of sequence models to infer the continuous cell cycle state from label-free brightfield images. 
Our analysis demonstrates that temporal sequence models can significantly improve the cell cycle prediction accuracy and enable the assessment of cell cycle state from brightfield images to a level comparable when using the more informative H2B channel. This suggests that brightfield morphological cues alone carry sufficient information for cell cycle inference.
Importantly, we found that causal state-space models substantially outperform commonly used fixed-history convolutional networks, demonstrating their potential for real-time computer vision and smart microscopy applications, where such causal inference is essential~\cite{simon_causalxtract_2025,mahecic2022event}. 
We note that the observed reduced accuracy on drug-treated cells underscores that creating general predictive cell state models for strong biological perturbations remains challenging.
However, our findings demonstrate that sequence models can be effective predictors of 
cellular dynamics in more controlled settings.  




\clearpage  


%
\begin{thebibliography}{36}
\providecommand{\natexlab}[1]{#1}
\providecommand{\url}[1]{\texttt{#1}}
\expandafter\ifx\csname urlstyle\endcsname\relax
  \providecommand{\doi}[1]{doi: #1}\else
  \providecommand{\doi}{doi: \begingroup \urlstyle{rm}\Url}\fi

\bibitem[Blasi et~al.()Blasi, Hennig, Summers, Theis, Cerveira, Patterson, Davies, Filby, Carpenter, and Rees]{blasi_label-free_2016}
Thomas Blasi, Holger Hennig, Huw~D. Summers, Fabian~J. Theis, Joana Cerveira, James~O. Patterson, Derek Davies, Andrew Filby, Anne~E. Carpenter, and Paul Rees.
\newblock Label-free cell cycle analysis for high-throughput imaging flow cytometry.
\newblock \emph{Nature Communications}, 7\penalty0 (1):\penalty0 10256.
\newblock ISSN 2041-1723.
\newblock Publisher: Nature Publishing Group.

\bibitem[Chu et~al.(2020)Chu, Abe, Yokota, Sudo, Nakamura, Chang, Fang, and Tsai]{chu_prediction_2020}
Slo-Li Chu, Kuniya Abe, Hideo Yokota, Kazuhiro Sudo, Yukio Nakamura, Yuan-Hsiang Chang, Liang-Che Fang, and Ming-Dar Tsai.
\newblock Prediction for morphology and states of stem cell colonies using a lstm network with progressive training microscopy images.
\newblock In \emph{2020 42nd Annual International Conference of the IEEE Engineering in Medicine \& Biology Society (EMBC)}, pages 1820--1823, 2020.
\newblock \doi{10.1109/EMBC44109.2020.9175759}.

\bibitem[Eulenberg et~al.()Eulenberg, Köhler, Blasi, Filby, Carpenter, Rees, Theis, and Wolf]{eulenberg_reconstructing_2017}
Philipp Eulenberg, Niklas Köhler, Thomas Blasi, Andrew Filby, Anne~E. Carpenter, Paul Rees, Fabian~J. Theis, and F.~Alexander Wolf.
\newblock Reconstructing cell cycle and disease progression using deep learning.
\newblock \emph{Nature Communications}, 8\penalty0 (1):\penalty0 463.
\newblock ISSN 2041-1723.
\newblock Publisher: Nature Publishing Group.

\bibitem[Grant et~al.()Grant, Kedziora, Limas, Cook, and Purvis]{grant_accurate_2018}
Gavin~D. Grant, Katarzyna~M. Kedziora, Juanita~C. Limas, Jeanette~Gowen Cook, and Jeremy~E. Purvis.
\newblock Accurate delineation of cell cycle phase transitions in living cells with {PIP}-{FUCCI}.
\newblock \emph{Cell Cycle}, 17\penalty0 (21):\penalty0 2496--2516.
\newblock ISSN 1538-4101, 1551-4005.

\bibitem[Gu and Dao(2023)]{gu_mamba_2024}
Albert Gu and Tri Dao.
\newblock Mamba: Linear-time sequence modeling with selective state spaces.
\newblock \emph{arXiv preprint arXiv:2312.00752}, 2023.

\bibitem[Gu et~al.(2022)Gu, Goel, and Ré]{gu_efficiently_2022}
Albert Gu, Karan Goel, and Christopher Ré.
\newblock Efficiently modeling long sequences with structured state spaces.
\newblock \emph{arXiv}, \penalty0 ({arXiv}:2111.00396), 2022.

\bibitem[He et~al.(2016)He, Zhang, Ren, and Sun]{he2016deep}
Kaiming He, Xiangyu Zhang, Shaoqing Ren, and Jian Sun.
\newblock {Deep Residual Learning for Image Recognition}.
\newblock In \emph{Proceedings of 2016 IEEE Conference on Computer Vision and Pattern Recognition}, CVPR '16, pages 770--778. IEEE, June 2016.

\bibitem[He et~al.()He, He, Kandel, Lee, Hu, Sobh, Anastasio, and Popescu]{he_cell_2022}
Yuchen~R. He, Shenghua He, Mikhail~E. Kandel, Young~Jae Lee, Chenfei Hu, Nahil Sobh, Mark~A. Anastasio, and Gabriel Popescu.
\newblock Cell cycle stage classification using phase imaging with computational specificity.
\newblock \emph{{ACS} Photonics}, 9\penalty0 (4):\penalty0 1264--1273.
\newblock Publisher: American Chemical Society.

\bibitem[Held et~al.()Held, Schmitz, Fischer, Walter, Neumann, Olma, Peter, Ellenberg, and Gerlich]{held_cellcognition_2010}
Michael Held, Michael H.~A. Schmitz, Bernd Fischer, Thomas Walter, Beate Neumann, Michael~H. Olma, Matthias Peter, Jan Ellenberg, and Daniel~W. Gerlich.
\newblock {CellCognition}: time-resolved phenotype annotation in high-throughput live cell imaging.
\newblock \emph{Nature Methods}, 7\penalty0 (9):\penalty0 747--754.
\newblock ISSN 1548-7105.
\newblock Publisher: Nature Publishing Group.

\bibitem[Hewamalage et~al.(2021)Hewamalage, Bergmeir, and Bandara]{hewamalage2021recurrent}
Hansika Hewamalage, Christoph Bergmeir, and Kasun Bandara.
\newblock Recurrent neural networks for time series forecasting: Current status and future directions.
\newblock \emph{International Journal of Forecasting}, 37\penalty0 (1):\penalty0 388--427, 2021.
\newblock ISSN 0169-2070.

\bibitem[Hochreiter and Schmidhuber(1997)]{hochreiter_long_1997}
Sepp Hochreiter and Jürgen Schmidhuber.
\newblock Long short-term memory.
\newblock \emph{Neural Computation}, 9\penalty0 (8):\penalty0 1735--1780, 1997.
\newblock ISSN 0899-7667.
\newblock Conference Name: Neural Computation.

\bibitem[Jin et~al.()Jin, Zou, and Huang]{jin_imbalanced_2021}
Xin Jin, Yuanwen Zou, and Zhongbing Huang.
\newblock An imbalanced image classification method for the cell cycle phase.
\newblock \emph{Information}, 12\penalty0 (6):\penalty0 249.
\newblock ISSN 2078-2489.
\newblock Number: 6 Publisher: Multidisciplinary Digital Publishing Institute.

\bibitem[Jose et~al.(2024)Jose, Roy, Moreno-Andrés, and Stegmaier]{jose_automatic_2024}
Abin Jose, Rijo Roy, Daniel Moreno-Andrés, and Johannes Stegmaier.
\newblock Automatic detection of cell-cycle stages using recurrent neural networks.
\newblock \emph{PloS one}, 19\penalty0 (3):\penalty0 e0297356, 2024.
\newblock ISSN 1932-6203.

\bibitem[Kastan and Bartek(2004)]{kastan_cell-cycle_2004}
Michael~B. Kastan and Jiri Bartek.
\newblock Cell-cycle checkpoints and cancer.
\newblock \emph{Nature}, 432\penalty0 (7015):\penalty0 316--323, November 2004.
\newblock ISSN 1476-4687.

\bibitem[Li et~al.()Li, Nichols, Browning, Longhi, Camplisson, Beliveau, and Noble]{li_predicting_2024}
Gang Li, Eva~K. Nichols, Valentino~E. Browning, Nicolas~J. Longhi, Conor Camplisson, Brian~J. Beliveau, and William~Stafford Noble.
\newblock Predicting cell cycle stage from 3d single-cell nuclear-stained images.
\newblock \emph{bioRxiv}.

\bibitem[Mahecic et~al.(2022)Mahecic, Stepp, Zhang, Griffi{\'e}, Weigert, and Manley]{mahecic2022event}
Dora Mahecic, Willi~L Stepp, Chen Zhang, Juliette Griffi{\'e}, Martin Weigert, and Suliana Manley.
\newblock Event-driven acquisition for content-enriched microscopy.
\newblock \emph{Nature methods}, 19\penalty0 (10):\penalty0 1262--1267, 2022.

\bibitem[Meert et~al.()Meert, Hendrickx, Van~Craenendonck, Robberechts, Blockeel, and Davis]{meert_dtaidistance_2020}
Wannes Meert, Kilian Hendrickx, Toon Van~Craenendonck, Pieter Robberechts, Hendrik Blockeel, and Jesse Davis.
\newblock {DTAIDistance}.

\bibitem[Moreno-Andrés et~al.()Moreno-Andrés, Bhattacharyya, Scheufen, and Stegmaier]{moreno-andres_livecellminer_2022}
Daniel Moreno-Andrés, Anuk Bhattacharyya, Anja Scheufen, and Johannes Stegmaier.
\newblock {LiveCellMiner}: A new tool to analyze mitotic progression.
\newblock \emph{{PLOS} {ONE}}, 17\penalty0 (7):\penalty0 e0270923.
\newblock ISSN 1932-6203.
\newblock Publisher: Public Library of Science.

\bibitem[Narotamo et~al.()Narotamo, Fernandes, Moreira, Melo, Seruca, Silveira, and Sanches]{narotamo_machine_2021}
Hemaxi Narotamo, Maria~Sofia Fernandes, Ana~Margarida Moreira, Soraia Melo, Raquel Seruca, Margarida Silveira, and João~Miguel Sanches.
\newblock A machine learning approach for single cell interphase cell cycle staging.
\newblock \emph{Scientific Reports}, 11\penalty0 (1):\penalty0 19278.
\newblock ISSN 2045-2322.
\newblock Publisher: Nature Publishing Group.

\bibitem[Preibisch et~al.()Preibisch, Saalfeld, and Tomancak]{preibisch_globally_2009}
Stephan Preibisch, Stephan Saalfeld, and Pavel Tomancak.
\newblock Globally optimal stitching of tiled 3d microscopic image acquisitions.
\newblock \emph{Bioinformatics}, 25\penalty0 (11):\penalty0 1463--1465.
\newblock ISSN 1367-4803.

\bibitem[Rappez et~al.()Rappez, Rakhlin, Rigopoulos, Nikolenko, and Alexandrov]{rappez_deepcycle_2020}
Luca Rappez, Alexander Rakhlin, Angelos Rigopoulos, Sergey Nikolenko, and Theodore Alexandrov.
\newblock {DeepCycle} reconstructs a cyclic cell cycle trajectory from unsegmented cell images using convolutional neural networks.
\newblock \emph{Molecular Systems Biology}, 16\penalty0 (10).
\newblock ISSN 1744-4292, 1744-4292.
\newblock {ZSCC}: 0000000.

\bibitem[Sakaue-Sawano et~al.()Sakaue-Sawano, Yo, Komatsu, Hiratsuka, Kogure, Hoshida, Goshima, Matsuda, Miyoshi, and Miyawaki]{sakaue-sawano_genetically_2017}
Asako Sakaue-Sawano, Masahiro Yo, Naoki Komatsu, Toru Hiratsuka, Takako Kogure, Tetsushi Hoshida, Naoki Goshima, Michiyuki Matsuda, Hiroyuki Miyoshi, and Atsushi Miyawaki.
\newblock Genetically encoded tools for optical dissection of the mammalian cell cycle.
\newblock \emph{Molecular Cell}, 68\penalty0 (3):\penalty0 626--640.e5.
\newblock ISSN 10972765.

\bibitem[Sakaue-Sawano et~al.(2008)Sakaue-Sawano, Kurokawa, Morimura, Hanyu, Hama, Osawa, Kashiwagi, Fukami, Miyata, Miyoshi, Imamura, Ogawa, Masai, and Miyawaki]{sakaue-sawano_visualizing_2008}
Asako Sakaue-Sawano, Hiroshi Kurokawa, Toshifumi Morimura, Aki Hanyu, Hiroshi Hama, Hatsuki Osawa, Saori Kashiwagi, Kiyoko Fukami, Takaki Miyata, Hiroyuki Miyoshi, Takeshi Imamura, Masaharu Ogawa, Hisao Masai, and Atsushi Miyawaki.
\newblock Visualizing spatiotemporal dynamics of multicellular cell-cycle progression.
\newblock \emph{Cell}, 132\penalty0 (3):\penalty0 487--498, 2008.
\newblock ISSN 0092-8674, 1097-4172.

\bibitem[Schmidt et~al.(2018)Schmidt, Weigert, Broaddus, and Myers]{schmidt_cell_2018}
Uwe Schmidt, Martin Weigert, Coleman Broaddus, and Gene Myers.
\newblock Cell detection with star-convex polygons.
\newblock In \emph{Medical Image Computing and Computer Assisted Intervention - {MICCAI} 2018 - 21st International Conference, Granada, Spain, September 16-20, 2018, Proceedings, Part {II}}, pages 265--273, 2018.

\bibitem[Simon et~al.()Simon, Comes, Tocci, Dupuis, Cabeli, Lagrange, Mencattini, Parrini, Martinelli, and Isambert]{simon_causalxtract_2025}
Franck Simon, Maria~Colomba Comes, Tiziana Tocci, Louise Dupuis, Vincent Cabeli, Nikita Lagrange, Arianna Mencattini, Maria~Carla Parrini, Eugenio Martinelli, and Herve Isambert.
\newblock {CausalXtract}, a flexible pipeline to extract causal effects from live-cell time-lapse imaging data.
\newblock \emph{{eLife}}, 13:\penalty0 RP95485.
\newblock ISSN 2050-084X.
\newblock Publisher: {eLife} Sciences Publications, Ltd.

\bibitem[Stallaert et~al.()Stallaert, Kedziora, Taylor, Zikry, Ranek, Sobon, Taylor, Young, Cook, and Purvis]{stallaert_structure_2022}
Wayne Stallaert, Katarzyna~M. Kedziora, Colin~D. Taylor, Tarek~M. Zikry, Jolene~S. Ranek, Holly~K. Sobon, Sovanny~R. Taylor, Catherine~L. Young, Jeanette~G. Cook, and Jeremy~E. Purvis.
\newblock The structure of the human cell cycle.
\newblock \emph{Cell Systems}, 13\penalty0 (3):\penalty0 230--240.e3.
\newblock ISSN 2405-4712.

\bibitem[Su et~al.(2024)Su, Ahmed, Lu, Pan, Bo, and Liu]{su2024roformer}
Jianlin Su, Murtadha Ahmed, Yu~Lu, Shengfeng Pan, Wen Bo, and Yunfeng Liu.
\newblock Roformer: Enhanced transformer with rotary position embedding.
\newblock \emph{Neurocomputing}, 568:\penalty0 127063, 2024.

\bibitem[Su et~al.(2017)Su, Lu, Chen, and Liu]{su_spatiotemporal_2017}
Yu-Ting Su, Yao Lu, Mei Chen, and An-An Liu.
\newblock Spatiotemporal joint mitosis detection using cnn-lstm network in time-lapse phase contrast microscopy images.
\newblock \emph{IEEE Access}, 5:\penalty0 18033--18041, 2017.

\bibitem[Tinevez et~al.(2017)Tinevez, Perry, Schindelin, Hoopes, Reynolds, Laplantine, Bednarek, Shorte, and Eliceiri]{tinevez_trackmate_2017}
Jean-Yves Tinevez, Nick Perry, Johannes Schindelin, Genevieve~M. Hoopes, Gregory~D. Reynolds, Emmanuel Laplantine, Sebastian~Y. Bednarek, Spencer~L. Shorte, and Kevin~W. Eliceiri.
\newblock {TrackMate}: An open and extensible platform for single-particle tracking.
\newblock \emph{Methods}, 115:\penalty0 80--90, 2017.
\newblock ISSN 1046-2023.

\bibitem[Ulicna et~al.()Ulicna, Kelkar, Soelistyo, Charras, and Lowe]{ulicna_learning_2023}
Kristina Ulicna, Manasi Kelkar, Christopher~J Soelistyo, Guillaume~T Charras, and Alan~R Lowe.
\newblock Learning dynamic image representations for self-supervised cell cycle annotation.
\newblock In \emph{Workshop on Computational Biology}.

\bibitem[Van Den~Oord et~al.(2016)Van Den~Oord, Dieleman, Zen, Simonyan, Vinyals, Graves, Kalchbrenner, Senior, Kavukcuoglu, et~al.]{van2016wavenet}
Aaron Van Den~Oord, Sander Dieleman, Heiga Zen, Karen Simonyan, Oriol Vinyals, Alex Graves, Nal Kalchbrenner, Andrew Senior, Koray Kavukcuoglu, et~al.
\newblock Wavenet: A generative model for raw audio.
\newblock \emph{arXiv preprint arXiv:1609.03499}, 12, 2016.

\bibitem[Vaswani et~al.(2017)Vaswani, Shazeer, Parmar, Uszkoreit, Jones, Gomez, Kaiser, and Polosukhin]{vaswani_attention_nodate}
Ashish Vaswani, Noam Shazeer, Niki Parmar, Jakob Uszkoreit, Llion Jones, Aidan~N Gomez, \L~ukasz Kaiser, and Illia Polosukhin.
\newblock Attention is all you need.
\newblock In I.~Guyon, U.~Von Luxburg, S.~Bengio, H.~Wallach, R.~Fergus, S.~Vishwanathan, and R.~Garnett, editors, \emph{Advances in Neural Information Processing Systems}, volume~30. Curran Associates, Inc., 2017.

\bibitem[Wang et~al.(2020)Wang, Douglas, Zhang, Kumari, Enuameh, Dai, Wallace, Watkins, Shu, and Xing]{wang_live-cell_2020}
Weikang Wang, Diana Douglas, Jingyu Zhang, Sangeeta Kumari, Metewo~Selase Enuameh, Yan Dai, Callen~T. Wallace, Simon~C. Watkins, Weiguo Shu, and Jianhua Xing.
\newblock Live-cell imaging and analysis reveal cell phenotypic transition dynamics inherently missing in snapshot data.
\newblock \emph{Science Advances}, 6\penalty0 (36):\penalty0 eaba9319, 2020.
\newblock Publisher: American Association for the Advancement of Science.

\bibitem[Weigert and Schmidt()]{weigert_nuclei_2022}
Martin Weigert and Uwe Schmidt.
\newblock Nuclei instance segmentation and classification in histopathology images with stardist.
\newblock In \emph{2022 {IEEE} International Symposium on Biomedical Imaging Challenges ({ISBIC})}, pages 1--4.

\bibitem[Wen et~al.(2023)Wen, Zhou, Zhang, Chen, Ma, Yan, and Sun]{wen2022transformers}
Qingsong Wen, Tian Zhou, Chaoli Zhang, Weiqi Chen, Ziqing Ma, Junchi Yan, and Liang Sun.
\newblock Transformers in time series: A survey.
\newblock In Edith Elkind, editor, \emph{Proceedings of the Thirty-Second International Joint Conference on Artificial Intelligence, {IJCAI-23}}, pages 6778--6786. International Joint Conferences on Artificial Intelligence Organization, 8 2023.
\newblock Survey Track.

\bibitem[Zhao et~al.()Zhao, De~Perez, Dimitrova, Kemp, and Anderson]{zhao_insights_2024}
Xinran Zhao, Alexander~Ruys De~Perez, Elena~S. Dimitrova, Melissa Kemp, and Paul~E. Anderson.
\newblock Insights into cellular evolution: Temporal deep learning models and analysis for cell image classification.
\newblock \emph{bioRxiv}.

\end{thebibliography}
 %


\newpage
%

\appendix
%
%

\section{Supplementary Methods}
\subsection{Cell culture}
\Fucci-RPE1 cells, kindly provided by Battich et al. [2020], were cultured at 37°C with 5\% CO2 in DMEM/F12 medium (Gibco 11320033), supplemented with 1 \% non-essential amino acids (NEAA) (Gibco 11140-035), 1\% penicillin-streptomycin (Sigma-Aldrich G6784), and 10\% fetal bovine serum (FBS) (Gibco 10437-028). In addition, the H2B-iRFP marker, driven by a PGK promoter, was introduced into the cells using the second-generation lentiviral system with a commercially available plasmid (Addgene: 90237).

\subsection{Imaging}
For imaging, H2B-\Fucci-RPE1 cells were seeded into 96-well plates and cultured under the conditions described above, with Fluorobrite medium (Gibco A1896701) replacing DMEM/F12. For the perturbation experiments, cells were treated with 10 nM Palbociclib (CDK4-6 inhibitor).  Images from four channels—Brightfield, H2B (far red), Cdt1 (red), and Geminin (green)—were acquired every 5 minutes using a PerkinElmer Operetta Microscope with a 20x/0.80 objective. Four or nine tiles per well were captured for each channel, with a 15\% overlap for subsequent stitching.

\subsection{Image preprocessing} 
Image preprocessing involved stitching the tiles \cite{preibisch_globally_2009} and applying background subtraction to fluorescent channels using a rolling ball algorithm. Cell nuclei were segmented on the H2B channel with a custom StarDist model \cite{weigert_nuclei_2022} , and tracked across frames using TrackMate \cite{tinevez_trackmate_2017}. Full cell cycle tracks (M-M, tracks encompassing one complete cell division cycle from one mitosis (M) to the next) were isolated using K-Means clustering of interpolated \Fucci signals. Our groundtruth labels were obtained by averaging the fluorescent channels over the nuclei area and taking the logarithm of this signal, with a smooth noise removal. 
The raw fluorescent \Fucci signal is not normalized for background noise (starting at $2^{5}$) and expresses a greater dynamic range in log scale as previously shown in DeepCycle \cite{rappez_deepcycle_2020}. However taking the logarithm increases the dynamic range of the background noise, leading to interesting questions about the proper scale of these tracks.
We average the pixels present in the nucleus for each \Fucci marker and then for the background noise normalization, we express the signal shifted and in units of an $\epsilon$, where $\epsilon$ can be k-th percentile of the signals distribution. We use the 1st percentile $\epsilon = P_1 = (P_{1,f_1}, P_{1,f_2})$. Then to deal with the increased dynamic range from the log, we take the Softplus with $\beta = 1$ of our new units of $\epsilon$ before applying the logarithm:
$$ F = (f_1, f_2), \quad \overline{F} = \dfrac{F}{A} = (\overline{f_1}, \overline{f_2})$$
$$ F^\prime = \dfrac{ \overline{F} - \epsilon}{\epsilon}, \quad \text{\Fucci} = log_2(Softplus(F^\prime)) $$

Where $F$ is the raw Fucci values from imaging of each nucleus pixel, $A$ is the area of the nucleus and $\text{Softplus}(x) = \dfrac{1}{\beta} * log(1 + exp(\beta * x))$. 




\newpage
\section{Supplementary Tables}
\setcounter{table}{0}
\renewcommand{\thetable}{B \arabic{table}}


\begin{table}[H]
    \floatconts
    {tab:params}
    {\caption{Number of parameters for each sequence model head.}}
    {
        \centering
    \begin{tabular}{ll}
    \toprule
    \textbf{Model} & \textbf{Parameters} \\
    \midrule
    MLP & 1.32 × 10\textsuperscript{6} \\
    Causal CNN & 0.94 × 10\textsuperscript{6} \\
    LSTM & 1.28 × 10\textsuperscript{6} \\
    Mamba & 1.11 × 10\textsuperscript{6} \\
    Transformer & 1.12 × 10\textsuperscript{6} \\
    \bottomrule
        \end{tabular}
    }
\end{table}



\begin{table}[H]
\floatconts
  {tab:sup_table_WT} %
  {\caption{\textbf{Side by side performance comparison of BF and H2B modalities at predicting \Fucci channels on \dataregular.} Both data modalities present similar results: sequence encoders outperform the single frame method. Moreover H2B only shows modestly better performance than BF.}}
  {
\centering
\begin{adjustbox}{width=1.\textwidth,center}
  \begin{tabular}{lllllllll}
    \toprule
    & \multicolumn{4}{c|}{Brightfield} & \multicolumn{4}{c}{Histone H2B} \\
    \textbf{Models} & $L_{1, FUCCI_1}$ & $L_{1, FUCCI_2}$ & $R^2$ & $DTW$ & $L_{1, FUCCI_1}$ & $L_{1, FUCCI_2}$ & $R^2$ & \DTW \\
\midrule
Single Frame & 0.193 ± 0.066 & 0.146 ± 0.045 & 0.459 ± 0.271 & 3.735 ± 0.863 & 0.183 ± 0.104 & 0.130 ± 0.064 & 0.491 ± 0.431 & 2.595 ± 1.201 \\
Causal CNN & 0.157 ± 0.078 & 0.122 ± 0.049 & 0.608 ± 0.294 & 2.468 ± 0.917 & 0.154 ± 0.105 & 0.118 ± 0.061 & 0.586 ± 0.415 & 2.165 ± 1.210 \\
LSTM & 0.108 ± 0.069 & 0.087 ± 0.047 & 0.749 ± 0.266 & 1.527 ± 0.814 & 0.079 ± 0.065 & 0.075 ± 0.044 & 0.833 ± 0.265 & 1.467 ± 1.161 \\
Causal Transformer & 0.121 ± 0.073 & 0.094 ± 0.048 & 0.720 ± 0.279 & 1.728 ± 0.811 & 0.079 ± 0.057 & 0.079 ± 0.042 & 0.839 ± 0.214 & 1.552 ± 0.955 \\
Mamba & 0.112 ± 0.072 & 0.091 ± 0.049 & 0.739 ± 0.282 & 1.444 ± 0.898 & 0.074 ± 0.056 & 0.075 ± 0.040 & 0.853 ± 0.215 & 1.426 ± 0.949 \\
Transformer & \textbf{0.066} \textbf{± 0.038} & \textbf{0.062} \textbf{± 0.037} & \textbf{0.892} \textbf{± 0.111} & \textbf{1.285} \textbf{± 0.553} & \textbf{0.056} \textbf{± 0.039} & \textbf{0.054} \textbf{± 0.033} & \textbf{0.912} \textbf{± 0.116} & \textbf{1.155} \textbf{± 0.612} \\
\bottomrule
  \end{tabular}
\end{adjustbox}}
\end{table}

\begin{table}[H]
\floatconts
  {tab:sup_table_drug}
  {\caption{\textbf{Performance metrics for  both brightfield and histone H2B modalities on \datadrug.}}
  }
  {
\centering
\begin{adjustbox}{width=1.\textwidth,center}
  \begin{tabular}{lllllllll}
    \toprule
    \textbf{Palbociclib} & \multicolumn{4}{c|}{Brightfield} & \multicolumn{4}{c}{Histone H2B} \\
    \textbf{Models} & $L_{1, FUCCI_1}$ & $L_{1, FUCCI_2}$ & $R^2$ & \DTW & $L_{1, FUCCI_1}$ & $L_{1, FUCCI_2}$ & $R^2$ & \DTW \\
\midrule
Single Frame & 0.239 ± 0.082 & 0.182 ± 0.056 & -0.297 ± 1.064 & 5.329 ± 1.147 & 0.183 ± 0.050 & 0.107 ± 0.048 & 0.260 ± 0.466 & 3.285 ± 0.820 \\
Causal CNN & 0.252 ± 0.113 & 0.161 ± 0.059 & -0.353 ± 1.459 & 4.323 ± 1.302 & 0.149 ± 0.042 & 0.125 ± 0.040 & 0.401 ± 0.376 & 3.077 ± 0.840 \\
LSTM & 0.424 ± 0.101 & 0.229 ± 0.045 & -1.663 ± 1.727 & 3.685 ± 1.678 & 0.140 ± 0.051 & 0.115 ± 0.040 & 0.503 ± 0.404 & \textbf{2.750} \textbf{± 0.838} \\
Causal Transformer & 0.326 ± 0.104 & 0.214 ± 0.049 & -0.728 ± 1.558 & 5.159 ± 1.562 & 0.132 ± 0.034 & 0.111 ± 0.037 & 0.628 ± 0.207 & 3.154 ± 0.872 \\
Mamba & 0.485 ± 0.090 & 0.259 ± 0.045 & -2.244 ± 1.949 & 3.563 ± 1.918 & 0.185 ± 0.078 & 0.134 ± 0.043 & 0.255 ± 0.651 & 2.896 ± 0.903 \\
Transformer & \textbf{0.147} \textbf{± 0.056} & \textbf{0.139} \textbf{± 0.048} & \textbf{0.408} \textbf{± 0.478} & \textbf{3.022} \textbf{± 0.985} & \textbf{0.074} \textbf{± 0.029} & \textbf{0.095} \textbf{± 0.031} & \textbf{0.789} \textbf{± 0.131} & 2.896 ± 1.201 \\
\bottomrule
  \end{tabular}
\end{adjustbox}}
\end{table}


\newpage
\section{Supplementary Figures}





\begin{figure}[h]
\floatconts
  {fig:histograms}
  {\caption{\textbf{Error Distribution of Predictions of \Fucci on Test Set with Brightfield.} \textbf{a.} Distribution of L1 errors across the different  models
\textbf{b.} Error Distributions with Q1, Median and Q3 Percentiles overlayed
\textbf{c.} Q1, Median and Q3 Error Predictions visualized per model.
  }} 
  {\includegraphics[width=1\linewidth]{Figures/histograms-representative-tracks.pdf}} 
\end{figure}


\begin{figure}[h]
\floatconts
  {fig:umap}
  {\caption{\textbf{Learned Latent Space Representations (UMAP).} Each frame of a track is represented as a dot in umap space, the coloring is the normalized time \textbf{a.} Single Frame (no history).
\textbf{b.} Transformer (full sequence).
  }} 
  {\includegraphics[width=1\linewidth]{Figures/umaps/umaps_better_res.pdf}} 
\end{figure}

\begin{figure}[h]
\floatconts
  {fig:visual_landmarks}
  {\caption{\textbf{Predicted $\Delta t_{G1/S}$ and $\Delta t_{S/G2}$ from  BF images for the different models.} 
  }} 
  {\includegraphics[width=1\linewidth]{Figures/visualize_landmark_predictions.pdf}} 
\end{figure}




\begin{figure}[t!]
\floatconts
  {fig:partial_track_fucci}
  {\caption{\textbf{Comparative Performance of Temporal Encoders in Predicting \Fucci1 and \Fucci2 from BF and H2B in partial cell cycle tracks.}
 Error maps showing the prediction error of the different models, assessed on the last frame of segments from the M-M track, spanning indices $\tau_1$ to $\tau_2$.  
 }}
{\includegraphics[width=1\linewidth]{Figures/partial_track_by_FUCCI.pdf}}
\end{figure}


\begin{figure}[h]
\floatconts
  {fig:predictions_h2b}
  {\caption{\textbf{Comparative Performance of Temporal Encoders in Predicting Continuous Cell Cycle States from H2B Imaging in Unperturbed RPE Cells.} \textbf{a)} Distribution of L1 errors across the different  models.
\textbf{b)} Example predictions of \Fucci signals from different models on two tracks: one with accurate predictions and one with poor predictions. The ground truth signal is shown in black.
\textbf{c)} Average prediction error and  \textbf{d.} ground truth standard deviation are plotted in function of   cell cycle phases. 
  }} 
  {\includegraphics[width=1\linewidth]{Figures/Predictions_h2b.pdf}} 
\end{figure}





\begin{figure}[t!]
\floatconts
  {fig:partial_track_h2b}
  {\caption{\textbf{Comparative Performance of Temporal Encoders in Predicting Continuous Cell Cycle States from H2B in partial cell cycle tracks.}
  Error maps showing the prediction error of the different models, assessed on the last frame of segments from the M-M track, spanning indices $\tau_1$ to $\tau_2$. 
 }}
{\includegraphics[width=1\linewidth]{Figures/Partial_track_h2b.pdf}}
\end{figure}





 

\end{document}
