
% \begin{figure}[t]
%     \small
%     \centering
%     \begin{subfigure}{0.49\linewidth}
%         \includegraphics[width=0.98\linewidth]{figure/Protein Hydrolysis.png}
%         \vskip -0.5em
%         \caption{Protein Hydrolysis }
%         \label{fig:pre1}
%     \end{subfigure}
%     \begin{subfigure}{0.49\linewidth}
%         \includegraphics[width=0.98\linewidth]{figure/activesite_blue.png}
%         \vskip -0.5em
%         \caption{Enzyme Active Site}
%         \label{fig:pre2}
%     \end{subfigure}
%     \vskip -1 em
%     \caption{Illustration of Enzyme-Catalyzed Protein Hydrolysis.}
%     \vskip -1 em
%     \label{fig:prelinminary}
% \end{figure}




\begin{table}[t]
    \small
    \centering
    \caption{Statistics of Cleavage Sites Data in MEROPS.}
    \resizebox{0.98\columnwidth}{!}{ % Adjust the width to column width
    \begin{tabular}{cccc}
    \toprule
      \#Proteolytic Enzyme   & \#Substrate & Enzyme–Substrate Ratio \\
      \midrule
         866 \ & 10146 & 11.45 \\
    \bottomrule
    \end{tabular}
    }
    \label{tab:data}
    \vskip -1em
\end{table}












\section{Problem Formulation}
In this section, we firstly introduce the preliminaries of enzyme-catalyzed protein hydrolysis. Then, we present the formal problem definition of protein cleavage site prediction with enzyme active-site knowledge. 
% \enyan{We should introduce the process of protein hydrolysis. And also explain the current framework of Protein Hydrolysis}

% \enyan{Also we will introduced the data form of Enzyme and substrate protein in aspect of ML.}





\subsection{Preliminaries of Protein Hydrolysis}

\textbf{Cleavage Sites in Protein Hydrolysis}. \textit{Protein hydrolysis} is a biochemical process where proteins are broken down into smaller fragments such as amino acids and peptides under the catalysis of proteolytic enzymes. As illustrated in Fig.~\ref{fig:bio process}, during the protein hydrolysis, proteolytic enzymes will firstly recognize specific amino acid sequences or structural motifs within unfold substrate proteins. Then, the enzymes catalyze the cleavage of peptide bonds at the \textit{cleavage site}, leading to the formation of smaller peptide fragments or individual amino acids. The positions of cleavage sites are governed by various factors including substrate's amino acid composition, spatial conformation, and unique properties of the enzyme~\cite{MechanismsFunction, specificityandcatalysis, Turk2001}. 
% Predicting the cleavage sites in enzyme-catalyzed protein hydrolysis can aid in designing proteolytic enzyme variants with enhanced specificity. 
In therapeutic contexts, precisely designed enzymes can degrade disease-associated proteins while minimizing off-target effects~\cite{TANDON2021102455, meghwanshi2020enzymes}. Moreover, cleavage site prediction could pinpoint key molecular regions for therapeutic intervention in the development of targeted peptide-based drugs~\cite{Radchenko2019}. 




% \enyan{draw a figure to visualize this process, of course based on the figure, you will have more comprehensive introduction about the protein hydrolysis. Also you will need to show what is the cleavage site}. 

% \textbf{Cleavage Site in Protein Hydrolysis}. In protein hydrolysis, proteins are broken down at the cleavage site, a position determined by the interactions between substrate proteins and the enzyme. Identifying this specific site is crucial because: 

% \enyan{We will need to illustrate the necessity of cleavage site. Specially, we need to mention that the cleavage site is the cleavage site is not a fixing sequence for each enzyme. It is determined by both substrates and enzyme structures. }.  

% \enyan{We discuss the applications of identifying cleavage sites. }

% \textbf{Cleavage Site in Protein Hydrolysis}. 



% Additionally, micro environmental conditions (e.g., pH and ion concentration) and the dynamic nature of enzyme-substrate interactions further shape the specificity and adaptability of cleavage site recognition. 
% Understanding these sites is essential for predicting enzyme specificity, designing targeted therapeutics, and advancing applications in protein digestion and synthesis.


% \enyan{Such as is written by ChatGPT, check whether they are right.}
% \enyan{Then, we give the current framework of predicting cleavage site of a single enzyme.}

% \textbf{Cleavage Site Resources}. Cleavage site identification relies on experimental techniques such as mass spectrometry-based proteomics, site-directed mutagenesis, and in vitro digestion assays. Complementary to these methods, databases like MEROPS provide extensive datasets of natural and synthetic protein substrates along with their associated cleavage data. 




\textbf{Current Framework of Cleavage Site Prediction}.  Recent studies have employed machine learning models to predict cleavage sites~\cite{WANG2024309, ScreenCap3, SitePrediction, DeepCleave, procleave}. However, these methods generally train an independent model for each enzyme, which only predict the cleavage sites of substrate proteins under the catalysis of one specific enzyme. Specifically, let $\mathcal{P}^s$ denote the  substrate protein, this enzyme-specific cleavage site predictor aims to learn the $f: \mathcal{P}^s \rightarrow \mathbf{c}^{e,s}$, where $\mathbf{c}^{e,s} \in \{0,1\}^{|\mathcal{P}^s|}$ denotes the labels of cleavage site with the enzyme $\mathcal{P}^e$.
However, the training of enzyme-specific model excludes the valuable interaction knowledge from other enzyme–protein systems. In addition, the enzyme-specific model cannot generalize to unseen enzymes, which significantly limits its application in de novo enzyme design. Therefore, it is crucial to develop a unified cleavage site predictor capable of identifying cleavage sites in substrate proteins across various enzymes.





\textbf{Data for Cleavage Site Prediction.}
Tab.~\ref{tab:data} presents statistics from the MEROPS cleavage site database. Due to the high cost of experimental assays, MEROPS~\cite{Merops} primarily includes 866 commonly used proteolytic enzymes, which collectively target 10,146 protein substrates. This limited enzyme coverage poses a significant challenge for developing a unified cleavage site predictor that generalizes across diverse enzyme–substrate systems.






% As shown in Tab, only 866 proteolytic enzymes that mediate protein hydrolysis in natural environments have been identified across various organisms. In contrast, biologists have annotated the cleavage sites of approximately 10146 substrate proteins under the catalysis of these enzymes.
% % With extensive efforts of biologists, cleavage sites within substrate proteins under the catalyzed enzyme have been annotated using experimental methods. 
% This extremely imbalanced ratio between enzymes and substrate proteins poses a significant challenge in developing a unified cleavage site predictor that can generalize across diverse enzyme–substrate systems.
% % Additionally, resources such as the Protein Data Bank (PDB)\cite{} and  like AlphaFold2~\cite{} provide comprehensive sequence and 3D structural information for these enzymes and their substrates.

% These frameworks primarily emphasize substrate-specific feature extraction while overlooking enzyme-specific information, which leads to several limitations:
% \begin{itemize}
%     \item \textbf{Data Dependency}: Each model $f_{E}$ is exclusively trained on substrate data $\mathcal{D}_E = \{(P_i, y_i)\}$, where $y_i$ denotes the ground truth cleavage sites. Since no knowledge is shared among enzymes, the robustness of $f_{E}$ is severely constrained, particularly when $\mathcal{D}_E$ is small or noisy.
%     \item \textbf{Lack of Generalization}: The enzyme-specific model $f_{E}$ cannot generalize to previously unseen enzymes $E_{\mathrm{new}} \notin \mathcal{E}_{\mathrm{train}}$. Consequently, such frameworks lack zero-shot prediction capabilities, making $f_{E_{\mathrm{new}}}(P; \Theta_{E_{\mathrm{new}}})$ inapplicable for enzymes not included in the training set.
% \end{itemize}








\subsection{Preliminaries of Active Sites in Enzyme}

% \enyan{We give the science background of activate site of Enzyme: 1. what is active site. 2. Why this is important information in Hydrolysis Prediction. }

% \enyan{Also we will introduced the data form of available data for Enzyme in aspect of ML. Such as the introduction of the current database of active site. }

During enzyme-catalyzed protein hydrolysis, the active site provides a specialized environment that lowers the activation energy required for peptide bond cleavage.  The illustration of active site of the protein can be found in Fig.~\ref{fig:bio process}. 
With the active sites, enzymes can selectively recognize and bind target substrates, which further enables the cleavage of specific peptide bonds within substrate proteins~\cite{activesites}. 
Therefore, the active sites plays a crucial role in substrate cleavage. 
Hence, incorporating active site information can benefit the modeling of the enzyme-catalyzed protein hydrolysis, thereby enhancing the cleavage site prediction. 

% The physicochemical properties of active sites such as their three-dimensional geometry, charge distribution, and hydrophobicity, govern the enzyme's substrate specificity and efficiency. 

% The active site of an enzyme is a highly specialized region that directly catalyzes biochemical reactions, creating a unique micro-environment that lowers the activation energy required for chemical transformations. plays a critical role in determining enzyme specificity, reaction rates, and substrate recognition. 




% Understanding the structure and function of active sites provides crucial insights into enzyme-mediated processes and lays the foundation for predictive modeling of enzymatic activity. 





% By integrating active site information into predictive models, it becomes possible to better represent the enzyme-substrate interaction landscape, thereby improving the accuracy and reliability of cleavage site predictions. This approach enables a deeper understanding of enzymatic specificity, which is critical for therapeutic target identification and biotechnological applications. 


\textbf{Resources for Active Sites.}
Thanks to the lower cost in annotating active sites of enzymes, UniProt~\cite{Uniprot} provides 10{,}749 high-quality proteolytic enzymes with labeled active sites across multiple organisms. 
Because most of the entries are synthetic substrates instead of natural proteins, these data cannot be directly utilized for cleavage site prediction.  
However the rich information of active sites would be helpful in modeling the enzyme to facilitate the cleavage site prediction in protein hydrolysis.



%thereby enabling the development of robust machine learning models for enzymatic function prediction and active site identification. 


\subsection{Problem Definition}
% \enyan{TO DO}

In protein hydrolysis, both enzyme $\mathcal{P}^e$ and substrate $\mathcal{P}^s$ are proteins composed of amino acid residues that fold into 3D structures. We denote a protein of length $N$ by $\mathcal{P}=(\mathbf{X}, \mathbf{R})$, where $\mathbf{X}\in\mathbb{R}^{N \times d}$ is the feature matrix of $N$ residues, $\mathbf{R} \in \mathbb{R}^{N \times 3}$ denotes the 3D positions of residues. We denote $\mathbf{c}^{e,s} \in \{0,1\}^{|\mathcal{P}_s|}$ as the cleavage site label for the substrate protein $\mathcal{P}^s$ under the catalysis of enzyme $\mathcal{P}^e$. The training data for cleavage site prediction can be represented as $\mathcal{D}_c=\{(\mathcal{P}^e_i, \mathcal{P}^s_i, \mathbf{c}^{e,s}_i)\}^{|\mathcal{D}_c|}_{i=1}$. 
In this work, we propose to enhance the cleavage site prediction with the active site information of enzymes. Hence, we will also utilize a set of enzymes labeled with active sites $\mathbf{a} \in \{0,1\}^{|\mathcal{P}_e|}$ , which is denoted as $\mathcal{D}_a=\{(\mathcal{P}^e_i, \mathbf{a}_i)\}_{i=1}^{|\mathcal{D}_a|}$. 

% In this would 
% In this paper, we will evaluate the performance of cleavage site prediction of test substrate proteins for both seen enzymes and unseen enzymes.  
During the test phase, we will predict the cleavage site for each pair of test enzymes and substrates $(\mathcal{P}^e_t, \mathcal{P}^s_t)$. The active sites of test enzymes will not be available for inference. And the test enzyme $\mathcal{P}^e_t$ can be either seen ,i.e, $\mathcal{P}^e_t \in \mathcal{D}_c$  or unseen, i.e, $\mathcal{P}^e_t \notin \mathcal{D}_c$, which correspond to the applications on well-studied enzymes and de novo enzymes, respectively. For each test substrate protein $\mathcal{P}^s_t$, its cleavage site with the test enzyme $\mathcal{P}^e_t$ is not included in the training set $\mathcal{D}_c$.With the above notations and descriptions, the formal definition of building a unified cleavage site predictor can be given by:
\begin{problem}
    Given the dataset $\mathcal{D}_c$ annotated  for  cleavage site prediction and the supplemented dataset $\mathcal{D}_a$ containing enzymes active sites, we aim to obtain a unified cleavage site predictor:
    \begin{equation}
        f: (\mathcal{P}^e, \mathcal{P}^s) \rightarrow \mathbf{c}^{e,s},
    \end{equation}
    which can accurately predict the cleavage site of test substrate proteins $\mathcal{P}^e_t$ under the catalysis of test proteolytic enzymes $\mathcal{P}^s_t$. Note that the test enzymes can be either seen or unseen during the training phase. 
\end{problem}

% 1. \textbf{Generalization Across Enzymes} Given a limited set of training enzymes $ \mathcal{E}_{\text{train}} $ with imbalanced cleavage data, the model must generalize to a disjoint set of unseen enzymes $ \mathcal{E}_{\text{test}} $, satisfying:$f(E_{new}, P) \approx C, \quad \forall E_{new} \in \mathcal{E}_{\text{test}},$ where $ C $ represents the ground truth cleavage sites. The scarcity of proteolytic enzyme datasets further underscores the need for robust learning strategies that effectively leverage minimal annotated data.


% 2.\textbf{Integration of Active Site Information}  Leverage detailed annotations of active site residues $ A $ to enhance enzyme representation. Active sites are defined as: $A = \{a_i \mid i \in \mathcal{I}_A\}, \quad \mathcal{I}_A \subset \{1, 2, \dots, m\},$ where $ m $ is the enzyme sequence length, and $ \mathcal{I}_A $ indexes active site residues. Integrating $ A $ enhances the representation of enzyme-specific features, thereby improving the modeling of enzyme-substrate interactions.

% 3. \textbf{Incorporation of Structural and Energetic Features} Combine structure $ \mathbf{S} $  and energetic frustration metrics $ \mathbf{E} $ to capture the complex biological context. For any residue pair $ (i, j) $, the distance $ D_{(i,j)} $ and energetic frustration $ E_{(i,j)} $ are defined as:$
%    D_{(i,j)} = \frac{1}{1 + \|r_i - r_j\|}, \quad 
%    E_{(i,j)} = \frac{E_{\text{actual},(i,j)} - \mu_{\text{random},(i,j)}}{\sigma_{\text{random},(i,j)}},
%   $  where $ r_i - r_j $ represents the distance between the Ca atoms of the two residues, and $ E_{ij} $ represents the relative energy between the two residues. These structural and energetic features are crucial for accurately modeling the spatial and energetic interactions that govern enzyme-substrate specificity.

% 3.\textbf{Unification of Structural and Energetic Features} To model the complex biological interactions, structural  $ \mathbf{S} $ and energetic frustration $ \mathbf{E} $ are integrated into a unified feature. Specifically, for any residue pair $ (i, j) $, the distance $ D_{(i,j)} $ and energetic frustration $ E_{(i,j)} $ are denoted as: $D_{(i,j)} = f_{\text{dist}}(r_i, r_j), \quad E_{(i,j)} = f_{\text{energy}}(E_{\text{actual},(i,j)}, E_{\text{random},(i,j)}),$  where $ f_{\text{dist}} $ and $ f_{\text{energy}} $ represent functions for distance normalization and energy standardization, $ E_{\text{actual}} $ represents the true energetic value, while $ E_{\text{random}} $ corresponds to the energy derived from a random configuration. 