\subsection{Notation}

Consider a set of $N$ taxa $\mathcal{X} = \{x_1, x_2,\ldots,x_N\}$. A nonempty subset $X$ of $\mathcal{X}$ is referred to as a \textit{clade} of $\mathcal{X}$. A clade represents a collection of taxa which share a common ancestor at a particular time in the past. Further, we represent evolutionary branching events with bipartitions $\{W,Z\}$ of the clade $X$.

We focus on \textit{ultrametric trees}, which consist of a rooted, binary tree topology $\tau$ and a set of coalescent times $\bft = \{t_n\}_{n=1}^{N-1}$, where there is one $t_n$ for each internal node in $\tau$. For ultrametric trees in particular, the branch length between a child and its parent is equal to the difference in coalescent times of the parent and child nodes. We measure $\bft$ in backwards time, so each $t_n$ is positive and represents time before present. 
%
The leaves of $\tau$ correspond to the genomes of each measured taxon $x \in \calX$. Additionally, an internal node $u$ of $\tau$ represents the (unobserved) genome of the most recent common ancestor of all taxa that have $u$ as a parent node. 
%
As a binary tree, $\tau$ contains a total of $N-1$ internal nodes (including the root). We can thus represent the tree with a collection of bipartitions: $\tau = \{\{W_n,Z_n\}\}_{n=1}^{N-1}$. In this representation, an internal node $u_n$ is the most recent common ancestor for all taxa $x \in W_n \cup Z_n$, and the $n$-th coalescent event is represented by $\{W_n,Z_n\}$.

Denote the set of possible characters within a set of aligned genome sequences by $\Omega$ (\eg, a DNA sequence may correspond to $\Omega = \{A,T,C,G\}$ and an RNA sequence to $\Omega=\{A,U,G,C\}$). Further, denote the set of measured genomes by $\bfY^{\ob} = \{Y^{\ob}_1,\ldots,Y^{\ob}_M\}$, where $Y^{\ob}_m = (Y_{m,x_1}, \ldots, Y_{m,x_N})$ corresponds to the base pairs at site $m$ for all observed taxa $x \in \calX$. In addition to the observed genomes $\bfY^{\ob}$, denote the unobserved genomes of all internal nodes as $\bfY^{\un} = \{Y^{\un}_1,\ldots,Y^{\un}_M\}$, where $Y^{\un}_m = (Y_{m,u_1}, \ldots, Y_{m,u_{N-1}})$ corresponds the base at site $m$ for all \textit{unobserved} internal nodes $u_1,\ldots,u_{N-1}$. Let the index of the root node be $N-1$ (so, $u_{N-1}$ is the root node). We denote the combined observed and unobserved genomes by $\bfY = \{\bfY^{\ob},\bfY^{\un}\}$. For further background on phylogenetics, we refer to~\citet{jotun}.

\subsection{Phylogenetic Likelihood}

For simplicity, we focus on the  \citet{Jukes:1969} model of evolution, but our methods apply more broadly. We define the stationary distribution as $\bfpi$ and a 4-by-4 matrix $P(b)$ with entry $(i,j)$ corresponding to the probability of transitioning from base $i$ to base $j$ given branch length $b$ under \citet{Jukes:1969}. The likelihood function of an observed set of genetic sequences $Y^{\ob}$ at site $m$ is thus:
%
\begin{gather}
    p(Y_m^{\ob}\!\mid\!\tau, \bft) = \sum_{Y_m^{\un}} \hspace{-0.1em}\bfpi(Y_{m,r}) \prod_{(u,v)}\hspace{-0.2em} \left(P(b_{u,v}(\tau,\bft))\right)_{{\textstyle\mathstrut}Y_{m,u},Y_{m,v}}.
\end{gather}
%
Here the product is over all edges $(u,v)$ in $\tau$. As is customary in the phylogenetic inference literature, we assume independence between sites, so the likelihood function for the observed genomes is:
%
\begin{equation}
    \label{eqn:p}
    p(\bfY^{\ob}\!\mid\!\tau, \bft) = \prod_{m=1}^M\hspace{-0.1em} p(Y^{\ob}_m\!\mid\!\tau, \bft).
\end{equation}

Equation (\ref{eqn:p}) can be evaluated in $\calO(NM)$ time using the pruning algorithm \citep{Felsenstein:1981}, which also is known as the sum-product algorithm \citep{Koller:2009}.

\subsection{Prior Distribution over Trees}

We use the Kingman coalescent \citep{Kingman:1982} with a constant effective population size $N_e$ as a prior distribution on $p(\tau,\bft)$:
%
\begin{equation}
    p(\tau,\bft) = \frac{2^{N-1}}{N!(N-1)!} \prod_{k=2}^N \lambda_k \exp \left(-\lambda_k (t_{k} - t_{k-1})\right).
\end{equation}

Here $\lambda_k = \binom{k}{2}/N_e$ is the rate at which taxa coalesce backwards in time.

\subsection{Variational Inference for Phylogenetic Trees}

Our goal is to infer a posterior distribution over tree structures and coalescent times given the observed genetic sequences: 

\begin{equation}
    p(\tau, \bft \mid \bfY^{\ob}) = \frac{p(\bfY^{\ob} \mid \tau, \bft) ~ p(\tau, \bft)}{p(\bfY^{\ob})}.
\end{equation}

Here $p(\bfY^{\ob})$ is an intractable normalization constant.

Variational inference involves defining a tractable family of probability densities parameterized by some variational parameters $\phi$. Then, the posterior density is approximated by a variational density $q_{\phi} (\tau, \bft)$ whose parameters $\phi$ are selected to minimize a divergence measure $D$ between the posterior $p(\cdot, \cdot \mid \bfY^{\ob})$ and $q_{\phi}$. Here we use the reverse KL divergence:

\begin{equation}
    D_{KL}(q_{\phi} \mid \mid p) = \bbE_{(\tau,\bft) \sim q_{\phi}}\left[\log\left(\frac{q_{\phi}\left(\tau,\bft\right)}{p(\tau,\bft \mid \bfY^{\ob})}\right)\right]\!.
\end{equation}

Evaluating the exact posterior $p(\tau,\bft \mid \bfY^{\ob})$ is difficult. Instead, we equivalently (up to a normalizing constant) maximize the evidence lower bound (ELBO), which is also known as the negative variational free energy in statistical physics and some areas of machine learning:

\begin{align}
     \phi^* = \argmax_{\phi} L(\phi), \qquad L(\phi) = \bbE_{q_{\phi}}\left[\log\left(\frac{p(\tau,\bft,\bfY^{\ob})}{q_\phi(\tau,\bft)}\right)\right]\!.
    \label{eqn:ELBO} 
\end{align}

This expectation over $q_{\phi}$ consists of a sum over tree structures $\tau$ and an integral over coalescent times $\bft$, forming the following objective function:

\begin{equation}
     L(\phi) = \sum_{\tau} \int_{\bft} q_{\phi}(\tau,\bft) \log\left(\frac{p(\tau,\bft,\bfY^{\ob})}{q_\phi(\tau,\bft)}\right) d\bft.
    \label{eqn:ELBO_int} 
\end{equation}

\begin{algorithm}[ht]
\caption{{\tt Single-Linkage Clustering}$(\bfT,\mathcal{X}_0)$}\label{alg:slc}
\begin{algorithmic}[1]
\STATE {\bfseries Input:} Distances $\bfT \in \bbR_{>0}^{\binom{N}{2}}$ and taxa set $\mathcal{X}_0 = \{\{x_1\},\{x_2\},\ldots,\{x_N\}\}$.
%
\vspace{1mm}
%
\FOR{$n = 1,\ldots,N-1$}
%
\STATE $w^*,z^* \leftarrow \argmin_{w,z} \{t^{\{w,z\}}: w, z \text{ not coalesced}\}$.
%
\STATE Set $W_n \in \mathcal{X}_0$ as the set containing $w^*$.
%
\STATE Set $Z_n \in \mathcal{X}_0$ as the set containing $z^*$. 
%
\STATE $t_n \leftarrow t^{\{w^*,z^*\}}$
%
\STATE Remove $W_n$, $Z_n$ from $\mathcal{X}_0$ and add $W_n \cup Z_n$ to $\mathcal{X}_0$.
\ENDFOR
\STATE $\tau \leftarrow \{\{W_n,Z_n\}\}_{n=1}^{N-1}$
\STATE $\bft \leftarrow \{t_n\}_{n=1}^{N-1}$
\STATE \textbf{Return } $(\tau,\bft)$
\end{algorithmic}
\end{algorithm}

\subsection{Matrix Representation of Tree Space}

One way to construct a phylogenetic tree is to use a distance matrix $\bfT$ (a symmetric $N \times N$ matrix with positive and finite off-diagonal entries) and the \textit{single-linkage clustering} algorithm, as described in Algorithm \ref{alg:slc}. We denote the distance between taxa $u$ and $v$ by $t^{\{u,v\}}$ and formulate the algorithm to return a representation of a phylogenetic tree using bipartitions and coalescent times that is consistent with our notation. We consider the distance matrix as an element of $\mathbb{R}^{\binom{N}{2}}_{>0}$ by identifying the off-diagonal elements. 
%
Algorithm \ref{alg:slc} is a naive implementation of single-linkage clustering with time complexity $\calO(N^3)$ and space complexity $\calO(N^2)$. \citet{Sibson:1973} introduce SLINK, an implementation with time complexity $\calO(N^2)$ and space complexity $\calO(N)$, and prove that both are optimal. 

\citet{Bouckaert:2024} use single-linkage clustering to perform variational inference over ultrametric trees. First, they note that if exactly $N-1$ entries of $\bfT$ are finite, then single-linkage clustering implies a bijection between $\bfT$ and $(\tau,\bft)$. Then, they specify exactly $N-1$ entries of $\bfT$ to be random and finite, setting all other entries to infinity. Next, they run an MCMC algorithm over trees to obtain an empirical distribution over coalescent times. Finally, they estimate the parameters associated with the $N-1$ finite entries of $\bfT$ using the MCMC-generated empirical distribution. This method has several notable setbacks. First, all entries of $\bfT$ must be specified to specify a distribution over all of tree space. Further, it is unclear how to select the $N-1$ best entries of $\bfT$ to cover the most posterior probability. To correct these issues, we present a gradient-based variational inference method for ultrametric trees based on single-linkage clustering that specifies a variational distribution over all of tree space.


\begin{figure}[ht]
    \centering
    \includegraphics[width=\linewidth]{variational_phylo.png}
    \caption{\emph{{\bf \emph{Diagram of the sampling process for VIPR}}. Two possible example matrices $\bfT$ could be drawn using $t^{\{u,v\}} \sim q_\phi^{\{u,v\}}$ and result in the same phylogenetic tree $(\tau,\bft) \sim q_\phi$ after single-linkage clustering. Entries of $\bfT$ that trigger a coalescence event are shown in bold. The form of $q_\phi^{\{u,v\}}$ is defined by the practitioner, while the expression for $q_\phi$ depends upon $q_\phi^{\{u,v\}}$ and is defined in Equation (\ref{eqn:q}).}}
    \label{fig:Phylo_diag}
\end{figure}