\vspace{-2mm}
\section{Experiment}\label{sec:experiment}
\vspace{-2mm}
We aim to evaluate the performance of the proposed method (\alg) across several tasks by investigating the effectiveness of refinement procedures compared to existing single-shot guidance methods in diffusion models. We begin by introducing the baselines and metrics used in our evaluation. Subsequently, we present our results in protein and DNA design. For further details and additional results, refer to \pref{sec:appendix}. The code is available at \href{https://github.com/masa-ue/ProDifEvo-Refinement}{https://github.com/masa-ue/ProDifEvo-Refinement}. 

\vspace{-2mm}
\paragraph{Baselines and our proposal.} We compare baselines that address reward-guided generation in diffusion models with \alg. Note that we primarily focus on settings where reward feedback is provided in a black-box manner. 
\vspace{-2mm}

\begin{table*}[!th]
    \centering
    \caption{The results for the protein design task show that our method consistently outperforms the baselines. Note that P50 and P95 represent the median and 95\% quantile of the rewards for generated designs, respectively. LL denotes the (estimated) per-residue log-likelihood. Values in parentheses represent the estimated 95\% standard deviation. } 
    \label{tab:all_results}
  \resizebox{\textwidth}{!}{    \begin{tabular}{c|ccc | ccc | ccc |ccc } 
  Task   & \multicolumn{3}{|c|}{ (a) ss-match }  & \multicolumn{3}{|c|}{ (b) cRMSD } & \multicolumn{3}{|c|}{ (c) globularity } & \multicolumn{3}{|c}{ (d) symmetry }\\ 
         &    P50 $\uparrow$  &  P95  $\uparrow$ & LL  $\uparrow$  & P50 $\downarrow$  &  P95  $\downarrow$ & LL  $\uparrow$  &  P50 $\uparrow$  &  P95  $\uparrow$ & LL  $\uparrow$ &  P50 $\uparrow$  &  P95  $\uparrow$ & LL  $\uparrow$  
         \\ \midrule  
    SMC  &   0.63 (0.04)  &  0.80   & -3.28  & 8.9 (0.7) &  5.1   & \textbf{-3.58}  & -2.79 (0.05) & -2.13  & -4.43 & -0.45 (0.03) & 0.21 & -3.30 \\   
    SVDD   &  0.66 (0.02) & 0.82  & \textbf{-3.03}  & 8.2 (0.4)  &  4.6  &  -3.59 &    -2.45 (0.02) & -2.00  &  -4.68 &   -0.33 (0.04)  & 0.36   &  -3.56   \\ 
    GA & 0.70 (0.00)  & 0.95   &  {-3.51}  &      6.3 (0.4)  &  3.01  &  -3.60 &   -1.35 (0.02)  &  -1.22 & \textbf{-4.38} &  0.21 (0.04)   &  0.44  &  \textbf{-3.07} \\  
 \rowcolor{lightgray}   \alg   &  \textbf{0.86} (0.01) & \textbf{0.96}  & -3.13 &   \textbf{1.68 (0.02)} & \textbf{0.96 }  & -3.51 &  \textbf{-1.29} (0.02) & \textbf{-1.15}  &  -4.45 &  \textbf{0.34} (0.01) &  \textbf{0.69} &  -3.08\\ 
    \end{tabular}}
\end{table*}

\begin{figure*}[!th]
    \centering
 \begin{minipage}{0.24\textwidth}  %
    \centering
    \includegraphics[width=0.41\textwidth]{images/ss_match_r15.png}
     \includegraphics[width=0.55\textwidth]{images/ss_match_EA.png}
    \subcaption{Generated proteins (\textcolor{green}{Green}) when optimizing \textbf{ss-match} are shown. \textcolor{red}{Red} represents the target secondary structures. The \textbf{ss-match} score for the left figure is 0.96, while for the right figure, it is 1.0. }
  \end{minipage} \hfill
 \begin{minipage}{0.24\textwidth}  %
    \centering
   \includegraphics[width=0.39\textwidth]{images/crmsd_EHEE_0.42.png}
   \includegraphics[width=0.55\textwidth]{images/crsmd_5kph_0.8.png}
    \subcaption{Generated proteins (\textcolor{green}{Green}) from \alg\,when \textbf{cRMSD} are shown. \textcolor{red}{Red} represents the target backbone structures. The \textbf{cRMSD} score for the left figure is 0.42, while for the right figure, it is 0.6.}
  \end{minipage} \hfill
   \begin{minipage}{0.18\textwidth}  %
    \centering
    \includegraphics[width=0.80\textwidth]{images/globularity.png}
    \subcaption{Generated proteins when optimizing \textbf{globularity}.}
  \end{minipage} 
  \begin{minipage}{0.30\textwidth}  %
    \centering
    \includegraphics[width=0.57\textwidth]{images/symmetric.png}
   \includegraphics[width=0.37\textwidth]{images/symmetric3.png}
    \subcaption{Generated proteins when optimizing \textbf{symmetry} }
  \end{minipage} 

     \caption{We visualize the sequences generated from \alg\,using ESMFold. }
    \label{fig:generated_results}
\end{figure*}

\begin{figure}[!th]
    \centering
     \includegraphics[width=0.48\linewidth]{images/cRMSD_5KPH.png}
     \includegraphics[width=0.48\linewidth]{images/cRMSD_6NJF.png}
    \caption{The refinement step from \alg\,when optimizing cRMSD in two target backbone structures is demonstrated. Recall that the first iteration corresponds to the result from SVDD. The Y-axis represents the median reward of generated samples (\textbf{Lower is better}). }
    \label{fig:refinement}
\end{figure}


\begin{itemize}
    \item \textbf{SVDD \citep{li2024derivative}}:  A representative single-shot, derivative-free guidance method (without refinement).
    \item \textbf{SMC \citep{wu2024practical}}: Another single-shot, representative derivative-free guidance method.
    \item \textbf{GA}: A na\"ive approach for sequence design that uses pre-trained diffusion models to generate mutated designs within a standard genetic algorithm (GA) pipeline \citep{hie2022high}. To ensure a fair comparison, we allocate the same computational budget as \alg\,below. 
    \item \textbf{\alg\,in \pref{alg:decoding2} (Ours)}. We set $K/T=10\%$ and $S=50$.  For initial designs, we use the results generated by SVDD in \pref{sec:protein} and designs that satisfy the constraints in \pref{sec:DNA}.
\end{itemize}
Note that we have used the same hyperparameters $\alpha,L$ across baselines (SMC, SVDD) and \alg. 

\vspace{-3mm}
\paragraph{Metrics.}  We report the top 95\% quantile (\textbf{P95}) and median of rewards (\textbf{P50}) from generated designs, as these are the primary metrics to optimize. Additionally, we present the estimated per-residue log-likelihood (\textbf{LL}) using the pre-trained diffusion models, which serves as a secondary metric that we aim to maintain at a moderately high value to preserve the naturalness of the designs. \footnote{We also report the diversity of generated designs. Since this metric is difficult to compare formally and secondary in the context of reward optimization, it is included in the Appendix.}



\subsection{Protein Design}\label{sec:protein}

We begin by outlining our tasks. First, we use EvoDiff \citep{alamdari2023protein}, a representative discrete diffusion model for protein sequences trained on the UniRef database, as our unconditional base model. Next, following existing representative works in protein design \citep{hie2022high,watson2023novo,ingraham2023illuminating}, we consider four reward functions related to structural properties, which take the generated sequence as input. For more details, refer to \pref{sec:appendix}. 

\vspace{-2mm}

\begin{table*}[!th]
    \centering
    \caption{The results for the DNA design task show that our method consistently outperforms the baselines.  } 
    \label{tab:all_results2}
  \resizebox{0.8\textwidth}{!}{    \begin{tabular}{c|ccc | ccc | ccc  } 
  Task   & \multicolumn{3}{c|}{ HepG2 }  & \multicolumn{3}{c|}{ K562 } & \multicolumn{3}{c}{ SKNSH}  \\ 
         &    P50 $\uparrow$  &  P95  $\uparrow$ & LL  $\uparrow$  & P50 $\downarrow$  &  P95  $\downarrow$ & LL  $\uparrow$  &  P50 $\uparrow$  &  P95  $\uparrow$ & LL  $\uparrow$ 
         \\ \midrule  
    SMC    & 1.2 (0.3)    & 1.6   & -1.15  & 1.0 (0.2) & 1.4 & \textbf{-1.21} & 0.8 (0.2) & 1.0 & -1.22   \\ 
   SVDD &  2.3 (0.2)  &  2.8   & \textbf{-1.08} &  1.3 (0.3) & 1.6 & -1.26   & 1.7 (0.3) &  2.0 &  \textbf{-1.21}\\ 
    GA   &  2.3 (0.4)  &  2.7 &  -1.21  & 2.2 (0.3) & 2.6  & -1.31 & 1.9 (0.4) & 2.5 & -1.28  \\
  \rowcolor{lightgray}   \alg &  \textbf{7.9} (0.2)  & \textbf{9.1}  &  -1.18 & \textbf{7.4} (0.2) & \textbf{8.9} & -1.25 & \textbf{5.5} (0.2) & \textbf{6.7} & -1.24 \\ 
    \end{tabular}}
\end{table*}

\begin{itemize}
    \item \textbf{ss-match}: We use Biotite \citep{kunzmann2018biotite} to predict the secondary structure (ss). We then calculate the mean matching probability across all residues between the predicted and reference secondary structures, where the target structure is represented by a sequence consisting of $a$ ($\alpha$-helices), $b$ ($\beta$-sheets), and $c$ (coils). A score of $1.0$ indicates perfect alignment.
    \item \textbf{cRMSD}: This is the constrained root mean square deviation against the reference backbone structure after structural alignment. Typically, $<2\r{A}$ indicates a highly similar structure. Note that a lower value is preferred. 
    \item \textbf{globularity (+ pLDDT)}: It reflects how closely the structure resembles a globular shape. Additionally, we optimize \textbf{pLDDT} to improve the stability of the structure.
    \item \textbf{symmetry (+pLDDT, hydrophobicity)}: It indicates the symmetry of the structure in the generated sequence. Additionally, we optimize \textbf{pLDDT} and \textbf{hydrophobicity} to improve the stability of the structure.
\end{itemize}
Note that each of the above rewards is computed after estimating the corresponding structure using ESMFold \citep{lin2023evolutionary}. Besides, for both \textbf{ss-match} and \textbf{cRMSD}, we use 10 reference proteins randomly chosen from datasets in \citet{dauparas2022robust} and report the mean of the results. 







\vspace{-2mm}

\paragraph{Results.} We present our performance in Table~\ref{tab:all_results} and visualize generated sequences in \pref{fig:generated_results}. Overall, our algorithm (\alg) consistently demonstrates superior performance in terms of rewards while maintaining reasonably high likelihood. Notably, as illustrated in \pref{fig:refinement}, for several challenging tasks, while one-shot guidance methods such as SVDD underperforms, our approach, with refinement steps, gradually yields improved results.






\subsection{Cell-Type-Specific Regulatory DNA Design} \label{sec:DNA}

We begin by outlining our tasks. Here, we focus on widely studied cell-type-specific regulatory DNA designs, which are crucial for cell engineering \citep{taskiran2024cell}. Specifically, our goal is to design enhancers (i.e., DNA sequences that regulate gene expression) that exhibit high activity levels in certain cell lines while maintaining low activity in others. 

Following existing works \citep{lal2024reglm,sarkar2024designing,gosai2023machine}, we construct reward functions as follows. Using datasets from \citet{gosai2023machine}, which measures the enhancer activity of $700$k DNA sequences (200-bp length) in human cell lines using massively parallel reporter assays (MPRAs), we trained oracles based on the Enformer architecture \citep{avsec2021effective} as rewards across three cell lines ($r_{\mathrm{H}}(\cdot)$ in HepG2 cell line,
$r_{\mathrm{K}}(\cdot)$ in K562 cell line , and $r_{\mathrm{S}}(\cdot)$ in SKNSH cell line). Then, we aim to respectively optimize the following:
\begin{align}\label{eq:reward_original}
    \bar r_{\mathrm{H}}(x)  &= r_{\mathrm{H}}(x)\mathrm{I}(r_{\mathrm{K}}(x)<c)\mathrm{I}(r_{\mathrm{S}}(x)<c)
\end{align}
where $c$ is a threshold. Here, optimizing $\bar r_{\mathrm{H}}$ means maximizing $r_{\mathrm{H}}$ while retaining $r_{\mathrm{K}},r_{\mathrm{S}}$ low. Then, similarly, we define $\bar r_{\mathrm{K}},\bar r_{\mathrm{S}}$ by exchanging their roles.  

Here are several additional points to note. First, as discussed in \pref{sec:hard_constraint}, directly using $\bar r_H, \bar r_K, \bar r_S$ in practice would lead to suboptimal performance. Therefore, we use log barrier reward functions for all methods. Additionally, for \textbf{GA} and \alg, we initialize the designs with samples that satisfy the constraints (e.g., $\mathrm{I}(r_{\mathrm{K}}(x)<c)\mathrm{I}(r_{\mathrm{S}}(x)<c))$). Recall that one of the advantages of our method is its ability to leverage designs from feasible regions that satisfy the constraints. Finally, we use pre-trained discrete diffusion models from \citet{wang2024fine} as the backbone unconditional diffusion models.


\begin{figure}[!th]
    \centering
     \includegraphics[width=0.48\linewidth]{images/RERD_HepG2.png}
     \includegraphics[width=0.48\linewidth]{images/sequence_DNA.png}
    \caption{(Left) The refinement step from \alg\,is demonstrated. The Y-axis represents the median reward of generated samples (\textbf{Higher is better}), (Right) Generated designs from IRAO. It is seen that the activity in the target cell line HepG2 is only high. }
    \label{fig:refinement_cell}
\end{figure}



\vspace{-2mm}
\paragraph{Results.} The results are presented in Table~\ref{tab:all_results2}. Our methods consistently exhibit superior performance in terms of rewards while maintaining a relatively high likelihood. Notably, while it has been reported that \textbf{SMC} and \textbf{SVDD} excel in optimizing individual rewards (e.g., $r_{\mathrm{H}}$ only) in existing works such as \citet{li2024derivative}, we have observed that they struggle with handling additional constraints. In contrast, as shown in \pref{fig:refinement_cell}, \alg\, effectively handles such constraints (i.e., ensuring cell-type specificity) by gradually refining the results, starting from designs in feasible regions. 


