\subsection{An Example Application}\label{sec:application}

To show a concrete application of our framework, we show that {\uindex} can be
used for \textit{long-read mapping}.
This is the problem of aligning long DNA or RNA sequencing reads (e.g., of length more than 1,000 base pairs) to a
reference genome.
Although long reads offer significant advantages over short reads in many
crucial tasks in bioinformatics,
such as genome assembly or structural variant detection,
their alignment is computationally costly.

\parag{Setup}
We run the following experiment.
As input, we take the full human genome (CHM13v2.0) and
a prefix of 450 PacBio HiFi long reads from the HG002-rep1 dataset\footnote{Downloaded from
\url{https://downloads.pacbcloud.com/public/revio/2022Q4/HG002-rep1/}.}.
These reads are approximately 99.9\% accurate and have an average length of
16kbp.

We partition (chunk) each read into patterns of length 256, and search each
of these patterns in our index. Short leftover suffixes are ignored.
Ideally, each read has then at least one pattern that
exactly matches the text, which can then be used to anchor an alignment.

\parag{Results}
We build the {\uindex} on top of the \method{libsais} suffix array in the
configuration where the minimizer space sequence $S$ and map $H$ are both stored.
We use $k=8$ and $\ell=128$, so that at least 2 and on average 4 minimizers are
sampled from every pattern.
This results in 53M minimizers, and the entire {\uindex} is built in 12 seconds.

Out of the 450 sequences, 445 have at least one matching pattern, and in total,
14\ 824 of the 28\ 243 patterns match (52\%).
As observed before, an issue with DNA is that it contains many long repetitive
regions. In particular, 160 of the patterns each match around 3820 times for a
total of 611k matches, while the 28k remaining patterns only match 27k in total.
Worse, there are 721M mismatches, i.e., candidate matches in sketch-space that
turn out not to be matches in the genome. Verifying these candidate matches
takes over 98\% of the time.

\parag{Limiting Matches}
Only very few (<200) patterns have more than 10 matches, and thus we stop
searching once we hit 10 matches. Further, most patterns have
relatively few mismatches, while a few patterns have a lot of mismatches. To
also avoid those negative effects, we generally only consider the first 100
matches in sketch-space. We still match 445 of 450 reads, while the number
of matched patterns goes down to 13\ 717. On the other hand, the number of
mismatches is now 663k (23 per query) and the number of matches is 25k (0.9 per
query).

The result is a query time of 8.7$\mu$s per pattern or 550$\mu$s per read, of which 33\% is sketching
the input, 18\% is locating the sketch, and the remaining 48\% is verification.