Several steps of the pipeline involve necessarily subjective choices made in order to precisely define replicable evaluation procedures and metrics.
Pending community feedback, these choices may be altered in subsequent MotifBench versions.

\subsection*{Problem specification, inputs, and outputs}
\paragraph{Placement of motif segments:}
Most previous evaluations have relied on pre-specified orderings and suggested ranges for the placements of sequence-contiguous motif segments within scaffolds (see, e.g.\ the  ``contig'' specifications in \citep{watson2022broadly}).
Such pre-specification simplifies the task and allows methods for image inpainting/outpainting to be more easily applied, as these methods typically assume a fixed mask.
Accordingly, we include as suggested motif-placement for each problem chosen as the placement of the motif within the experimentally validated scaffold in the PDB from which the motif is derived.

However, ideal placements of motif segments may not be known in general for realistic problems and so it may be beneficial to automate this choice.
Furthermore, it may be helpful to vary this hyper-parameter to achieve larger numbers of diverse solutions,
success rates, and novelty (e.g.\ by random sampling \citep[see e.g.][]{watson2022broadly} or dynamically throughout the course of generation \citep[see e.g.][]{wu2024practical}).

\paragraph{Scaffold length:}
By contrast with the placement of motif segments, the lengths of the scaffolds to be generated are specified as part of each problem.
The scaffold length is also a subjective hyper-parameter whose choice could be automated in principle.
However, we restrict to a single length to avoid length-dependent biases that can appear in several parts of the evaluation pipeline, e.g. diversity and novelty.
In the problem specifications in \cref{table:test_cases}, these scaffold lengths are subjectively chosen based on the length of the native scaffolds and what seemed plausible to the authors.

\paragraph{Number of sequences per backbone:}
Most previous evaluations have relied on the design of several sequences for each backbone and evaluated generations as successes if the metrics of one or more of the generated sequences passed the cutoffs.
This choice reflects that computational filtering is a practical technique that is commonly in protein engineering,
and addresses the limitation that the stochastic generations of common fixed backbone sequence design methods sometimes fail to identify an adequate sequence on the first generation.
However using a larger number of sequences also increases computational cost and risks higher false-positive rates.
MotifBench specifies eight sequences for each scaffold as a practical balance that has been used across several previous works \citep{trippe2022diffusion,yim2023se,watson2022broadly}.

\subsection*{Choices of thresholds in evaluation}
We next discuss several subjective thresholds in the definition of evaluation metrics.

\paragraph{motifRMSD and scRMSD thresholds:}
The success criteria in MotifBench includes thresholds on the precision to which the motif and full scaffold must be recapitulated.
The 1{\AA} threshold on motif recapitulation is chosen to demand atomic precision;
by comparison, the atomic radius of a hydrogen atom is about 1.2{\AA}.
The 2{\AA} threshold on backbone recapitulation is set as a coarser level of precision that demands the overall backbone structure can be designed. 

\paragraph{Structural similarity thresholds for clustering and novelty:}
We adopt default settings of Foldseek-search and Foldseek-cluster for simplicity.
However, these methods have several parameters that impact their behavior for which different settings could in principle better align the number of unique solutions and novelty metrics with the efficacy of a design method in application.

\paragraph{Filtering on structure prediction confidence:}
Predictions of high confidence in the accuracy of protein structure prediction outputs has been included in previous motif-scaffolding evaluations as a proxy for the quality of designed scaffolds.
However, this criterion is partly largely redundant with the designability metric of scRMSD$<2${\AA} and is highly dependent on the accuracy of confidence head for specified structure prediction method.
Therefore, we dropped this criterion for simplicity.

\subsection*{Software choices}
MotifBench makes several choices of open sources methods in the evaluation pipeline on the basis of 
demonstrated predictive power for viability in \emph{in vitro} experiments,
computational efficiency and convenience,
and prior community adoption.

\paragraph{Fixed-backbone sequence design (inverse folding) method:}
A number of inverse folding methods exist that could in principle be used for the the sequence design step \citep[e.g.][]{dauparas2022robust,liu2022rotamer,anand2022protein}.
MotifBench specifies ProteinMPNN \citep{dauparas2022robust} with default parameters (including sampling temperature set to 0.1) for its precedent in past work \citep[e.g.][]{trippe2022diffusion}, its significant experimental validation \citep{dauparas2022robust,watson2022broadly}, and for its familiarity to the authors.

\paragraph{Structure prediction method:}
Several public software packages provide an accurate prediction of a protein structure from its amino acid sequence \citep{baek2021accurate,jumper2021highly,lin2022language,wu2022omegafold}. 
MotifBench specifies folding using ESMFold \citep{lin2022language} for its simplicity of implementation and computational efficiency.
However, the predictions of any computational structure prediction method may be incorrect and should be interpreted with this fact in mind;
and in some cases \texttt{ESMFold} is known to be less predictive of experimental viability as compared to AlphaFold2 \citep{jumper2021highly} when run with no multiple sequence alignment input \citep{martin2023validation}.
As such, the choice of ESMFold in MotifBench should not be interpreted as an endorsement for its use as an \emph{in silico} filter in protein design campaigns and we encourage users to report results by both ESMFold and AlphaFold2 (with no MSA input) for an orthogonal test when not computationally burdensome.

\paragraph{Structural clustering and similarity method:}
MotifBench specifies the use of \texttt{Foldseek-cluster} \citep{barrio2023clustering} and \texttt{Foldseek-search} \citep{van2024fast} for clustering and novelty evaluation (version \texttt{8.ef4e960}).
We choose these approaches as a alternative to the more widely used TMscore \citep{zhang2005tm} for their computational speed.

A limitation of Foldseek-Cluster is that it sometimes raises opaque errors during a ``prefilter'' step on certain scaffold sets.
We found that this error can be resolved by (1) adding an additional backbone unrelated to the design task to scaffold set, (2) re-running Foldseek-Cluster on the augmented set, and (3) removing the unrelated protein from the clustering results to limit its impact on the \emph{number of unique solutions} metric.

