% TODO:
% - Unify tissue section / tissue slice
% - molecular vs. transcriptomics vs. genomics
% - sequencing vs. assay

\documentclass{naturep}
\usepackage[utf8]{inputenc}
\usepackage{lineno}
\usepackage{setspace}
\usepackage{color}
\usepackage{graphicx}
\usepackage{booktabs}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{hyperref}
\usepackage{adjustbox}
\usepackage{subcaption}
\usepackage{tablefootnote}
\usepackage[margin=0.6in]{geometry}
\usepackage{verbatim}
\usepackage{caption}
\usepackage{paralist}
\usepackage{multirow}
\newcommand{\andrew}[1]{\textcolor{blue}{#1}}
\newcommand*{\tabindent}{\hspace{5mm}}

\newcommand{\ours}{\textsc{VORTEX}}
\newcommand{\TODO}[1]{{\textcolor[rgb]{0,0,1}{#1}}}

\newcommand\Heading[1]{
  \noindent\textbf{\Large{#1}}
}

\newcommand\heading[1]{
  \noindent\textbf{\large{#1}}
}

\newcommand\hheading[1]{
  \noindent\textbf{#1}
}



\title{\raggedright{\textbf{AI-driven Non-Destructive 3D Spatial Transcriptomics}}}
\title{\textbf{AI-driven 3D Spatial Transcriptomics}}
\author{\raggedright Cristina Almagro-PÃ©rez$^{\ast,1,2,3,4}$, Andrew H. Song$^{\ast,1,2,3}$, Luca Weishaupt$^{1,2,3,4}$, Ahrong Kim$^{1,5}$, Guillaume Jaume$^{1,2,3}$, Drew F.K. Williamson$^{1,6}$, Konstantin Hemker$^{1, 7}$, Ming Y. Lu$^{1,2,3,8}$,  Kritika Singh$^{9}$, Bowen Chen$^{1,2,3}$, Long Phi Le$^{1}$, Alexander S. Baras$^{10,11}$, Sizun Jiang$^{12,13,14}$, Ali Bashashati$^{15,16}$, Jonathan T.C. Liu$^{17}$, and Faisal Mahmood$^{\dagger,1,2,3,18}$}
    
\makeatletter
\let\saved@includegraphics\includegraphics
\AtBeginDocument{\let\includegraphics\saved@includegraphics}
\renewenvironment*{figure}{\@float{figure}}{\end@float}
\makeatother

\newcommand{\blue}[1]{{\textcolor[rgb]{0,0,1}{#1}}}

\begin{document}
%TC:ignore
\maketitle
\begin{affiliations}
 \item Department of Pathology, Mass General Brigham, Harvard Medical School, Boston, MA, USA
 \item Cancer Program, Broad Institute of Harvard and MIT, Cambridge, MA, USA
 \item Data Science Program, Dana-Farber Cancer Institute, Boston, MA, USA
\item Harvard-MIT Division of Health Sciences and Technology, Massachusetts Institute of Technology, Cambridge, MA, USA
\item Department of Pathology, Pusan National University, Busan, South Korea
\item Department of Pathology and Laboratory Medicine, Emory University School of Medicine, Atlanta, GA, USA
\item Department of Computer Science \& Technology, University of Cambridge, Cambridge, UK
\item Department of Electrical Engineering and Computer Science, Massachusetts Institute of Technology, Cambridge, MA, USA
\item Rutgers Robert Wood Johnson Medical School, New Brunswick, NJ
\item Department of Pathology, Johns Hopkins Hospital, Baltimore, MD, USA
\item Department of Biomedical Engineering, Johns Hopkins Hospital, Baltimore, MD, USA 
\item Broad Institute of Harvard and MIT, Cambridge, MA, USA
\item Center for Virology and Vaccine Research, Beth Israel Deaconess Medical Center, Harvard Medical School, Boston, MA, USA
\item Department of Pathology, Dana Farber Cancer Institute, Boston, MA, USA
\item Department of Pathology and Laboratory Medicine, University of British Columbia, Vancouver, BC, Canada
\item School of Biomedical Engineering, University of British Columbia, Vancouver, BC, Canada
\item Department of Mechanical Engineering, Bioengineering, and Laboratory Medicine \& Pathology, University of Washington, Seattle, WA, USA
\item Harvard Data Science Initiative, Harvard University, Cambridge, MA, USA
\item[$^{\ast}$] Equal contribution\\
\textbf{$\dagger$ Corresponding author}: Faisal Mahmood (FaisalMahmood@bwh.harvard.edu)
\end{affiliations}

\clearpage

\Heading{Abstract}

\input{sections/0-abstract}


% \newpage
%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Section 1: Introduction %
%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \detailtexcount{main} 
%TC:endignore
%\section{Main}
% \linenumbers
\clearpage
\begin{spacing}{1.35}
% \linenumbers
\Heading{Introduction}
\input{sections/1-introduction}

\Heading{Results}
\input{sections/2-results}

\Heading{Discussion}
\input{sections/3-discussion}
\clearpage
\Heading{Online Methods}





\input{supp/0-supp}
\input{supp/1-additional}

\end{spacing}


\clearpage



\newpage
\setcounter{figure}{0}
\renewcommand{\figurename}{\textbf{Extended Data Figure}}
\renewcommand{\thefigure}{\arabic{figure}}

\setcounter{table}{0}
\renewcommand{\tablename}{\textbf{Extended Data Table}}

\clearpage
\input{supp/2-figures}
\clearpage
% online methods
\input{tables/vortex_training}

\clearpage

\begin{nolinenumbers}
\Heading{References}
\bibliographystyle{nature}
\bibliography{main}
\end{nolinenumbers}

\end{document}


