\documentclass{melba}
\usepackage{subcaption}
% melba class has several options:
% - 'arxiv' in arXiv pre-print in submission (disable line numbers)
% - 'accepted' for MELBA _accepted_ papers **only**;
%                  to be used in conjunction with 'accepted'
% - 'specialissue' for MELBA accepted papers that are part of a special issue.
%                  to be used in conjunction with 'accepted'

\usepackage{mwe} % to get dummy images, only for the example
% Can be removed for actual manuscripts

% often used packages
\usepackage{amsmath,amsfonts}
% \usepackage[pass, showframe]{geometry}  % Draw borders on pdf, useful to debug figures placement

% add packages here

% Definitions of handy macros can go here
\newcommand{\dataset}{{\cal D}}
\newcommand{\fracpartial}[2]{\frac{\partial #1}{\partial  #2}}


% Header and footer (will be filled at publication)
\melbaid{YYYY:NNN}  % This is provided upon by the publishing editor
\doi{10.59275/j.melba.2024-AAAA}
\melbaauthors{Name1 and Name2}  % Note: this one is also used to set the pdf 'authors' metadata
\email{author@institute.tld}
\volume{2}
\firstpageno{1337}  % Communicated by the publishing editor
\melbayear{YYYY}  % The publication year
\datesubmitted{yyyy-m1-d1}  % Date submitted to MELBA: mm/yyyy
\datepublished{yyyy-m2-d2}  % Today's date: mm/yyyy

% The following is optionnal, only if you are publishing in a special issue
% The information is available on the README.md of this repository
% https://github.com/melba-journal/submission#special-issues
\melbaspecialissue{Medical Imaging with Deep Learning (MIDL) 2020}
\melbaspecialissueeditors{Marleen de Bruijne, Tal Arbel, Ismail Ben Ayed, Hervé Lombaert}


% Short headings should be {running head} and {authors last names}
\ShortHeadings{MELBA Journal Sample Article}{Name1 and Name2}


% Title
% If the title spans several lines, authors could decide where the title should be split using \\
% This will cause a warning from the hyperref package, when it sets the title as pdf 'title' metadata
\title{Automated Muscle and Fat Segmentation in Computed
Tomography for Comprehensive Body Composition Analysis}

% \firstname and \surname are optionnal, (simply using \name is doable), but can be useful
% to clarify names such as \firstname FIRSTNAME MIDDLE NAME \surname FAMILYNAME or composed names

% Notice that spacs left out (between name and orcid and affiliations will be displayed)
\author{
	\firstname Yaqian \surname Chen\aff{1},
	\name Hanxue \surname Gu\aff{1},
        \name Yuwen \surname Chen\aff{1},
        \name Jicheng \surname Yang\aff{1},
        \name Haoyu \surname Dong\aff{1},
        \name Joseph Y. \surname Cao\aff{2},
        \name Adrian \surname Camarena\aff{3},
        \name Christopher \surname Mantyh\aff{3},
        \name Roy \surname Colglazier\aff{2},
        \name Maciej A. \surname Mazurowski\aff{1,2,4,5},
}
% Authors are free to use either a new line (\\) or a simple comma (,) between affiliations.
\affiliations{% <- trailing '%' to avoid unwanted indent
	\num 1 \addr Department of Electrical and Computer Engineering, Duke University, Durham, NC 27708 \\
	\num 2 \addr Department of Radiology, Duke University, Durham, NC 27708\\
	\num 3 \addr Department of Surgery Duke University School of Medicine, Durham, NC 27708\\
        \num 4 \addr Department of Biostatistics \& Bioinformatics, Duke University, Durham, NC 27708\\
        \num 5 \addr Department of Computer Science, Duke University, Durham, NC 27708\\
}

\abstract{%   <- trailing '%' for backward compatibility of .sty file
    Body composition assessment using CT images can potentially be used for a number of clinical applications, including the prognostication of cardiovascular outcomes, evaluation of metabolic health, monitoring of disease progression, assessment of nutritional status, prediction of treatment response in oncology, and risk stratification for surgical and critical care outcomes. While multiple groups have developed in-house segmentation tools for this analysis, there are very limited publicly available tools that could be consistently used across different applications. To mitigate this gap, we present a publicly accessible, end-to-end segmentation and feature calculation model specifically for CT body composition analysis.
    Our model performs segmentation of skeletal muscle, subcutaneous adipose tissue (SAT), and visceral adipose tissue (VAT) across the chest, abdomen, and pelvis area in axial CT images. It also provides various body composition metrics, including muscle density, visceral-to-subcutaneous fat (VAT/SAT) ratio, muscle area/volume, and skeletal muscle index (SMI), supporting both 2D and 3D assessments. The model is shared for public use.
    To evaluate the model, the segmentation was applied to both internal and external datasets, with body composition metrics analyzed across different age, sex, and race groups. The model achieved high dice coefficients on both internal and external datasets, exceeding 89\% for skeletal muscle, SAT, and VAT segmentation. The model outperforms the benchmark by 2.40\% on skeletal muscle and 10.26\% on SAT compared to the manual annotations given by the publicly available dataset. Body composition metrics show mean relative absolute errors (MRAEs) under 10\% for all measures. Furthermore, the model provided muscular fat segmentation with a Dice coefficient of 56.27\%, which can be utilized for additional analyses as needed. 
    Our code is available at~\url{https://github.com/mazurowski-lab/CT-Muscle-and-Fat-Segmentation.git}.
    }

\keywords{Deep learning, Segmentation, Muscles, Subcutaneous Fat, Visceral Fat, Body Composition}

% Indendation is not mandatory, but usually makes the code more readable
\begin{document}

% top matter
\twocolumn[\maketitle]
% comment the preceedings and uncomment the following if the authors list + abstract is longer than one page
% \maketitle
% \twocolumn


% Introduction (or first section)
% \rule{\textwidth}{1pt}
\section{Introduction}\label{sec:intro}
Correlating body composition metrics based on computed tomography (CT) images with disease and clinical variables, such as cancer \citep{rutten2016loss, kumar2016muscle, deluche2018impact, yoshikawa2020sarcopenia, iwase2016impact, boer2020impact}, cachexia \citep{ali2014sarcopenia, fearon1990body, al2023body, baracos2010body} and frailty \citep{falsarella2015body, reinders2017body, villareal2004physical}, is becoming a widely adopted approach to leverage medical imaging data for real-world clinical applications \citep{tolonen2021methodology}. By measuring body composition, such as quantity and location of fat as well as quantity and quality of muscle, clinicians are able to gain valuable insights into a patient’s physiological status \citep{prado2014lean}. This information enables them to assess disease progression \citep{baracos2013clinical}, evaluate treatment efficacy \citep{bates2022ct}, and predict clinical outcomes \citep{weston2019automated}. 

Several key metrics are frequently utilized in body composition analysis, including muscle density, the visceral-to-subcutaneous fat (VAT/SAT) ratio, muscle area or volume, and the skeletal muscle index (SMI). Most studies in this field measure these metrics from a single slice, most commonly at the third lumbar vertebra (L3) \citep{arayne2023comparison}, while others employ volumetric analysis \citep{connelly2013volumetric}. However, regardless of the approach, extracting these metrics relies on effective segmentation models to accurately identify and quantify various tissues within the body. Traditional methods, such as pixel thresholding based on Hounsfield units (HU) \citep{wang2020artificial} and fuzzy c-means clustering \citep{wang2020artificial, christ2011fuzzy}, often require manual adjustments and are time-intensive \citep{wang2020artificial}. Furthermore, pixel thresholding algorithms cannot differentiate between visceral fat, subcutaneous fat, and intramuscular fat—an essential distinction when measuring the VAT/SAT ratio. Deep learning segmentation is a response to these limitations, and some groups have developed in-house segmentation models customized for their private datasets \citep{fu2020automatic, lee2017pixel, weston2019automated, wang2017two, hemke2020deep, koitka2021fully}. These models typically lack public accessibility and are designed for specific tasks. Furthermore, we observed inconsistency in how muscular fat (both intra-muscular and inter-muscular fat) is utilized in research. While some studies include muscular fat as part of skeletal muscle measurements \citep{hou2024enhanced, blanc2020abdominal, weston2019automated}, others classify it under VAT \citep{camus2014prognostic, wirtz2021ct, connelly2013volumetric}, and a smaller subset considers it part of SAT \citep{ozturk2020relationship}. 

In order to advance the research on the relationship of imaging-based body composition with disease and clinical variables, a robust, thoroughly validated, and publicly available tissue segmentation model and body composition variable calculation is needed. This model will allow different research groups to test the model with their data and correlate the unified body composition measurements with the clinical outcomes of their interest, building consistent scientific evidence of the importance of body composition in human health. 

Toward this goal, we developed a segmentation model using nnU-Net \citep{isensee2021nnu, isensee2024nnu} which identifies the areas of skeletal muscle, SAT, VAT, and muscular fat. For the model training and evaluation, we collected 813 CT volumes of chest, abdomen, and pelvis for 483 patients from Duke University Health System. In both training and test datasets, we included volumes from different years (from 2016 to 2019), various scanners, and diverse patient demographics to ensure the model's generalizability. Additionally, we incorporated the publicly available Sparsely Annotated Region and Organ Segmentation (SAROS) dataset \citep{koitka2023saros, clark2013cancer} for segmentation evaluation, demonstrating the generalizability of our proposed model. Furthermore, we also analyze the relationships between the four body composition metrics (muscle density, VAT/SAT ratio, muscle area/volume, and SMI) with respect to three key demographic variables: age, sex, and race (shown in Section \ref{sec:Body composition analysis}). Notably, to facilitate wide use of the model, we have made it publicly available. 

The main contributions of this work are summarized as follows:
\begin{itemize}
    \item To the best of our knowledge, this is the first publicly available deep-learning model designed to segment skeletal muscle, SAT, VAT, and muscular fat across the chest, abdomen, and pelvis on CT.
    \item We provide end-to-end standardized and publicly available measurements for four common body composition metrics, including muscle density, visceral-to-subcutaneous fat (VAT/SAT) ratio, muscle area/volume, and SMI on both L3  for 2D and T12 to L4 for 3D measurement.
    \item Our model outperforms TotalSegmentator \citep{wasserthal2023totalsegmentator} and Enhanced segmentation \citep{hou2024enhanced} by 2.40\% on skeletal muscle and 10.26\% on SAT compared to the manual annotations given by publicly available dataset SAROS.
    \item We perform the statistical analysis to correlate four metrics (muscle density, VAT/SAT ratio, muscle area/volume, and SMI) with three patient demographic variables: age, sex, and race in both 2D and 3D settings. 
\end{itemize}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Related works
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Make sure to put your work into context and include apporpriate citations.
% We do not have limits on citation counts.
\section{Related Works}
\subsection{Body composition analysis using CT}
Body composition plays a crucial role in influencing physical performance \citep{hernandez2024relationship, falsarella2015body}, metabolic health \citep{trouwborst2024body, kakinami2022body}, and disease outcomes \citep{rutten2016loss}. Imaging offers an objective, quantitative approach to its analysis through various techniques, including CT, magnetic resonance imaging (MRI), and ultrasonography \citep{hou2024enhanced, tan2024quantitative, sharafi2024quantitative, xu2024value}. Among all the modalities, CT offers high spatial resolution, faster acquisition times, and superior contrast between tissues \citep{zhang2021improving}, making it particularly suitable for assessing visceral and subcutaneous fat, skeletal muscle, and organ-specific fat deposits \citep{wathen2013vivo}. 

During body composition calculation on CT, several key metrics are frequently utilized, including muscle density, the VAT/SAT ratio, muscle area/volume, and the SMI. Muscle density in CT provides insights into muscle quality, which is linearly influenced by muscular fat content \citep{engelke2018quantitative}. A reduction in muscle density is often associated with increased fat infiltration within the muscle, known as myosteatosis \citep{chang2024prognostic}, which compromises muscle function and structural integrity. This reduction serves as a critical indicator of sarcopenia, a condition characterized by the progressive loss of skeletal muscle mass and strength, as well as frailty and diminished physical performance, particularly in aging populations \citep{cawthon2015assessment}. 

The VAT/SAT ratio, on the other hand, is a key metric for assessing metabolic risk \citep{kaess2012ratio, oh2017visceral}. While visceral adipose tissue (VAT) is strongly associated with metabolic disturbances and cardiovascular risk \citep{vasamsetti2023regulation}, its volume alone may reflect both overall fat mass and an individual's tendency to store fat viscerally \citep{kaess2012ratio}. In contrast, the VAT/SAT ratio offers a more precise assessment, as it accounts for the balance between visceral and subcutaneous fat, providing insight that is independent of total body fat percentage \citep{kaess2012ratio}.

Muscle area/volume and SMI are essential measurements of total muscle quantity and its proportionality to body size. These metrics provide critical information about an individual’s muscle reserves, which are vital for mobility, metabolic function, and overall health status \citep{chen2023really}. Studies highlight them as significant markers of nutritional status \citep{risch2022assessment}, which are crucial for recovery from illness, mortality, and treatment-related complications, such as the length of hospital stays and the rate of readmissions \citep{schuetz2021management, kaegi2021evaluation, guenter2021malnutrition}. Furthermore, they also serve as important factors in assessing metabolic health \citep{cruz2019sarcopenia, prado2014lean, martin2013cancer, dodds2015epidemiology}, as lower muscle mass is associated with insulin resistance and impaired glucose metabolism.

The collection of these metrics pictures the clear body condition of patients, showcasing a comprehensive overview of their muscle composition, fat distribution, and overall physiological status.

\subsection{Traditional methods for body composition segmentation}
Most early studies on body composition analysis rely on semi-automated threshold-based segmentation using predefined Hounsfield unit (HU) ranges to differentiate lean muscle mass from adipose tissue \citep{lee2017pixel, ji2022thresholds}. Despite its simplicity, threshold-based segmentation presents significant challenges due to the overlapping HU values between different tissue types, such as SAT and skin, as well as muscle and adjacent organs \citep{lee2017pixel}. The method is also highly susceptible to image noise \citep{sehgal2022ct, diwakar2020ct}, which can significantly compromise tissue classification accuracy, particularly in low-quality or artifact-prone scans. As a result, the method typically requires manual correction based on visual analysis by highly skilled radiologists and is impractical on large datasets due to the expense and time required.

To overcome these limitations, researchers have developed various advanced segmentation algorithms, including rule-based \citep{kamiya2009automated, kamiya2011automated}, clustering-based \citep{positano2009accurate, positano2004accurate, christ2011fuzzy}, and finite-element-method-based \citep{popuri2015body} approaches. Kamiya et al. proposed a rule-based expert system for segmenting the psoas major and rectus abdominis muscles from CT images, approximating muscle shapes with simple quadratic functions \citep{kamiya2009automated, kamiya2011automated}. Positano et al. utilize a fuzzy c-mean algorithm to make unsupervised classification of image pixels on MRI \citep{positano2009accurate, positano2004accurate}. Karteek and the team developed a novel FEM deformable model for muscle and fat segmentation from CT \citep{popuri2015body}.

However, these methods primarily focus on extracting specific muscle groups from CT or MRI scans and are unable to differentiate between visceral fat, subcutaneous fat, and intramuscular fat—an essential distinction in many body composition analysis tasks \citep{staley2019visceral, torres2013nutritional, iwase2016impact}. A potential approach to address these challenges is the use of deep learning-based segmentation algorithms.

\subsection{Deep learning-based models for body composition segmentation}
Deep learning-based segmentation has been proven to be a reliable technique in various clinical applications \citep{gu2024segmentanybone, dong2024segment, wasserthal2023totalsegmentator, mazurowski2023segment}. While networks offer high accuracy, reduce human labor, and provide greater generalizability compared to traditional segmentation algorithms, it is straightforward to apply deep learning-based segmentation algorithms for body composition analysis.

The majority of current deep learning-based segmentation models for body composition are still based on convolutional neural networks (CNNs) \citep{nowak2020fully}. U-Net and its variants are among the most widely used architectures in this domain, providing precise segmentation of body composition components such as skeletal muscle, SAT, and VAT \citep{paris2020body, weston2019automated}. However, these models are typically not publicly accessible and are often designed for specific tasks \citep{mai2023systematic}. While a few commercial models are available \citep{cespedes2020evaluation, mai2023systematic, lee2021deep}, they are often associated with high costs and limited customization options. TotalSegmentator \citep{wasserthal2023totalsegmentator}, a recently published general CT segmentation model based on nnU-Net, also supports muscle and fat segmentation. However, studies have shown that its performance in segmenting muscle, SAT, and VAT can be further improved, and its non-commercial license restricts broader usage. Therefore, there remains a significant need for publicly accessible, transparent, and generalizable segmentation models for body composition analysis.
% A methodological, model, or similar section often comes here.
\section{Methods}
\label{sec:method}
\subsection{Datasets} \label{sec:Datasets}
In this study, we utilize two datasets: an internal dataset collected from Duke Hospital and the publicly available SAROS dataset \citep{koitka2023saros, clark2013cancer} that integrates four publicly available CT datasets from TCIA with sparse annotations. The internal dataset is exclusively used for model implementation to ensure flexible and permissive licensing requirements. Both datasets are utilized for segmentation evaluation. The datasets are diverse in institutions, years, and patient demographics, providing the general and reliable evaluation and analysis results in this study.

\subsubsection{Dataset 1: Internal dataset}
\label{dataset:internal}
For this project, we collected 8948 CT volumes from Duke University Health System, spanning January 2016 to November 2019, including chest, abdomen, and pelvis exams. From the initial collection, we further identified 1927 volumes from 854 patients based on two criteria: (1) axial view exam was available (2) the volumes were original axial acquisitions, not derived from multiplanar reconstructions (MPR) or reformatted from other planes. These criteria were selected to align the model with real-world clinical scenarios. Among the identified studies, 483 patients were designated for segmentation model development, with 453 patients used for training and 30 patients used for testing, while the remaining 371 patients were allocated for further analysis of the relationship between body composition measurements and demographics. Noticeably, for 371 patients that were assigned for demographic analysis, only non-contrast-enhanced volumes were selected to ensure the analysis consistency.

To mitigate the potential bias in the testing and analysis process, only a single volume was randomly selected if patients had multiple eligible CT volumes. This approach ensured that the testing and analysis datasets provided an unbiased representation of the patient population.

Moreover, to best utilize our limited annotation resources and ensure data variability and model generalization, the training dataset was constructed by randomly sampling slices from the volumes. The selected slices were annotated by four Duke students under the guidance of experienced radiologists. To ensure the model accuracy, all the slices in the testing dataset are modified and approved by the radiologists.

\subsubsection{Dataset 2: SAROS dataset}
\label{dataset:external}
The SAROS dataset \citep{koitka2023saros, clark2013cancer} is a comprehensive collection of CT imaging volumes available on TCIA, featuring sparse annotations for 13 body region labels and six body part labels. The 13 annotations for body regions include the abdominal cavity, thoracic cavity, bones, brain, breast implants, mediastinum, muscles, parotid glands, submandibular glands, thyroid glands, pericardium, spinal cord, and subcutaneous tissue. The six body parts are the left arm, right arm, left leg, right leg, head, and torso, comprising a total of 900 CT volumes from 882 unique patients.

Given the torso label overlap with chest, abdomen, and pelvis regions, which are the focus body regions for our study, we utilize the slices with torso labels for our model evaluation by comparing the model segmented results with annotated skeletal muscle and SAT annotations. Furthermore, to ensure the flexible use of our model, we only selected a subset of the SAROS dataset covered under a commercial license for evaluation, more details for collection selection are shown in Appendix Section \ref{sec:Data collections from SAROS}. As a result, in total, 650 CT volumes from 632 unique patients with  CT slices are selected for segmentation model evaluation.

\subsubsection{Patient demographics}
The below table provides a demographic overview of patients from the five dataset collections, derived from two sources: the internal dataset (including internal training, internal testing, and demographic analysis collections, shown in Section \ref{dataset:internal}) and the external dataset (SAROS collections, in Section \ref{dataset:external}). Notably, the training, testing, and demographic analysis collections from the internal dataset are entirely separate from one another, ensuring a clean environment for segmentation model development. Patients' ages and races are unknown for the SAROS dataset due to de-identification.
\begin{table*}[ht!]
\resizebox{\textwidth}{!}{%
\centering
\renewcommand{\arraystretch}{1.2} % Adjust this value to increase row height
\resizebox{\textwidth}{!}{%
\begin{tabular}{|c|c|c|c|c|c|c|}
\hline
\multicolumn{3}{|c|}{} & \textbf{Internal Training} & \textbf{Internal Testing} & \textbf{Demographic Analysis} & \textbf{SAROS Dataset} \\ \hline
\multicolumn{3}{|c|}{Number of patients} & 453 & 30 & 371 & 632 \\ \hline
\multicolumn{3}{|c|}{Number of studies} & 783 & 30 & 371 & 650 \\ \hline
\multicolumn{3}{|c|}{Number of slices} & 1863 & 636 & 183972 & 10038 \\ \hline
\multirow{7}{*}{Demographics} & \multicolumn{2}{|c|}{Age} & 60.1 (3 - 89) & 58.5 (5 - 84)  & 58.3 (0.25-7) & - \\ \cline{2-7} 
 & \multirow{2}{*}{\centering Sex} & Female & 51.88\% (235) & 53.33\% (16) & 52.02\% (193) & 54.75\% (346) \\ \cline{3-7} 
 &  & Male & 48.12\% (218) & 46.67\% (14) & 47.98\% (178) & 45.25\% (286) \\ \cline{2-7} 
 & \multirow{5}{*}{\centering Race} & White & 67.11\% (304) & 63.33\% (19) & 68.19\% (253) & - \\ \cline{3-7} 
 &  & Black/ African American & 25.83\% (117) & 30.00\% (9) & 23.99\% (89) & - \\ \cline{3-7} 
 &  & Asian & 1.99\% (9) & 0\% (0) & 1.62\% (6) & - \\ \cline{3-7} 
 &  & American Indian & 0.66\% (3) & 0\% (0) & 1.08\% (4) & - \\ \cline{3-7} 
 &  & Other & 4.42\% (20) & 3.33\% (1) & 5.12\% (19) & - \\ \hline
\end{tabular}%
}}
\caption{\textbf{Patient demographics for four collections:} Table presents demographic details for four collections, including Internal Training, Internal Testing, Demographic Analysis, and SAROS Dataset Testing. For sex and race, the absolute number of patients is shown in parentheses alongside the percentages. For age, the mean values for ages are provided in years along with the minimum and maximum age in parentheses. Notably, the youngest patient, recorded as 3 months old, is consistently represented as 0.25 years.}
\label{tab:sample_table}
\end{table*}

\subsection{Segmentation algorithm}
While the nnU-Net \citep{isensee2021nnu, isensee2024nnu} has been proven to be one of the most powerful segmentation models in many bodies’ regions on medical imaging, in this work, we utilize the 2D nnU-Net with ResEnc presets on skeletal muscle, SAT, VAT, and muscular fat segmentation. Five-cross validation is utilized to select the best-performance model based on the average Dice coefficient across all four labels. In the subsections below, we introduce the overall architecture of the 2D nnU-Net and the Dice coefficient measurement.  In this work, the algorithm was trained and evaluated on an NVIDIA RTX 3090 GPU, ensuring efficient computation and high performance.

\subsubsection{nnU-Net}
nnU-Net \citep{isensee2021nnu} is a highly adaptable semantic segmentation method designed to automatically configure an optimized U-Net-based pipeline for any given dataset. Recent updates to the nnU-Net methodology have introduced enhancements to the U-Net baseline, emphasizing the importance of using advanced CNN architectures like ResNet and ConvNeXt variants, leveraging the robust nnU-Net framework, and employing model scaling for improved performance. By analyzing the nnU-Net's performance on multiple medical imaging datasets, the nnU-Net ResEnc XL has been shown to surpass the vanilla nnU-Net by an average of 0.93\% \citep{isensee2024nnu}. Therefore, in this work, we follow their findings and adopt the newly published nnU-Net ResEnc XL for our model development.

\subsubsection{Dice coefficient} \label{Method:Dice coefficient}
Dice coefficient is commonly used for image segmentation tasks to evaluate segmentation accuracy by measuring the overlap between predicted and ground truth regions. It ranges from 0 to 1, where a value of 1 indicates perfect overlap and 0 signifies no overlap. The mathematic formula for the Dice coefficient is Equation \eqref{dice}
\begin{equation}\label{dice}
    \text{Dice} = \frac{2 |A \cap B|}{|A| + |B|}
\end{equation}
where A represents the model predicted mask and the B represents the set of pixels in the ground truth. 

\subsubsection{Mean relative absolute error}\label{Method:Mean relative absolute error}
MRAE measures the absolute mean for relative errors across all data points (shown in Equation \eqref{MRAE}, which is normally applied to measure the relative error between prediction and ground truth. Lower MRAE values indicate better model performance, reflecting smaller deviations between predicted and actual values.
\begin{equation}\label{MRAE}
\text{MRAE} = \frac{1}{n} \sum_{i=1}^{n} \left| \frac{A_i - B_i}{A_i} \right|
\end{equation}
where $A_i$ represents the ground truth values, $B_i$ represents the predicted values, and $n$ is the total number of data points.

\subsection{Body composition metrics}
Our model is capable of measuring four commonly used body composition metrics: muscle density, VAT/SAT ratio, muscle area/volume, and SMI in both 2D and 3D settings. By analyzing previous studies on body composition analysis, we selected the third lumbar vertebral level (L3) for 2D body composition measurements and the region spanning the twelfth thoracic vertebral level (T12) to the fourth lumbar vertebral level (L4) for 3D measurements. L3 is considered the most commonly used standard for body composition assessment in multiple clinical applications, including rectal cancer assessment \citep{han2020association, arayne2023comparison}, sarcopenia evaluation \citep{amini2019approaches, pickhardt2020automated, pickhardt2020automate}, and obesity research \citep{liu2023fully, malietzis2015role}. For the 3D body composition measurement, T12 is selected as the beginning of the 3D measurement region following the approach of previous studies \citep{demerath2007approximation, tong2014optimization}. This focus is particularly relevant for assessing visceral adipose tissue (VAT) and subcutaneous adipose tissue (SAT); however, recent studies related to sarcopenia and rectal cancer also pay increasing attention to T12 \citep{fernandez2024ia, arayne2023comparison, soh2024prognostic}. L4 is selected as the ending point since 92.62\% of our internal abdominal CT volumes include L4, while only 78.07\% include L5.

TotalSegmentator \citep{wasserthal2023totalsegmentator} is utilized for automatically extracting, T12, L3, and L4. We select the slice with the largest L3 label among all slices with L3 mask for our 2D body composition measurement. For 3D measurement, we extract the portion between the first slice with T12 label and the last slice with L4. In the subsequent sections, we detail the calculations for muscle density, VAT/SAT ratio, muscle area/volume, and SMI.

\textbf{Muscle density} measures the average Hounsfield Unit (HU) values within the segmented skeletal muscle area (SMA) with higher values indicating leaner muscle and lower values (typically from -29 to 29 HU \citep{salam2023opportunistic}) suggesting fat infiltration. Muscle density is the crucial biomarker for muscle quality \citep{looijaard2016skeletal, cleary2015does, wang2021muscle} and is frequently associated with evaluations of sarcopenia and myosteatosis \citep{cawthon2015assessment, sergi2016imaging, tagliafico2022sarcopenia}.

\textbf{VAT/SAT ratio} is more commonly related to obesity-related health risks, such as diabetes, cardiovascular disease, and metabolic syndrome \citep{piche2018overview, frank2019determinants, goossens2017metabolic, ladeiras2017ratio, tanaka2021distinct}. A higher ratio indicates a predominance of visceral adipose tissue (VAT) over subcutaneous adipose tissue (SAT), reflecting an unfavorable fat distribution pattern \citep{ladeiras2017ratio}. Visceral fat is metabolically active and associated with chronic inflammation, insulin resistance, and dyslipidemia, which contribute to the development and progression of these conditions \citep{hardy2012causes, chait2020adipose, bansal2023visceral}. 

\textbf{Muscle area/volume} assesses the total skeletal muscle within the region of interest (ROI). Specifically, for 2D measurement, this metric, also referred to as skeletal muscle area (SMA), is calculated by multiplying the number of pixels within the segmented skeletal muscle mask by the area ($m^2$) represented by each pixel. For 3D measurement, the segmented skeletal muscle area/volume is determined by multiplying both the pixel size and the slice thickness ($m^3$). The SMA is one of the standard metrics for muscle quantity evaluation \citep{goodpaster2000composition, sinelnikov2016measurement, vella2020skeletal} and has been demonstrated to be highly correlated with patients' post-operative recovery and survival rates \citep{antoniou2019effect, bradley2022relationship, antoniou2019effect}. 3D muscle area/volume provides better representation of the entire muscle \citep{momose2017ct}. 

\textbf{SMI} is another commonly used metrics for muscle quantity measurement. This metrics normalizes the muscle cross-sectional area (CSA) by dividing it by the individual's height squared ($m^2$).

\section{Segmentation results}
This section presents the segmentation performance of our model on three selected labels: skeletal muscle, SAT, and VAT, as well as four key body composition metrics: muscle density, muscle area/volume, SMI, VAT/SAT ratio. The evaluation is provided both qualitatively, through visual comparisons, and quantitatively, using the Dice coefficient and MRAE to measure the overlap between the manually annotated ground truth and the model's segmentation. The quantitative analysis highlights the performance of our model on both our internal dataset and the publicly available SAROS dataset \citep{koitka2023saros, clark2013cancer}, benchmarking it against TotalSegmentator \citep{wasserthal2023totalsegmentator} and the internal tool \citep{hou2024enhanced}. 

\subsection{Qualitative evaluation}
Figure \ref{fig1:2D_qualitative_evaluation} presents the L3 segmentation results and their corresponding body composition metrics (muscle density, VAT/SAT ratio, muscle area/volume, and SMI) for selected patients. The samples are categorized based on their body composition metric values into five groups: Low, Moderately Low, Moderate, Moderately High, and High, with cut-off points set at the 20th, 40th, 60th, and 80th percentiles of the entire population. For each body composition metric, one sample is randomly selected from each category for visualization. Each column in Figure \ref{fig1:2D_qualitative_evaluation} corresponds to a specific category, with patients arranged from low to high values across the columns, ensuring a consistent representation of the metric's progression. Each row, in turn, highlights a specific body composition metric. The first row illustrates muscle density, the second depicts the VAT/SAT ratio, the third represents muscle area/volume, and the fourth corresponds to the SMI. Notably, as shown in figure \ref{fig1:2D_qualitative_evaluation} result, there's no simple correlation between the four body composition metrics. For example, a patient with the highest muscle density in the first row does not exhibit the highest muscle area/volume. More precise relationship analysis based on Pearson Correlation coefficient for each body composition metrics pair is shown in Appendix \ref{sec:Body composition metrics relationship}.

\clearpage
\begin{figure*}[t]
    \centering
    % Subfigure 1
    \begin{subfigure}[b]{0.9\textwidth}
        \centering
        \includegraphics[width=\linewidth]{fig/Qualitative_evaluation_muscle_density_2d.png}
    \end{subfigure}
     \vspace{0.05cm}
    
    % Subfigure 2
    \begin{subfigure}[b]{0.9\textwidth}
        \centering
        \includegraphics[width=\linewidth]{fig/Qualitative_evaluation_vat_sat_ratio_2d.png}
    \end{subfigure}
     \vspace{0.05cm}
    
    % Subfigure 3
    \begin{subfigure}[b]{0.9\textwidth}
        \centering
        \includegraphics[width=\linewidth]{fig/Qualitative_evaluation_muscle_volume_2d.png}
    \end{subfigure}
     \vspace{0.05cm}
    
    % Subfigure 4
    \begin{subfigure}[b]{0.9\textwidth}
        \centering
        \includegraphics[width=\linewidth]{fig/Qualitative_evaluation_SMI.png}
    \end{subfigure}
    \vspace{0.05cm}
    
    \caption{\textbf{Qualitative evaluation of our segmentation model:} Figure shows segmentation results of the abdominal L3 slice. Each row represents a specific body composition metric (in bold), with five patients arranged from left to right in categories: Low, Moderately Low, Moderate, Moderately High, and High. For example, in the first row (muscle density), the first patient has low muscle density, and the fifth has high muscle density. The second, third, and fourth rows show the VAT/SAT ratio, muscle area/volume, and SMI, respectively, following the same left-to-right order. In the segmentation, \textit{dark blue} shows skeletal muscle, \textit{light blue} SAT, \textit{yellow} VAT, and \textit{maroon} muscular fat.}
    \label{fig1:2D_qualitative_evaluation}
\end{figure*}
\clearpage

In figure \ref{fig2:3D_qualitative_evaluation}, three volumes are randomly selected from the demographic analysis dataset to demonstrate the algorithm's performance for 3D body composition measurement. For each volume, four slices corresponding to T12, L1, L2, and L4 are displayed on the left, along with their automatically generated segmentation masks. L3 segmentation is not included in this figure, as it is fully demonstrated in Figure \ref{fig1:2D_qualitative_evaluation}. Additionally, the stacked slices from T12 to L4 are visualized in the sagittal view, shown on the right side of the figure.

\subsection{Quantitative evaluation} \label{sec:Quantitative evaluatio}
Quantitative segmentation model evaluation is demonstrated with three experiments. Subsection \ref{sec:Internal dataset evaluation} demonstrate the model segmentation performance on our internal dataset and public available SAROS dataset based on the Dice coefficient between model segmentation and manual annotation for skeletal muscle, SAT, VAT, and muscular fat. The performance on multiple body location has been analyzed, specifically we evaluate the segmentation result on both L3, T12-L4, and every slice among the selected volume (chest, abdomen, and pelvis). \ref{sec:External dataset evaluation} compares our segmentation performance with two of external segmentation model \citep{wasserthal2023totalsegmentator,hou2024enhanced} on the public available SAROS dataset \citep{koitka2023saros, clark2013cancer}. The evaluation focus on skeletal muscle and SAT performance on the location from to align with the evaluation proposed by \citep{clark2013cancer}. Section \ref{sec:Analysis metric evaluation} evaluates the segmentation accuracy by analyzing the mean error between automatically measured body composition metrics and those derived from manual annotations. Specifically, we present the mean error for metrics such as muscle density, VAT/SAT ratio, muscle area/volume, and SMI in both 2D and 3D settings. For more flexible use of muscular fat (including intra-muscular and inter-muscular fat), we also provide the Dice coefficient for muscular fat alone in Section \ref{sec:muscular fat segmentation}, along with the Dice coefficients for different applications. Specifically, we calculate the Dice coefficients for muscular fat + VAT and muscular fat + SAT, compared with the manual annotations in our internal dataset.

\begin{figure*}[ht!]
    \centering
    % Left Subfigure
    \begin{subfigure}[b]{0.65\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/Qualitative_evaluation_3D_sample5.png} % Replace with your image
    \end{subfigure}
    % Right Subfigure
    \begin{subfigure}[b]{0.65\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/Qualitative_evaluation_3D_sample3.png} % Replace with your image
    \end{subfigure}
    
    \begin{subfigure}[b]{0.65\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/Qualitative_evaluation_3D_sample2.png} % Replace with your image
    \end{subfigure}
    \caption{\textbf{Qualitative evaluation of our segmentation model for 3D setting:} Four slices corresponding to T12, L1, L2, and L4 are displayed on the left, while the stacked slices from T12 to L4 are visualized in the sagittal view on the right side of the figure. The visualization volumes are ordered by increasing of 2D muscle density (in bold). In the segmentation, \textit{dark blue} shows skeletal muscle, \textit{light blue} SAT, \textit{yellow} VAT, and \textit{maroon} muscular fat.}
    \label{fig2:3D_qualitative_evaluation}
    \label{fig:main}
\end{figure*}
\subsubsection{Internal segmentation performance} \label{sec:Internal dataset evaluation}
For our internal evaluation, without comparisons to other methods, we utilize the Dice coefficient to compare our auto-segmented labels with manual annotations for skeletal muscle, subcutaneous adipose tissue (SAT), and visceral adipose tissue (VAT). Table \ref{tab:internal_evaluation} summarizes the segmentation performance for both internal and external datasets, using the Dice coefficient (Section \ref{Method:Dice coefficient}) and MRAE (Section \ref{Method:Mean relative absolute error}). The model consistently performs best at the L3 slice, with higher average Dice coefficients (93.19\%) and lower MRAE (5.31\%) compared to the T12-L4 sub-volume and all slices. Among the tissues, the model achieves the highest segmentation accuracy for SAT, which consistently shows superior Dice scores and lower MRAEs compared to skeletal muscle and VAT. The external dataset follows a similar trend, with the best performance observed at L3 (average Dice of 91.91\% and MRAE of 6.15\%) and the highest accuracy for SAT. 

Notably, there is a label inconsistency between the annotations in our internal dataset and those in the SAROS dataset. Specifically, the SAROS annotation includes skin as part of the SAT label. To address this discrepancy, we applied a simple post-processing step to our model by dilating our SAT segmentation to include the skin. The detailed process for this post-processing is described in Appendix \ref{sec:Post-processing for label inconsistencies}. However, the post-processing step only mimics the inclusion of skin in the segmentation, which still leaves a gap between the two segmentation approaches.

\begin{table*}[t]
\centering
\resizebox{\textwidth}{!}{%
\begin{tabular}{c|c|c|c|c|c|c|c|c}
\toprule
\multicolumn{9}{c}{Internal Dataset} \\
\midrule
 & \multicolumn{2}{c|}{Skeletal Muscle} & \multicolumn{2}{c|}{SAT} & \multicolumn{2}{c|}{VAT} & \multicolumn{2}{c}{Average} \\
 \midrule
 & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) \\
\midrule
L3 & 92.33 $\pm$ 4.42 & 7.02 $\pm$ 8.16 & 94.55 $\pm$ 5.21 & 3.29 $\pm$ 3.77 & 92.70 $\pm$ 5.00 & 5.61 $\pm$ 6.19 & 93.19 $\pm$ 4.88 & 5.31 $\pm$ 6.04 \\
\midrule
T12-L4 & 91.95 $\pm$ 5.76 & 6.47 $\pm$ 7.46 & 93.38 $\pm$ 8.57 & 5.44 $\pm$ 8.68 & 91.29 $\pm$ 8.93 & 7.02 $\pm$ 10.49 & 92.21 $\pm$ 7.09 & 6.31 $\pm$ 8.88 \\
\midrule
All Slices & 91.85 $\pm$ 3.37 & 4.13 $\pm$ 4.10 & 94.06 $\pm$ 4.25 & 3.06 $\pm$ 3.97 & 89.45 $\pm$ 7.09 & 5.91 $\pm$ 5.51 & 91.79 $\pm$ 4.90 & 4.37 $\pm$ 4.53 \\
\toprule
\multicolumn{9}{c}{External Dataset} \\
\midrule
 & \multicolumn{2}{c|}{Skeletal Muscle} & \multicolumn{2}{c|}{SAT} & \multicolumn{2}{c|}{VAT} & \multicolumn{2}{c}{Average} \\
\midrule
 & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) & Dice $\uparrow$ (\%) & MRAE $\downarrow$ (\%) \\
\midrule
L3 & 91.70 $\pm$ 3.45 & 8.49 $\pm$ 5.26 & 92.12 $\pm$ 4.54 & 3.80 $\pm$ 3.37 & - & - & 91.91 $\pm$ 3.99 & 6.15 $\pm$ 4.32 \\
\midrule
T12-L4 & 91.03 $\pm$ 3.77 & 6.61 $\pm$ 4.24 & 92.59 $\pm$ 4.35 & 4.39 $\pm$ 3.32 & - & - & 91.81 $\pm$ 4.06 & 5.50 $\pm$ 3.78 \\
\midrule
All Slices & 89.68 $\pm$ 2.75 & 7.35 $\pm$ 3.80 & 90.27 $\pm$ 4.83 & 3.60 $\pm$ 3.39 & - & - & 89.98 $\pm$ 3.79 & 5.48 $\pm$ 3.60 \\
\bottomrule
\end{tabular}
}
\caption{ \textbf{Internal segmentation Performance:} Segmentation performance for skeletal muscle, subcutaneous adipose tissue (SAT), and visceral adipose tissue (VAT) across the internal and external datasets, reported using Dice scores ($\uparrow$) and MRAE ($\downarrow$). Results are provided for L3, T12-L4, and all slices, highlighting the model's superior performance at the L3 slice and on SAT compared to skeletal muscle and VAT. The "Average" column provides the mean Dice score and MRAE across the reported tissues. VAT performance is unavailable due to the absence of VAT annotation in the SAROS dataset.}
\label{tab:internal_evaluation}
\end{table*}

\subsubsection{Comparison with benchmark models} \label{sec:External dataset evaluation}
In this experiment we choose the algorithm proposed by Hou et al. \citep{hou2024enhanced} and TotalSegmentator model \citep{wasserthal2023totalsegmentator} as the benchmark. Performance is evaluated by the Dice coefficient compared with segmented mask and the public available skeletal muscle and SAT on SAROS dataset \citep{koitka2023saros, clark2013cancer}. To compare our method with the chosen benchmarks, we follow the instructions provided in \citep{hou2024enhanced}, constraining the analysis to the abdomen section, specifically L1–L5 and T9–T12 following the instructions provided \citep{hou2024enhanced}. The performance results are illustrated in Table \ref{tab:external_evaluation}. As a result, our model outperform the enhanced segmentation model \citep{hou2024enhanced} by 2.40\% for skeletal muscle and 10.26\% for SAT. Additionally, it surpasses the TotalSegmentator \citep{wasserthal2023totalsegmentator} by 7.81\% for skeletal muscle and 14.36\% for SAT. Notably, due to license restrictions, our evaluation dataset is a large subset of theirs, with 650 commercially licensed volumes used in our study compared to 900 volumes in theirs. However, due to the considerable amount of data and data overlap, it is still representative of the original dataset, ensuring the confidence of our advancements.

\begin{table*}[t]
\centering
\begin{tabular}{c|c|c}
\toprule
 & Skeletal Muscle (\%) & SAT (\%) \\
\midrule
TotalSegmentator \citep{wasserthal2023totalsegmentator} & 83.2 $\pm$ 4.6 [80.5, 86.4] & 80.8 $\pm$ 10.4 [76.7, 87.7]  \\
\midrule
Enhanced Segmentation \citep{hou2024enhanced} & 87.6 $\pm$ 3.3 [85.6, 90.0] & 83.8 $\pm$ 10.9 [80.7, 90.5] \\
\midrule
Ours & \textbf{89.7 $\pm$ 3.2 [88.3, 91.7]} & \textbf{92.4 $\pm$ 3.7 [91.0, 94.7]} \\
\bottomrule
\end{tabular}
\caption{\textbf{Comparison with benchmark models:} Segmentation performance for skeletal muscle and subcutaneous adipose tissue (SAT) across the external dataset is reported using Dice scores. The scores are presented as mean, standard deviation, and interquartile range (IQR). The bolded result highlights the best-performing model among the evaluated models.} % Add your caption
\label{tab:external_evaluation}
\end{table*}

\subsubsection{Analysis metric evaluation} \label{sec:Analysis metric evaluation}
For the metric evaluation, we utilize our auto-segmentation to measure four selected body composition metrics: muscle density, VAT/SAT ratio, muscle volume, and SMI, in both 2D and 3D settings on our internal test dataset. These results are compared with the body composition metrics derived from manual annotations. The MRAE result is presented in Table \ref{tab:metric_analysis}.

The model demonstrates the best performance in measuring muscle density among four body compositions, with an MRAE lower than 5\% on both internal and external datasets for both 2D and 3D measurements. Across all body composition metrics for both datasets, the model achieves an MRAE lower than 10\%, showcasing its robustness.

\begin{table*}[t]
\centering
\resizebox{\textwidth}{!}{%
\begin{tabular}{c|c|c|c|c}
\toprule
\multicolumn{5}{c}{Internal Dataset} \\
\midrule
& Muscle density (\% among range -29 to +150 HU) & VAT/SAT ratio (\%) & Muscle area/volume (\%) & SMI (\%) \\
\midrule
2D & 2.43 $\pm$ 1.67 & 5.43 $\pm$ 5.26 & 9.81 $\pm$ 10.27 & 9.81 $\pm$ 10.27 \\
\midrule
3D & 4.11 $\pm$ 2.94 & 5.09 $\pm$ 4.82 & 8.44 $\pm$ 8.03 & - \\
\toprule
\multicolumn{5}{c}{External Dataset} \\
\midrule
& Muscle density (\% among range -29 to +150 HU) & VAT/SAT ratio (\%) & Muscle area/volume (\%) & SMI (\%) \\
\midrule
2D & 4.47 $\pm$ 2.44 & - & 9.20 $\pm$ 5.31 & - \\
\midrule
3D & 4.71 $\pm$ 2.22 & - & 6.61 $\pm$ 4.24 & - \\
\bottomrule
\end{tabular}%
}
\caption{\textbf{Analysis metric evaluation performance:} The performance of our segmentation model on both internal and external datasets is evaluated by comparing the four body composition metrics automatically calculated by our model with the ground truth measured from manual annotations. The evaluation is based on MRAE ($\downarrow$).}
\label{tab:metric_analysis}
\end{table*}

\subsubsection{Muscular fat segmentation} \label{sec:muscular fat segmentation}
During the literature review, we observed inconsistencies in how muscular fat (both intra-muscular and inter-muscular fat) is classified in research. While some studies include muscular fat as part of skeletal muscle measurements \citep{hou2024enhanced, van2018percentiles}, others classify it under VAT \citep{camus2014prognostic, wirtz2021ct, connelly2013volumetric}, and a smaller subset considers it part of SAT \citep{ozturk2020relationship, magudia2021population}. Consequently, we attempted to segment muscular fat as a separate label in our segmentation model.

In this section, we present the Dice coefficient performance for muscular fat segmented independently, muscular fat included as part of SAT, muscular fat included as part of VAT, and muscular fat included as part of muscle. Notably, for all skeletal muscle evaluations discussed in the previous sections, we follow the methodology adopted in prior studies, where muscular fat is evaluated as part of muscle segmentation (Section \ref{sec:Internal dataset evaluation}\ref{sec:External dataset evaluation}). Although the segmentation of muscular fat itself demonstrates a relatively low Dice coefficient (56.27 $\pm$ 10.33\%) compared to manual annotations on our internal dataset, incorporating muscular fat into other labels—specifically muscle, SAT, and VAT, as is common in body composition measurements—results in high Dice coefficients across all slices (91.85 $\pm$ 3.37\%, 92.35 $\pm$ 4.6\%, and 85.19 $\pm$ 6.73\%, respectively).

\section{Body composition vs. demographic analysis} \label{sec:Body composition analysis}
In the following sections, we analyze the body composition measurements generated by our algorithm, highlighting their relationships with patients' age, sex, and racial groups. The results produced by our algorithm are compared with previous body composition findings reported in leading medical journals. 
% In Section \ref{sec}, we also analyze the prediction results for the selected criteria. 
The aim of this section is to demonstrate the accuracy of our body composition metrics calculation. While the calculated metric values may differ slightly from those reported in previous studies due to variations in population distribution, the trends shown in this analysis strongly resemble those established before.

\subsection{Body composition metrics vs. age}
To ensure a sufficient sample size for analysis, we divided the age range into six distinct groups, each containing at least 20 instances. The observed trends in muscle area, SAT, VAT area, and SMI with increasing age closely align with findings from previous studies \citep{magudia2021population}, with both the trends and absolute measurement values showing strong consistency across age groups. Specifically, muscle density decreases with age, while VAT and the VAT/SAT ratio increase.

\begin{figure*}[htbp]
    \centering
    % Row 1
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/muscle_density_2d.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/muscle_density_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/SAT_area_2d.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/SAT_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/VAT_area_2d.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/VAT_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/VAT_SAT_ratio_2d.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/VAT_SAT_ratio_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 3
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/muscle_area_2d.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/muscle_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 4
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/age/SMI.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth} % Empty subfigure
    \end{subfigure}
    \caption{Body composition metrics vs. age categories.}
    \label{fig:body composition vs. age}
\end{figure*}

\subsection{Body composition metrics vs. sex}
The left column sub-figures in Figure \ref{fig:body composition vs. sex} illustrate muscle density, SAT/VAT ratio, muscle area, and SMI versus gender, respectively, as measured at the L3 level. Similarly, the corresponding measurements for muscle density, SAT/VAT ratio, muscle volume as measured from T12 to L4, are shown in the right column sub-figures. The mean and standard deviation of these body composition metrics are consistent with those reported in previous studies \citep{van2018percentiles, graffy2019deep}. Our measurement also aligns with the previous findings that compared to female, male typically have a higher muscle density, VAT/SAT ratio, muscle area, SMI \citep{kammerlander2021sex, van2018percentiles, graffy2019deep}.

\begin{figure*}[htbp]
    \centering
    % Row 1
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/muscle_density_2d.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/muscle_density_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/VAT_area_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/VAT_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/SAT_area_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/SAT_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/VAT_SAT_ratio_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/VAT_SAT_ratio_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 3
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/muscle_area_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/muscle_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 4
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/sex/SMI.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth} % Empty subfigure
        % Intentionally left blank
    \end{subfigure}
    \caption{Body composition metrics vs. sex.}
    \label{fig:body composition vs. sex}
\end{figure*}

\subsection{Body composition metrics vs. race group}
For the race groups, to ensure sufficient data size for analysis, we conducted the analysis based only on two race groups: Caucasian/White and Black or African American. The relationship between race groups and multiple body composition metrics is demonstrated in Figure \ref{fig:body composition vs. race}. Few previous studies have exclusively analyzed body composition across different races, limiting our ability for direct comparisons. However, several studies have examined the combined impact of both sex and race. For example, \citep{magudia2021population} demonstrates that Black or African American individuals, on average, have larger muscle areas and higher SMI for both males and females. Similarly, \citep{beasley2009body} reports an average abdominal visceral fat area of 152.0 for White individuals and 129.9 for Black individuals, as well as an average abdominal subcutaneous fat area of 266.0 for White individuals and 312.1 for Black individuals. Although our study and \citep{beasley2009body} have different population distributions, with the latter being limited to healthy elderly adults, the measurement differences between the two studies for all metrics are within 10 \%.

\begin{figure*}[htbp]
    \centering
    % Row 1
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/muscle_density_2d.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/muscle_density_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/SAT_area_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/SAT_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/VAT_area_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/VAT_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/VAT_SAT_ratio_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/VAT_SAT_ratio_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 3
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/muscle_area_2d.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/muscle_area_3d.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 4
    \begin{subfigure}[b]{0.36\textwidth}
        \includegraphics[width=\linewidth]{fig/race/SMI.png}
    \end{subfigure}
    \begin{subfigure}[b]{0.36\textwidth} % Empty subfigure
        % Intentionally left blank
    \end{subfigure}
    \caption{Body composition metrics vs. race.}
    \label{fig:body composition vs. race}
\end{figure*}

\section{Discussion and future work}
To mitigate the gap that there are few publicly available deep learning-based CT segmentation and body composition measurement models for abdominal muscle and fat, we built this model based on nnU-Net ResEnc XL. This model is able to segment skeletal muscle, subcutaneous adipose tissue (SAT), and visceral adipose tissue (VAT) across the chest, abdomen, and pelvis in axial CT images. It additionally automatically measures muscle density, visceral-to-subcutaneous fat (VAT/SAT) ratio, muscle area/volume, and SMI. All the code will be made publicly available at https://github.com/mazurowski-lab/CT-Muscle-and-Fat-Segmentation.git.

This study highlights the strong capability of our model in segmenting skeletal muscle, SAT, and VAT across the chest, abdomen, and pelvis in axial CT images. As detailed in Section \ref{sec:Internal dataset evaluation}, the model achieves an average Dice score of 91.79 ± 4.90\% across all slices and all four labels in the internal dataset. In the external dataset, the average Dice score is 89.98 ± 3.79\% .

The model demonstrates even better performance when segmenting the L3 slice and the T12–L4 region. For the L3 slice, it achieves an average Dice score of 93.19 ± 4.88 in the internal dataset and 91.91 ± 3.99\%  in the external dataset. For the T12–L4 region, the average Dice score is 92.21 ± 7.09\% on internal dataset and 91.81 ± 4.06\% on external dataset.

When compared to previous methods, our model shows significant improvements, outperforming the recently published in-house segmentation model \citep{hou2024enhanced} by 2.40\% for skeletal muscle and 10.26\% for SAT. Additionally, it surpasses the TotalSegmentator \citep{wasserthal2023totalsegmentator} by 7.81\% for skeletal muscle and 14.36\% for SAT. These evaluations are based on manual annotations from the publicly available SAROS dataset. A detailed comparison with benchmark models is provided in Section \ref{sec:External dataset evaluation}.

Apart from segmentation performance, our model also demonstrates high accuracy in measuring commonly used body composition metrics, including muscle density, visceral-to-subcutaneous fat ratio, muscle area/volume, and SMI in both 2D and 3D settings. The average MRAE for all metrics is below 10\%. As detailed in Section \ref{sec:Analysis metric evaluation}, the model achieves its best performance in measuring muscle density, with an MRAE of less than 5\% compared to manual annotations across internal and external datasets in both 2D and 3D settings.

Furthermore, utilizing our model, we performed body composition metrics analysis across different age, sex, and race groups on 371 randomly selected patients from Duke Hospital. The results demonstrate clear differences in muscle density, adipose tissue distribution, and SMI among patients of different ages, sexes, and races. With increasing age, there was a noticeable decline in muscle density and SMI, coupled with an increase in visceral adipose tissue (VAT) and the VAT/SAT ratio, indicating age-related muscle loss and fat redistribution. Sex-based comparisons revealed that males generally had higher muscle density, muscle volume, and SMI, while females exhibited higher subcutaneous adipose tissue (SAT) levels. Additionally, race-based analysis showed that Black or African American individuals had higher muscle mass and SAT but lower VAT levels and VAT/SAT ratios compared to Caucasian/White individuals. All the findings also align with results from previous studies, highlighting the robustness of our model for both segmentation and body composition measurement.

Despite the promising results, this study has several limitations that present opportunities for further development and improvement. First, the scope of the study is restricted to three body regions and relies solely on axial views. To enhance the generalizability and clinical utility of our approach, future work will focus on expanding the analysis to additional body regions, such as the hip, leg, and shoulder, which are also commonly assessed in body composition studies.

While our model demonstrates reasonable performance on sagittal and coronal views by stacking segmented axial slices and extracting intersections across different planes, as illustrated in Figure \ref{fig2:3D_qualitative_evaluation}, this approach has inherent limitations. For example, since axial slices are segmented independently, which may lead to inconsistencies between adjacent slices. We recognize the potential benefits of directly incorporating sagittal and coronal views into the training and evaluation pipeline, which may improve segmentation accuracy and consistency across all anatomical planes.

Lastly, although the segmentation model includes a muscular fat label, its performance is comparatively lower than that of the other three labels. This discrepancy is primarily due to variability in annotation granularity among different annotators. To enhance annotation consistency in future versions, we will establish clear annotation standards. Specifically, we will define muscular fat regions by applying a Hounsfield Unit (HU) threshold between -220 and -50 for fat tissue, as suggested by Chougule et al. \citep{chougule2018clinical}, and retain only contiguous fat regions comprising more than six pixels.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Mandatory Sections. Please complete, especially for final publication
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Acknowledgements.
% Please include any funding, intellectual contributions not included in the authorship, and any other acknowledgements.
\acks{This work was supported through a partnership between the Duke Departments of Surgery and Radiology and the Duke Spark Initiative for AI in Medical Imaging.}

% Ethical Standards.
% Please edit with the appropriate ethics considerations for your work. Include any pertinent IRB information, etc.
%
% Please note that the submission requirements included:
% The work presented must follow appropriate ethical standards in conducting research and writing the manuscript, following all applicable laws and regulations regarding treatment of animals or human subjects.
\ethics{The research protocol was approved by the Duke Health System Institutional Review Board (IRB) with ethical standards for research and manuscript preparation, adhering to all relevant laws and regulations concerning the treatment of human subjects and animals.}

% Conflict of Interest
% Declaration of possible conflicts of interest: Authors must disclose any financial, organisational, commercial or personal conflicts of interest that might bias their work.
% If no conflicts, please say "We declare we don't have conflicts of interest."
\coi{We declare we don't have conflicts of interest.}

% Data availability
\data{The external dataset utilized for model evaluation and analysis in this study is publicly accessible \citep{koitka2023saros, clark2013cancer}. However, the internal dataset is currently unavailable, as de-identifying the data requires an extensive institutional review process. Readers interested in evaluating the accuracy of the method can access and utilize the publicly available datasets, which are readily accessible and user-friendly. The code for this study is also publicly available at https://github.com/mazurowski-lab/CT-Muscle-and-Fat-Segmentation.git.}

\bibliography{refs}

% Manual newpage inserted to improve layout of sample file - not
% needed in general before appendices.
% \newpage

% Appendix is optional
\clearpage
\appendix
\section{Data collections from SAROS} \label{sec:Data collections from SAROS}
While only a subset of the collections within SAROS is provided with a commercially permitted license (specifically CC BY 3.0 and CC BY 4.0), we exclusively utilized this subset in external evaluation to ensure maximum flexibility for our model. In this section, we provide a detailed list of the dataset collections used in this study, shown in Table \ref{tab:SAROS_data_collection}, including the collection name, scan region (Abdomen, Thorax, Whole-body) assigned by SAROS \citep{koitka2023saros, clark2013cancer}, and their license type.

\begin{table*}[t]
\centering
\resizebox{\textwidth}{!}{%
\begin{tabular}{c|c|c|c|c|c}
\toprule
Collection & Number of studies & Abdomen & Thorax & Whole-body & License \\
\toprule
ACRIN-FLT-Breast$^{1,2}$ & 32 & 0 & 0 & 32 & CC BY 3.0 \\
\midrule
ACRIN-NSCLC-FDG-PET$^{3,4}$ & 129 & 0 & 78 & 51 & CC BY 3.0 \\
\midrule
Anti-PD-1\_Lung$^{5}$ & 12 & 0 & 0 & 12 & CC BY 3.0 \\
\midrule
C4KC-KiTS$^{6,7}$ & 175 & 175 & 0 & 0 & CC BY 3.0 \\
\midrule
CPTAC-CM$^{8}$ & 1 & 0 & 0 & 1 & CC BY 3.0 \\
\midrule
CPTAC-LSCC$^{9}$ & 3 & 0 & 0 & 3 & CC BY 3.0 \\
\midrule
CPTAC-LUAD$^{10}$ & 1 & 0 & 0 & 1 & CC BY 3.0 \\
\midrule
CPTAC-PDA$^{11}$ & 8 & 0 & 0 & 8 & CC BY 3.0 \\
\midrule
CPTAC-UCEC$^{12}$ & 26 & 25 & 0 & 1 & CC BY 3.0 \\
\midrule
LIDC-IDRI$^{13,14}$ & 133 & 0 & 133 & 0 & CC BY 3.0 \\
\midrule
NSCLC Radiogenomics$^{15,16,17,18}$ & 7 & 0 & 0 & 7 & CC BY 3.0 \\
\midrule
Pancreas CT$^{19,20}$ & 58 & 58 & 0 & 0 & CC BY 3.0 \\
\midrule
Soft-tissue-Sarcoma$^{21,22}$ & 6 & 0 & 0 & 6 & CC BY 3.0 \\
\midrule
TCGA-LIHC$^{23}$ & 33 & 32 & 0 & 1 & CC BY 3.0 \\
\midrule
TCGA-LUAD$^{24}$ & 2 & 0 & 0 & 2 & CC BY 3.0 \\
\midrule
TCGA-LUSC$^{25}$ & 3 & 0 & 0 & 3 & CC BY 3.0 \\
\midrule
TCGA-STAD$^{26}$ & 2 & 2 & 0 & 0 & CC BY 3.0 \\
\midrule
TCGA-UCEC$^{27}$ & 1 & 0 & 0 & 1 & CC BY 3.0 \\
\midrule
COVID-19-NY-SBU$^{28}$ & 1 & 0 & 0 & 1 & CC BY 4.0 \\
\midrule
Lung-PET-CT-Dx$^{29}$ & 17 & 0 & 15 & 2 & CC BY 4.0 \\
\bottomrule
In total & 650 & 292 & 226 & 132 & - \\
\bottomrule
\end{tabular}
}
\caption{Dataset collections from SAROS. References: 
$^{1}$\citep{kostakoglu2015phase}, $^{2}$\citep{Kinahan2017}, 
$^{3}$\citep{machtay2013prediction}, $^{4}$\citep{Kinahan2019}, 
$^{5}$\citep{Madhavi2019}, $^{6}$\citep{heller2021state}, $^{7}$\citep{Heller2019}, 
$^{8}$\citep{CPTAC2018}, $^{9}$\citep{CPTAC2018LSCC}, $^{10}$\citep{CPTAC2018LUAD}, 
$^{11}$\citep{CPTAC2018_v2}, $^{12}$\citep{CPTAC2019UCEC}, 
$^{13}$\citep{armato2011lung}, $^{14}$\citep{Armato2015}, 
$^{15}$\citep{Napel2014}, $^{16}$\citep{Bakr2017}, $^{17}$\citep{bakr2018radiogenomic}, $^{18}$\citep{gevaert2012non}, 
$^{19}$\citep{Roth2016}, $^{20}$\citep{roth2015deeporgan}, 
$^{21}$\citep{vallieres2015radiomics}, $^{22}$\citep{Vallieres2015}, 
$^{23}$\citep{Erickson2016}, $^{24}$\citep{Albertina2016}, 
$^{25}$\citep{Kirk2016}, $^{26}$\citep{Lucchesi2016}, 
$^{27}$\citep{Erickson2016UCEC}, $^{28}$\citep{Saltz2021}, $^{29}$\citep{Li2020}.}
\label{tab:SAROS_data_collection}
\end{table*}

\section{Post-processing for label inconsistencies} \label{sec:Post-processing for label inconsistencies}
Label post-processing is performed in two steps. First, a 5×5 structuring element is applied to morphologically dilate the SAT label. Second, the expanded region is constrained to ensure that (1) it does not overlap with any previously labeled areas and (2) it remains within the abdominal region. The abdominal boundary is determined by thresholding at -800 Hounsfield Units (HU), as skin typically exhibits an HU value around this level \citep{chougule2018clinical, villa2012hounsfield}. Figure \ref{fig:appendix:dilation} demonstrates the dilation result with  on two randomly selected slices. Notably, the post-processing step merely approximates the inclusion of skin in the segmentation, leaving a remaining discrepancy between the two approaches.

\begin{figure*}[t]
    \centering
    \captionsetup{font=footnotesize} % Reduce caption font size

    % ----- Subfigure 1 -----
    \begin{subfigure}[b]{0.19\textwidth} % Slightly wider to avoid wrapping
        \centering
        \includegraphics[width=\textwidth]{fig/appendix/volume1/raw_slice.png}
        \includegraphics[width=\textwidth]{fig/appendix/volume2/raw_slice.png}
        \caption{Raw slice}
    \end{subfigure}
    % ----- Subfigure 2 -----
    \begin{subfigure}[b]{0.19\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/appendix/volume1/ground_truth.png}
        \includegraphics[width=\textwidth]{fig/appendix/volume2/ground_truth.png}
        \caption{Ground Truth}
    \end{subfigure}
    % ----- Subfigure 3 -----
    \begin{subfigure}[b]{0.19\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/appendix/volume1/original_prediction.png}
        \includegraphics[width=\textwidth]{fig/appendix/volume2/original_prediction.png}
        \caption{Our prediction} % Fixed width
    \end{subfigure}
    % ----- Subfigure 4 -----
    \begin{subfigure}[b]{0.19\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/appendix/volume1/dilation.png}
        \includegraphics[width=\textwidth]{fig/appendix/volume2/dilation.png}
        \caption{\makebox[2.5cm][c]{Prediction + dilation}}
    \end{subfigure}
    \begin{subfigure}[b]{0.19\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/appendix/volume1/substraction.png}
        \includegraphics[width=\textwidth]{fig/appendix/volume2/substraction.png}
        \caption{Dilation difference}
    \end{subfigure}

    \caption{\textbf{Dilation examples for two randomly selected slices:} The first column shows raw data without a mask. The second column displays the ground truth from the SAROS dataset. The third column is our original prediction, which excludes skin for SAT. The fourth column demonstrates our prediction after dilation, and the last column illustrates the area added by dilation (in yellow). The yellow arrow highlights the difference introduced by the dilation and blue mask shows SAT.}
    \label{fig:appendix:dilation}
\end{figure*}

\section{Body composition metrics relationship} \label{sec:Body composition metrics relationship}
Pearson correlation coefficient (r) is the most common method of measuring a linear correlation between two variables, with its definition shown in Equation \eqref{r}, where $Cov(X, Y)$ represents the covariance of $X$ and $Y$, and $\sigma_X$, $\sigma_Y$ are standard deviations of $X$ and $Y$ respectively. This section demonstrates the correlation between four selected body composition metrics: muscle density, VAT/SAT ratio, muscle area, and SMI based both on Pearson correlation coefficient and scatter plots. The 2D and 3D measurements of the same metrics are highly correlated, as shown in Figure \ref{fig:self-correlation}, with all three Pearson correlation coefficients exceeding 0.96. In this experiment, we utilize only the body composition metrics measured in 2D settings (at the L3 level). The results are shown in Figure \ref{fig:correlation}, as we can observe, except for the relationship between muscle area and SMI (with $r$ equals to 0.94), all other pairs of metrics show insignificant or no linear correlation, with $r$ having an absolute value smaller than 0.2. Scatter plots also do not show a clear monotonic relationship.

\begin{equation} \label{r}
    r = \frac{\text{Cov}(X, Y)}{\sigma_X \sigma_Y}
\end{equation}

\begin{figure*}[t]
    \centering
    \captionsetup{font=footnotesize} % Reduce caption font size

    % ----- Subfigure 1 -----
    \begin{subfigure}[b]{0.31\textwidth} % Slightly wider to avoid wrapping
        \centering
        \includegraphics[width=\textwidth]{fig/self_correlation/muscle_density.png}
    \end{subfigure}
    % ----- Subfigure 2 -----
    \begin{subfigure}[b]{0.31\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/self_correlation/VAT_SAT_ratio.png}
    \end{subfigure}
    % ----- Subfigure 3 -----
    \begin{subfigure}[b]{0.31\textwidth}
        \centering
        \includegraphics[width=\textwidth]{fig/self_correlation/muscle_area.png}
    \end{subfigure}
    \caption{Scatter plots illustrating the relationships between 2D and 3D settings of body composition metrics: Muscle density 2D (HU), VAT/SAT ratio 2D, and Muscle area 2D (cm²). Each subplot represents a specific metric pair, displaying the distribution of data points alongside the calculated Pearson correlation coefficient (Pearson r) to quantify the strength and direction of their linear relationship.}
    \label{fig:self-correlation}
\end{figure*}


\begin{figure*}[htbp]
    \centering
    % Row 1
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\linewidth]{fig/correlation/muscle_density_vs_VAT_SAT_ratio.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\linewidth]{fig/correlation/muscle_density_muscle_area.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\linewidth]{fig/correlation/muscle_density_vs_SMI.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\linewidth]{fig/correlation/VAT_SAT_ratio_vs_muscle_area.png}
    \end{subfigure}
    \vspace{1mm}
    % Row 2
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\linewidth]{fig/correlation/VAT_SAT_ratio_vs_SMI.png}
    \end{subfigure}
    \vspace{1mm}
    \begin{subfigure}[b]{0.49\textwidth}
        \includegraphics[width=\linewidth]{fig/correlation/muscle_area_vs_SMI.png}
    \end{subfigure}
    \vspace{1mm}
    \caption{Scatter plots illustrating the relationships between pairs of body composition metrics: Muscle density 2D (HU), VAT/SAT ratio 2D, Muscle area 2D (cm²), and Skeletal Muscle Index (cm/m²). Each subplot represents a specific metric pair, displaying the distribution of data points alongside the calculated Pearson correlation coefficient (Pearson r) to quantify the strength and direction of their linear relationship.}
    \label{fig:correlation}
\end{figure*}

\end{document}
