\subsection{Drug Discovery and Development}

\paragraph{Subtopic:} 
Integrating Foundation Models: How can foundation models streamline specific stages like target identification, lead optimization, and clinical candidate selection, and what measurable impacts have been observed?  Mitigating Risks in AI Integration: What strategies address risks like data bias, overfitting, or lack of interpretability when integrating AI into drug discovery pipelines? Discovering Novel Biological Knowledge: How have ML models uncovered new biological mechanisms or pathways, and what are the key challenges in simulating complex processes like protein-ligand interactions? Enhancing Chemical Design and Synthesis: How can AI improve understanding of computational compound generation/modification and synthesis planning to accelerate the design of novel, drug-like molecules?

\paragraph{Chairs: }
\textit{Michael Craig, Zuheng Xu, and Geoffrey Woollard}

\paragraph{Background:}

AI and ML continue to rapidly advancing in drug discovery and development, accompanied by growing biological and chemical data and improved computational power. AI/ML is increasingly present in science broadly considered in terms of perspectives, language, tools, and community (both research and business). Computational biology, broadly considered, has become increasingly integrated with AI/ML communities in statistical inference, simulation, high performance computing, numerics, and data science \citep{Cranmer2020,Lavin2021,Hennig2022,Wang2024}. In recent years, there has been increased effort to integrate AI/ML methods into key stages of the drug discovery pipeline, including target identification \citep{hassanzadeh2016deeperbind}, lead compound discovery \citep{macedo2024medgan}, lead optimization \cite{pmlr-v108-korovina20a}, and candidate selection, and thus demonstrate tangible impacts. Furthermore, there is an overly-familiar claim that AI/ML perspectives will help the discovery of novel biological mechanisms and pathways, shedding light on complex interactions such as protein-ligand binding \citep{rezaei2020deep} and protein design \citep{TheNobelCommitteeforChemistry2024}.

The activity at NeurIPS 2024 showed attempts to develop foundation models (meaning large models trained on large amounts of diverse data and prioritizing generality), for target identification, lead optimization, protein design, and clinical candidate selection. Drug discovery and development industry commentators have summarized the recent 10-15 years of history, distinguishing trends towards highly relevant data generation leveraged by generative AI and iterative feedback loops. Generative AI and foundation models at the molecular and tissue scales are openly available \citep{Takeda2023,Ingraham2023_chroma,Wohlwend2024_boltz,Chen2024,Hayes2024,Lu2024,M.Bran2024,batatia2023foundation}.
%\footnote{\url{https://github.com/IBM/materials},\url{https://github.com/generatebio/chroma},\url{https://github.com/jwohlwend/boltz},\url{https://huggingface.co/proteinglm},\url{https://github.com/evolutionaryscale/esm},\url{https://github.com/ur-whitelab/chemcrow-public},\url{https://github.com/ACEsuit/mace-mp}}. 
The question ``how to systematically leverage large language models and AI agents to coordinate and automate the action of agents in drug discovery and development?'' is not only raised, but is has received ambitious answers that that signal towards fuller automated decision making across the entire drug discovery and development pipeline \citep{swanson2024virtual}. However, integrating AI/ML into this hybrid scientific-strategic-business-regulatory process also presents challenges, such as data bias, overfitting, the multiple scales (length, time, energy) of measurements, the complex relationships between experimental and observational modalities, complex relationships between experimental modalities, lack of interpretability, and translation into actionable decisions. These challenges can be mitigated through strategies such as diverse data curation, simulation-based inference, explainable AI, and interdisciplinary validation \citep{Administration2023}.
%\footnote{Also see links at \url{https://www.fda.gov/about-fda/center-drug-evaluation-and-research-cder/artificial-intelligence-drug-development}}. 

While some voices emphasize a certain humility in the face of biological complexity \citep{Castaldo2024}, and a more discerning financing climate, others continue a positive narrative \citep{Chitnis2024}. One report, by authors from a Strategic Management Consulting firm ``that works with the worldâ€™s leading biopharmaceutical companies'', systematically documents successes \citep{KPJayatunga2024} through assessing the number of clinical assets in company pipelines. The authors looked into over 100 companies through an online databases that document the history of clinical assets, and followed up manually with finer grained desk research through press releases, company websites, \url{clinicaltrials.gov} and analyst reports. They find a substantially increased Phase I success rate over historical industry averages, and a comparable Phase II success rate.

Such topics formed the foundation of our roundtable discussion.

\paragraph{Discussion:}
We first summarize the professional biographies of the participants, and then summarize the discussion in sections on streamlining drug discovery, mitigating AI/ML integration risks, discovering novel biological knowledge, enhancing chemical design and synthesis, bridging science and strategy, before finishing with future directions.

Participants represented diverse expertise, from clinical areas like computational phenotyping, single cell RNA sequencing, digital pathology, functional molecular and clinical trials data integration, as well as the sub-nano scale of biomolecular simulation and structural biology. A group of about twelve were present during the full hour, and most had doctorates in the area of statistics, computer science, or some form of computational science and were mostly postdoctoral researchers or industry scientists; there were also a few doctoral students and one undergraduate.

\subsubsection{Streamlining Drug Discovery}  
The roundtable discussion touched on leveraging self-supervised learning, large-scale imaging, and high-dimensional phenotype modeling to decode biological systems and automate workflows. Participants opined that while measurable impacts include faster hit identification and integration of multimodal data, challenges remain in bridging biological complexity and technical scalability. Participants admitted current bottlenecks of data acquisition and a lack of experimental automation, but the discussion focused on the informal risk assessment and communication between ``computational" researchers and ``experimentalists" and associated communication challenges. Participants had deep expertise and technical focus, and perhaps because of this, the larger picture of how it all comes together in a successful pipeline remained largely elusive.   

\subsubsection{Mitigating AI Integration Risks}  
Participants noted risks like data bias, lack of interpretability, and slow validation cycles in drug discovery. Strategies to mitigate these include cross-disciplinary training to improve communication between machine learning practitioners and domain scientists, company culture surrounding decision making and risk aversion, using robust evaluation metrics, and adopting active learning approaches for iterative validation. The slow feedback loops in biological experiments amplify the difficulty, highlighting the need for well-grounded models and transparency in decision-making, and active learning perspectives like Bayesian optimization, where the cost of obtaining new data is itself modeled to optimize exploit/explore tradeoffs \citep{garnett_bayesoptbook_2022}.

\subsubsection{Discovering Novel Biological Knowledge} 
Participants commented on to what extent AI/ML has revealed new mechanisms, such as gene-gene or gene-drug interactions, from image and transcriptomics data. Digital twin models were highlighted as promising for simulating patient-level drug responses, albeit limited by challenges in verification and cross-scale integration. Simulating complex biophysical mechanisms at multiple scales, like how protein-ligand dynamics gives rise to cellular states, seems computationally intractable due to chaotic multi-scale interactions familiar to computational fluid dynamics. Despite this, this area is receiving attention, for example from the Chan Zuckerberg Institute for Advanced Biological Imaging's cryogenic electron tomography community data challenge \citep{Harrington2024} on a highly visible online platform. Participants discussed the pros and cons for hybrid approaches combining traditional physics-based methods with data-driven simulations and observations to advance understanding and do inference, and noted the NeurIPS 2024 workshop on Data-driven and Differentiable Simulations, Surrogates, and Solvers (\url{https://d3s3workshop.github.io/}).

\subsubsection{Enhancing Chemical Design and Synthesis}  
Participants offered anecdotes on to what extent AI/ML aids in computational compound generation and synthesis planning by predicting properties like solubility and toxicity, and to what extent this informs decision making criteria of domain expert experimentalists like medicinal chemists. Integration of simulation frameworks, akin to those in aerodynamics or chip design, was suggested to enable more real-time decision around drug prototypes. Furthermore, some natural language information like synthesis or even economic/fiscal aspects of regulatory reimbursement could perhaps benefit from large language model approaches.

\subsubsection{Bridging Science and Strategy}
The discussion emphasized fostering collaboration across disciplines and industries. Much of the discussion was taken up by anecdotal knowledge of how teams work together in different settings (startups, contract research organizations, clinical research centers, academic biomedical research institutes, internal consulting teams in large pharmaceutical companies). Participants appreciated learning from each other in a way that is not formally taught but typically gained through job experience and networking at meetings such as this one.

\subsubsection{Future Outlook} 
Participants suggested truly AI-native information-first teams either as an independent startup, or within larger pharmaceutical companies could accelerate progress with their agility and organization. Addressing cultural inertia and building credibility in AI/ML predictions were identified as crucial for broader adoption. Who will be the first to demonstrate never before seen success, which will drive cultural change within scientific teams?

