% What is homologous genes and non-homologous genes?
% To validate whether the gene embeddings encoded by GeneCompass retained homology information, we randomly selected a total of 2000 B cells from human and mouse corpora, and compared the cosine similarity between the embeddings of homologous and non-homologous genes in different species.

% gene homology across species

% Do homologous genes have similar Genotypes? (To make it biologically reasonable)
% Generative quality: How well the model can generate high-quality images? (FID or other metrics)
% Generative consistency: Could the model generate images that match to the DNA? 
% Generative diversity: Could the model generate novel samples beyond the training set?
% How to balance the diversity and consistency? (Ajaust the guidance weight.)
% How to balance the consistency and quality? (Why we propose the new method)

\section{Results}
\label{sec:experiments}
In this section, we conduct extensive experiments to answer the following questions: 
\begin{itemize}[leftmargin=5.5mm]

  \item \textbf{Performance (Q1):} 
  Could the model generate phenotypic images that match the DNA?
  % CLIPScore，SR，IIS
  % FID
  % clustering results
  \item \textbf{Model Analysis (Q2):} What is the impact of each module on the model's overall performance?
  \item \textbf{Generalization (Q3):} Could our proposed method generalize across unseen species?
\end{itemize}




\begin{table*}[t]
\centering
\caption{Summary of CLIBDScore and success rate at different thresholds of ground-truth images, as well as images generated by our model and other non-diffusion and diffusion-based baselines. Our method outperforms the baselines across all evaluation metrics.}
\label{table:Q1}
\begin{adjustbox}{width=1\linewidth}
    \begin{tabular}{c|l|c|c|cc|cc|cc|cc}
    \toprule
    % \hline
    \textbf{Metric} & \textbf{Rank} & \textbf{GT} & \textbf{Random} &  \multicolumn{2}{c|}{\textbf{DF-GAN}} & \multicolumn{2}{c|}{\textbf{Stable Diffusion}} & \multicolumn{2}{c|}{\textbf{ControlNet}} &  \multicolumn{2}{c}{\chl \textbf{G2PDiffusion}} \\
     & & & & Abs. & Rel. & Abs. & Rel. & Abs. & Rel. & \chl Abs. & \chl Rel. \\
    % \hline
    \toprule
    \multirow{6}{*}{CLIBDScore} & Top-1 & \multirow{6}{*}{0.512} & \multirow{6}{*}{0.005} & 0.054 & 0.106 & 0.100 & 0.195 & 0.107 & 0.209 & \chl \textbf{0.182} & \chl \textbf{0.356} \\
    & Top-5 &&& 0.154 & 0.301 & 0.219 & 0.428 & 0.228 & 0.445 & \chl \textbf{0.302} &  \chl \textbf{0.590} \\
    & Top-10 &&& 0.181 & 0.354 & 0.254 & 0.496 & 0.265 & 0.518 & \chl \textbf{0.358} & \chl \textbf{0.700} \\
    & Top-20 &&& 0.224 & 0.438 & 0.292 & 0.570 & 0.307 & 0.600 & \chl \textbf{0.397} & \chl \textbf{0.776} \\
    & Top-50 &&& 0.276 & 0.539 & 0.338 & 0.660 & 0.351 & 0.686 & \chl \textbf{0.455} & \chl \textbf{0.889} \\
    & Top-100 &&& 0.314 & 0.614 & 0.367 & 0.718 & 0.384 & 0.750 &\chl  \textbf{0.480} & \chl \textbf{0.938} \\
    % \hline
    \toprule
    \multirow{6}{*}{Success Rate} & Top-1 & \multirow{6}{*}{96.4\%} & \multirow{6}{*}{4.4\%} & 5.6\% & 5.8\% & 11.5\% & 11.9\% & 12.4\% & 12.9\% &\chl \textbf{31.7\%} & \chl \textbf{32.8\%} \\
    & Top-5 &&& 18.7\% & 19.4\% & 36.6\% & 38.0\% & 39.1\% & 40.6\% & \chl \textbf{65.8\%} & \chl \textbf{68.3\%} \\
    & Top-10 &&& 32.1\% & 33.3\% & 43.5\% & 45.1\% & 47.0\% & 48.7\% & \chl \textbf{81.1\%} & \chl \textbf{84.1\%} \\
    & Top-20 &&& 40.9\% & 42.4\% & 55.7\% & 57.7\% & 57.8\% & 60.0\% & \chl \textbf{90.4\%} & \chl \textbf{93.8\%} \\
    & Top-50 &&& 48.1\% & 49.9\% & 68.7\% & 71.3\% & 70.7\% & 73.4\% & \chl \textbf{93.0\%} & \chl \textbf{96.5\%} \\
    & Top-100 &&& 52.6\% & 54.6\% & 74.8\% & 77.6\% & 77.0\% & 79.8\% & \chl \textbf{94.0\%} & \chl \textbf{97.5\%} \\
    % \toprule
    % \hline
    
    \bottomrule
    % \hline
    \end{tabular}
\end{adjustbox}
\vspace{-3mm}
\end{table*}


\vspace{-1mm}




\subsection{Performance (Q1)}

\begin{figure}[h]
    \centering
    \vspace{-1em}
    \includegraphics[width=1\linewidth]{img/contra-image.pdf}
    \vspace{-1.5em}
    \caption{Generative results. All methods can generate visually reasonable images with different the DNA-image consistency.}
    \vspace{-2em}
    \label{fig:Q1_Qualitative_results}
\end{figure}

\paragraph{Qualitative Results.} 
Fig.~\ref{fig:Q1_Qualitative_results} shows the qualitative results of various methods. Our method, G2PDiffusion, stands out by producing the most resonable phenotype predictions from DNA inputs, thanks to the carefully designed evolutionary conditioner and dynamic aligner. DF-GAN, on the other hand, struggles to generate high-quality images and often fails to capture the precise characteristics of the ground truth phenotypes. Although Stable Diffusion and ControlNet could generate visually appealing images, they lack the ability to align these images closely with the true phenotypes. 



\begin{table}[H]
\centering
\caption{PES scores comparison across DF-GAN, Stable Diffusion, ControlNet, and our proposed G2PDiffusion. G2PDiffusion achieves the highest PES scores at all evaluated ranks.}
\label{tab:pes_comparison}
\begin{adjustbox}{width=1\linewidth}
    \begin{tabular}{l|c|c|c|c}
    \toprule
    Rank & DF-GAN & Stable Diffusion & ControlNet &  \chl G2PDiffusion \\
    \midrule
    Top-1   & 0.021 & 0.062 & 0.061 & \chl \textbf{0.152} \\
    Top-5   & 0.134 & 0.207 & 0.212 & \chl \textbf{0.291} \\
    Top-10  & 0.167 & 0.240 & 0.254 & \chl \textbf{0.346} \\
    Top-20  & 0.216 & 0.288 & 0.299 &\chl  \textbf{0.405} \\
    Top-50  & 0.276 & 0.349 & 0.359 & \chl \textbf{0.478} \\
    Top-100 & 0.301 & 0.389 & 0.403 & \chl \textbf{0.511} \\
    \bottomrule
    \end{tabular}
\end{adjustbox}
\end{table}

% \vspace{-2em}
\paragraph{Quantitative results.}
For quantitative evaluation, we consider the three metics: CLIBDScore, Success Rate, and Phenotype Embedding Similarity (as shown in Table \ref{table:Q1} and Table \ref{tab:pes_comparison}).
In addition to reporting absolute scores, we also calculate relative scores by dividing each score by the ground truth score (shown as Abs. and Rel. in the table). We summary that:
(a) Compared to the random baseline, all deep learning methods demonstrate non-trivial potential in deciphering phenotypes from genotype and environment. (b) Diffusion models consistently outperform DF-GAN, as their multi-step generation process progressively refines the generated phenotypes, making it easier to capture the complex genotype-phenotype relationships. (c) The proposed G2PDiffusion demonstrates significantly higher performance than other models across all metrics. For example, in the Top-5 success rate, our model achieves a score of 65.8\%, notably outperforming Stable Diffusion (36.6\%) and ControlNet (39.1\%). Furthermore, our method shows remarkable improvements with a Top-10 success rate of 81.1\% and a Top-100 rate of 94.0\%, indicating strong alignment with ground truth images. These results highlight the effectiveness of our approach in accurately generating phenotype images from DNA sequences. (d) The compared PES scores show that G2PDiffusion generates morphological phenotypes with higher biological relevance in the phenotype embedding space, demonstrating that incorporating genotype-environment interaction and evolutionary constraints helps align generated images with real phenotypic variation.
% With the dynamic aligner, G2PDiffusion outperforms the consistency score of ground truth data, which may provide insights for biologist to mining  genotype-phenotype relationships.

\begin{table*}[t]
\caption{Summary of CLIBDScore and success rate evalutions at different thresholds on the unseen set. }
\label{table:Q2_unseen_compare}
\centering
{\resizebox{\textwidth}{!}{
\begin{tabular}{l|cc|cc|cc|cc|cc|cc}
\toprule
% \hline
\textbf{Method} & \multicolumn{2}{c|}{\textbf{Top-1}} & \multicolumn{2}{c|}{\textbf{Top-5}} & \multicolumn{2}{c|}{\textbf{Top-10}} & \multicolumn{2}{c|}{\textbf{Top-20}}   & \multicolumn{2}{c|}{\textbf{Top-50}}  & \multicolumn{2}{c}{\textbf{Top-100}} \\
& Score. & Acc. & Score. & Acc. & Score. & Acc. & Score. & Acc. & Score. & Acc. & Score. & Acc.\\
% \hline
\toprule
DF-GAN & 0.045 & 4.2\% & 0.110 & 12.5\% & 0.130 & 18.3\% & 0.155 & 22.8\% & 0.180 & 33.7\% & 0.190 & 38.4\% \\
Stable Diffusion & 0.068 & 6.4\% & 0.162 & 19.3\% & 0.185 & 28.7\% & 0.210 & 37.5\% & 0.235 & 48.2\% & 0.250 & 53.1\% \\
ControlNet & 0.072 & 7.1\% & 0.155 & 18.4\% & 0.180 & 29.2\% & 0.205 & 40.3\% & 0.235 & 51.7\% & 0.250 & 56.3\% \\
\chl Ours &\chl  \textbf{0.081} &\chl  \textbf{8.8\%} &\chl  \textbf{0.184} &\chl  \textbf{25.0\%} &\chl  \textbf{0.228} &\chl  \textbf{41.4\%} &\chl  \textbf{0.263} &\chl  \textbf{55.1\%} &\chl  \textbf{0.313} &\chl  \textbf{75.5\%} &\chl  \textbf{0.340} &\chl  \textbf{80.3\%} \\
% \hline
\bottomrule
\end{tabular}
}}
\vspace{-1em}
\end{table*}

\subsection{Model Analysis (Q2)}
% \paragraph{Effectiveness of Environment-enhanced DNA Conditioner}
\paragraph{Effects of Environment-aware MSA Conditioner and Dynamic Alignment.}
We investigate the impact of environment-aware MSA conditioner and dynamic alignment sampling mechanism, as shown in Table \ref{tab:ablation}. In particular, we replace the environment-aware MSA encoder with the simplest DNABERT\cite{ji2021dnabert} and remove the dynamic alignment sampling mechanism to construct our baseline.

\begin{table}[h]
\centering
\caption{Ablation studies of environment-aware MSA conditioner and dynamic alignment sampling mechanism.}
\label{tab:ablation}
\begin{adjustbox}{width=1\linewidth}
    \begin{tabular}{l|cc|cc|cc}
    \toprule
    \multicolumn{1}{c|}{Methood} & \multicolumn{2}{c|}{CLIBDScore} & \multicolumn{2}{c|}{Success Rate} & \multicolumn{2}{c}{PES} \\
    % \cmidrule(lr){2-3} \cmidrule(lr){4-5} \cmidrule(lr){6-7}
    & Top-1 & Top-5 & Top-1 & Top-5 & Top-1 & Top-5 \\
    \midrule
    Baseline & 0.100 & 0.219 & 11.50\% & 26.60\% & 0.062 & 0.187 \\
    + Conditioner & 0.125 & 0.235 & 16.73\% & 28.21\% & 0.098 & 0.254 \\
    + Alignment & 0.167 & 0.289 & 27.14\% & 51.24\% & 0.137 & 0.268 \\
    + Both & 0.182 & 0.302 & 31.70\% & 65.80\% & 0.152 & 0.291 \\
    \bottomrule
    \end{tabular}
\end{adjustbox}
\vspace{-3mm}
\end{table}

The ablation results show that both the environment-aware MSA conditioner and the dynamic alignment sampling mechanism contribute to model performance. We summary that incorporating evolutionary context and environment-aware sequence representations helps the model capture biologically meaningful genotype-phenotype relationships. Meanwhile, the dynamic alignment sampling mechanism further enhances the biological relevance of generated phenotypes to the DNA sequences.





\paragraph{Effects of Evolutional-Alignments Retrieval}

We investigate the influence of the retrieved MSA for G2PDiffusion through an ablation study on variable $m$, which denotes the number of retrieved sequence alignments.
From the results in Table \ref{tab:ablation_m}, we observe that: 
(a) increasing $m$ from 0 to 1 leads to significant improvements across all evaluation metrics, indicating that incorporating homologous sequence alignments provides evolutionary context, which enhances the quality of phenotype generation; (b) the best performance is achieved when $m$ is set to 1 or 2, where the retrieved sequences exhibit high similarity to the target, enabling effective integration of conserved evolutionary signals into the generation process; (c) however, further increasing $m$ introduces more distant sequences with lower relevance, which inevitably introduces noise and reduces the overall generation quality. 


\begin{table}[h]
\centering
\caption{The effect of hyper-parameter $m$. The top 2 results are highlighted with \textbf{bold text} and \underline{underlined text}, respectively.}
\label{tab:ablation_m}
\begin{adjustbox}{width=1\linewidth}
    \begin{tabular}{l|cc|cc|cc}
    \toprule
    \multicolumn{1}{c|}{Methood} & \multicolumn{2}{c|}{CLIBDScore} & \multicolumn{2}{c|}{Success Rate} & \multicolumn{2}{c}{PES} \\
    % \cmidrule(lr){2-3} \cmidrule(lr){4-5} \cmidrule(lr){6-7}
    & Top-1 & Top-5 & Top-1 & Top-5 & Top-1 & Top-5 \\
    \midrule
    $m$=0 & 0.178 & 0.284 & 27.17\% & 53.10\% & 0.151 & 0.284 \\
    $m$=1 & \textbf{0.193} & \underline{0.299} & \textbf{36.23\%} & \textbf{65.80\%} & \underline{0.143} & \textbf{0.293} \\
    $m$=2 & \underline{0.182} & \textbf{0.302} & \underline{31.70}\% & \textbf{65.80\%} & \textbf{0.152} & \underline{0.291} \\
    $m$=3 & 0.166 & 0.285 & 29.90\% & 58.87\% & 0.128 & 0.271 \\
    $m$=4 & 0.176 & 0.296 & 29.40\% & 62.34\% & 0.142 & 0.280 \\
    \bottomrule
    \end{tabular}
\end{adjustbox}
\end{table}




\subsection{Generalization to Unseen Species (Q3)}

% In previous section, we conduct experiments on \textbf{closed-world setting} of BIOSCAN-5M dataset, in which all species have been established scientific names. 




To investigate the generalization capability of our method, we evaluate its performance on unseen species in the dataset, called the \textbf{open-world scenario}. In this case, species do not have scientific names in the dataset.

\begin{figure}[!htb]
    \centering
    % \vspace{-1em}
    \includegraphics[width=1\linewidth]{img/unseen.pdf}
    % \vspace{-1.5em}
    \caption{Generative results on unseen species.}
    \vspace{-1.5em}
    \label{fig:unseen}
\end{figure}

Results in Table~\ref{table:Q2_unseen_compare} show that our model maintains high performance on these unseen species, though not as high as on the seen species. 
We show some prediction results for unseen species in Fig.~\ref{fig:unseen}, where most of these predictions can closely match the ground truth phenotypes (the first  three rows). It is an interesting that generative models can produce different view's images for the same species given the same genotype and environment conditions. There are also some predictions that retain the essential traits, although not perfectly match the ground truth. As shown in the last two rows, the model retain key features such as the insect's body color, shape patterns and the overall wing structure.
These findings show the potential of our approach to explore genotype-phenotype relationships, uncover species-specific traits, even in challenging or under-explored species.



% where the model effectively generates the insect's body and limb colors as well as the shape and details of the wings.



% \subsection{Evaluating Phenotype Space Consistency (Q4)}
% \paragraph{Validating Species Distances in the Embedding Space}

% \paragraph{Assessing the Clustering Quality of the Embedding Space Using t-SNE Plots}


% \subsection{Variation Analysis (Q3)}


% \paragraph{Mutation Effect.} As shown in Fig.~\ref{fig:mutation}, we randomly mutate the DNA sequence by 10\% to 50\% and alter the sample's latitude to examine how both genetic and environmental factors influence the phenotype. We find that the model is sensitive to DNA mutations; specifically, a higher mutation rate results in worse visual quality, indicating that the model can learn the critical dependency between the DNA sequence and the phenotype. Interestingly, when we change the latitude value, the closer to the equator, the smaller the specie sizes are. This phenomenon meet the Bergmann's Rule \cite{mcqueen2022thermal}, which states that within a species, individuals living in colder climates tend to have larger body sizes compared to those in warmer climates. This trend is thought to be an adaptation to temperature regulation: larger bodies have a smaller surface area-to-volume ratio, which helps to conserve heat, making it easier for animals to survive in colder environments. In contrast, smaller body sizes are more common in warmer climates, where heat dissipation is more important for survival.

% \begin{figure}[!htb]
%     \centering
%     \vspace{-0.5em}
%     \includegraphics[width=1\linewidth]{img/mutate.pdf}
%     % \vspace{-2em}
%     \caption{Mutation effects of the genotype and environment.}
%     \vspace{-3.5mm}
%     \label{fig:mutation}
% \end{figure}

% % 相似同源基因 是否表现型相同？生成模型的效果？

% % 这里展示图片, 相似的同源基因序列会生成相似的图片么? 结合生物里面分析同源序列的方法, 根据序列相似度为90%, 50%, 20%, 生成样本, 最好能找到关于同源基因的文献: 物种A1和A2是同源的，他们生成的结果相似. 




% % \subsection{Environmental Variation Analysis (Q4)}



% % \subsection{Ablation Study (Q3)}
% % To further examine the impact of individual components in our proposed method, we conducted an ablation study focusing on the alignment guidance and retrieval augmentation module.

% % \paragraph{Alignment Guidance}
% % % 这里说明为什么用planc?-->直接用clip guidance会损害图像质量. 

% % \paragraph{Retrieval Augmentation}
% % % 对不同类的物种的success rate进行计算; 统计不同类物种的训练样本数量, unbalanced data
% % % 绘制 不同属的success_rate


% % \subsection{Genotype-Phenotype Discovery (Q4)}
% % \paragraph{Homologous Analysis}




% % 也可以直接通过目科属种, 观察同属于一个类的DNA是否能生成形状上相似/相关的图像. 可以画一个进化的树状图. 
