\section{Evaluation Metrics}
\label{sec:eval}
Comprehensive evaluation is essential for applying \proteinllms, which are assessed on tasks like structure prediction, function prediction, and sequence generation. Appendix~\ref{sec:eval} provides detailed descriptions of structure and function prediction metrics, as well as sequence generation metrics for generative \proteinllms.

\subsection{Structure Prediction Metrics}
\texttt{Root Mean Square Deviation (RMSD)} measures the distance between predicted and actual atomic coordinates, with lower values indicating better accuracy~\cite{li2013difficulty}. \texttt{Global Distance Test (GDT-TS)} calculates the percentage of alpha-carbon atoms within 1, 2, 4, and 8 \r{A} thresholds, reflecting structural similarity~\cite{zemla2003lga}. \texttt{Template Modeling (TM) Score} evaluates global structural similarity (scores between 0 and 1) via 
\begin{gather}
\text{TM}=\max \left[ \frac{1}{L_{\text{tgt}}} \sum_{i}^{L_{\text{com}}} \frac{1}{1+\left( \frac{d_{i}}{\scriptscriptstyle d_{0}(L_{\text{tgt}})} \right)^{2}} \right], \\
d_0(L_{\text{tgt}}) = 1.24 \sqrt[3]{L_{\text{tgt}} - 15} - 1.8.
\end{gather}
\texttt{Local Distance Difference Test (lDDT)} quantifies local accuracy by comparing interatomic distances~\cite{mariani2013lddt}, and \texttt{Predicted Local Distance Difference Test (pLDDT)} provides per-residue confidence scores (0â€“100) without a reference structure, as used in AlphaFold~\cite{guo2022alphafold2,jumper2021highly}.

\subsection{Function Prediction Metrics}

Protein function prediction determines biological roles, including biomolecular interactions~\cite{radivojac2013large}. Machine learning metrics include classification measures (precision, recall, F-1 score, accuracy, AUC) and generative metrics such as BLEU~\cite{papineni2002bleu}, ROUGE-L~\cite{lin2004rouge}, and METEOR~\cite{banerjee2005meteor}. These evaluation methods offer quantitative benchmarks crucial for model validation and biological inference.

\texttt{Subcellular Localization} predicts proteins' cellular positions to infer functions~\cite{briesemeister2010yloc,holm2020dali}. \texttt{Homology Detection} identifies evolutionary relationships using sequence alignment methods like BLAST~\cite{altschul1990basic} or deep learning approaches such as TM-vec~\cite{hamamsy2024protein}. \texttt{Stability} and \texttt{Solubility} assessments evaluate whether a protein can function effectively in its environment~\cite{cheng2006prediction,hebditch2017protein}, while \texttt{Mutation Effect Prediction} gauges the impact of amino acid changes on protein properties~\cite{mansoor2022accurate}. These integrative metrics underpin the development of robust protein prediction systems and support advancements in drug design and molecular biology.
