\section{Evaluation Metrics}
\label{sec:app_eval}
Comprehensive and accurate evaluation is essential for understanding and applying \proteinllms. Currently, these models are commonly assessed on tasks such as structure prediction, function prediction, and sequence generation.

\subsection{Structure Prediction Metrics}
Structure prediction evaluates how accurately a model predicts a proteinâ€™s three-dimensional structure from its sequence~\cite{kuhlman2019advances}. Common metrics include:

\smallskip\noindent \textit{\textbf{Root Mean Square Deviation (RMSD)}} measures the distance between the predicted and actual atomic coordinates. Lower RMSD indicates higher structural accuracy~\cite{li2013difficulty}.


\smallskip\noindent \textit{\textbf{Global Distance Test (GDT-TS)}} calculates the percentage of alpha-carbon atoms within thresholds (1, 2, 4, and 8 \r{A}) of the reference structure after iterative superimposition~\cite{zemla2003lga}.

GDT-TS usually uses thresholds of 1, 2, 4, and 8 \r{A}. The higher the GDT-TS score, the closer the predicted structure is to the reference structure.


\smallskip\noindent \textit{\textbf{Template Modeling (TM) Score}} evaluates the global structural similarity of proteins with values ranging from 0 to 1~\cite{zhang2004scoring}. 
\begin{gather}
\text{TM}=\max \left[ \frac{1}{L_{\text{tgt}}} \sum_{i}^{L_{\text{com}}} \frac{1}{1+\left( \frac{d_{i}}{\scriptscriptstyle d_{0}(L_{\text{tgt}})} \right)^{2}} \right], \\
d_0(L_{\text{tgt}}) = 1.24 \sqrt[3]{L_{\text{tgt}} - 15} - 1.8.
\end{gather}
Here, $L_{\text{tgt}}$ is the length of the target protein amino acid sequence. $L_{\text{com}}$ is the number of residues in the template and target structures. $d_i$ represents the distance between the $i$-th residue pair in the template structure and the target structure. Higher scores indicate closer similarity. 

\smallskip\noindent \textit{\textbf{lDDT}}, Local Distance Difference Test, evaluates the local accuracy of protein structure prediction by comparing distances between atom pairs in the predicted structures and those in the reference structures~\cite{mariani2013lddt}.

A distance is considered preserved if it falls within a specified threshold. lDDT is calculated as the proportion of preserved distances, with higher values indicating better local accuracy. 

\smallskip\noindent \textit{\textbf{pLDDT}}, Predicted Local Distance Difference Test, is a per-residue measure of local confidence~\cite{guo2022alphafold2}. pLDDT evaluates the local quality of the predicted structure without a reference structure. Its computation usually relies on models such as AlphaFold \cite{jumper2021highly}, which learns patterns from large-scale protein data. Scores range from 0 to 100, with higher scores indicating greater confidence and more accurate predictions.

\subsection{Function Prediction Metrics}
Protein function prediction aims to determine biological roles, including interactions with other biomolecules \cite{radivojac2013large}. The evaluation methods involve machine learning performance metrics and biomedical relevance validation.

Machine learning evaluation metrics can be categorized into classification task metrics and generative task metrics. For classification tasks, such as protein classification and interaction prediction, standard metrics can be adopted, such as precision, recall, F-1 scores, accuracy, and area under the curve (AUC). For generative tasks, such as question answering, evaluation is performed by measuring the alignment between the LLM's output and the ground truth using metrics such as BLEU \cite{papineni2002bleu}, ROUGE-L \cite{lin2004rouge}, and METEOR \cite{banerjee2005meteor}.


In addition to machine learning metrics, there are also biometric-related evaluation metrics:

\smallskip\noindent \textit{\textbf{Subcellular Localization}} refers to the specific location of proteins within a cell \cite{briesemeister2010yloc}. The location of a protein is closely related to the function it performs, so by predicting the subcellular localization of a protein, it is possible to speculate on the biological function it may have \cite{holm2020dali}.

\smallskip\noindent \textit{\textbf{Homology Detection}} aims to identify proteins that share an evolutionary relationship (homologous) with the target protein, usually reflected in similarities in  sequences, structure, and functions. Traditional methods such as BLAST \cite{altschul1990basic} 
perform sequence alignment to identify homologs by comparing the query sequence against a database.

Recent deep learning approaches such as TM-vec \cite{hamamsy2024protein} focus on structural similarity and generate vector representations of proteins.

\smallskip\noindent \textit{\textbf{Stability}} of the protein is critical for many applications, such as drug development. Predicting the stability of a protein can help determine whether the protein can perform its function efficiently in the cellular environment \cite{cheng2006prediction}.

\smallskip\noindent \textit{\textbf{Solubility}} reflects the solubility characteristics of a protein in a particular solvent. Predictions of solubility can help to understand whether a protein can exist and function properly within a cell \cite{hebditch2017protein}.

\smallskip\noindent \textit{\textbf{Mutation Effect Prediction}} of proteins refers to the assessment of the impact on various properties, structures, and functions of proteins when their amino acid sequences are changed \cite{mansoor2022accurate}. Commonly used methods include molecular dynamics-based methods, deep learning-based prediction models, and structural comparison methods.

\subsection{Sequence Generation Metrics} 
Protein sequence generation is the process of creating new protein sequences using specific methods, models, or algorithms \cite{anand2022protein}. Common evaluation methods include:

\smallskip\noindent \textit{\textbf{Perplexity (PPL)}} can be used to measure how accurately a model predicts amino acids \cite{hesslow2022rita}. The lower the perplexity, the more accurate the prediction.

\smallskip\noindent \textit{\textbf{Novelty}} refers to the degree of uniqueness of the generated protein sequence compared to a database of known protein sequences \cite{truong2023poet}.

\smallskip\noindent \textit{\textbf{Fr\'echet Protein Distance (FPD)}} is used to measure the similarity between the distribution represented by the generated protein sequence and the distribution of the real protein sequence~\citep{jiang2008protein}, denoted as: 
\begin{equation}
\delta_{\mathcal{F}}(f, g) = \inf_{\alpha, \beta} \max_{s \in [0,1]} \text{dist}(f(\alpha(s)), g(\beta(s)))
\end{equation}
where $\alpha$ and $\beta$ are continuous non-decreasing functions. The sequence distribution can be denoted by $f$ and $g$.

\smallskip\noindent \textit{\textbf{Diversity}} is designed to evaluate the degree of difference between a range of protein sequences generated by a model. Rich diversity means that the model is capable of generating a variety of different sequences. Common methods include Shannon Entropy \cite{bywater2015prediction} and Hamming Distance \cite{mcgee2021generative}.

\smallskip\noindent \textit{\textbf{Foldability}} focuses on whether the generated protein sequence can be folded into a stable three-dimensional structure. Measuring foldability is usually performed with tools such as RoseTTAFold \cite{baek2021accurate} or computational methods based on physicochemical principles \cite{magliery2015protein} to predict the likelihood that the generated sequence will form a stable structure.

\smallskip\noindent \textit{\textbf{Recovery}} is focused on the ability of a model to predict the corresponding sequence for a given structure accurately \cite{watson2023novo}. Evaluating recovery includes methods sequence comparison, structure comparison, functionality comparison, etc.

