\newpage
\section{Functional Annotation Generation}
\label{app:dna2func}
 \begin{figure*}[ht!]
    \centering
    \includegraphics[width=0.93\linewidth]{figures/dataset.drawio.pdf}
    \vspace{-2em} % Adjust this value as needed
    \caption{\textbf{Seq2Func Dataset Construction and Evaluation Pipeline.} Genomic sequences from 20 species are annotated with functional descriptions using an LLM, generating 300,000 DNA-function pairs. The evaluation pipeline compares model predictions against ground truth, with an LLM determining transcription type accuracy.}
    \label{fig:seq2func}
% https://drive.google.com/file/d/1wDsMfefBgGdnAlqFbvc8slthBj4nuMys/view?usp=sharing
\end{figure*}


\paragraph{Dataset Construction Process}

The Seq2Func dataset is constructed using genomic sequences sourced from the \textbf{NCBI Reference Sequence (RefSeq)} database, incorporating sequences from \textbf{20 different species}. The selected species and their genome assemblies include:

\begin{itemize}
\item GCF-000001405.40, human
\item GCF-000001635.27, mouse
\item GCF-036323735.1, rat
\item GCF-028858775.2, Chimpanzee
\item GCF-018350175.1, cat
\item GCF-011100685.1, dog
\item GCF-000696695.1, burgud
\item GCF-000003025.6, pig
\item GCF-016772045.2, sheep
\item GCF-016699485.2, chicken
\item GCF-025200985.1, fly
\item GCF-000002035.6, Zebrafish
\item GCF-029289425.2, Pygmy Chimpanzee
\item GCF-029281585.2, Western gorilla
\item GCF-028885655.2, Pongo abelii
\item GCF-028885625.2, Bornean orangutan
\item GCF-003339765.1, Rhesus monkey
\item GCF-037993035.1, Macaca fascicularis
\item GCF-003668045.3, Chinese hamster
\item GCF-041296265.1, horse
\end{itemize}

The construction process follows these key steps:

\paragraph{Sequence Extraction and Functional Annotation}

\begin{itemize}
\item The raw dataset includes a diverse set of RNA sequences with different functional annotations.
\item The original dataset consists of the following RNA types and their counts:
\begin{itemize}
\item mRNA: 220397
\item tRNA: 20674
\item snRNA: 18996
\item snoRNA: 15577
\item lncRNA: 14291
\item miRNA: 12238
\item primary-transcript: 8644
\item rRNA: 6225
\item transcript: 3860
\item ncRNA: 477
\item guide-RNA: 143
\item antisense-RNA: 28
\item RNase-P-RNA: 9
\item V-gene-segment: 7
\item scRNA: 7
\item Y-RNA: 6
\item telomerase-RNA: 5
\item vault-RNA: 4
\item SRP-RNA: 4
\item RNase-MRP-RNA: 3
\item C-gene-segment: 2
\item D-gene-segment: 1
\end{itemize}
\item To refine the dataset, only \textbf{seven functional RNA types} are retained:
\begin{itemize}
\item mRNA
\item tRNA
\item snRNA
\item snoRNA
\item lncRNA
\item miRNA
\item rRNA
\end{itemize}
\end{itemize}

\paragraph{Annotation Enhancement Using a Large Language Model (LLM)}

\begin{itemize}
\item A \textbf{LLM Annotator} is employed to extend and refine functional annotations.
\item The model is prompted with: \textit{Describe the functions of the given gene/RNA precisely and concisely.''}
    \item For example, an initial annotation like Lnc RNA 2789'' is expanded into:
\begin{quote}
``It is Lnc RNA 2789 that doesn't encode proteins but regulates various biological processes.''
\end{quote}
\end{itemize}

\paragraph{Final Dataset Composition}

\begin{itemize}
\item The final dataset consists of \textbf{300,000 DNA-function pairs}, where each DNA sequence is paired with an enhanced functional description.
\item These functionally annotated sequences serve as high-quality input for machine learning models.
\end{itemize}




\begin{table}[h]
    \centering
    \renewcommand{\arraystretch}{1.5}
    \begin{tabular}{|p{5cm}|p{10cm}|}
        \hline
        \textbf{Usage}  & \textbf{Prompt} \\
        \hline
          Used  by \textbf{LLM (Annotator)} for extending the short annotation to detailed function annotation& 
          \textbf{(System Prompt)} You are a helpful assistant that answers functions of given gene/rna precisely and concisely.
         \\
        \hline
        Used by \textbf{LLM (Evaluator)} for deciding the type of DNA given its function description & 
        \textbf{(System Prompt)} You are a helpful assistant that determines the type of RNA based on the given function description. When deciding the functoin. Your answer should to only be one of the ['mRNA', 'tRNA', 'snRNA', 'snoRNA', 'lnc\-RNA', 'miRNA', 'rRNA'] \\
        \hline
        Given to \textbf{GPT4o@zeroshot} for performing zeroshot DNA function prediction & 
        \textbf{(System Prompt)} You are a helpful assistant that determines the type of RNA based on the given DNA sequence. Your answer should only be one of the mRNA, tRNA, snRNA, snoRNA, lnc\-RNA, miRNA, rRNA\\
        \hline
       
    \end{tabular}
    \caption{Prompt Used during dataset construction, evaluation and zeroshot prediction}
    \label{tab:prompt-for-dna2func}
\end{table}


\begin{table}[h]
    \centering
    \renewcommand{\arraystretch}{1.5}
    \begin{tabular}{|p{4cm}|p{4cm}|p{4cm}|p{4cm}|}
        \hline
        \textbf{Ground Truth} & \textbf{GPT4o} & \textbf{Olmo@ft} & \textbf{OmniDNA@ft} \\
        \hline
        \multicolumn{4}{|p{16cm}|}{\textbf{(Input)} TGGCAGAGATATG...AAGAAAAAAGAAAATGAACTTGGTGCAGGCA (DNA of length 1000) } \\
        \hline
       mRNA:CD209, also known as DC-SIGN (Dendritic Cell Specific Intercellular adhesion molecule-3-grabbing Nonintegrin), is a type I transmembrane protein primarily expressed on dendritic cells and some other immune cells. It functions as a receptor that binds and internalizes pathogens, including viruses and bacteria, thereby facilitating their presentation to  \colorbox{lightgrey}{T cells} and playing a critical role in the \colorbox{lightgrey}{immune response}. CD209 is involved in the recognition of mannose-rich ligands, mediating cellular adhesion, and influencing various aspects of \colorbox{lightgrey}{immune activation} and tolerance. Additionally, it can \colorbox{lightgrey}{modulate inflammatory} responses and has been implicated in HIV-1 and other pathogen infections. & The provided DNA sequence likely represents a genomic region involved in regulatory or coding functions. Based on its composition, it may contain promoter elements, coding regions, untranslated regions (UTRs), or repetitive sequences.  & The mRNA for the putative POM121-like protein 1-like likely encodes a protein that is similar to POM121, which is a component of the nuclear pore complex. This protein may play a role in nucleocytoplasmic transport, facilitating the movement of molecules between the nucleus and cytoplasm. The "putative" designation suggests that its function is inferred from homology to other known proteins, but it may not yet be fully characterized in terms of its specific biological roles or mechanisms. & C-C motif chemokine ligand 22 ( CCL22 ) is a chemokine involved in immune responses . It primarily attracts and \colorbox{lightgreenRGB}{activates T cells} , particularly regulatory T cells ( Tregs ), and plays a role in modulating inflammation and \colorbox{lightgreenRGB}{immune responses}. CCL22 is produced by various cell types , including macrophages and dendritic cells , and is implicated in various diseases , including autoimmune disorders and cancer. \\
        \hline
    \end{tabular}
    \caption{Comparison of Ground Truth, GPT4o, Olmo@ft, and OmniDNA@ft in RNA type determination (Example 1)}
    \label{tab:dna2func-1}
\end{table}



\begin{table}[h]
    \centering
    \renewcommand{\arraystretch}{1.5}
    \begin{tabular}{|p{4cm}|p{4cm}|p{4cm}|p{4cm}|}
        \hline
        \textbf{Ground Truth} & \textbf{GPT4o} & \textbf{Olmo@ft} & \textbf{OmniDNA@ft} \\
        \hline
\multicolumn{4}{|p{16cm}|}{\textbf{(Input)} GAGGTAGTAGATTGTATAG} \\
        \hline
       \colorbox{lightgrey}{miR}-3596d is a microRNA found in rats (Rattus norvegicus). It is involved in the regulation of gene expression at the post-transcriptional level, primarily by binding to complementary sequences in messenger RNAs (mRNAs), leading to their \colorbox{lightgrey}{degradation} or \colorbox{lightgrey}{inhibition} of \colorbox{lightgrey}{translation}. While specific target genes may vary, miRNAs like rno-miR-3596d generally play roles in various biological processes such as development, cell differentiation, proliferation, and apoptosis. Specific roles or targets for rno-miR-3596d would require experimental validation or further studies to elucidate its functions in rat biology. & The DNA sequence GAGGTAGTAGATTGTATAG could serve various functions depending on its genomic context. Generally, it could be:
	1.	A regulatory element – If found in promoter or enhancer regions, it may influence gene expression.
	2.	A binding site – It could be a recognition site for transcription factors, such as ETS-family proteins (which recognize GGAA/T motifs).
	3.	A coding or non-coding region – If within a gene, it might encode a peptide or function as a non-coding RNA.
	4.	A structural or repeat element – It could contribute to chromatin organization or genome stability.For precise functional annotation, its location within a genome and experimental validation would be needed. & hsa-miR-4684 is a microRNA (miRNA) found in humans. It is involved in the regulation of gene expression by binding to complementary sequences in target messenger RNAs (mRNAs), leading to their degradation or inhibition of translation. miR-4684 may play roles in various biological processes, including development, cell proliferation, and differentiation. Specific target genes and pathways regulated by hsa-miR-4684 may vary, and ongoing research aims to elucidate its precise functions & gga - miR - 65 77 - 5p is a microRNA ( miRNA ) found in chickens ( Gallus gallus ). MicroRNAs are small , non - coding RNA molecules that play a crucial role in regulating gene expression by binding to complementary sequences on target messenger RNAs ( mRNAs ), leading to their \colorbox{lightgreenRGB}{degradation} or \colorbox{lightgreenRGB}{inhibition} of \colorbox{lightgreenRGB}{translation} . Specifically , gga - miR - 65 77 - 5p may be involved in various biological processes , including development , differentiation , and response to stress , although specific target genes and functions may require further study \\
        \hline
    \end{tabular}
    \caption{Comparison of Ground Truth, GPT4o, Olmo@ft, and OmniDNA@ft in RNA type determination (Example 2)}
    \label{tab:dna2func-2}
\end{table}



\begin{table}[h]
    \centering
    \renewcommand{\arraystretch}{1.5}
    \begin{tabular}{|p{4cm}|p{4cm}|p{4cm}|p{4cm}|}
        \hline
        \textbf{Ground Truth} & \textbf{GPT4o} & \textbf{Olmo@ft} & \textbf{OmniDNA@ft} \\
        \hline
\multicolumn{4}{|p{16cm}|}{\textbf{(Input)} 
TGCTCACTTCAGCAGCACATGTGCTAAAAT
TGGAATGACACAAAGAAGATTAGCATGGCCGC
TGAGCAAGGATGACACGCACATTTGTGAATTGTTCCATAAAAAA} \\
        \hline
       nRNA:U6 spliceosomal RNA is a small nuclear RNA (snRNA) that plays an essential role in the splicing of pre-mRNA in eukaryotic cells. It is a core component of the spliceosome, the complex responsible for the removal of introns from pre-mRNA transcripts. U6 engages in base-pairing interactions with other spliceosomal RNAs and pre-mRNA, facilitating the assembly and function of the spliceosome. Its primary roles include catalyzing the cleavage of introns and the ligation of exons during the splicing process. U6 is also involved in the activation of the catalytic core of the spliceosome, making it crucial for proper gene expression. & The given DNA sequence likely encodes a functional genetic element, such as a protein-coding region or a regulatory sequence. To determine its precise function, I will need to analyze it for potential coding regions, motifs, or regulatory elements. & U6 spliceosomal RNA is a small nuclear RNA (snRNA) that plays a crucial role in the splicing of pre-mRNA in eukaryotic cells. It is a core component of the spliceosome, the complex responsible for removing introns from pre-mRNA transcripts. U6 RNA is involved in the catalysis of the splicing reaction and helps in the formation of the active site of the spliceosome, facilitating the precise excision of introns and the ligation of exons.  & U6 spliceosomal RNA is a small nuclear RNA ( snRNA ) that plays a crucial role in the splicing of pre - mRNA in eukaryotic cells . It is a component of the spliceosome , the complex responsible for removing introns from pre - mRNA and joining exons together . U6 RNA is involved in the catalytic activity of the spliceosome and is essential for the recognition of splice sites and the formation of the active site for splicing . It also interacts with other snRNAs. \\
        \hline
    \end{tabular}
    \caption{Comparison of Ground Truth, GPT4o, Olmo@ft, and OmniDNA@ft in RNA type determination (Example 3)}
    \label{tab:dna2func-3}
\end{table}