\begin{spacing}{1.2}
\noindent\textbf{
% Under construction
\noindent A comprehensive three-dimensional (3D) map of tissue architecture and gene expression is crucial for illuminating the complexity and heterogeneity of tissues across diverse biomedical applications\cite{erturk2024deep}. However, most spatial transcriptomics (ST) approaches remain limited to two-dimensional (2D) sections of tissue\cite{staahl2016visualization, marx2021method, moses2022museum}. Although current 3D ST methods hold promise, they typically require extensive tissue sectioning, are complex, are not compatible with non-destructive 3D tissue imaging technologies, and often lack scalability\cite{wang2023construction, tang2024search, schott2024open, mo2024tumour}. Here, we present VOlumetrically Resolved Transcriptomics EXpression (VORTEX), an AI framework that leverages 3D tissue morphology and minimal 2D ST to predict volumetric 3D ST. By pretraining on diverse 3D morphologyâ€“transcriptomic pairs from heterogeneous tissue samples and then fine-tuning on minimal 2D ST data from a specific volume of interest, VORTEX learns both generic tissue-related and sample-specific morphological correlates of gene expression. This approach enables dense, high-throughput, and fast 3D ST, scaling seamlessly to large tissue volumes far beyond the reach of existing 3D ST techniques. By offering a cost-effective and minimally destructive route to obtaining volumetric molecular insights, we anticipate that VORTEX will accelerate biomarker discovery and our understanding of morphomolecular associations and cell states in complex tissues.  Interactive 3D ST volumes can be viewed at \url{https://vortex-demo.github.io/}. 
}
\end{spacing}
