
% Theapeutic response & Cancer treatment
Understanding intratumoral morphological and molecular heterogeneity in human tissue is critical for developing personalized treatments and predicting therapeutic responses\cite{song2023artificial, marusyk2020intratumor, vitale2021intratumoral, fu2021spatial, bagaev2021conserved, arora2023spatial}.
Spatially-resolved transcriptomics (ST) provides expression profiles for many genes at high spatial resolution on two-dimensional (2D) tissue sections\cite{staahl2016visualization, rao2021exploring, rodriques2019slide, marx2021method, moses2022museum, palla2022spatial, ren2024spatial}. 
By analyzing ST with its associated high-resolution tissue morphology, researchers can holistically characterize intratumoral heterogeneity with multimodal views, investigate how changes in molecular profile influence underlying morphology, and vice versa.

The molecular and morphological traits captured within a 2D tissue section only represent a small fraction of the tissue volume and the patient\cite{liu2021harnessing, song2023artificial, braxton20243d, mo2024tumour, erturk2024deep, wang20243d, mathur2024glioblastoma}. Therefore, increasing attention has recently been directed toward extending molecular characterization from within a single tissue section to many adjacent tissue sections or across a larger volume. Recent three-dimensional (3D) pathology studies, fueled by substantial advances in high-resolution 3D tissue imaging modalities such as micro computed tomography (microCT) or open-top light-sheet microscopy\cite{withers2021x, liu2021harnessing, bishop2024end} showed that 3D morphological characterization can lead to better patient prognostication or cancer biomarker discovery\cite{song2024analysis, xie2022prostate, erturk2024deep}. 
Parallel efforts have been devoted to creating 3D molecular atlases of tissue, either with \textit{in situ} sequencing\cite{wang2018three, wang2021easi, fang2024three, sui2024scalable,doi:10.1126/science.adq2084} or by registering serial sections of 2D ST data meticulously obtained from a single tissue volume\cite{dong2022deciphering, vickovic2022three, zeira2022alignment, zhou2023integrating,wang2023construction, lin2023multiplexed, schott2024open, tang2024search,shu2024efficient}. While promising, \textit{in situ} approaches remain limited in terms of capture area and depth, and require long processing times (e.g., capture area of $3\times 4\, mm^2$ and depth of $\sim 200 \mu m$). Serial section-based approaches provide discontinuous coverage along the axial dimension (i.e., 2.5D ST characterization) of thick tissues. Such approaches are impractical for scaling to whole-volume transcriptomic profiling in terms of cost and effort, with up to several days of processing for a single clinical sample.

An AI-based computational predictive framework offers an attractive alternative for characterizing the molecular landscapes of tissue specimens.
Evidence of the close relationship between spatially variable genes and underlying tissue morphology \cite{edsgard2018identification, svensson2018spatialde, sun2020statistical, binder2021morphological, ash2021joint, hu2021spagcn, song2023artificial} suggests that such \textit{morphomolecular} links can be modeled, especially when leveraging the powerful capabilities of deep learning. Coupled with the increasing availability of paired high-resolution 2D tissue images and 2D ST data\cite{moses2022museum, marx2021method, jaume2024hest, chen2024stimagekm}, recent AI-based frameworks have demonstrated success in directly learning morphomolecular links and predicting transcript expression and localization from morphological data alone\cite{he2020integrating, hu2021spagcn, bergenstraahle2022super, xie2024spatially, chung2024accurate, coleman2024unlocking, kueckelhaus2024inferring, zhang2024inferring, lee2024Path}. 
However, these models are exclusively restricted to 2D tissue sections, and designing 3D ST prediction frameworks based on 3D tissue morphology necessitates additional consideration. 

Here, we present an AI-based computational framework called $\ours$, \textbf{VO}lumetrically \textbf{R}esolved \textbf{T}ransciptomics \textbf{EX}pression. $\ours$ enables scalable and efficient 3D ST prediction of large volumes from 3D pathology datasets.
$\ours$ is pretrained on 3D morphology and 2D ST data pairs from diverse volumes of the same cancer and is further fine-tuned on data pairs from a specific volume of interest. This learning paradigm takes advantage of both generic morphomolecular links prevalent across diverse volumes and volume-specific links that are difficult to learn due to inter-volume heterogeneity.
A distinguishing feature of $\ours$ is its ability to adapt to different 3D imaging modalities and tissue sizes for 3D ST prediction.
To handle diverse non-destructive 3D tissue imaging approaches\cite{withers2021x, palermo2025investigating, glaser2017light, bishop2024end} with ST data confined to 2D tissue sections, $\ours$ performs cross-modal registration and integration between 3D tissue images, 2D tissue images, and 2D ST. Furthermore, $\ours$ can easily scale up to performing ST predictions in large tissue volumes, vastly exceeding typical ST capture areas with little additional cost and processing time. To highlight the versatility of $\ours$, we also demonstrate it for ST predictions in 2.5D tissue images constructed from serial 2D tissue sections, a commonly utilized approach that is compatible with current histopathology workflows.