\section{Problem Formulation}
\label{sec:background}

In this section, we introduce the objective, data, and mathematical formulation of cellular morphology prediction.

\subsection{Objective}

Let $\mathcal{X}$ denote the cell image space and $\mathcal{C}$ the perturbation space. Let $p_0$ represent the original cell distribution and $p_1$ represent the distribution of cells after a perturbation $c \in \mathcal{C}$. Cellular morphology prediction aims to learn a generative model $p_\theta: \mathcal{X} \times \mathcal{C} \to \mathcal{P}(\mathcal{X})$, which, given an unperturbed cell image $x_0 \sim p_0$ and a perturbation $c \in \mathcal{C}$, predicts the resulting conditional distribution $p(x_1 | x_0, c)$. From this distribution, new images can be sampled to simulate the effects of the perturbation, such that $x_1 \sim p_1$ (Figure~\ref{fig:overview}a).

The input space consists of multi-channel microscopy images, where $\mathcal{X} \subset \mathbb{R}^{H \times W \times C}$. Here, $H$ and $W$ represent the image height and width, while $C$ denotes the number of channels, each highlighting different cellular components through specific fluorescent markers (analogous to RGB channels in natural images, but capturing biological structures like mitochondria, nuclei, and cellular membranes).

The perturbation space $\mathcal{C}$ includes two types of biological interventions: chemical (drugs) and genetic (gene modifications). Chemical perturbations involve compounds that target specific cellular processes --- for example, affecting DNA replication or protein synthesis. 
Genetic perturbations can turn off gene expression (CRISPR) or upregulate gene expression (ORF).

This generative model enables in silico simulation of cellular responses, which traditionally require time-intensive and costly wet-lab experiments. Such computational modeling could revolutionize drug discovery by enabling rapid virtual drug screening and advance personalized medicine through digital cell twins for treatment optimization.

\subsection{Data}

Cell morphology data are collected through high-content microscopy screening (Figure~\ref{fig:overview}b) \cite{perlman2004multidimensional}. In this process, biological samples are prepared in multi-well plates containing hundreds of independent experimental units (wells). Selected wells receive interventions --- either chemical compounds or genetic modifications --- while control wells remain unperturbed. After a designated period, cells are fixed using chemical fixatives and stained with fluorescent dyes to highlight key structures like the nucleus, cytoskeleton, and mitochondria. An automated microscope then captures multiple images per well. This process is called cell painting. Modern automated high-content screening systems have enabled large-scale data collection, resulting in datasets of terabyte to petabyte images from thousands of perturbation conditions \cite{fay2023rxrx3, chandrasekaran2023jump}.

However, the cell painting process has limitations: cell painting requires cell fixation, which is destructive, making it impossible to observe the same cells dynamically during a perturbation. This creates a fundamental constraint: we cannot obtain paired samples $\{(x_0, x_1)\}$ showing the exact same cell without and with treatment. Instead, we must work with unpaired data $(\{x_0\}, \{x_1\})$, where $\{x_0\}$ represents control images and $\{x_1\}$ represents treated images, to learn the conditional distribution $p(x_1 | x_0, c)$.

One solution is to leverage the distribution transformation from control cells to perturbed cells within the same batch to learn conditional generation. Control cells serve as a crucial reference by providing prior information to separate true perturbation effects from confounding factors such as batch effects. Variations in experimental conditions across different runs (batches) introduce systematic biases unrelated to the perturbation itself. For instance, images from one batch may consistently differ in pixel intensities from those in another. Therefore, meaningful comparisons require analyzing treated and control samples from the same batch. As shown in Figure~\ref{fig:overview}b, this approach helps distinguish true biological responses, like changes in nuclear size, from batch-specific artifacts, like changes in color.

\input{tabs/cellflow}

\subsection{Mathematical Formulation}

\begin{center}
\begin{tikzpicture}[->, >=stealth, node distance=1.6cm, font=\sffamily]
    \node[draw, circle, fill=gray!30] (b) {$B$};
    \node[draw, circle, fill=gray!30, right of=b] (c) {$C$};
    \node[draw, circle, fill=gray!30, below left=of b] (x0tilde) {$\tilde{X_0}$};
    \node[draw, circle, below of=b] (x0) {$X_0$};
    \node[draw, circle, fill=gray!30, below of=c] (x1) {$X_1$};
    \draw (b) -- (x0);
    \draw (c) -- (x1);
    \draw (x0) -- (x1);
    \draw (b) -- (x0tilde);
\end{tikzpicture}
\vspace{-10pt}
\end{center}

Let us formalize our learning problem in light of the experimental constraints described before. Our objective is to learn a conditional distribution $p(x_1|x_0,c)$ that models the cellular response to perturbation. However, due to the destructive nature of imaging, we cannot observe paired samples $\{(x_0,x_1)\}$. We propose a probabilistic graphical model to address this challenge.

In our graph, random variable $B$ denotes the experimental batch, $C$ denotes the perturbation condition, $X_0$ represents the unobservable basal cell state, $\tilde{X_0}$ represents control cells from the same batch, and $X_1$ denotes the perturbed cell state. From our experimental setup, we have access to the control distribution $p(\tilde{x_0}|b)$ from unperturbed cells and the perturbed distribution $p(x_1|c,b)$ from treated cells.

We propose to leverage the distributional transition from $p(\tilde{x_0}|b)$ to $p(x_1|c,b)$ to learn the individual-level trajectory $p(x_1|x_0,c)$, as shown in Figure~\ref{fig:overview}c. There are two key reasons. First, there exists a natural connection between $p(x_1|c,b)$ and $p(x_0|b)$ through the marginalization $p(x_1|c,b)= \int p(x_1|x_0,c) p(x_0|b) dx_0$. Second, while $p(x_0|b)$ is not directly tractable, we can approximate it using $p(\tilde{x_0}|b)$ since both the ground-truth $X_0$ distribution and control distribution $\tilde{X_0}$ follow the same batch-conditional distribution: $x_0 \sim p(\cdot|b)$ and $\tilde{x_0} \sim p(\cdot|b)$. 

Our approach of learning $p(x_1|\tilde{x_0},c)$ by conditioning on same-batch control images improves upon existing methods that ignore control cells and learn only $p(x_1|c)$. Intuitively, conditioning on $\tilde{x_0}$ allows the model to initiate the transition from a distribution more closely aligned with the underlying $x_0$, leading to a better approximation of the true distribution $p(x_1|x_0,c)$. We formalize this intuition in the following proposition, with proof provided in Appendix~\ref{sec:theory}:

\begin{prop}
    \label{prop:cond}
    Given random variables $B$, $C$, $X_0$, $\tilde{X_0}$, and $X_1$ following the graphical model above with joint distribution $p(b, c, x_0, \tilde{x_0}, x_1)$, the distribution $p(x_1|x_0,c)$ can be better approximated by the conditional distribution $p(x_1|\tilde{x_0},c)$ than $p(x_1|c)$ in expectation. Formally,
    \begin{equation}
    \begin{split}
    & \mathbb{E}_{p(x_0, \tilde{x_0}, c)}\left[ D_{\text{KL}}(p(x_1|x_0,c)||p(x_1|\tilde{x_0},c)) \right] \nonumber\\
    \leq  & \mathbb{E}_{p(x_0, c)}\left[ D_{\text{KL}}(p(x_1|x_0,c)||p(x_1|c)) \right]
    \end{split}
    \end{equation}
\end{prop}
