\begin{table}[!h]
\centering
\caption{\textbf{Dataset for $\ours$ training and evaluation}.}
\tiny
\begin{tabular}{|l|l|p{2cm}|l|l|l|l|l|}
\hline
\textbf{Organ/}\textbf{Dataset}                                                      & \textbf{Disease state}                                                                                                                                         & \textbf{Data source}                                                                                          & \textbf{Imaging modality}                                                                                         & \textbf{Dataset dimensions}                                                                                                                                                                                                                                                & \textbf{Protocol}                                                               & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}Spot diameter \\ (µm)\end{tabular}}} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}Distance between \\ spots (µm)\end{tabular}}} \\ \hline
Prostate I                                                            & Prostate adenocarcinoma                                                                                                                                        & Internal                                                                                                      & \begin{tabular}[c]{@{}l@{}}FFPE H\&E-stained \\ histological sections and \\ microCT images\end{tabular}          & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 65,715\\ -Num. Sections: 16\\ -Num. Patients: 11\\ -Num. Genes: 17,943\end{tabular}                                                                        & Visium                                                                          & 55                                                                                          & 100                                                                                                  \\ \hline
Prostate II                                                           & \begin{tabular}[c]{@{}l@{}}-Healthy (N=1)\\ -Prostate acinar cell \\  carcinoma (N=1)\\ -Prostate adenocarcinoma \\ with invasive carcinoma (N=1)\end{tabular} & 10x Genomics (blocks 1E333\_Tn17, 1E500\_Tp12 Section 1, and 1D1061-Tp11 Section 1)                                            & \begin{tabular}[c]{@{}l@{}}FFPE H\&E \\ histological sections\end{tabular}                                        & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 9,957 \\ -Num. Sections: 3 \\ -Num. Patients: 3\\ -Num. Genes: 17,943\end{tabular}                                                                                                                                                 & Visium                                                                          & 55                                                                                          & 100                                                                                                  \\ \hline
Prostate III                                                            & Prostate adenocarcinoma                                                                                                                                        & \href{https://data.mendeley.com/datasets/svw96g68dv/1}{\url{https://data.mendeley.com/datasets/svw96g68dv/1}}                                                               & \begin{tabular}[c]{@{}l@{}}Frozen H\&E-stained \\ histological sections\end{tabular}                              & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 58,906\\ -Num. Sections: 22\\ -Num. Patients: 2 \\ -Num. Genes: 33,538\end{tabular}                                                                                                                                                & Visium                                                                          & 55                                                                                          & 100                                                                                                  \\ \hline
Prostate IV                                                            & Prostate adenocarcinoma                                                                                                                                        & \href{https://data.mendeley.com/datasets/mdt8n2xgf4/1}{\url{https://data.mendeley.com/datasets/mdt8n2xgf4/1}}                                                              & \begin{tabular}[c]{@{}l@{}}Frozen H\&E-stained \\ histological sections from\\  core needle biopsies\end{tabular} & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 3,969\\ -Num. Sections: 24\\ -Num. Patients: 3\\ -Num. Genes (mean): 24,945 \\ (Range: 19,537 - 30,157)\end{tabular}                                                                                                               & \begin{tabular}[c]{@{}l@{}}Spatial \\ Transcriptomics\end{tabular}              & 100                                                                                         & 200                                                                                                  \\ \hline
Breast I                                                              & HER2+ breast cancer                                                                                                                                            & \begin{tabular}[c]{@{}l@{}}Andersson et al. \tiny \cite{andersson2021spatial} \\ Data retrieved from \\ \href{https://huggingface.co/datasets/MahmoodLab/hest}{HEST-1k database} \end{tabular} & \begin{tabular}[c]{@{}l@{}}Frozen H\&E-stained \\ histological sections\end{tabular}                              & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 13,299\\ -Num. Spots with associated \\  3D morphology: 6,577\\ -Num. Sections: 36\\ -Num. Sections with associated \\ 3D morphology: 16\\ -Num. Patients: 8\\ -Num. Genes (mean): 15,364 \\ (Range: 14,861 - 15,842)\end{tabular} & \begin{tabular}[c]{@{}l@{}}Spatial \\ Transcriptomics\end{tabular}              & 100                                                                                         & 200                                                                                                  \\ \hline
Breast II                                                              & Invasive ductal carcinoma                                                                                                                                      & \begin{tabular}[c]{@{}l@{}}B. He et al. \cite{he2020integrating}, \\ Data retrieved from \\ HEST-1k database.\end{tabular}        & \begin{tabular}[c]{@{}l@{}}Frozen H\&E-stained \\ histological sections\end{tabular}                              & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 23,699\\ -Num. Sections: 56\\ -Num. Patients: N/A\\ -Num. Genes (mean): 18,010 \\ (Range: 16,744 - 19,729)\end{tabular}                                                                                                            & \begin{tabular}[c]{@{}l@{}}Spatial \\ Transcriptomics\end{tabular}              & 100                                                                                         & 200                                                                                                  \\ \hline
Breast III                                                              & Invasive ductal carcinoma                                                                                                                                      & \begin{tabular}[c]{@{}l@{}}10x Genomics. \\ Data retrieved from \\ HEST-1k database.\end{tabular}                 & \begin{tabular}[c]{@{}l@{}}Frozen (N=2) \& \\ FFPE (N=2) H\&E-stained \\ histological sections\end{tabular}       & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 20,236\\ -Num. Sections: 4\\ -Num. Patients: 3\\ -Num. Genes (mean): 27,971 \\ (Range: 17,943 - 36,601)\end{tabular}                                                                                                               & Visium                                                                          & 55                                                                                          & 100                                                                                                  \\ \hline
Breast IV                                                              & Invasive ductal carcinoma                                                                                                                                      & \begin{tabular}[c]{@{}l@{}}PL Stalh et al \cite{staahl2016visualization}. \\ Data retrieved from \\ HEST-1k database.\end{tabular}        & \begin{tabular}[c]{@{}l@{}}H\&E-stained \\ histological sections\end{tabular}                                     & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 1,029\\ -Num. Sections: 4\\ -Num. Patients: 1\\ -Num. Genes: 14851 \\ (Range: 14789 - 14929)\end{tabular}                                                                                                                          & \begin{tabular}[c]{@{}l@{}}Spatial \\ Transcriptomics\end{tabular}              & 100                                                                                         & 200                                                                                                  \\ \hline
CRC I                                                        & Colorectal cancer                                                                                                                                              & \begin{tabular}[c]{@{}l@{}}A. Valdeolivas et al. \cite{valdeolivas2024profiling}, \\ Data retrieved from \\ HEST-1k database.\end{tabular}  & \begin{tabular}[c]{@{}l@{}}Frozen H\&E-stained \\ histological sections\end{tabular}                              & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 20,708\\ -Num. Sections: 14\\ -Num. Patients: 7\\ -Num. Genes: 36,601\end{tabular}                                                                                                                                                 & Visium                                                                          & 55                                                                                          & 100                                                                                                  \\ \hline
\begin{tabular}[c]{@{}l@{}}CRC II\end{tabular} & Healthy                                                                                                                                                        & \begin{tabular}[c]{@{}l@{}}R. Mirzazadeh et al \cite{mirzazadeh2023spatially}, \\ Data retrieved from \\ HEST-1k database.\end{tabular}   & \begin{tabular}[c]{@{}l@{}}Frozen H\&E-stained \\ histological sections\end{tabular}                              & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 11,049\\ -Num. Sections: 5\\ -Num. Patients: N/A\\ -Num. Genes (mean): 21,674 \\ (Range: 17,943 - 36,601)\end{tabular}                                                                                                             & \begin{tabular}[c]{@{}l@{}}RNA-Rescue \\ Spatial\\ Transcriptomics\end{tabular} & 55                                                                                          & 100                                                                                                  \\ \hline
CRC III                                                        & \begin{tabular}[c]{@{}l@{}}-Colon adenocarcinoma (N=6) \\ -Colorectal\\  adenocarcinoma (N=1)\end{tabular}                                                     & \begin{tabular}[c]{@{}l@{}}10x Genomics. \\ Data retrieved from \\ HEST-1k database.\end{tabular}                 & \begin{tabular}[c]{@{}l@{}}Frozen (N=1) \& \\ FFPE (N=6) H\&E-stained \\ histological sections\end{tabular}       & \begin{tabular}[c]{@{}l@{}}-Num. Spots: 40,285\\ -Num. Sections: 7\\ -Num. Patients: 4\\ -Num. Genes (mean): 20,709 \\ (Range: 17,943 - 36,601)\end{tabular}                                                                                                               & Visium                                                                          & 55                                                                                          & 100                                                                                                  \\ \hline
\end{tabular}
\label{tab:dataset}
\end{table}



% \href{https://www.10xgenomics.com/datasets/normal-human-prostate-ffpe-1-standard-1-3-0}{\url{https://www.10xgenomics.com/datasets/normal-human-prostate-ffpe-1-standard-1-3-0}},  \href{https://www.10xgenomics.com/datasets/human-prostate-cancer-adenocarcinoma-with-invasive-carcinoma-ffpe-1-standard-1-3-0}{\url{https://www.10xgenomics.com/datasets/human-prostate-cancer-adenocarcinoma-with-invasive-carcinoma-ffpe-1-standard-1-3-0}}, \href{https://www.10xgenomics.com/datasets/human-prostate-cancer-acinar-cell-carcinoma-ffpe-1-standard}{\url{https://www.10xgenomics.com/datasets/human-prostate-cancer-acinar-cell-carcinoma-ffpe-1-standard}} 

\clearpage

\begin{table}[h!]
    \centering
    \caption{Marker genes for different cancer cohorts.}
    \begin{tabular}{|p{3cm}|p{12cm}|} 
        \toprule
        \textbf{Dataset} & \textbf{Marker Genes} \\
        \midrule
        Prostate & \textbf{OncotypeDX}: AZGP1, KLK2, FAM13C, FLNC, GSN, TPM2, BGN, COL1A1, SFRP4 \\ 
                 & \textbf{Decipher}: CAMK2N1, G6PD, ATM, LASP1, PRDX4, PCNA, STMN1, ERG, NFIB, ANO7 \\
        \midrule
        Breast & \textbf{HER2DX}: RRM2, FGFR4, BAG1, PHGDH, MLPH, MYC \\
        & \textbf{Additional genes}: CDH1, ERBB3, ESR1, KMT2C, MAP2K4, MDM2, PTEN, ARID1A, ERBB2, FGFR1, GATA3, KMT2D, MAP3K1, TP53 \\
        \midrule
        Colorectal & APC, TP53, KRAS, SMAD4, BRAF, VCAN, ARID1A, SOX9, NRAS, KDR, FBXW7, MET, PTEN, BIRC6, ACVR2A, RNF43, UBR5, SETD1B, KMT2C, ZFP36L2, BMPR2, EFEMP2, FBN1, SPARC, SCD, RNF43, MMP1, PLAU, CXCL14, AXIN2 \\
        \bottomrule
    \end{tabular}
    \label{tab:marker_genes}
\end{table}






\clearpage
%% STAGE I OF TRAINING
\begin{table}[h]
    \centering
    \caption{\textbf{Hyperparameters in $\ours$ pretraining (stage I).} CoCa training between H\&E and ST.}
    \scalebox{0.95}{
        \begin{tabular}{l|l}
            \toprule
            Hyperparameter & Value \\
            \midrule
            \midrule
            GPU & 2$\times$ 24GB GeForce RTX 3090 \\
            Batch size per GPU & 256 \\
            AdamW $\beta$ & (0.9, 0.999) \\
            Num. trainable blocks H\&E encoder & 3 \\
            Num. trainable blocks ST encoder & 3 \\
            H\&E encoder learning rate & 0.00001 \\
            ST encoder learning rate & 0.00001 \\
            ST predictor learning rate & 0.00001 \\
            Adversarial discriminator (batch correction) learning rate & 0.0001 \\ 
            H\&E encoder weight decay & 0.01 \\
            ST encoder weight decay & 0.01 \\
            ST predictor weight decay & 0.01 \\
            Adversarial discriminator (batch correction) weight decay & 0.01 \\
            Learning rate schedule & Linear (warmup period)-cosine \\
            Learning rate (start) & 0 \\
            Learning rate (post warmup) & 1e-5\\
            Learning rate (final) & 0 \\
            Warmup epochs & 5 \\
            Total epochs & 25 \\
            Contrastive Loss Temperature & 0.1 \\
            Automatic mixed precision & bfloat16 \\
            \bottomrule
            \end{tabular}}
\label{tab:stageI}
\end{table}

\clearpage
%% STAGE II OF TRAINING
\begin{table}[h]
    \centering
    \caption{\textbf{Hyperparameters in $\ours$ training (stage II).} CoCa training between 3D tissue image, H\&E and ST.}
    \scalebox{0.95}{
        \begin{tabular}{l|l}
            \toprule
            Hyperparameter & Value \\
            \midrule
            \midrule
            GPU & 1$\times$ 24GB GeForce RTX 3090 \\
            Batch size per GPU & 128 \\
            AdamW $\beta$ & (0.9, 0.999) \\
            Num. trainable blocks H\&E encoder & 0 \\
            Num. trainable blocks volumetric image encoder & 0/3 (MicroCT/Serial Sections) \\
            Num. trainable blocks ST encoder & 0 \\
            Volumetric image encoder encoder learning rate & 0.00001 \\
            ST predictor learning rate & 0.00001 \\
            Volumetric image encoder weight decay & 0.01 \\
            ST predictor weight decay & 0.01 \\
            Learning rate schedule & Cosine \\
            Learning rate (start) & 0 \\
            Learning rate (post warmup) & 1e-5\\
            Learning rate (final) & 0 \\
            Epochs & 15 \\
            Contrastive Loss Temperature & 0.1 \\
            Automatic mixed precision & bfloat16 \\
            \bottomrule
            \end{tabular}
            }
\label{tab:stageII}
\end{table}


\clearpage
%% STAGE III OF TRAINING
\begin{table}[h]
    \centering
    \caption{\textbf{Hyperparameters in $\ours$ fine-tuning (stage III).} CoCa training between volumetric image, H\&E and ST.}
    \scalebox{0.95}{
        \begin{tabular}{l|l}
            \toprule
            Hyperparameter & Value \\
            \midrule
            \midrule
            GPU & 1$\times$ 24GB GeForce RTX 3090 \\
            Batch size per GPU & 16 \\
            AdamW $\beta$ & (0.9, 0.999) \\
            Num. trainable blocks H\&E encoder & 3 \\
            Num. trainable blocks volumetric image encoder & 0/3 (MicroCT/Serial Sections) \\
            Num. trainable blocks ST encoder & 3 \\
            H\&E encoder learning rate & 0.00001 \\
            Volumetric image encoder encoder learning rate & 0.00001 \\
            ST encoder learning rate & 0.00001 \\
            ST predictor learning rate & 0.00001 \\
             H\&E encoder weight decay & 0.01 \\
             Volumetric image encoder weight decay & 0.01 \\
            ST encoder weight decay & 0.01 \\
            ST predictor weight decay & 0.01 \\
            Learning rate schedule & Cosine \\
            Learning rate (start) & 0 \\
            Learning rate (post warmup) & 1e-5\\
            Learning rate (final) & 0 \\
            Epochs & 10 \\
            Contrastive Loss Temperature & 0.1 \\
            Automatic mixed precision & bfloat16 \\
            \bottomrule
            \end{tabular}
            }
\label{tab:stageIII}
\end{table}


\clearpage

\begin{table}[h!]
    \centering
    \caption{\textit{Spatial filter genes} for molecular query design in the prostate cancer cohort.}
    \begin{tabular}{|p{4cm}|p{12cm}|} 
        \hline
        \textbf{Molecular Query} & \textbf{Gene Sets} \\
        \hline
        \multirow{2}{4cm}{Molecular Query 1: \textit{PI3K/AKT/mTOR} pathway} 
        & \textbf{\textit{Genes of interest}}: SPON2, TFF3 \\  
        & \textit{\textbf{Correlated genes}}: TSPAN1, SERP1, ERGIC1, SPDEF, STEAP2, NKX3-1, HSPA5, CORO1B, NCAPD3, TMBIM6, KRT18, CAMKK2, FOXA1, DHRS7, SERF2, ABHD2, ALDH1A3, APRT, CANT1, P4HB, DCXR, FASN, HMG20B, FXYD3, KLK3, KLK2, FAM3B, SMS \\  
        \hline
        \multirow{2}{4cm}{Molecular Query 2: \textit{Myogenesis}}  
        & \textbf{\textit{Genes of interest}}: TPM2, TAGLN \\  
        & \textit{\textbf{Correlated genes}}: LMOD1, EMILIN1, ACTG2, LIMS2, FN1, DES, COL6A3, MYLK, WDR1, SYNPO2, PALLD, PDLIM3, MAP1B, DPYSL3, SPARC, ACTB, CALD1, CLU, OGN, GSN, PTGDS, SVIL, ACTA2, SORBS1, ILK, ACTN1, TPM1, SYNM, TGFB1I1, KANK2, CNN1, HSPB6, DSTN, MYL9, JPH2, PCP4, SMTN, LGALS1, FLNA \\  
        \hline
        \multirow{2}{4cm}{Molecular Query 3: \textit{Tumor suppression}}  
        & \textbf{\textit{Genes of interest}}: MSMB, ACPP \\  
        & \textit{\textbf{Correlated genes}}: ADIRF, SCD, RDH11, NEFH \\  
        \hline
    \end{tabular}
    \label{tab:query_genes}
\end{table}

% \begin{table}[h!]
%     \centering
%     \caption{\textit{Spatial filter genes} for molecular query design in the prostate cancer cohort.}
%     \begin{tabular}{|p{4cm}|p{12cm}|} 
%         \toprule
%         \textbf{Molecular Query} & \textbf{Gene Sets} \\
%         \midrule
%         Molecular Query 1: \textit{PI3K/AKT/mTOR} pathway & \textbf{\textit{Genes of interest}}: SPON2, TFF3 \\ 
%                  & \textit{\textbf{Correlated genes}}: TSPAN1, SERP1, ERGIC1, SPDEF, STEAP2, NKX3-1, HSPA5, CORO1B, NCAPD3, TMBIM6, KRT18, CAMKK2, FOXA1, DHRS7, SERF2, ABHD2, ALDH1A3, APRT, CANT1, P4HB, DCXR, FASN, HMG20B, FXYD3, KLK3, KLK2, FAM3B, SMS \\
%         \midrule
%         Molecular Query 2: \textit{Myogenesis} & \textbf{\textit{Genes of interest}}: TPM2, TAGLN \\
%         & \textit{\textbf{Correlated genes}}: LMOD1, EMILIN1, ACTG2, LIMS2, FN1, DES, COL6A3, MYLK, WDR1, SYNPO2, PALLD, PDLIM3, MAP1B, DPYSL3, SPARC, ACTB, CALD1, CLU, OGN, GSN, PTGDS, SVIL, ACTA2, SORBS1, ILK, ACTN1, TPM1, SYNM, TGFB1I1, KANK2, CNN1, HSPB6, DSTN, MYL9, JPH2, PCP4, SMTN, LGALS1, FLNA \\
%         \midrule
%         Molecular Query 3: \textit{Tumor suppression} & \textbf{\textit{Genes of interest}}: MSMB, ACPP \\
%         & \textit{\textbf{Correlated genes}}: ADIRF, SCD, RDH11, NEFH \\
%         \bottomrule
%     \end{tabular}
%     \label{tab:query_genes}
% \end{table}


