\begin{table}[t]
    \centering
    \vspace{-1.0em}
    \caption{\textbf{Overview of Bio-Foundation Models}. Unlike DNA models NT-2500M or EVO-7B using BERT or generative pre-training with pure DNA sequences, our Life-Code models the Central Dogma with the unified nucleotide input format. We consider four aspects of evaluations: The DNA tasks report the averaged MCC (\%) over the Promoter Detection (all) and Covid Variants Classification, the RNA task considers the Splicing Site Reconstructed in the GUE benchmark, the Protein tasks report the averaged SRCC (\%) Protein Fitness Prediction with DMS, and the Central Dogma evaluation considers the translation between CDS and Amino Acids with top-1 accuracy (\%). Full results in Table~\ref{tab:app_overall}.
    }
    % \vspace{-0.5em}
    \setlength{\tabcolsep}{1.0mm}
\resizebox{1.0\linewidth}{!}{
    \begin{tabular}{l|ccgb}
    \toprule
% 
Method         & NT-2500M & EVO-7B & LucaOne       & \bf{Life-Code}\\ \hline
\# Params      & 2500M    & 7B     & 110M          & 350M          \\
Learning Tasks & MLM      & MLM    & MLM+SL        & MLM+KD        \\
Data Source    & DNA      & DNA    & Multi-omics   & DNA+AA        \\ \hline
DNA Task       & 82.0     & 73.6   & \textbf{83.3} & 82.6          \\
RNA Task       & 89.4     & 87.5   & 89.1          & \bf{89.8}     \\
Protein Task   & 7.1      & 28.2   & 32.9          & \textbf{34.1} \\
Central Dogma  & $-$      & 75.5   & 84.8          & \textbf{85.6} \\
% 
    \bottomrule
    \end{tabular}
    }
    \label{tab:overall}
    \vspace{-1.0em}
\end{table}
