\begin{figure*}[ht]
\centering
\begin{forest}
  for tree={
    forked edges,
    grow=east,
    reversed=true,
    anchor=base west,
    parent anchor=east,
    child anchor=west,
    base=middle,
    font=\scriptsize,
    rectangle,
    draw=black,
    edge=black!50, 
    rounded corners,
    align=center,
    minimum width=1em,
    s sep=5pt,
    inner xsep=2.5pt,
    inner ysep=1pt
  },
  [Protein Large Language Models,rotate=90,anchor=north,edge=black!50,fill=myblue,draw=black
    [Problem,edge=black!50, fill=mypurple, minimum height=1.2em
      [Protein Understanding \& Prediction,text width=9.6em,fill=mypurple
        [\textbf{Pretraining Dataset:} UniRef Clusters~\cite{suzek2015uniref}{,} Pfam~\cite{finn2006pfam}{,} \\UniProtKB~\cite{boutet2016uniprotkb,m1999edittotrembl},text width=23em, fill=mypurple]
       [\textbf{Benchmark Dataset:} CASP~\cite{kryshtafovych2019critical}{,} TAPE\citep{rao2019evaluating}{,} \\ProteinLMBench~\cite{shen2024fine}{,} PEER~\cite{xu2022peer},text width=23em, fill=mypurple]
      ]
      [Protein Engineering \& \\Generation  \& Translation ,text width=6.8em,fill=mypurple
        [\textbf{Pretraining Dataset:} AlphaFoldDB~\cite{tunyasuvunakool2021highly},text width=25.8em, fill=mypurple]
       [\textbf{Benchmark Dataset:} ProteinGym~\cite{notin2024proteingym}{,} ProteinLMBench~\cite{shen2024fine},text width=25.8em, fill=mypurple]]
      ]
    [Method, edge=black!50, fill=myred, minimum height=1.2em
      [LLM Methods for \\ProteinUnderstanding \\and Prediction, edge=black!50, fill=myred
        [Protein Sequence Models, text width=6.6em, fill=myred
          [ UniRep~\citep{alley2019unified}{,} SSA \citep{bepler2019learning}{,}\\ MuPIPR \citep{zhou2020mutation}{,} ProtTrans\citep{elnaggar2021prottrans}{,}\\ AminoBERT\citep{chowdhury2022single}{,} Provis \citep{vig2020bertology}{,}  \\xTrimoPGLM~\citep{chen2024xtrimopglm}{,} CSCS~\citep{hie2021learning}{,}\\TCR-BERT \citep{wu2024tcr}{,} PeptideBert \citep{guntuboina2023peptidebert}{,}\\
          ESM-1b \citep{rives2021biological}{,} ESM-1v \citep{meier2021language}{,} \\
          AlphaFold2 \citep{jumper2021highly}{,} ESM-IF \citep{hsu2022learning}{,}\\
          ESM-2 \citep{lin2023evolutionary}{,} ProtTrans \citep{elnaggar2021prottrans}{,} \\
          ProteinLM \citep{xiao2021modeling}{,} 
 ProteinBERT~\citep{brandes2022proteinbert}{,}\\
          MSA Transformer \citep{rao2021msa}{,} Tranception \citep{notin2022tranception}{,} \\vcMSA~\citep{mcwhite2023leveraging}{,} Poet~\citep{truong2023poet} , text width=18.9em, fill=myred]
        ]
        [Structure-Integrated \&\\ Knowledge-Enhanced \\Models, text width=5.7em, fill=myred
          [OntoProtein \citep{zhang2022ontoprotein}{,} ESM-GearNet \citep{zhang2023systematic}{,}\\
          SaProt \citep{su2023saprot}{,} ProteinBERT \citep{brandes2022proteinbert}{,}\\ProteinCLIP \citep{wu2024proteinclip} , text width=19.8em, fill=myred]
        ]
        [Protein Description \&\\ Annotation Models, text width=5.6em, fill=myred
          [ProTranslator~\citep{xu2022protranslator}{,} BioTranslator~\citep{xu2023multilingual}{,}\\ProtST~\cite{xu2023protst}{,} Prot2Text~\citep{abdine2024prot2text}{,} \\BioT5~\citep{pei2023biot5}{,} ProtChatGPT~\citep{wang2024protchatgpt}{,}\\ProteinChat~\citep{guo2023proteinchat}{,} ProteinGPT~\cite{xiao2024proteingpt},text width=19.8em, fill=myred]
        ]
      ]
      [LLM Methods for \\Protein Engineering\\\& Generation \& Translation, edge=black!50, fill=myred
        % [Encoder-based Models, text width=6em, fill=myred
        %   [ProtST~\cite{xu2023protst}, text width=7.0em, fill=myred]
        % ]
        [Protein Engineering Models, text width=7.4em, fill=myred
        [
        ProteinDT~\citep{liu2023text}{,} PLMeAE~\cite{plmeae}{,} \\ Toursynbio~\cite{shen2024toursynbio}, text width=16.3em, fill=myred
        ]
        ]
        [Protein Generation Models, text width=7.2em, fill=myred
          [ProGen \citep{madani2023large}{,} ProtGPT2 \citep{ferruz2022protgpt2}{,}\\
          ProGen2 \citep{nijkamp2023progen2}{,} ProLLaMA~\citep{lv2024prollama}{,}\\Ankh \citep{elnaggar2023ankhoptimizedproteinlanguage}{,}
          PAAG \citep{yuan2024annotation} {,} \\ Pinal \citep{dai2024toward} {,} IgLM \citep{shuai2023iglm}{,}\\
          PALM-H3 \citep{he2024novo}{,}  LM-D~\citep{10.5555/3618408.3620189}, text width=16.6em, fill=myred]
        ]
        [Protein Translation Models, text width=7em, fill=myred
          [ProstT5 \citep{heinzinger2023bilingual}{,} Fold2Seq \citep{cao2021fold2seq} {,} \\
          ProtAgents \citep{ghafarollahi2024protagents},
           text width=16.8em, fill=myred]
        ]
      ]
      [Traditional experimental methods, edge=black!50, fill=myred
        [X-ray \citep{jones1986using}{,} NMR \citep{shukla2023biomolecular}{,} Cryo-EM \citep{lyumkis2019challenges}{,}\\
        CryoDRGN~\citep{zhong2021cryodrgn}{,} CryoGAN~\citep{gupta2021cryogan}{,}\\
        CryoSTAR~\citep{li2024cryostar}{,} E2gmm~\citep{chen2021deep}, text width=24.1em, fill=myred]
      ]
    ]
    % [Capability,edge=black!50, fill=mygreen, minimum height=1.2em
    %   [Structure Prediction,text width=5.5em,fill=mygreen
    %     [SSA \citep{bepler2019learning}{,} ESM-1b \citep{rives2021biological}{,} ESM-1v \citep{meier2021language}{,} \\AlphaFold2~\citep{jumper2021highly}{,}
    %     ESM-IF~\citep{hsu2022learning}{,} ESM-2 \citep{lin2023evolutionary}{,}\\AminoBERT\citep{chowdhury2022single}{,} 
    %     Prottrans \citep{elnaggar2021prottrans}{,} ProtST~\cite{xu2023protst}{,}\\ProteinBERT~\citep{brandes2022proteinbert}{,} 
    %     MSA Transformer~\citep{rao2021msa}{,} Tranception \citep{notin2022tranception}, text width=28em, fill=mygreen]
    %   ]
    %   [Function Prediction,text width=5.5em,fill=mygreen
    %     [OntoProtein \citep{zhang2022ontoprotein}{,}  ESM-GearNet \citep{zhang2023systematic}{,}ProTranslator~\citep{xu2022protranslator}{,}\\ Prot2Text~\citep{abdine2024prot2text}{,}
    %     SaProt \citep{su2023saprot}, text width=28em, fill=mygreen]
    %   ]
    %   [Sequence Generation,text width=5.5em,fill=mygreen
    %     [ProLLaMA~\citep{lv2024prollama}{,} ProGen \citep{madani2023large}{,} ProtGPT2 \citep{ferruz2022protgpt2}{,} \\ProteinDT~\citep{liu2023text}{,}ProGen2 \citep{nijkamp2023progen2}{,}
    %     IgLM \citep{shuai2023iglm}{,}\\ PALM-H3 \citep{he2024novo}, text width=24em, fill=mygreen]
    %   ]
    %   [Sequence Translation,text width=5.5em,fill=mygreen
    %     [ProTranslator~\citep{xu2022protranslator}{,}ProstT5 \citep{heinzinger2023bilingual}{,} Fold2Seq \citep{cao2021fold2seq}{,} \\Ankh \citep{elnaggar2023ankhoptimizedproteinlanguage}{,}
    %     ProtChatGPT \citep{wang2024protchatgpt}{,} ProteinChat~\citep{guo2023proteinchat}{,} \\ProteinGPT \citep{xiao2024proteingpt}{,}BioT5~\citep{pei2023biot5}{,} BioTranslator~\citep{xu2023multilingual}, text width=26em, fill=mygreen]
    %   ]
    % ]
    [Evaluation Metrics,edge=black!50, fill=mygreen, minimum height=1.2em
        [Structure Prediction Metrics, text width=7.6em,fill=mygreen
        [RMSD~\cite{li2013difficulty}{,} GDT-TS~\cite{zemla2003lga}{,} TM-Score~\cite{zhang2004scoring}{,}\\lDDT~\cite{mariani2013lddt}{,} pLDDT ~\cite{guo2022alphafold2}, text width=22.3em, fill=mygreen]
        ]
        [Function Prediction Metrics, text width=7.6em,fill=mygreen
        [Accuracy{,} Precision{,} Recall{,} F1-score{,} AUC{,} BLEU \cite{papineni2002bleu}{,} \\ROUGE-L \cite{lin2004rouge}{,} METEOR \cite{banerjee2005meteor}{,} \\Subcellular Localization~\cite{briesemeister2010yloc}{,} Stability~\cite{cheng2006prediction}{,}\\Homology Detection~\cite{altschul1990basic,hamamsy2024protein}{,} \\ Solubility~\cite{hebditch2017protein}{,} Mutation Effect Prediction~\cite{mansoor2022accurate},text width=22.3em, fill=mygreen]
	]
        [Sequence Generation Metrics, text width=7.6em,fill=mygreen
	    [Perplexity~\cite{hesslow2022rita}{,} Diversity~\cite{bywater2015prediction,mcgee2021generative}{,}\\Novelty~\cite{truong2023poet}{,} Fr\'echet Protein Distance~\citep{jiang2008protein}{,} \\Foldability~\cite{baek2021accurate,magliery2015protein}{,} Recovery~\cite{watson2023novo}, text width=22.3em, fill=mygreen]
            ]
      ]
  ]
\end{forest}
\caption{Taxonomy of Protein Large Language Models.}
% \YS{what's the difference between problem and capability? People might think where does protein folding fit in, due to the popularity of AlphaFold.}\WJ{added detailed dataset and benchmark for problems}
\label{fig:taxonomy}
\end{figure*}

