\clearpage


% Define a new command to center rotated tables
\newcommand{\centeredrottable}[1]{%
    \begin{center}
    \vspace*{\textheight}  % Adjust this value to move table up/down
    \begin{table*}[!h]
    \centering
    #1
    \end{table*}
    \end{center}
}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Rotated Table I: LLM Methods for Protein Understanding and Prediction
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \renewcommand{\arraystretch}{1.2}
\begin{table*}[htbp]
% \vspace{6cm}
  \centering
  \caption{LLM Methods for Protein Understanding and Prediction: Protein Sequence Models}
  \label{tab:protein-engineering-generation-1}
  \small
  \renewcommand{\tabularxcolumn}[1]{m{#1}}
  \begin{tabularx}{\linewidth}{%
      >{\centering\arraybackslash}m{2.5cm}  % Model
      >{\centering\arraybackslash}m{0.7cm}  % Time
      >{\centering\arraybackslash}m{1.5cm}  % Base Model
      >{\centering\arraybackslash}m{5cm}  % Dataset
      >{\centering\arraybackslash}X}         % Keywords
    \toprule
    \textbf{Model} & \textbf{Time} & \textbf{Base Model} & \textbf{Dataset} & \textbf{Keywords} \\
    \midrule
    UniRep~\citep{alley2019unified} & 2019 & BiLSTM & UniRef50 & Representation learning, Stability prediction, Functional effects of mutations \\
    \midrule
    \citet{bepler2019learning} & 2019 & BiLSTM & SCOPe ASTRAL, Pfam, PDB, TOPCONS, CASP12 & Structural property prediction, Soft symmetric alignment, Transmembrane \\ %1
    \midrule
    MuPIPR \citep{zhou2020mutation} & 2020 & BiLSTM & STRING, PDB, SKP1402m, SKP1102s & Protein–Protein Interactions (PPI), binding affinity, buried surface area \\ %1
    \midrule
    CSCS \citep{hie2021learning} & 2020 & BiLSTM & IRD,LANL HIV database,ViPR,NCBI Virus,GISAID  & Viral escape patterns, Constrained Semantic Change Search \\ %1
    \midrule
    ProtTrans~\citep{elnaggar2021prottrans} & 2021 & Transformer-XL, XLNet, BERT, Albert, Electra, T5 & UniRef, BFD & Protein secondary structure, sub-cellular localization, membrane vs. water-soluble \\ %1
    \midrule
    ESM-1b~\citep{rives2021biological} & 2021 & Transformer & Uniparc & Large-scale pretraining, protein structure, functional effects of mutations \\ %1
    \midrule
    ESM-1v~\citep{meier2021language} & 2021 & ESM-1b & Uniref90 & Functional effects of mutations, zero-shot prediction \\ %1
    \midrule
    ESM-2, ESMFold~\citep{lin2023evolutionary} & 2023 & Transformer & UniRef, PDB, CAMEO, CASP14, MGnify, trRosetta Dataset & Atom-level resolution structure prediction \\ %1
    \midrule
    AminoBERT~\citep{chowdhury2022single} & 2022 & BERT & ProteinNet12, SCOPe ASTRAL & Single-sequence protein structure prediction \\ %1
    \midrule
    TCR-BERT~\citep{wu2024tcr} & 2021 & BERT & VDJdb, PIRD, LCMV dataset & TCR–antigen binding \\ %1
    \midrule
    MSA Transformer~\citep{rao2021msa} & 2021 & Transformer & UniRef50, UniClust30, CASP13, CAMEO & Multiple sequence alignment, evolutionary relationships \\ %1
    \midrule
    Tranception~\citep{notin2022tranception} & 2022 & Transformer & UniRef & Homologous sequences retrieval, fitness prediction \\ %1
    \midrule
    XTrimoPGLM~\citep{chen2024xtrimopglm} & 2024 & Transformer & UniRef90, ColabFoldDB, UniProt, AlphaFold Database, PDB & 100B parameters, Unified Protein Language Model \\ %1
    \midrule
    TurNuP \citep{kroll2023turnover} & 2022 & ESM-1b &BRENDA, UniProt, Sabio-RK &Turnover number predictions, Differential Reaction Fingerprints \\ %1
    \midrule
    CLEAN \citep{yu2023enzyme} & 2023 & ESM-1b &UniProt,SwissProt   & Contrastive Learning, Enzymatic function prediction \\ %1
    \midrule
    DeepTMP \citep{lin2023deep} & 2023 & ESM-MSA-1b & PDB, PDBTM, UniRef30, BFD & Transfer learning, Transmembrane protein complexes,Inter-chain Contact Prediction \\ %1
    \midrule
    vcMSA~\citep{mcwhite2023leveraging} & 2023 & ProtT5-XL-UniRef50 & Quantest2, HOMSTRAD, UniRef50 & MSA identification, Reciprocal Best Hits\\ %1
    \midrule
    Poet~\citep{truong2023poet} & 2023 & Transformer  & UniRef50, UniRef100, ProteinGym & Homologous Sequences, Retrieval-augmented LM \\ %1
    \bottomrule
  \end{tabularx}
\end{table*}


\clearpage

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Rotated Table II: LLM Methods for Protein Understanding and Prediction
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{table*}[htbp]
  \centering
  \caption{LLM Methods for Protein Understanding and Prediction: Structure-Integrated and Knowledge-Enhanced Models}
  \label{tab:PUP_structure_knowledge}
  \small
  \renewcommand{\tabularxcolumn}[1]{m{#1}}
  \begin{tabularx}{\linewidth}{%
      >{\centering\arraybackslash}m{2.5cm}  % Model
      >{\centering\arraybackslash}m{0.7cm}  % Time
      >{\centering\arraybackslash}m{2.0cm}  % Base Model
      >{\centering\arraybackslash}m{4cm}  % Dataset
      >{\centering\arraybackslash}X}         % Keywords
    \toprule
    \textbf{Model} & \textbf{Time} & \textbf{Base Model} & \textbf{Dataset} & \textbf{Keywords} \\
    \midrule
    ProteinBERT~\citep{brandes2022proteinbert} & 2022 & BERT & UniRef90, TAPE & GO annotations, protein structures, post-translational modifications, biophysical properties \\ %2
    \midrule
    OntoProtein~\citep{zhang2022ontoprotein} & 2022 & ProtBert, Bert & ProteinKG25, UniRef100, TAPE, STRING, SHS27k, SHS148k & Knowledge graphs, gene ontology, PPI, structure prediction \\ %2
    \midrule
    ProteinCLIP~\citep{wu2024proteinclip} & 2024 & ESM2, ProtT5, Text-Embedding-3-Large & UniProt & Contrastive learning, PPI, homology identification \\ %2
    \midrule
    SaProt~\citep{su2023saprot} & 2023 & ESM2 & AlphaFoldDB, UniProt, ProteinGym, ClinVar, thermostability, metal ion binding, DeepLoc, TAPE, PEER, FLIP, PDB & Structure-aware vocabulary, Foldseek \\ %2
    \midrule
    ESM-GearNet ~\citep{zhang2023systematic} & 2023 & GVP, GearNet, CDConv &  AlphaFold Database, GO~\citep{gligorijevic2021structure}, Atom3D & Structural encoders for protein modeling \\ %2
    \midrule
    SES-Adapter\citep{tan2024simple} & 2024 & ESM2, ProtBert, ProtT5, Ankh & GO \citep{gligorijevic2021structure} & Parameter-Efficient Fine-Tuning, Structure Representation \\ %2
        \midrule
     PromptProtein ~\citep{wang2022multi}, &2023 & Transformer & UniRef50, PDB, STRING, GO~\citep{gligorijevic2021structure}& Prompt Learning, Multi-level of structures\\ %2
          \midrule
     SI-pLMs ~\citep{sun2024structure} & 2024 & BERT & Pfam, PDB, AlphaFold Database & Variant Effect Prediction, Structural Information\\ %2
     \midrule
     \citet{zhang2024structure} & 2024 & ESM-2 & SCOPe, GO and EC \citep{gligorijevic2021structure}, Swiss-Prot & Remote Homology Detection, Structural Information \\ %2
         \midrule
   S-plm~\citep{wang2025s} & 2025 &ESM3, discrete diffusion & BPTI, RMSD, Apo/holo, Fold-switch, ATLAS  & Contrastive Learning, Structural Information \\ %2
    \midrule
    \citet{wu2023integration} & 2023 & ESM-2, MSA-Transformer, GVP-GNN, EGNN, SE(3)-Transformer, Schnet, DimeNet& CASP, DB5.5, DIPS, PDBbind & Geometric Deep Learning \\ %2
    \midrule 
    CCPL~\citep{zheng2024ccpl} & 2023-2024 & GVP-GNN, ESM-2 & PDB,  AlphaFoldDB, ProteinGym, trRosetta, CASP14, CATH, Ts 50\&Ts500 & Contrastive Learning, Structure-
Sequence Pairing \\ %2
    \midrule
    KeAP~\citep{zhou2023protein} & 2023 & ProteinKG25 & ProteinNet,TAPE  & Knowledge Graph, Contrastive Learning \\ %2
    \midrule
    MolBind~\citep{xiao2024molbind} & 2024 & SciBERT, GIN, Uni-Mol  & MolBind-M4, CASF-2016  & Contrastive Learning, Protein-text-molecule Alignment \\ %2
    \bottomrule
  \end{tabularx}
\end{table*}


\clearpage


\begin{table*}[htbp]
  \centering
  \caption{LLM Methods for Protein Understanding and Prediction: Protein Description and Annotation Models}
  \label{tab:PUP_description_annotation}
  \small
  \renewcommand{\tabularxcolumn}[1]{m{#1}}
  \begin{tabularx}{\linewidth}{%
      >{\centering\arraybackslash}m{2.5cm}  % Model
      >{\centering\arraybackslash}m{0.7cm}  % Time
      >{\centering\arraybackslash}m{2.0cm}  % Base Model
      >{\centering\arraybackslash}m{4cm}  % Dataset
      >{\centering\arraybackslash}X}         % Keywords
    \toprule
    \textbf{Model} & \textbf{Time} & \textbf{Base Model} & \textbf{Dataset} & \textbf{Keywords} \\
    \midrule
    ProtST~\cite{xu2023protst} & 2023 & ProtBert, PubMedBERT, etc.& ProtDescribe & Multimodal learning, protein function annotation, zero-shot text-to-protein retrieval \\ %3
    \midrule
    ProtChatGPT~\cite{wang2024protchatgpt} & 2024 & ESM-1b, Transformer & PDB-QA, ProteinKG25 &Protein Q\&A, cross-modal protein retrieval, qualitative dialogs \\ %3
    \midrule
    ProteinChat~\cite{guo2023proteinchat} & 2023 & ESM-IF1, Vicuna-13B & RCSB-PDB Protein Description & Interactive protein inquiries, automated protein understanding \\ %3
       \midrule
    Prot2Text~\citep{abdine2024prot2text} & 2024 & RGCN, ESM2, GPT2  & SwissProt & Multimodality, textual function prediction \\ %3
    \midrule
    ProTranslator~\citep{xu2022protranslator} & 2022 & DeepGOCNN, Transformer & CAFA3, SwissProt, GOA, Reactome, KEGG, MSigDB & Function annotation based on text description, text description generation \\ %3
     \midrule
    BioTranslator~\citep{xu2023multilingual} & 2023 & PubMedBERT & GOA, Swiss-Prot, CAFA3, STRING, GeneCards, Tabula Muris, Tabula Sapiens, Tabula Microcebus, GDSC, STITCH, Monarch Initiative, Reactome & Multimodality, text-to-bio-identity translation \\ %3
    \midrule
    BioT5~\citep{pei2023biot5} & 2023 & T5 & ZINC20, UniRef, C4, PubMed articles, PubChem, ChEBI20, SwissProt, MoleculeNet, PEER, BindingDB, BioSNAP, HPRD, Yeast PPI dataset & SELFIES-based molecular representation, wrapped text for bio-entities \\ %3
    \midrule
    BioT5+~\citep{pei2024biot5+} & 2024 & T5 & MoleculeNet, ChEBI-20, PEER, BioSNAP, BindingDB & Multi-task instruction tuning, Molecular\\
    \midrule
    ProLLaMA~\citep{lv2024prollama} & 2024 & LLaMA2 & UniRef, InterPro & Instruction understanding, protein understanding and generation \\ %3
     \midrule
    ProteinGPT~\cite{xiao2024proteingpt} & 2024 & ESM-2, ESM-IF1, Vicuna, LLaMA-2, LLaMA-3 & ProteinQA & Multimodal, interactive protein Q\&A \\ %3
    \midrule
    ProLLM\citep{jin2024prollm} & 2024 & Flan-T5-large & Human, STRING, Mol-Instructions & Chain-of-Thought, PPI \\
    \bottomrule
  \end{tabularx}
\end{table*}

\clearpage
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Rotated Table 2: Generative LLM Methods for Protein
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{table*}[htbp]
% \vspace{6cm}
  \centering
  \caption{LLM Methods for Protein Engineering, Generation and Translation}
  \label{tab:generative-llm-protein}
  \small
  \renewcommand{\tabularxcolumn}[1]{m{#1}}
  \begin{tabularx}{\linewidth}{%
      >{\centering\arraybackslash}m{2.5cm}  % Model
      >{\centering\arraybackslash}m{1cm}  % Time
      >{\centering\arraybackslash}m{2.5cm}  % Base Model
      >{\centering\arraybackslash}m{3cm}  % Dataset
      >{\centering\arraybackslash}X}         % Keywords
    \toprule
    \textbf{Model} & \textbf{Time} & \textbf{Base Model} & \textbf{Dataset} & \textbf{Keywords} \\
    \midrule
    ProGen~\cite{madani2023large} & 2020 & Transformer & UniParc, UniProtKB, Swiss-Prot, TrEMBL & Controllable protein generation, de novo protein design \\
     \midrule
    ProGen2~\cite{nijkamp2023progen2} & 2022 & Autoregressive & UniRef50 & Protein generation, de novo protein design \\
     \midrule
    ProtGPT2~\cite{ferruz2022protgpt2} & 2022 & Autoregressive & UniRef50 & Autoregressive transformer, BPE tokenization, zero-shot protein generation \\
     \midrule
    ProLLaMA~\cite{lv2024prollama} & 2024 & LLaMA2 & UniRef50, InterPro & Multi-task, instruction tuning \\
     \midrule
    IgLM~\cite{shuai2023iglm} & 2023 & GPT-style Transformer & OAS Training Data, Thera-SAbDab & Infilling, conditioned generation, controllable diversity \\
     \midrule
    PALM-H3~\cite{he2024novo} & 2024 & ESM2, RoFormer & Observed Antibody Space, CoV-AbDab, BioMap & Strong generalization to novel proteins, interpretability, antibody \\
     \midrule
    ProstT5~\cite{heinzinger2023bilingual} & 2023 & T5, ProtT5 & 3Di from AlphaFoldDB, CASP12/14, NetSurfP2.0 & Bilingual LM, Foldseek, inverse folding \\
     \midrule
    Fold2Seq~\cite{cao2021fold2seq} & 2021 & Transformer & CATH 4.2 & Inverse protein design, fold-level representation \\
     \midrule
    Ankh~\cite{elnaggar2023ankhoptimizedproteinlanguage} & 2023 & T5 & UniRef50, CASP12/14, NetSurfP-2.0, DeepSF, etc & Contact prediction, secondary structure, fold classification, efficiency \\
     \midrule
    ProteinDT~\cite{liu2023text} & 2023 & ProtBert, SciBERT, ProteinDiff, T5 & SwissProtCLAP & Multimodal learning, text-to-protein generation, autoregressive \\
    \midrule
    PLMeAE~\cite{plmeae} & 2025 & ESM-2 & GB1, UBC9 dataset, Ubiquitin  & Protein engineering, automatic biofoundry\\
    \midrule
    ESM-IF~\citep{hsu2022learning} & 2022 & GVP, GNN, Transformer & UniRef50, CATH  & Inverse folding, AlphaFold2 augmented dataset \\
    \midrule
    ESM-3~\citep{hayes2025simulating} & 2024 & Transformer & UniProt, PDB, AlphaFoldDB, Pfam, InterPro, MGnify, JGI, GO Consortium & Multimodal Learning, Evolutionary Simulation\\
    \midrule
    PAAG~\citep{yuan2024annotation} & 2024 & ProtBERT, SciBERT & ProtAnnotation & Text alignment, annotation\\
        \midrule
    Pinal~\citep{dai2024toward} & 2024 & T2struct, SaProt-T & SwissProt, UniRef50-ProTrek & Multi-step, functional labels\\
    \midrule
    ProtAgents~\citep{ghafarollahi2024protagents} & 2024 & GPT-4, Chroma,  OmegaFold & 
GPTProteinPretrained & Multi-agent, de novo protein design, protein folding\\
\midrule
Toursynbio~\citep{shen2024toursynbio} & 2024 & InternLM2-7B & ProteinLMDataset & Multi-modal, agent, interactive\\
\midrule
LM-DESIGN~\citep{10.5555/3618408.3620189} & 2024 & ESM-1b, ESM-2, ProteinMPNN & CATH 4.2, CATH 4.3, TS50, TS500 & De novo protein design, protein folding\\
    \bottomrule
  \end{tabularx}
\end{table*}

\clearpage
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Rotated Table 3: Summary of Datasets for Protein Language Model
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \renewcommand{\arraystretch}{1.2}

\begin{table*}[htbp]
% \vspace{6cm}
  \centering
  \caption{Summary of Datasets for Protein Language Model}
  \label{tab:protein-datasets}
  \small
  \renewcommand{\tabularxcolumn}[1]{m{#1}}
  \begin{tabularx}{\linewidth}{%
      >{\centering\arraybackslash}m{2cm}  % Category
      >{\centering\arraybackslash}m{4cm}    % Dataset
      >{\centering\arraybackslash}m{1cm}  % Last Update
      >{\centering\arraybackslash}m{1cm}    % Scale
      >{\centering\arraybackslash}X}         % Keywords
    \toprule
    & \textbf{Dataset} & \textbf{Last Update} & \textbf{Scale} & \textbf{Keywords} \\
    \midrule
    \multirow{17}{*}{Pretraining} 
      & UniProtKB/Swiss-Prot~\cite{boutet2016uniprotkb} & 2025 & 573K & Manually curated, high-quality annotations, reviewed \\
      & UniProtKB/TrEMBL~\cite{m1999edittotrembl} & 2025 & 253M & Computationally annotated, unreviewed, automated predictions \\
      & UniRef Clusters~\cite{suzek2015uniref} & 2025 & $>$250M & Clustered sequences, reduced redundancy, hierarchical organization \\
      & Pfam~\cite{finn2006pfam} & 2024 & 22k & Protein families, HMMs, functional domains \\
      & PDB~\cite{bank1971crystallography} & 2025 & 231K & Protein structures, crystallography, molecular modeling \\
      & BFD~\cite{steinegger2018clustering} & 2021 & 2.5B & Massive protein database, sequence clustering, structure prediction \\
      & UniParc~\cite{bairoch2005universal} & 2025 & $>$250M & Non-redundant, protein sequence archive, database cross-referencing \\
      & PIR~\cite{barker2001protein} & 2025 & 513M & Protein sequence database, functional annotation, evolutionary classification \\
      & AlphaFoldDB~\cite{tunyasuvunakool2021highly} & 2025 & $>$200M & Predicted protein structures, deep learning, proteome coverage \\
     \midrule
    \multirow{13}{*}{Benchmark} 
      & CASP~\cite{kryshtafovych2019critical} & 2024 & N/A & Protein structure prediction, modeling competitions \\
      & ProteinGym~\cite{notin2024proteingym} & 2024 & 2.7M & Protein mutations, deep mutational scanning \\
      & TAPE~\cite{rao2019evaluating} & 2021 & $\sim$120K & Protein embeddings, sequence modeling \\
      & CATH~\cite{orengo1997cath} & 2024 & $>$150M & Structure classification, evolutionary relationships, domain hierarchy \\
      & PEER~\cite{xu2022peer} & 2022 & $>$60K & Protein understanding, multi-task benchmark, sequence evaluation \\
      & ExplorEnz~\cite{mcdonald2009explorenz} & 2025 & 8K & Enzyme classification, EC numbering, catalytic reactions \\
      & HIPPIE~\cite{schaefer2012hippie} & 2022 & 39K & Human protein interactions, network analysis \\
      & ProteinLMBench~\cite{shen2024fine} & 2024 & 893K & Protein language understanding, multiple-choice QA, model evaluation \\
    \bottomrule
  \end{tabularx}
\end{table*}

\clearpage
