Decades of research have generated extensive network biology data, revealing that systems, from cells to organisms, can be considered molecular networks and have been reported in extensive literature \cite{Network_1,Network_2,Network_3}.
These data resources have been compiled into functional biological knowledge bases, such as KEGG \cite{kanehisa2000kegg} and STRING \cite{DataSTRING}, framing bio-network resources \cite{GRN,PPI} that document various interactions between molecules (e.g., genes or proteins) to describe how molecular behaviors relate to biological systems.
These knowledge bases are now widely used to mine and interpret wet-lab experimental data, enabling researchers to study disease mechanisms \cite{DB_1}, drug interactions \cite{DB_2}, and potential therapeutic targets \cite{DB_3}.


% Several biological knowledge bases have been established that compile such data into 
% These networks capture interactions between molecules, such as genes or proteins, and describe how molecular behaviors relate to biological systems.
% These networks, which capture molecular (gene or protein) interactions, that describe how molecular behaviors relate to biological systems, have been compiled into various knowledge bases, such as KEGG \cite{kanehisa2000kegg} and STRING \cite{DataSTRING}, framing functional bio-networks \cite{GRN,PPI}. that describe how molecular behaviors relate to biological systems.
% These knowledge bases are now widely used to mine and interpret wet-lab experimental data, enabling researchers to study disease mechanisms \cite{DB_1}, drug interactions \cite{DB_2}, and potential therapeutic targets \cite{DB_3}.

\begin{figure}[t]
\centering 
\includegraphics[width=\linewidth]{FIG/stroy_fig.pdf}
\caption{Biological knowledge bases lack specificity when integrating experimental data. This example illustrates two experimental datasets with different mutations that share the same disease pathway structure, yet fail to reveal the distinct interactions responsible for these differences.}
\label{fig:story_fig}
\end{figure}


While biological knowledge bases are comprehensive and continuously updated, a main concern remains: they lack specificity for experimental data.
The bases cannot provide the information on which interactions are more relevant to the given data, even though their main purpose is to interpret the data.
% These databases cannot provide targeted interactions within a bio-network that are more relevant to the data, even though their primary purpose is to interpret such data.
As shown in Figure \ref{fig:story_fig}, two experimental datasets from the same disease with different mutations share the same network structure, but fail to reveal which distinct interactions account for these differences.
However, inferring key molecular interactions is crucial for understanding the potential roles of genes or proteins, potentially accelerating new biomarker discovery \cite{bionetworkperspective2024}.
% when retrieving these databases, particularly interpreting experimental data, \textit{they lack specificity regarding which interactions within a bio-network are more relevant to these given data} \cite{bionetworkperspective2024}. 
% \yd{can we explain this question in more plain text? it may be difficult for some reviewers who are not familiar with the background knowledge to understand this problem.} 
% \yd{maybe we explicitly mention something is a critical/significant concern under this problem, and we expand the introduction for the gene stuff.}
Conceptually, in-lab experimental data, such as amino acid sequences or gene expression profiles, is typically generated under specific research objectives or experiments, often focusing on proteins or genes of interest.
However, knowledge bases provide only generic networks.
They collect “meaningful” molecular interactions via automated text mining and manual curation \cite{KEGG2024}, but do not incorporate the specific data that “confer such meaningfulness” into the network-building process.
As a result, they cannot provide data-specific interactions, and this generalization can lead to misinterpretations of experimental results.
% When faced with newly discovered genes or proteins, understanding their potential role in a biological context remains challenging.
% {\color{blue}as shown in Figure. 1,} 
% for example, mapping different patients’ gene expression data with differential features onto cancer network resources cannot readily yield sub-networks or unique interactions to explain patient conditions, such as distinct cancer subtypes \cite{bionetworkperspective2024,GeSubNet2024}.
A few complementary bio-tools, such as BLAST \cite{BLAST1}, combined with downstream analytic methods \cite{enrichment}, can help infer bio-networks.
However, these tools are not user-friendly for non-experts, as they require domain expertise for implementation and information management \cite{BLAST2}.
Moreover, the downstream analysis generally provides empirical explanations for the data, and sometimes, extensive in-lab evaluation, such as pairwise examinations among hundreds to thousands of genes \cite{zaman2013signaling}, is needed.
There is thus an urgent need for a method that integrates bio-network resources with experimental data to infer targeted interactions, thereby facilitating the efficient use of knowledge bases.
% \yd{it could be better to imply what could be the potential output. For example, we may say there lacks a comprehensive benchmark for people to compare performances and choose the appropriate models / there lacks a method that is efficient enough for people to deploy models in industrial businesses, etc.}


% \yd{we need to make the audience know that the research works below are all under the same context and problem settings we highlighted above. it reads like we are opening a new direction so we need to introduce a bit at the beginning.}


Bio-network inference, including computational and learning-based methods, has emerged as a promising solution for mining targeted interactions.
Such bio-network inference has attracted great attention from both computational biology and machine learning researchers \cite{wu2021gaerf,kang2022lr}. 
Existing methods typically form bio-network topology as graph data, embed experimental data as node features, and define proper objective functions to infer meaningful subgraphs.
Computational methods typically incorporate statistical node-centric metrics in graph theory, such as node degrees \cite{RSS,RSS1}, centrality \cite{MDS,MDS1}, betweenness \cite{NACHER201657}, or PageRank scores \cite{PMID:21149343,PPR1}, to evaluate the importance of nodes within in a graph.
The edges connecting these top-ranked nodes, or those highlighted during node evaluations, can be identified as more targeted interactions.
However, such objectives lack explicit inference of interactions, i.e., edges, and are often computationally intractable for large bio-networks \cite{BG_1}.
% They generally use simple measures, such as overall similarity, to integrate node features into the evaluation. 
Moreover, computational methods cannot fully exploit experimental data, as the statistical metrics do not integrate node features into scoring schemes.
By contrast, machine learning methods, particularly graph neural networks (GNNs), model both network topology and node-level attributes for subgraph inference in a data-driven manner \cite{GNNS1, GNNS2}.
These methods set up objective functions such as link prediction and graph reconstruction to explicitly infer targeted interactions. 
More importantly, experimental data can directly influence the objective functions through node aggregation in GNN learning, ensuring that the model outputs are more specific to the data.
However, most existing works are task-specific, and their objectives aim to learn the general graph structure accurately, including irrelevant interactions \cite{wu2021gaerf, zhao2021csgnn}.
Some works propose to gradually infer subgraph structure, mitigating the constraints of prior general bio-network information \cite{kang2022lr,graph1}.
Nonetheless, they do not explicitly learn the distinct interactions unique to different experimental data.
Furthermore, existing methods typically require downstream biological analysis to interpret the model outputs or interactions.
We acknowledge that current explorations in bio-network inference remain nascent.
% Furthermore, these methods can contribute to broader biological studies, as node attributes are flexible and can incorporate various data types, such as multi-omics profiles or structural annotations.
% Recent literature has shown remarkable achievements in some studies, such as gene regulatory network \cite{GRN} and gene-disease association \cite{GeSubNet2024}. 
% \yd{the general purpose of this paragraph is ``these works seem excellent but are actually problematic. they cannot handle the problem we highlighted well (in which way?). So, existing explorations remain nascent."}

In this paper, we study a novel and critical problem of developing a bio-network inference framework that explicitly generates data-specific interactions while maintaining biological plausibility.
We note that this is a non-trivial task. 
In essence, we mainly face three challenges. 
$\rm(\hspace{.18em}1\hspace{.18em})$ 
\textit{Qualitative interaction inference objective.}
The first challenge involves formulating a new objective that not only learns but also qualitatively assesses interactions directly.
The model outputs, therefore, enable the direct interpretation of targeted interactions, eliminating the need for downstream analysis.
$\rm(\hspace{.18em}2\hspace{.18em})$ 
\textit{Pathway modeling.}
One primary focus of knowledge bases is on ``pathways'' \cite{KEGG2024}, i.e., connected multi-interactions, representing a sequence of events where one protein interaction triggers the next, ultimately leading to a defined outcome.
However, existing works often treat all interactions equally and uniformly, focusing primarily on isolated interactions.
A reliable method should explicitly incorporate pathway information into the modeling process to maintain biological plausibility.
$\rm(\hspace{.18em}3\hspace{.18em})$ 
\textit{ML-oriented evaluation.}
Currently, there is no standardized quantitative evaluation framework tailored for machine learning models.
Most evaluations depend on qualitative methods, such as enrichment analysis \cite{enrichment}, to determine whether the resulting interactions are biologically relevant, which requires domain expertise.
A quantitative evaluation method, designed to directly assess the model outputs, would bridge the gap between computational and biological disciplines.

In this paper, we present \method, a deep learning framework for inferring targeted pathways for bio-networks.
To tackle the above challenges,
$\rm(\hspace{.18em}1\hspace{.18em})$ we formulate bio-network inference as a subgraph learning and explanation task.
Subgraphs, contributing most significantly to the learning objective, can be identified as targeted interactions.
$\rm(\hspace{.18em}2\hspace{.18em})$ To ensure these subgraphs capture high-order pathways, technically, we propose two novel models:
\classifier, a hybrid learning model, combines graph neural networks (GNNs) with state-space sequence modeling (Mamba) to learn both local interactions and global pathway-level dependencies;
\explainer, a novel variation of GNNExplainer, designed as a subgraph explanation module, identifies objective-critical pathways.
We take amino acid (AA) sequences as the reference experimental data since many biological databases organize pathway information at the protein level \cite{PDB1}. 
% Several studies, such as AlphaFold \cite{AlphaFold}, focus on AA sequences, but effectively handling such data remains a challenge. 
We utilize a large protein language model (pLM) (ESM-2 \cite{ESM}) to encode AA sequences into graph embeddings, driving \method to mine targeted pathways effectively.
$\rm(\hspace{.18em}3\hspace{.18em})$ We propose an evaluation workflow that directly incorporates model-derived weights of subgraphs to quantitatively assess their biological significance.
Overall, our contributions are:
\begin{itemize}[left=0pt]
\item \textbf{Formulating Bio-network Inference Problem.}
We formulate and make an initial investigation on a novel research problem of inferring data-specific pathways for bio-networks.

\item \textbf{Proposing New Framework.}
\method consists of \classifier and \explainer, tailored for pathway-level modeling, can interpret data directly.
It achieves the best fidelity+ and fidelity- across 10 baselines, showing both sufficiency and necessity.
% Our experiments on the largest knowledge pathway base, KEGG \cite{KEGG2024}, confirm that \method can infer pathways across bio-networks while maintaining biological meaningfulness.


% \method is an effective architecture that combines a VQ-VAE and Neo-GNN, achieving average improvements of 30.6\%, 21.0\%, 20.1\%, and 56.6\% across three metrics on four cancer datasets. More advanced models can be easily integrated into \method.

\item \textbf{Impacting Biological Relevance.}
We propose ML-oriented biological evaluations and a new metric.
The experiments involving 301 bio-networks evaluations demonstrate that pathways inferred by \method maintain biological meaningfulness.

% Biological evaluation results demonstrate that the pathways extracted by \method are biologically meaningful and achieve the best performance across all three aspects of biological function: Breadth, Depth, and Reliability.

\item \textbf{Datasets.}
We collected all available human networks from the KEGG, constructed machine-learning-ready datasets, and will release them soon.

% We preprocessed the raw data by matching gene IDs, filtering noisy nodes and edges, and constructing a machine-learning-ready, labeled dataset for experiments and evaluation.
% We release our dataset with this paper to support continued investigation.
\end{itemize}


\begin{comment}
    

% \yd{we shall transfer the criticism against existing works to challenges in handling the mentioned problem: theoretically, why it is a difficult problem does not rely on what existing works have done.}
Despite their great potential, existing explorations in developing integrative methods for generating meaningful interactions remain nascent; we acknowledge several limitations from modeling, biological plausibility, and evaluation perspectives. \yd{probably we can name each challenge like: \\https://dl.acm.org/doi/abs/10.1145/3637528.3671744}
$\rm(\hspace{.18em}1\hspace{.18em})$
Existing works are typically task-specific \cite{GNNS2,GeSubNet2024}, and their objectives aim to learn the general graph structure accurately, including irrelevant interactions.
This approach does not explicitly learn the distinct interactions unique to different experimental data.
Sometimes, the downstream analysis is still required \cite{zhao2021csgnn}.
$\rm(\hspace{.18em}2\hspace{.18em})$ The primary focus of knowledge bases is on ``pathways'', i.e., connected multi-interactions, representing a sequence of events where one gene/protein interaction triggers the next, ultimately leading to a defined outcome \cite{KEGG2024}.
However, existing methods treat all links uniformly, assuming that each interaction contributes equally to the learning objective.
As a result, they focus only on isolated interactions and are unable to capture the higher-order connectivity of pathways, often violating biological plausibility.
$\rm(\hspace{.18em}3\hspace{.18em})$ Currently, there is no standardized quantitative evaluation framework tailored for deep learning models.
Most evaluations depend on qualitative methods, such as enrichment analysis \cite{enrichment}, to determine whether the resulted interactions are biologically relevant.
A quantitative evaluation method, designed to directly assess the interactions inferred by models, would bridge the gap between computational and biological disciplines.

% \noindent\textbf{Novelty and Contributions.}
% In this paper, we present \method, a deep learning framework that integrates knowledge bases and experimental data to learn and explain data-specific pathways within bio-networks explicitly.
% To tackle the above limitations,
% $\rm(\hspace{.18em}1\hspace{.18em})$ Instead of end-to-end link prediction, we formulate pathway mining as a subgraph learning and explanation task.
% The subgraph, represented by experimental data and contributing most significantly to the learning objective, can be identified as targeted interactions.
% $\rm(\hspace{.18em}2\hspace{.18em})$ To ensure these interactions capture high-order pathways, technically, we propose two novel models incorporated into a large foundation model.
% $\rm(\hspace{.18em}$2-1$\hspace{.18em})$ \classifier, a hybrid learning model, combines graph neural networks (GNNs) with state-space sequence modeling (Mamba) to learn both local interactions and global pathway-level dependencies. 
% $\rm(\hspace{.18em}$2-2$\hspace{.18em})$  \explainer, a novel variation of GNNExplainer, designed as a subgraph explanation module, identifies objective-critical nodes and edges using trainable, directed pathway masking.
% We take amino acid (AA) sequences as the reference experimental data since many biological databases organize information at the protein level \cite{PDB1}, and AA sequences play a central role in determining protein function. 
% Several studies, such as AlphaFold \cite{AlphaFold}, focus on AA sequences, but effectively handling such data remains a challenge. 
% Here, $\rm(\hspace{.18em}$2-3$\hspace{.18em})$ we utilize a large protein language model (pLM) (ESM-2 \cite{ESM}) to encode AA sequences into graph embeddings, driving \method to mine targeted pathways effectively.
% $\rm(\hspace{.18em}3\hspace{.18em})$ We designed an evaluation workflow that directly incorporates model-derived weights from extracted subgraphs to quantitatively assess their biological significance.
% Our experiments on the largest knowledge pathway base, KEGG \cite{KEGG2024}, confirm that \method can generate pathways across various bio-networks while maintaining biological meaningfulness.
% Overall, our contributions are as follows:
% \begin{itemize}[left=0pt]
% \item \textbf{Formulating Bio-Database Retrieval Problem.}
% We formulate the task of mining data-specific interactions within bio-networks as a graph learning and explanation problem, with a particular focus on identifying interactions that form biological pathways.

% \item \textbf{Proposing Data-Driven Pathway Mining Framework.}
% \method consists of \classifier and \explainer, tailored for pathway-level modeling, with the support of a protein language model (pLM) to provide biologically meaningful explanations.

% \item \textbf{Impacting Broad Biological Relevance.}
% Biological analysis results demonstrate that the subgraphs extracted by \method are biologically meaningful and achieve the best performance across all three aspects of biological function: Breadth, Depth, and Reliability.

% \item \textbf{Open Datasets.}
% We collected all available human pathway data from the KEGG online database.
% We preprocessed the raw data by matching gene IDs, filtering noisy nodes and edges, and constructing a machine-learning-ready, labeled dataset for experiments and evaluation.
% We release our dataset with this paper to support continued investigation.
% \end{itemize}



% Since bio-networks can be represented as graph data, researchers have framed the task of mining and explaining targeted interactions as a graph learning problem \cite{wu2021gaerf,zhao2021csgnn,kang2022lr}.
% Specifically, they treat compiled molecular interactions in knowledge bases as edges, embed experimental data as node features, and define suitable objective functions to infer subgraphs.
% Existing methods can generally be categorized into biologically computational methods and graph neural network (GNN)-based methods.
% Computational methods typically incorporate well-established statistical or combinatorial strategies to identify functionally relevant interactions







% Several graph learning methods have been proposed for mining bio-networks \cite{wu2021gaerf,zhao2021csgnn,kang2022lr}.
% These methods treat bio-networks as graphs, embed experimental data as node features, and define learning objectives such as link prediction or reconstruction to capture meaningful graph structures.
% However, (1) existing works are typically task-specific \cite{GNNS2,GeSubNet2024}, and their objectives aim to learn the general graph structure accurately, including irrelevant interactions.
% They do not explicitly learn targeted interactions driven by the data.
% Moreover, (2) link-wise modeling treats all links uniformly, assuming that each interaction contributes equally to the learning objective.
% In plain words, these methods focus only on isolated interactions and are unable to capture the higher-order connectivity of pathways, often violating biological plausibility.




% Decades of research on network biology have revealed that various systems, from cells and organisms to ecosystems, can be represented as molecular networks, compiled into knowledge bases \cite{DataSTRING,KEGG2024}.
% % through both automated text mining and manual curation \cite{DataSTRING,KEGG2024}.
% These resources integrate relationships among genes, proteins, and fundamental biological entities, framing functional networks, such as gene regulatory \cite{GRN} and protein-protein interactions \cite{PPI}, to describe the underlying molecular mechanisms of biological systems. 
% Researchers can further study disease mechanisms {\color{blue}Cite}, drug interactions{\color{blue}Cite}, and potential therapeutic targets {\color{blue}Cite} with these as bases.


% Biological knowledge databases, such as KEGG \cite{kanehisa2000kegg}, which integrate diverse molecular network information (e.g., genomic or proteomic), are widely used to analyze biological systems and interpret wet-lab experimental data.
% While these resources are comprehensive and continuously updated, they lack specificity regarding which interactions in a bio-network are most relevant to specific experimental data \cite{bionetworkperspective2024}.
% This work introduces a deep learning method to integrate the knowledge bases with experimental data to generate targeted interactions.
% Moreover, we focus on a more challenging task: mining meaningful \textit{pathways} in bio-networks, which represent sequences of connected molecular interactions.






% We are interested in the problem of retrieving bio-networks from knowledge bases, particularly when incorporating and interpreting experimental data. 
% In-lab data, such as amino acid sequences or gene expression profiles, is typically generated under specific research objectives or experimental conditions, often focusing on particular proteins or genes.
% % focus on specific proteins or genes identified through experiments.
% However, knowledge bases provide only generic networks, which cannot interpret data-specific interactions, {\color{blue}as shown in Figure. 1.}
%  This generalization can lead to misinterpretations of experimental results.
% While several complementary tools, such as BLAST \cite{BLAST1,BLAST2}, combined with downstream biological analysis \cite{enrichment}, can help quantify these interactions, such empirical analyses require extensive in-lab examinations\cite{GeSubNet2024} and expertise for proper evaluation.

% We propose to tackle this issue with a novel data-driven method that automates the integration of knowledge bases and experimental data to generate more targeted interactions.
% More importantly, we focus on providing biologically sensible explanations of interactions, specifically by mining connected multi-interactions, i.e., pathways, rather than isolated interactions.
% Such pathways represent a sequence of events where one molecular interaction triggers the next, ultimately leading to a defined outcome {\color{blue}Cite}.

% Several graph learning methods have been proposed for mining bio-networks \cite{wu2021gaerf,zhao2021csgnn,kang2022lr}.
% These methods treat bio-networks as graphs, embed experimental data as node features, and define learning objectives such as link prediction or reconstruction to capture meaningful graph structures.
% However, (1) existing works are typically task-specific \cite{GNNS2,GeSubNet2024}, and their objectives aim to learn the general graph structure accurately, including irrelevant interactions.
% They do not explicitly learn targeted interactions driven by the data.
% Moreover, (2) link-wise modeling treats all links uniformly, assuming that each interaction contributes equally to the learning objective.
% In plain words, these methods focus only on isolated interactions and are unable to capture the higher-order connectivity of pathways, often violating biological plausibility.

% \noindent\textbf{Novelty and Contributions.}
% In this paper, we present \method, a deep learning framework that integrates knowledge bases and experimental data to learn and explain data-specific pathways within various bio-networks explicitly.
% Technically, $\rm(\hspace{.18em}1\hspace{.18em})$ we formulate pathway mining as a subgraph learning and explanation task.
% The subgraph, represented by experimental data and contributing most significantly to the learning objective, can be identified as targeted interactions.
% $\rm(\hspace{.18em}2\hspace{.18em})$ To ensure these interactions capture high-order pathways, we propose two novel models.
% The first is \classifier, a hybrid learning model that combines graph neural networks (GNNs) with state-space sequence modeling (Mamba) to learn both local interactions and global pathway-level dependencies. 
% The second is \explainer, a novel variation of GNNExplainer, designed as a subgraph explanation module that identifies objective-critical nodes and edges using trainable, directed pathway masking.
% We take amino acid (AA) sequences as the reference experimental data since many biological databases organize information at the protein level \cite{PDB1}, and AA sequences play a central role in determining protein function. 
% Several studies, such as AlphaFold \cite{AlphaFold}, focus on AA sequences, but effectively handling AA sequence data remains a challenge. 
% Here, $\rm(\hspace{.18em}3\hspace{.18em})$ we utilize a large protein language model (pLM) (ESM-2 \cite{ESM}) to encode AA sequences into graph embeddings, driving \method to mine targeted pathways effectively.
% Our experiments on the largest knowledge pathway base, KEGG \cite{KEGG2024}, confirm that \method can generate pathways across various bio-networks while maintaining biological meaningfulness.
% Overall, our contributions are as follows:
% \begin{itemize}[left=0pt]
% \item \textbf{Formulating Bio-Database Retrieval Problem.}
% We formulate the task of mining data-specific interactions within bio-networks as a graph learning and explanation problem, with a particular focus on identifying interactions that form biological pathways.

% \item \textbf{Proposing Data-Driven Pathway Mining Framework.}
% \method consists of \classifier and \explainer, tailored for pathway-level modeling, with the support of a protein language model (pLM) to provide biologically meaningful explanations.

% \item \textbf{Impacting Broad Biological Relevance.}
% \askFillIn{fill in biological analysis experiments and results, referring to our GeSubNet.}

% \item \textbf{Open Datasets.}
% \askFillIn{fill in the description of our collected DB.}

% \end{itemize}





% Specifically, we focus on proteins, as many biological databases organize their information at the protein level \cite{PDB1}, making protein-centric data, such as amino acid (AA) sequences {\color{blue}Cite}, essential for reliable retrieval and functional interpretation.
% To this end, we acknowledge three challenges: First, current databases do not fully incorporate AA sequence data when building bio-networks, and standard text-mining tools are ill-suited for handling sequence information. Thus, a specialized learning strategy that jointly leverages AA sequences and bio-network structures is needed.
% Second, complexity of targeted interactions.
% Targeted interactions often extend beyond simple protein–protein links to form longer interaction chains or entire pathways. Incorporating these higher-order connections into models can provide deeper insights and more robust explanations of experimental data.
% Third, requirement for specialized models.
% Although tools like BLAST can align sequences, large-scale machine learning approaches are necessary to capture the intricate relationships in AA sequence data. This makes integrating AA sequences into existing network databases computationally intensive and technically demanding, thereby preventing straightforward mapping of experimental AA sequence data onto these networks.

% \textbf{Contributions and Novelty.} We present \method, a deep learning framework designed to explain extended functional pathways in the KEGG biological database. 
% Technically, (1) we formulate this challenge as a graph learning and explanation task, where various functional networks are represented as graphs and AA sequences are embedded as node features. 
% By adopting a graph-based explainable learning approach, we can identify pathways that contribute most significantly—thereby highlighting these targeted interactions.
% (2) To achieve this, we first utilize a large protein language model (pLM) to encode AA sequences into graph embeddings. 
% (3) Next, we propose two modules tailored for long-pathway analysis:
% \classifier, a hybrid architecture that combines graph neural networks (GNNs) with state-space sequence modeling (Mamba) to capture both local interactions and global pathway-level dependencies, and \explainer, a subgraph learning module that identifies functionally critical nodes and edges through trainable pathway masks.
% Overall, our contributions are fourfold:

% \begin{itemize}[left=0pt]
% \item \textbf{Formulating New Database Retrival Problem.}


% %\item \textbf{Developing Novel ML-based Framework.}
% %\item \textbf{Developing Novel Automated Data Integration Solution.}
% \item \textbf{Proposing automated data integration methodology.}

% \item \textbf{Impacting Broad Biological Relevance.}

% \item \textbf{Integrated Datasets.}

% \end{itemize}







% The text-mining tool has no interface to incorporate experimental data.
% Hence we form a classification task.
% BLAST is compute-intensive, we use LPM.
% Form a graph-learning and GNNExplainer.


% % such as KEGG \cite{kanehisa2000kegg}, integrate genomic, chemical, and network information to analyze biological systems and interpret large-scale datasets, including protein or transcriptomic data.
% % These databases consist of molecular building blocks (genes and molecules), molecular networks (interactions and reactions), and mechanisms to link building blocks to networks.
% % While these knowledge bases are comprehensive and continuously updated, they do not directly provide specialized components tailored for different experimental datasets.
% % This work introduces a deep learning method to integrate network corpora from these databases with wet-lab experimental data, enabling the explanation of more targeted, data-specific networks.

% % Gene network databases, such as KEGG \cite{kanehisa2000kegg}, are widely used resources that serve as baseline references and tools for studying and modeling biological systems. 


% % Biological networks, such as gene regulatory and protein-protein interaction networks, represent the relationships between genes, proteins, and other fundamental biological entities. 
% % These networks are essential for understanding the underlying mechanisms of complex biological systems, including how genes and proteins work together to regulate cellular processes, respond to environmental changes, and maintain necessary homeostasis. 
% % Researchers can further study disease mechanisms, drug interactions, and potential therapeutic targets with these as bases.

% % Gene network databases are widely used resources that serve as baseline references and tools for studying and modeling biological systems. 
% % KEGG, for instance, is the largest database of biological pathways, it builds its gene pathway networks through both automated text mining and manual curation. 
% % Relationships between genes and proteins are extracted from published literature and integrated into structured, classified pathways, which are then used to represent various biological process types such as metabolism, signaling, and disease progression.
% % The most common way to utilize KEGG resources is to select a classic network belonging to a specific class (e.g., cancer pathways, metabolic pathways) in the database as a reference for the biological system under investigation. 
% % This reference helps researchers map experimental gene data to known pathways, providing a foundation for understanding how genes behave (expression) under specific conditions. 
% % Databases like KEGG have been widely used in studies such as pathway enrichment analysis, drug-target interaction prediction, metabolic engineering, and disease pathway mapping.

% % Despite their importance in supporting biological research and bioinformatics applications, these databases face notable challenges in their construction and application. 
% % A major limitation is their failure to incorporate amino acid (AA) sequence data into the graph-building process, even though AA sequences play a central role in determining protein function, structure, and evolutionary conservation. 
% % Proteins are the key players in carrying out gene functions, and their AA sequences determine their ability to interact with other molecules, bind substrates, catalyze reactions, and respond to environmental signals.
% % This is because incorporating such long-text information (the AA sequences) is challenging for several reasons: AA sequences cannot be directly mined using standard text-mining tools, and their analysis requires specialized models. 
% % For example, tools like BLAST are used to align sequences, and large-scale machine learning models are needed to handle the complex relationships within sequence data. 
% % These requirements make the integration of AA sequences into network databases computationally intensive and technologically demanding.
% % Due to this dilemma, experimental AA sequence data cannot be easily mapped onto existing networks in these databases. 
% % This limitation delays researchers’ ability to explore the biological significance of newly discovered networks by comparing them to the baseline networks in the database. 
% % Without this comparison, understanding the exact role of a newly discovered gene or protein in a specific biological context becomes challenging. 
% % Consequently, critical insights into novel network functions, such as classifying new genes into existing functional categories (e.g., signaling, metabolism, disease-related), are often left unexplored.

% As biological network data grows rapidly—from both computational and experimental sources—this issue becomes increasingly important to resolve. 
% New experimental techniques generate vast amounts of sequence data, while computational methods produce more accurate networks of gene functions and protein interactions. 
% However, without incorporating sequence-level information into these models, the reference networks provided by databases are forced to rely heavily on basic graph structures, which often offer limited functional insights.
% This limitation arises from the nature of biological networks, where the same topological relationships between gene sets—such as positive or negative regulation loops—can commonly exist and contribute to the formation of gene networks with different functions. 
% For instance, genes involved in different metabolic pathways might share similar regulatory relationships, making it difficult to distinguish their specific roles based solely on graph structure. 
% However, the true biological functions lie in the sequences of proteins that carry out these gene functions, making sequence-level information crucial for more accurate functional characterization of networks. 
% Moreover, relying only on basic graph structures constrains the interpretation of existing networks and hinders researchers from comparing and integrating new experimental findings at the sequence level into the broader biological context or linking new experimental discoveries to known pathways and functional categories, limiting the scope of their analysis.

% New approaches that incorporate AA sequences into database biological network modeling are needed to address these limitations, enabling more detailed and scalable analyses. 
% Incorporating sequence-level information will allow for a more precise understanding of how proteins interact, how mutations impact function, and how new discoveries fit into the existing biological knowledge base. 
% Therefore, in this research, we propose a novel approach that combines pre-trained large-scale sequence models (which capture the vast diversity of sequence data) with graph learning models (which model the complex relationships between proteins) to tackle this challenge. 

% % Pathway mapping is a bioinformatics approach used to link molecular data, such as genes, proteins, or metabolites, to known physical pathways \citep{NM_1,NM_3}.
% % These pathways represent complex networks of biochemical reactions or signaling processes that occur within cells, governing various functions like metabolism, cell communication, and disease progression. \citep{GenIn,biomarker}.

% % To perform pathway mapping, researchers typically analyze in-lab experimental results or data using a knowledge repository, with KEGG being a well-known database. 
% % To begin, they prepare a list of genes or proteins of interest, ensuring they are in formats compatible with KEGG (such as Entrez gene IDs or identifiers) \citep{kanehisa2000kegg}. 
% % These are then input into the KEGG Mapper tool, which enables the visualization of how the proteins are distributed across various cellular processes.
% % The mapping results are displayed as interactive pathway maps, where the proteins are highlighted, providing clear insight into their roles within specific biological contexts \citep{becskei2000engineering}.
% % However, the initial design of knowleadge database aims to serve broader research needs, several challenges remain when using pathway mapping.

% % \textbf{- Overlapping pathways across categories.}
% % Many proteins participate in multiple biological processes, meaning they can appear in different pathway classes. 
% % For instance, a metabolic pathway, such as insulin signaling, might be relevant to cancer pathway of PI3K-Akt.
% % This overlap can cause redundancy when the same pathways appear in different contexts.\\
% % \textbf{- Pathway complexity.}
% % KEGG pathways often represent complex biological systems. 
% % As a result, a single pathway (e.g., cell cycle) can be broken down into multiple sub-pathways or listed under different classes, leading to some repetition when analyzing data in different contexts.

% % Typically, further analysis is needed to carefully interpret the KEGG outputs. 
% % For example, using KEGG in combination with other databases or more dynamic tools (e.g., Reactome, GO) \citep{DataSTRING} can mitigate redundancy issues and provide a more nuanced understanding.
% % In this paper, we address this issue of KEGG by taking a statistical approach using deep learning. 
% % To date,
% % % Deep learning can fully exploit complex data patterns, making data-driven decisions  
% % % Our contributions are as follows:

% % \textbf{- Problem formulation.}
% % We set up mitigating redundancy as a classification task, more importantly, we propose an \emph{explainer} to identify the key features, i.e., links, in a pathway that contribute to class assignment. 
% % This provides insights into how specific pathways relate to certain classes.\\
% % \textbf{- Framework.}
% % We hence propose \method, which ($i$)forms each pathway as a graph, ($ii$)introduces a GNN model that incorporates graph classification to learn probabilistic relationships between each path and various biological classes, and ($iii$)proposes a GNN explainer to further quantify the importance of each feature for class assignment.
% % \\
% % \textbf{- Dataset.}
% % We construct a database for model training and experiments, including 301 pathway samples from six biological systems.
% % Each pathway sample is transformed into graphical representation, where each node represents a protein and edges indicate functional pathways collected in KEGG.\\
% % \textbf{- Simulation experiments.}
% % we xxxx.







\end{comment}
