\begin{figure*}[t]
    \centering
    \includegraphics[width=0.99\linewidth]{FIG/systemov.pdf}
    \caption{
    % {\color{red}
    % Rikuto: need to revise.
    % }
    Overview of ExPath. ExPath comprises three components. (1) Data encoding with a large protein language model and RWPE for node attributes (AA sequences) learning; (2) PathMamba combining graph neural networks with state-space sequence modeling (Mamba) to capture both local interactions and global pathway-level dependencies for pathway information learning; and (3) PathExplainer identifies functionally critical nodes and edges through trainable pathway masks for targeted pathway inference.}
    \label{fig:systemOV}
\end{figure*}


\subsection{Pathways in Biological Knowledge Bases}\label{subsec:pathway}
% Biological pathways are structured representations of molecular interactions and metabolites that drive diverse biological functions. 
Knowledge bases organize these pathways as interconnected graphs, where nodes represent proteins and edges denote interactions (e.g., enzymatic reactions, regulatory effects). 
These pathways help researchers understand how different molecules work together rather than looking at just one interaction at a time.
For example, while protein-protein interactions (PPI) \cite{PPI} capture pair-wise molecular associations, pathways show how many proteins are linked in complex networks. 
This helps scientists see the “big picture,” revealing how different parts of a cell cooperate to perform important tasks \cite{KEGG2024}.
Learning from amino acid (AA) sequence data is challenging due to its inherent complexity. 
Even slight variations can lead to significant structural changes, potentially disrupting protein functionality within pathways. 
Several studies focus on extracting meaningful features from AA sequences, like like AlphaFold \cite{AlphaFold}.
In this paper, we investigate the mapping of AA sequence data to corresponding pathway bio-networks.


% Even slight variations in AA sequences can lead to significant structural changes, potentially disrupting protein functionality within pathways. 
% Learning models like AlphaFold \cite{AlphaFold} also highlight the importance of structure prediction for extracting meaningful features from AA sequences. 
% If we can learn an AA sequence-suitable pathway, we can better understand functional associations between sequence encoding and regulatory mechanisms in the pathway, as well as applications like designing more effective enzyme modifications for metabolic engineering and drug development.



\subsection{Problem Setting}\label{subsec:data}
\noindent\textbf{Definition 1 (Amino Acid sequence data).}
Consider a collection of $M$ AA sequences, denoted as 
$\mathbf{S} = \left\{ S^{(m)} \right\}_{m=1}^{M}$,
where each sequence $S^{(m)}$ has a length $L^{(m)}$ (which may vary across $m$) and is represented as
$
S^{(m)} = \left[s_1^{(m)}, s_2^{(m)}, \dots, s_{L^{(m)}}^{(m)}\right]$.
Here, each amino acid $s_i^{(m)}$ belongs to the standard set of $20$ canonical amino acids.


    % Let us consider a collection of $M$ AA sequences, denoted by 
    % $
    % \mathbf{S} = \Bigl\{S^{(m)}\Bigr\}_{m=1}^{M},
    % $
    % where each sequence $s^{(m)}$ is of length $L^{(m)}$ (potentially varying across $m$) and can be written as
    % $
    % S^{(m)} = \bigl[s_1^{(m)}, s_2^{(m)}, \dots, s_{L^{(m)}}^{(m)}\bigr].
    % $
    % Each amino acid $s_i^{(m)}$ belongs to a standard set of $20$ canonical amino acids. 

\noindent\textbf{Definition 2 (Knowledge bio-networks).}
% A \textit{pathway} refers to a series of interactions or biochemical reactions , which collectively achieve a specific biological function. For instance, pathways can regulate metabolic processes, gene expression, or signal transduction.
The bio-networks can be represented as a graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$, where $\mathcal{V}$ denotes the vertices (e.g., AA sequences) and $\mathcal{E}$ is the set of edges, representing molecular interactions. 
Let $\mathbf{G} = {\{\mathcal{G}^{(m)}\}}_{m=1}^{M}$ denote a dataset comprising $M$ bio-networks. 
Each $\mathcal{G}^{(m)}$ is associated with a label $y^{(m)} \in \mathbf{Y}$, indicating its primary biologically functional class (e.g., {metabolism}, {genetic information processing}, and {human diseases}).

% The bio-network, as compiled in knowledge databases, can be represented as a graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$, where $\mathcal{V}$ denotes the set of vertices (e.g., AA sequences) and $\mathcal{E}$ is the set of edges representing their interactions or regulatory links.
% Let $\mathbf{G} = \{\mathcal{G}^{(m)}\}_{m=1}^{M}$ denote a dataset of $M$ such pathway networks.
% Each pathway network $\mathcal{G}^{(m)}$ is associated with a label $y^{(m)} \in \mathbf{Y}$ that describes its primary biological function class (e.g., \textit{Metabolism}, \textit{Genetic Information Processing}, \textit{Human Diseases}, etc.).

% \noindent\textbf{Definition 2 (Knowledge-based classification).}
% Pathways are systematically cataloged into different function classes $\mathbf{Y} = \{y_1, y_2, \dots, y_{|\mathbf{Y}|}\}$.
% This classification is curated based on current biological knowledge, and each pathway typically belongs to one or more categories reflecting its main functional or disease-related role.
% Our goal is to leverage this curated information to build a pathway classification model while enabling deeper insights into \textit{which substructures} within each pathway drive its class membership.

\noindent\textbf{Problem (Targeted Pathway Inference).}
We frame this research problem as a two-phase graph learning and subgraph explanation task.
Given $\mathbf{G}$ and the labels $\mathbf{Y}$, we focus on two main tasks:
\begin{itemize}[left=0pt]
    \item \textbf{Task-1 (Classification):} 
    Train a classifier $F(\cdot)$ to predict the function label $y^{(m)}$ of an unseen bio-network $\mathcal{G}^{(m)}$, driven by the node features, i.e., AA sequence data $S$.
    To perform accurate classification, it can effectively learn various pathways in biological knowledge bases.
    
    \item \textbf{Task-2 (Explanation):} 
    For each predicted function class $y \in \mathbf{Y}$, employ an explainer $E(\cdot)$ to extract a class-specific subgraph $\hat{\mathcal{G}} \subseteq \mathcal{G}$ that highlights the most influential pathways contributing to the classification outcome.
\end{itemize}


In plain words, a subgraph learned from various biological pathways and capable of accurately predicting its associated bio-network can be considered as representing \textit{targeted pathways}.
Let $\hat{\mathcal{G}}_y = E(\mathcal{G}, y;S)$ denote subgraphs for bio-network $y$, integrated to experimental data $S$.
Here, we hold three hypotheses for our method:
\begin{itemize}[left=0pt]
    \item \textbf{Hypothesis-1.} 
    The classifier $F(\cdot)$  leverages bio-network topology and AA seq features to achieve high prediction accuracy while maintaining balanced performance across all classes.
    
    \item \textbf{Hypothesis-2.} The explainer $E(\cdot)$ infers subgraphs that retain high fidelity for class discrimination and exhibit distinct structures across classes, reflecting unique biological mechanisms.
    % The subgraph $\hat{\mathcal{G}}_y$ typically involves \textit{fewer} critical nodes and edges than the entire pathway graph $\mathcal{G}$, i.e., $|\hat{\mathcal{G}}_y| \ll |\mathcal{G}|$.
    % Furthermore, these subgraphs should exhibit distinctive structures when comparing different functions $y \in \mathbf{Y}$.
    \item \textbf{Hypothesis-3.} 
    Both $F(\cdot)$ and $E(\cdot)$ must maintain biologically plausible mechanisms, ensuring that the inferred subgraphs capture high-order pathways and hold biological meaningfulness.
    % Each function-specific subgraph must remain biologically plausible, preserving essential interactions relevant to pathway-level mechanisms.
    % Such subgraphs can provide deeper insights into the key regulatory modules and mechanisms that underlie each function class.
\end{itemize}

% To this end, we propose two novel models corresponding to Hypotheses-1 and 2, respectively. 
% These models are designed with pathway-level considerations to address Hypothesis-3, as detailed in Section \ref{sec:moethod}. 
% Moreover, we evaluate \method from both machine learning and biological perspectives, as detailed in Section \ref{sec:exp}.