\subsection{Pathology}
\label{app:tasks:pathology}
\begin{tcolorbox}[title={\texttt{conch\_extract\_features}}]
Perform feature extraction on an input image using CONCH.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{input\_image} (\texttt{str}): Path to the input image\\  Example: \texttt{'/mount/input/TUM/TUM-TCGA-ACRLPPQE.tif'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{features} (\texttt{list}): The feature vector extracted from the input image, as a list of floats
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/mahmoodlab/CONCH}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{lu2024conch}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{musk\_extract\_features}}]
Perform feature extraction on an input image using the vision part of MUSK.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{input\_image} (\texttt{str}): Path to the input image\\  Example: \texttt{'/mount/input/TUM/TUM-TCGA-ACRLPPQE.tif'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{features} (\texttt{list}): The feature vector extracted from the input image, as a list of floats
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/lilab-stanford/MUSK}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{xiang2025musk}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{pathfinder\_verify\_biomarker}}]
Given WSI probability maps, a hypothesis of a potential biomarker, and clinical data, determine (1) whether the potential biomarker is significant for patient prognosis, and (2) whether the potential biomarker is independent among already known biomarkers.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{heatmaps} (\texttt{str}): Path to the folder containing the numpy array (\textasciigrave{}*.npy\textasciigrave{}) files, which contains the heatmaps of the trained model (each heatmap is HxWxC where C is the number of classes)\\  Example: \texttt{'/mount/input/TCGA\_CRC'}
\item \texttt{hypothesis} (\texttt{str}): A python file, which contains a function \textasciigrave{}def hypothesis\_score(prob\_map\_path: str) -\textgreater{} float\textasciigrave{} which expresses a mathematical model of a hypothesis of a potential biomarker.  For a particular patient, the function returns a risk score.\\  Example: \texttt{'/mount/input/mus\_fraction\_score.py'}
\item \texttt{clini\_table} (\texttt{str}): Path to the CSV file containing the clinical data\\  Example: \texttt{'/mount/input/TCGA\_CRC\_info.csv'}
\item \texttt{files\_table} (\texttt{str}): Path to the CSV file containing the mapping between patient IDs (in the PATIENT column) and heatmap filenames (in the FILENAME column)\\  Example: \texttt{'/mount/input/TCGA\_CRC\_files.csv'}
\item \texttt{survival\_time\_column} (\texttt{str}): The name of the column in the clinical data that contains the survival time\\  Example: \texttt{'OS.time'}
\item \texttt{event\_column} (\texttt{str}): The name of the column in the clinical data that contains the event (e.g. death, recurrence, etc.)\\  Example: \texttt{'vital\_status'}
\item \texttt{known\_biomarkers} (\texttt{list}): A list of known biomarkers. These are column names in the clinical data.\\  Example: \texttt{['MSI']}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{p\_value} (\texttt{float}): The p-value of the significance of the potential biomarker
\item \texttt{hazard\_ratio} (\texttt{float}): The hazard ratio for the biomarker
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/LiangJunhao-THU/PathFinderCRC}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{liang2023pathfinder}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{stamp\_extract\_features}}]
Perform feature extraction using CTransPath with STAMP on a set of whole slide images, and save the resulting features to a new folder.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{output\_dir} (\texttt{str}): Path to the output folder where the features will be saved\\  Example: \texttt{'/mount/output/TCGA-BRCA-features'}
\item \texttt{slide\_dir} (\texttt{str}): Path to the input folder containing the whole slide images\\  Example: \texttt{'/mount/input/TCGA-BRCA-SLIDES'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{num\_processed\_slides} (\texttt{int}): The number of slides that were processed
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/KatherLab/STAMP}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{elnahhas2024stamp}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{stamp\_train\_classification\_model}}]
Train a model for biomarker classification. You will be supplied with the path to the folder containing the whole slide images, alongside a path to a CSV file containing the training labels.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{slide\_dir} (\texttt{str}): Path to the folder containing the whole slide images\\  Example: \texttt{'/mount/input/TCGA-BRCA-SLIDES'}
\item \texttt{clini\_table} (\texttt{str}): Path to the CSV file containing the clinical data\\  Example: \texttt{'/mount/input/TCGA-BRCA-DX\_CLINI.xlsx'}
\item \texttt{slide\_table} (\texttt{str}): Path to the CSV file containing the slide metadata\\  Example: \texttt{'/mount/input/TCGA-BRCA-DX\_SLIDE.csv'}
\item \texttt{target\_column} (\texttt{str}): The name of the column in the clinical data that contains the target labels\\  Example: \texttt{'TP53\_driver'}
\item \texttt{trained\_model\_path} (\texttt{str}): Path to the *.pkl file where the trained model should be saved by this function\\  Example: \texttt{'/mount/output/STAMP-BRCA-TP53-model.pkl'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{num\_params} (\texttt{int}): The number of parameters in the trained model
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/KatherLab/STAMP}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{elnahhas2024stamp}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{uni\_extract\_features}}]
Perform feature extraction on an input image using UNI.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{input\_image} (\texttt{str}): Path to the input image\\  Example: \texttt{'/mount/input/TUM/TUM-TCGA-ACRLPPQE.tif'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{features} (\texttt{list}): The feature vector extracted from the input image, as a list of floats
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/mahmoodlab/UNI}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{chen2024uni}


\end{tcolorbox}

\subsection{Radiology}
\label{app:tasks:radiology}
\begin{tcolorbox}[title={\texttt{medsam\_inference}}]
Use the trained MedSAM model to segment the given abdomen CT scan.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{image\_file} (\texttt{str}): Path to the abdomen CT scan image.\\  Example: \texttt{'/mount/input/my\_image.jpg'}
\item \texttt{bounding\_box} (\texttt{list}): Bounding box to segment (list of 4 integers).\\  Example: \texttt{[25, 100, 155, 155]}
\item \texttt{segmentation\_file} (\texttt{str}): Path to where the segmentation image should be saved.\\  Example: \texttt{'/mount/output/segmented\_image.png'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \textit{empty dict}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/bowang-lab/MedSAM}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{ma2024medsam}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{nnunet\_train\_model}}]
Train a nnUNet model from scratch on abdomen CT scans. You will be provided with  the path to the dataset, the nnUNet configuration to use, and the fold number  to train the model on.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{dataset\_path} (\texttt{str}): The path to the dataset to train the model on (contains dataset.json, imagesTr, imagesTs, labelsTr)\\  Example: \texttt{'/mount/input/Task02\_Heart'}
\item \texttt{unet\_configuration} (\texttt{str}): The configuration of the UNet to use for training. One of '2d', '3d\_fullres', '3d\_lowres', '3d\_cascade\_fullres'\\  Example: \texttt{'3d\_fullres'}
\item \texttt{fold} (\texttt{int}): The fold number to train the model on. One of 0, 1, 2, 3, 4.\\  Example: \texttt{0}
\item \texttt{output\_folder} (\texttt{str}): Path to the folder where the trained model should be saved\\  Example: \texttt{'/mount/output/trained\_model'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \textit{empty dict}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/MIC-DKFZ/nnUNet}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{isensee2020nnunet}


\end{tcolorbox}

\subsection{Omics}
\label{app:tasks:genomics_proteomics}
\begin{tcolorbox}[title={\texttt{cytopus\_db}}]
Initialize the Cytopus KnowledgeBase and generate a JSON file containing a nested dictionary with gene set annotations organized by cell type, suitable for input into the Spectra library.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{celltype\_of\_interest} (\texttt{list}): List of cell types for which to retrieve gene sets\\  Example: \texttt{['B\_memory', 'B\_naive', 'CD4\_T', 'CD8\_T', 'DC', 'ILC3', 'MDC', 'NK', 'Treg', 'gdT', 'mast', 'pDC', 'plasma']}
\item \texttt{global\_celltypes} (\texttt{list}): List of global cell types to include in the JSON file.\\  Example: \texttt{['all-cells', 'leukocyte']}
\item \texttt{output\_file} (\texttt{str}): Path to the file where the output JSON file should be stored.\\  Example: \texttt{'/mount/output/Spectra\_dict.json'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{keys} (\texttt{list}): The list of keys in the produced JSON file.
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/wallet-maker/cytopus}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{kunes2023cytopus}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{esm\_fold\_predict}}]
Generate the representation of a protein sequence and the contact map using Facebook Research's pretrained esm2\_t33\_650M\_UR50D model.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{sequence} (\texttt{str}): Protein sequence to for which to generate representation and contact map.\\  Example: \texttt{'MKTVRQERLKSIVRILERSKEPVSGAQLAEELSVSRQVIVQDIAYLRSLGYNIVATPRGYVLAGG'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{sequence\_representation} (\texttt{list}): Token representations for the protein sequence as a list of floats, i.e. a 1D array of shape L where L is the number of tokens.
\item \texttt{contact\_map} (\texttt{list}): Contact map for the protein sequence as a list of list of floats, i.e. a 2D array of shape LxL where L is the number of tokens.
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/facebookresearch/esm}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{verkuil2022esm1}


\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{hie2022esm2}


\end{tcolorbox}

\subsection{Other}
\label{app:tasks:imaging}
\begin{tcolorbox}[title={\texttt{retfound\_feature\_vector}}]
Extract the feature vector for the given retinal image using the RETFound pretrained vit\_large\_patch16 model.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{image\_file} (\texttt{str}): Path to the retinal image.\\  Example: \texttt{'/mount/input/retinal\_image.jpg'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{feature\_vector} (\texttt{list}): The feature vector for the given retinal image, as a list of floats.
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/rmaphoh/RETFound_MAE}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{zhou2023retfound}


\end{tcolorbox}

\label{app:tasks:llms}
\begin{tcolorbox}[title={\texttt{medsss\_generate}}]
Given a user message, generate a response using the MedSSS\_Policy model.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{user\_message} (\texttt{str}): The user message.\\  Example: \texttt{'How to stop a cough?'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{response} (\texttt{str}): The response generated by the model.
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/pixas/MedSSS}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{jiang2025medsss}


\end{tcolorbox}

\begin{tcolorbox}[title={\texttt{modernbert\_predict\_masked}}]
Given a masked sentence string, predict the original sentence using the pretrained ModernBERT-base model on CPU.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{input\_string} (\texttt{str}): The masked sentence string. The masked part is represented by "[MASK]"".\\  Example: \texttt{'Paris is the [MASK] of France.'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{prediction} (\texttt{str}): The predicted original sentence
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/AnswerDotAI/ModernBERT}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{warner2024modernbert}


\end{tcolorbox}

\label{app:tasks:3d_vision}
\begin{tcolorbox}[title={\texttt{flowmap\_overfit\_scene}}]
Overfit FlowMap on an input scene to determine camera extrinsics for each frame in the scene.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{input\_scene} (\texttt{str}): Path to the directory containing the images of the input scene (just the image files, nothing else)\\  Example: \texttt{'/mount/input/llff\_flower'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{n} (\texttt{int}): The number of images (frames) in the scene
\item \texttt{camera\_extrinsics} (\texttt{list}): The camera extrinsics matrix for each of the n frames in the scene, must have a shape of nx4x4 (as a nested python list of floats)
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/dcharatan/flowmap}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{smith2024flowmap}


\end{tcolorbox}

\label{app:tasks:tabular}
\begin{tcolorbox}[title={\texttt{tabpfn\_predict}}]
Train a predictor using TabPFN on a tabular dataset. Evaluate the predictor on the test set.

\vspace{.5em}
\textbf{Arguments:}
\begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{train\_csv} (\texttt{str}): Path to the CSV file containing the training data\\  Example: \texttt{'/mount/input/breast\_cancer\_train.csv'}
\item \texttt{test\_csv} (\texttt{str}): Path to the CSV file containing the test data\\  Example: \texttt{'/mount/input/breast\_cancer\_test.csv'}
\item \texttt{feature\_columns} (\texttt{list}): The names of the columns to use as features\\  Example: \texttt{['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension', 'radius error', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness error', 'concavity error', 'concave points error', 'symmetry error', 'fractal dimension error', 'worst radius', 'worst texture', 'worst perimeter', 'worst area', 'worst smoothness', 'worst compactness', 'worst concavity', 'worst concave points', 'worst symmetry', 'worst fractal dimension']}
\item \texttt{target\_column} (\texttt{str}): The name of the column to predict\\  Example: \texttt{'target'}
\end{itemize}

\vspace{.5em}
\textbf{Returns:} \begin{itemize}[topsep=0pt,parsep=-1pt,partopsep=0pt]
\item \texttt{roc\_auc} (\texttt{float}): The ROC AUC score of the predictor on the test set
\item \texttt{accuracy} (\texttt{float}): The accuracy of the predictor on the test set
\item \texttt{probs} (\texttt{list}): The probabilities of the predictor on the test set, as a list of floats (one per sample in the test set)
\end{itemize}
\tcblower
\setlength{\hangindent}{\widthof{\faGithub~}}
\faGithub~\url{https://github.com/PriorLabs/TabPFN}

\vspace{.5em}\setlength{\hangindent}{\widthof{\faFile*[regular]~}}\faFile*[regular]~\bibentry{hollmann2025tabpfn}


\end{tcolorbox}

