\subsection{Analyzing Multi-omics Data}\label{sec:td_multiomics}

%\subsubsection{Applications}
% \AB{General applications of TD in multi-omics data. Mostly focused on integrating multi-omics data such as MONTI~\cite{jung2021monti}, etc. or this recent paper~\cite{mitchel2024coordinated}.
% We might have to look into approximate TD methods such as probabilistic TD, etc. as well. 
% We should also review this paper~\cite{hore2016tensor}. We need to also look at mono-omic applications for TD such as captured clinical records~\cite{luo2017tensor}, gene expression~\cite{taguchi2022tensor}.} 

The application of tensor decomposition in multi-omics data analysis has significantly advanced in recent years, offering scalable and interpretable solutions for integrating complex biological datasets~\cite{hore2016tensor, amin2023tensor, lee2018gift, taguchi2022adapted, leistico2021epigenomic, taguchi2023tensor, wang2023probabilistic, tsuyuzaki2023sctensor, jung2021monti, taguchi2018tensor, fang2019tightly, taguchi2020tensor}. 
Compared to traditional approaches (e.g., clustering, PCA, correlation analysis), which typically analyze relationships between only one or two variables at a time, TD simultaneously captures complex interactions across multiple omics layers, preserving high-order biological structures~\cite{jung2021monti}. As multi-omics studies continue to generate increasingly high-dimensional and heterogenous data, TD techniques have evolved to address challenges related to dimensionality reduction, data sparsity, and the extraction of biologically meaningful patterns. 

One major development in tensor decomposition for multi-omics datasets has been the integration of probabilistic models to address sparsity and noise. Traditional TD methods assume homogenous data distributions, which limits their ability to model multiomics data with intrinsic variability. Probabilistic TD methods, like \texttt{SCOIT} (Single Cell Multiomics Data Integration with Tensor decomposition)~\cite{10.1093/nar/gkad570}, leverage statistical distributions such as Gaussian, Poisson, and negative binomial models to better capture variability across different omics layers. These models enhance downstream analyses such as cell clustering, gene expression integration, and regulatory network inference, making them particularly valuable for single-cell transcriptomics and epigenomics. To improve biological interpretability, non-negative TD methods enforce non-negativity constraints on the factorized components, ensuring that all contributions to latent factors remain positive. This is particularly useful in biomedical research, where feature contributions must be biologically meaningful. A prime example is \texttt{MONTI} (Multi-Omics Non-negative Tensor Decomposition for Integrative Analysis)~\cite{jung2021monti}, which selects key molecular features by identifying non-negative latent factors that drive disease progression. By preserving additive relationships in multi-omics features, \texttt{MONTI} has been effective in distinguishing molecular signatures associated with cancer subtypes, enhancing patient stratification.

Structural constraints can also be incorporated in TD to improve the separability of multi-omics data while preserving both local and global values. \texttt{BioSTD} (Strong Complementarity Tensor Decomposition Model)~\cite{gao2023biostd} uses t-SVD to factor multi-omics data from the Cancer Genome Atlas (TCGA) and enforces a strong complementarity constraint to maintain high-dimensional spatial relationships across omics layers. This improves coordination between different types of omics data and ensures that redundant features do not obscure biologically significant patterns. Such an approach is particularly beneficial when analyzing high-dimensional datasets in precision oncology, where feature redundancy can compromise the interpretability of predictive biomarkers. Multi-omics integration often faces challenges due to the unequal feature dimensions across omics layers, where transcriptomics datasets typically contain tens of thousands of features, while proteomics and metabolomics datasets are comparatively sparse. \texttt{GSTRPCA} (Irregular Tensor Singular Value Decomposition Model)~\cite{Cui2024GSTRPCA} overcomes this limitation by maintaining the original data structure while incorporating low-rank and sparsity constraints using a tensor-PCA based model. This enables effective feature selection and clustering without requiring aggressive pre-processing or feature imputation, thereby preserving biologically relevant information.

Handling missing values in multi-omics datasets remains a fundamental challenge, particularly in epigenomics, where experimental limitations often result in incomplete data across different cell types. \texttt{PREDICTD} (Parallel Epigenomics Data Imputation with Cloud-Based Tensor Decomposition)
~\cite{Durham2018PREDICTD} addresses this issue by constructing a tensor where dimensions represent cell types, genomic regions, and epigenomic marks, allowing the decomposition process to identify latent factors that capture underlying regulatory patterns. By applying CP decomposition, the method imputes missing values using latent factor reconstruction. This cloud-based framework has been validated in large-scale projects such as ENCODE and the Roadmap Epigenomics Project, highlighting its utility in functional genomics. Although tensor decomposition effectively reduces dimensionality, one of its key challenges is ensuring biological interpretability. Guided and Interpretable Factorization for Tensors~\cite{lee2018gift} (\texttt{GIFT}) enhances interpretability by incorporating functional gene set information as a regularization term within the decomposition process. It adapts Tucker decomposition to factorize with prior biological knowledge, \texttt{GIFT} ensures that extracted components correspond to meaningful gene regulatory patterns. Applied to datasets such as the PanCan12 dataset from TCGA, \texttt{GIFT} has successfully identified relationships between gene expression, DNA methylation, and copy number variations across cancer subtypes, outperforming traditional tensor decomposition approaches in terms of accuracy and scalability. 

In terms of scalability, TD has evolved to accommodate large-scale genomic data through parallelized optimization techniques. An example is \texttt{SNeCT} (Scalable Network Constrained Tucker Decomposition)~\cite{Choi2020SNeCT} which extends traditional Tucker decomposition by incorporating prior biological networks into the decomposition process. This method constructs a tensor from multi-platform genomic data while integrating pathway-based constraints to improve feature selection and classification performance. By leveraging parallel stochastic gradient descent, \texttt{SNeCT} efficiently processes large-scale datasets such as TCGA’s PanCan12 dataset. Its ability to integrate gene association networks ensures that the resulting factor matrices maintain biological relevance, making it particularly effective for cancer subtype classification and patient stratification. Standard TD approaches often treat time as a discrete variable, limiting their ability to model continuous temporal changes in biological processes. Temporal Tensor Decomposition~\cite{Shi2024TEMPTED} (\texttt{TEMPTED}) overcomes this limitation by integrating time as a continuous variable within the decomposition framework. This method enables the characterization of dynamic transcriptional and epigenomic changes across different time points, making it particularly useful for analyzing developmental processes and disease progression. \texttt{TEMPTED} also accounts for non-uniform temporal sampling and missing data, providing a more comprehensive approach to longitudinal multi-omics analysis. More recently, Mitchel \textit{et al.} have introduced a computational approach called single-cell interpretable tensor decomposition (\texttt{scITD}), that utilizes tensor decomposition to analyze single-cell gene expression data~\cite{mitchel2024coordinated}. This approach enables the identification of coordinated transcriptional variations across multiple cell types, facilitating the discovery of common patterns among individuals, in turn facilitating patient stratification.

As tensor decomposition methodologies for multi-omics datasets continue to evolve, future advancements will likely focus on improving real-time data analysis capabilities, refining computational efficiency, and expanding the applicability of TD in emerging fields such as spatial transcriptomics and single-cell multi-omics. The continued refinement of these methods will be critical for advancing precision medicine and uncovering novel insights into complex biological systems.


% Precision medicine has been revolutionized by genetic analysis, enabling clinicians to identify individuals at high risk of disease and implement targeted, genetically informed interventions. However, multi-omics data, encompassing genomics, phenomics, transcriptomics, proteomics, metabolomics, etc. present unique challenges: they are inherently noisy, sparse, high-dimensional, and plagued by batch effects. Although there are emerging methods for integrating these diverse modalities, there is a lack of a standardized, user-friendly approach for multi-omics integration~\cite{tarazona2021undisclosed}. Moreover, current computational models still struggle to accurately predict phenotype outcomes from genotype data due to the complex, multi-layered interactions among genetic variants, epigenetic modifications, and environmental influences~\cite{ritchie2015methods}. Tensor decompositions is particularly well-suited for analyzing such data because it captures the multidimensional relationships within these datasets effectively. A tensor is a multi-dimensional array that can naturally represent relationships across different omics layers. Instead of analyzing two-dimensional matrices (e.g., samples × genes), tensors allow us to model higher-order interactions~\cite{chang2021gene}. In comparison to traditional approaches (e.g., clustering, PCA, correlation analysis), which typically analyze relationships between only one or two variables at a time, TD, however, simultaneously captures complex interactions across multiple omics layers, preserving high-order biological structures~\cite{jung2021monti}.  Hence, TD methods have been widely applied in both single-omic analysis such as in genomics~\cite{hore2016tensor, amin2023tensor}, transcriptomics~\cite{lee2018gift, taguchi2022adapted}, epigenomics~\cite{leistico2021epigenomic}, single-cell analysis~\cite{taguchi2023tensor, wang2023probabilistic, tsuyuzaki2023sctensor}, as well as in multi-omics studies~\cite{jung2021monti, taguchi2018tensor, fang2019tightly, taguchi2020tensor}. 


% The triumph of utilizing genetics in precision medicine is the ability to flag individuals at very high risk of disease, with the ability to enact specific gene-informed medical management. Despite their promise in identifying biologically relevant outputs, omics data are noisy, sparse (i.e., contain many zeros), high-dimensional, and contain batch effects. Particularly, there are no standard or user-friendly tools that exist to integrate or link data from all these modalities. The area under the curve (AUC - a measure of the predictive power) for many genetic risk scores has plateaued below 0.7 \os{(Shehab: Citations?)}, demonstrating the limits in linking genomes and outcomes based on classical computational methods. Multi-omics datasets, which integrate data from various omics layers such as genomics, transcriptomics, proteomics, and metabolomics, are inherently high-dimensional and complex. Tensor decomposition is particularly well-suited for analyzing such data because it captures the multidimensional relationships within these datasets effectively. A tensor is a multi-dimensional array that can naturally represent relationships across different omics layers. Instead of analyzing two-dimensional matrices (e.g., samples × genes), tensors allow us to model higher-order interactions. In comparison to traditional approaches (e.g., clustering, PCA, correlation analysis), which typically analyze relationships between only one or two variables at a time, tensor decomposition, however, simultaneously captures complex interactions across multiple omics layers, preserving high-order biological structures. 


% Tensor decomposition has been applied in multi-omics data processing, including integrative analyzes, disease or phenotype subtyping, feature extraction, biomarker discovery, gene regulation studies, as well as temporal or longitudinal data analyses \cite{luo2017tensor}. Hore et al. introduced a Bayesian method designed to analyze multi-tissue gene expression data, with a focus on uncovering gene networks associated with genetic variation by decomposing the three-dimensional matrix (tensor) of gene expression measurements into latent components \cite{hore2016tensor}. The study uncovered several gene networks related to genetic variation, providing insights into the genetic architecture of gene expression across multiple tissues, thus highlighting the potential of tensor decomposition methods in elucidating complex gene regulatory mechanisms. Jung et al. developed MONTI (Multi-Omics Non-negative Tensor Decomposition for Integrative Analysis), a multi-omics non-negative tensor decomposition framework tailored for integrative analysis of multi-omics data \cite{jung2021monti}. Through integrating multi-omics data in a gene-centric manner, MONTI enabled the detection of cancer subtype-specific features and other clinical features, which in turn facilitates the identification of biologically meaningful associations. In another study, Wang et al. introduced SCOIT (Single Cell Multiomics Data Integration with Tensor decomposition), a probabilistic tensor decomposition framework designed to extract embeddings from single cell multiomics data \cite{10.1093/nar/gkad570}. The framework was applied to eight single-cell multi-omics datasets from various sequencing protocols, covering DNA methylation data, RNA expression data, proteomics data, and chromatin accessibility data. SCOIT achieved superior clustering accuracy compared to nine state-of-the-art methods on heterogeneous datasets. More recently, Mitchel et al. have introduced a computational approach called single-cell interpretable tensor decomposition (scITD), a computational method that utilizes tensor decomposition to analyze single-cell gene expression data \cite{mitchel2024coordinated}. This approach enables the identification of coordinated transcriptional variations across multiple cell types, facilitating the discovery of common patterns among individuals, in turn facilitating patient stratification.


% \subsubsection{Challenges}
% % \AB{Phase transition/computational hardness discussions with plots following from the notebook shared with you}

% Multi-omics data analysis presents several challenges, particularly in handling, integration, and interpretation. One of the major difficulties lies in the computational complexities associated with large-scale multi-omics datasets. These datasets are often high-dimensional high dimensions due to the large number of features across multiple omics layers. Tensor decomposition methods, widely used for multi-omics integration, often have difficulties with scalability due to increasing computational demands. Therefore, efficient algorithms and scalable frameworks are required to handle these datasets effectively, as tensor decomposition can be both computationally intensive and time-consuming. 

% To illustrate these computational challenges, we performed tensor decomposition on a multi-omics dataset with a tensor size of (264, 10772, 31), where the dimensions represent the number of samples (264), genomic features (10,772), and metabolomic features (31). Figure~\ref{fig:tensor_decomposition} presents the results of this analysis. The first plot shows execution time (in seconds) as the decomposition rank varies from 2 to 30, revealing an initial increase followed by a decline after rank 14. The second plot illustrates memory usage (in MB), which steadily rises with increasing decomposition rank. These results highlight a critical trade-off between rank selection and computational efficiency, emphasizing the need for optimized feature selection strategies to balance computational feasibility and biological relevance when applying tensor decomposition to multi-omics data. Another key challenge is the heterogeneity of multi-omics data. These datasets comprise diverse data types, including continuous, binary, and categorical variables, complicating integration efforts. This heterogeneity complicates the integration process, as traditional tensor decomposition methods such as Tucker and CANDECOMP / PARAFAC (CP) may not adequately model complex interactions and different data types \cite{xu2015bayesian}. Beyond computational and integration challenges, interpreting the latent factors extracted from tensor decomposition in a biologically relevant context can be challenging. Moreover, multi-omics data are often noisy and contain outliers. This can distort the results of the tensor decomposition. However, there are new methods such as SCOIT \cite{10.1093/nar/gkad570} which attempt to address this challenge by incorporating various distributions to model noise and sparsity in the data. 

% Finally, multi-omics data often contain missing values or sparse entries, which can significantly impact the performance of tensor decomposition methods. Several traditional approaches require complete data or extensive preprocessing to handle missing data, which can be time-consuming and may introduce biases \cite{xu2015bayesian, taguchi2021tensor}. While tensor decomposition provides a powerful framework for multi-omics integration,  overcoming computational constraints, handling data heterogeneity, improving interpretability, mitigating noise, and addressing missing data challenges remain critical areas for future research and methodological advancements.

% \begin{figure}[ht]
%     \centering
%     \subfloat{\includegraphics[width=0.45\textwidth]{figures/memory_usage.png}}
%     \hspace{0.5cm} % Adds some spacing between images
%     \subfloat{\includegraphics[width=0.45\textwidth]{figures/execution_time.png}}
%     \caption{Memory usage and execution time for tensor decomposition for random dataset. \os{(Shehab: Can we add more details? How the dataset was generated?)}}
%     \label{fig:tensor_decomposition}
% \end{figure}


