\section{Related Works}
\subsection{Domain Generalization}
Domain generalization (DG) is proposed to construct models that perform well on unseen domains without access to their data during training. Recent advancements in DG have achieved remarkable progress~\cite{wang2022generalizing}, with approaches generally categorized into three main groups: data manipulation~\cite{shankar2018generalizing,yue2019domain,zhou2021domain}, representation learning~\cite{ghifary2015domain,li2018domain,shao2019multi,sicilia2023domain,zhang2022towards}, and learning strategy~\cite{chen2022discriminative,tian2022neuron,carlucci2019domain}. While our method falls within the realm of representation learning, it significantly differs from existing domain adversarial learning-based methods. Specifically, we introduce an asymmetric adaptive clustering constraint, enabling the model to better capture the nuances of real-world drug response data. The most closely related work is the SSDG\cite{jia2020single}, a single-sided domain generalization approach for face anti-spoofing. However, our method distinguishes itself through the incorporation of contrastive learning-based asymmetric clustering constraints and the proposal of a latent independent projection module, allowing the encoder to learn non-redundant features and thereby enhance the model's generalization capability.


\subsection{Drug Response Prediction}
The prediction of clinical drug responses has drawn considerable attention from machine learning community. Some studies employed patient drug response data to fine-tune models initially trained on cell line datasets. For instance, CODE-AE~\cite{he2022context} is based on domain separation network~\cite{bousmalis2016domain}to extract shared features between cell lines and patients. It was trained on cell line drug sensitivity data and then used to predict the drug response for tumor patients. Precily ~\cite{chawla2022gene} integrated signaling pathway and drug feature to predict drug responses \textit{in vitro} and \textit{in vivo}. In contrast, drug response prediction at the single-cell level is still in its infancy, due to the scarcity of drug response data in single-cell context. Only a few studies leveraged domain adaptation between bulk RNA-seq (source domain) and scRNA-seq (target domain) data to predict drug sensitivity of individual cells. For example, scDEAL ~\cite{chen2022deep} aligned the bulk RNA-seq and scRNA-seq features by minimizing the maximum mean discrepancy (MMD)~\cite{gretton2012kernel} in the latent space for single-cell drug response prediction. SCAD~\cite{zheng2023enabling} adopted adversarial domain adaptation to learn drug-gene signatures from the GDSC dataset~\cite{yang2012genomics} for inferring drug sensitivity in single cells. Distinct from the existing methods, we aim to establish a predictive model with superior generalizability across distinct target domains, spanning both single-cell and patient-level datasets, without access to target-domain data during training.