\section{Related Work}
\subsection{Deep Learning and Antibiotic Resistance}

Several recent works \cite{chakraborty_deep_2022,kim_machine_2022, ren_prediction_2022}, have identified both the potential and the challenges that exist within the intersection of deep learning and AR prediction, and significant effort is already underway in the area. 

Many methods \cite{tharmakulasingam_explainable_2022, tharmakulasingam_rectified_2023, macesic_predicting_2020, jin_predicting_2024, kuang_accurate_2022} utilize full gene sequences as input data. These methods either utilize dimensionality reduction on the sequences or convert the sequences to genes through existing algorithms. Gene sequences are rich in information, but they require whole-genome sequencing, which is often impractical in clinical routine.

Some attempts \cite{her_pan-genome-based_2018, hyun_machine_2020, kavvas_machine_2018} have been made to identify important AR genes. In  \cite{her_pan-genome-based_2018}, they use a genetic algorithm to select the best performing genes. However, to predict multiple resistances, their method requires a large subset of genes, not achievable by a single PCR test. Others \cite{hyun_machine_2020, kavvas_machine_2018} manage to identify new AR genes along with important existing ones, however, only while utilizing full genome sequences.

\subsection{Feature Selection Methods}

Selecting an optimal subset of genes for testing, based on the predictive performance of a model, can be framed as a black-box optimization problem, where the prediction model serves as the black box. A wide range \cite{sxue_survey_evolutionary_2016} of methods has been proposed to address such problems, including evolutionary algorithms, search-based techniques and heuristic approaches. For instance, evolutionary algorithms have been explored \cite{salimans2017evolutionstrategiesscalablealternative} as alternatives to reinforcement learning for optimizing policy parameters. Several black-box optimization techniques \cite{pudjihartono_review_feature_selection_2022} and heuristic methods \cite{ALIREZANEJAD20201173} have been used to identify key gene features. Although adaptable to our problem, exploring these methods fully is beyond the scope of this study.


\subsection{Active Feature Acquisition}

The challenge of selecting the optimal subset of genes also falls under the umbrella of active feature acquisition (AFA) problems. The majority of research \cite{nam_reinforcement_2021, li_active_2021, fahy_dynamic_2019, yin_reinforcement_2020, shim_joint_2018} in that area considers the problem as a Markov Decision Process (MDP), where at each state, one can either decide to observe another feature or make a prediction. The INVASE framework \cite{yoon_invase_2018} observes all features before selecting the relevant features, which requires all tests to be performed and eliminates the potential gain of only selecting a subset. ODIN \cite{zannone_odin_2019} and many other methods \cite{yu_deep_2023, yin_reinforcement_2020, li_active_2021, nam_reinforcement_2021, shim_joint_2018} assume that once an action has been taken, the new state and the result of that action can be observed (often for a cost). This additional information improves subsequent choices but is not applicable to our problem since doctors would not test genes one at a time but instead test for a set of genes all at once.