
MotifBench comprises 30 test problems.
We specify these problems in \Cref{table:test_cases}, which has the following columns:
\begin{itemize}
\item{\textbf{PDB ID}: The Protein Data Bank identifier of the experimentally characterized structure from which the motif extracted.}
\item{\textbf{Group:} The problem group into which the motif is assigned. This grouping is defined based on the number of contiguous segments that comprise the motif: Group 1 motifs have only one segment, Group 2 motifs have 2 segments, and Group 3 motifs have 3 or more segments. We provide an additional visualization of motifs by group in \Cref{fig:group_1_problems,fig:group_2_problems,fig:group_3_problems}. }
\item{\textbf{Length}: The number of residues required in each scaffold.}
\item{\textbf{Motif Residues}: The chain ID and indices of residues that comprise the motif. Discontiguous residue ranges are separated by semicolons.}
\item{\textbf{Redesign Indices}: The indices of residues within motif segments for which the amino acid type is not constrained to match its identity in the reference protein, and will be ``redesigned'' during inverse-folding. This column is included because in cases where side-chain atoms are not involved in protein function, the motif-scaffolding problem may be made easier by allowing alternative amino acid types to be chosen for these positions during fixed-backbone sequence design.  This field is blank when no positions are allowed to be redesigned.}
\item{\textbf{Description:} A short explanation and reference for the source of the motif.}
\end{itemize}


\begin{footnotesize}
%\begin{longtable}{|l|l|l|l|p{3.2cm}|p{3.2cm}|p{3.2cm}|}
\begin{table}[htbp]
\footnotesize
\caption{MotifBench test-cases.}\label{table:test_cases} 
%\caption[]{Benchmark V1.0 test-cases (continued)} \\ \hline
\begin{tabularx}{\textwidth}{|l|l|l|l|p{3.2cm}|p{3.3cm}|p{3.2cm}|}
\hline
\# & PDB  ID & Group & Length & Motif Residues & Positions where residue type may be designed & Description \\ \hline
\hline
1  & 1LDB & 1 & 125 & A186-206 & & Lactate dehydrogenase \citep{hayes2024simulating} \\ \hline
2  & 1ITU & 1 & 150 & A124-147 &  & Renal dipeptidase \citep{hayes2024simulating} \\ \hline
3  & 2CGA & 1 & 125  & A184-194	& & Strained chymotrypsinogen loop that undergoes conformation change \citep{wang1985bovine} \\ \hline
4  & 5WN9 & 1 & 75 & A170-189 & A170-175;A188-189 & RSV G-protein 2D10 site \citep{yang2021bottom} \\ \hline
5  & 5ZE9 & 1 & 100 & A229-243 & & P-loop \citep{hayes2024simulating} \\ \hline
6  & 6E6R & 1 & 75 & A25-35 & A25-35 & Ferredoxin Protein \citep{trippe2022diffusion} \\ \hline
7  & 6E6R & 1 & 200 & A25-35 & A25-35 & Ferredoxin Protein \citep{trippe2022diffusion} \\ \hline
8  & 7AD5 & 1 & 125 & A99-113 &  & Orphan protein \citep{wu2022omegafold} \\ \hline
9  & 7CG5 & 1 & 125 & A6-20 &  & Orphan protein \citep{wu2022omegafold} \\ \hline
10  & 7WRK & 1 & 125 & A80-94 & & Orphan protein \citep{wu2022omegafold} \\ \hline
\hline
11  & 3TQB & 2 & 125 & A37-51;A65-79 & & Parallel beta strand and loop \citep{tkaczuk2013structural} \\ \hline
12  & 4JHW & 2 & 100 & F63-69;F196-212 & F63;F69;F196;F198; F203;F211-212 & RSV F-protein Site 0 \citep{sesterhenn2020novo} \\ \hline
13  & 4JHW & 2 & 200 & F63-69;F196-212 & F63;F69;F196;F198; F203;F211-212  & RSV F-protein Site 0 \citep{sesterhenn2020novo} \\ \hline
14  & 5IUS & 2 & 100 & A63-82;A119-140 & A63;A65;A67;A69;\ A71-72;A76;A79-80;A82;A119-123;\ A125;A127;A129-130;\ A133;A135;A137-138;A140 & PD-L1 binding interface on PD-1 \citep{watson2022broadly} \\ \hline
15  & 7A8S & 2 & 100 & A41-55;A72-86 &  & Orphan protein \citep{wu2022omegafold} \\ \hline
16  & 7BNY & 2 & 125 & A83-97;A111-125 &  & Orphan protein \citep{wu2022omegafold} \\ \hline
17  & 7DGW & 2 & 125 & A22-36;A70-84 &  & Orphan protein \citep{wu2022omegafold} \\ \hline
18  & 7MQQ & 2 & 100 & A80-94;A115-129 & & Orphan protein \citep{wu2022omegafold} \\ \hline
19  & 7MQQ & 2 & 200 & A80-94;A115-129 &  & Orphan protein \citep{wu2022omegafold} \\ \hline
20  & 7UWL & 2 &  175 & E63-73;E101-111 & E63-73;E101-103;E105-111 & IL17-RA interface to IL17-RB \citep{wilson2022organizing} \\ \hline
\hline
21  & 1B73 & 3 & 125 & A7-8;A70;A178-180 & A179 & Glutamate racemase active site \citep{ribeiro2018mechanism} \\ \hline
22  & 1BCF & 3 & 125 & A18-25;A47-54;A92-99;A123-130  & A19-25;A47-50;A52-53; A92-93;A95-99;A123-126;A128-129 & Di-iron binding motif \citep{watson2022broadly} \\ \hline
23  & 1MPY & 3 & 125 & A153;A199;A214;A246;\ A255;A265 & & Catechol deoxygenase active site \citep{ribeiro2018mechanism} \\ \hline
24  & 1QY3 & 3 & 225 & A58-71;A96;A222  & A58-61;A63-64;A68-71  & GFP pre-cyclized state \citep{hayes2024simulating}.\footnote{The deposited structure includes an inactivating mutation (R96A) mutation.  Following \citep{hayes2024simulating}, this mutation must be reverted in generated scaffolds.} \\ \hline
25  & 2RKX & 3 & 225 & A9-11;A48-50;A101;A128;\ A169;A176;A201;A222-224 & A10;A49;A223 & De novo designed Kemp eliminase \citep{rothlisberger2008kemp} \\ \hline
26  & 3B5V & 3 & 200 & A51-53;A81;A110;A131;\ A159;A180-184;A210-211;A231-233 & A52;A181;A183;A232 & De novo designed retro-aldol enzyme \citep{jiang2008novo} \\ \hline
27  & 4XOJ & 3 & 150 & A55;A99;A190-192 & A191  & Trypsin catalytic triad and oxyanion hole \citep{du2024conformational} \\ \hline
28  & 5YUI & 3 & 75 & A93-97;A118-120;A198-200 & A93;A95;A97;A118;A120  & Carbonic anhydrase active site \citep{watson2022broadly} \\ \hline
29  & 6CPA & 3 & 200 & A69-72;A127;A196;\ A248;A270 & A70-71 & Carboxypeptidase active site \citep{ribeiro2018mechanism} \\ \hline
30  & 7UWL & 3 & 175 & E63-73;E101-111;E132-142;E165-174 & E63-73;E101-103;E105-111;E132-142;E165-174 & IL17-RA interface to IL17-RB \citep{wilson2022organizing} \\ \hline
\end{tabularx}
\end{table}
\end{footnotesize}

These test cases are derived from several sources.
\begin{itemize}
    \item{Eight problems are motifs considered in published protein design papers collected in an earlier benchmark set \citep[Table S9 of ][]{watson2022broadly}. Compared to the 25 problems in the this earlier benchmark, we drop one problem that involves multiple chains (6VW1), two problems for which we observed ESMFold to dramatically increase the success rate compared to AlphaFold2 with no MSA input (1QJG, 1PRW), and several problems that are already readily solved by existing methods.  The problems we include this set are 5WN9 (4), 6E6R (6 and 7), 4JHW (12 and 13), 5IUS (14), 1BCF (22), and 5YUI (28).}
\item Eight problems are fragments of ``orphan'' proteins selected in \citep[Table S11 of ][]{watson2022broadly} from structures in \citep{wu2022omegafold} with little sequence or structure homology other known proteins.  These were selected from among an initial set of 25 problems for structural diversity and greater difficulty.  These problems are 7AD5 (8), 7CG5 (9), 7WRK (10), 78AS (15), 7BNY (16), 7DGW(17), 7MQQ (18-19).
\item{Four problems are obtained from the test cases considered in another protein design paper \citep{hayes2024simulating}.  These problems are 1LDB (1), 1ITU (2), 5ZE9 (5), and 1QY3 (25).  Problem the PDB entry 1QY3 is crystal structure of the green fluorescent protein (GFP) with a mutation (Arg $\rightarrow$ Ala at residue 96) which prevents the formation of the GFP fluorophore.  Our motif definition involves the reversion of this mutation to the native Arginine.}
\item{One problem, 2CGA (3) was identified as a difficult single-segment case in a strained conformation; this motif is loop in chymotrypsinogen that has a documented conformation change after a cleavage of its native scaffold and activation to chymotrypsin \citep{wang1985bovine}.}
\item{One problem, 3TQB (11) was selected because it includes a parallel beta strand structure not elsewhere represented in the benchmark \citep{tkaczuk2013structural}.}
\item{Six problems are additional enzyme active sites: 
\begin{itemize}
    \item{Three are from natural enzymes in the ``Mechanism and Catalytic Site Atlas'' \citep{ribeiro2018mechanism}: 1B73 (21), 1MPY (23) and 6CPA (29). }
    \item{Two are active sites of \emph{de novo} designed enzymes, 2RKX (22) \citep{rothlisberger2008kemp} and 3BV5 (26) \citep{jiang2008novo}.  }
    \item{One is the active residues of a serine protease described in a structural study \citep{du2024conformational}: 4XOJ (27).}
\end{itemize}
For the motifs from natural enzymes, the motif residues are chosen from among those that are documented as involved in the catalytic mechanism and or can be observed from the experimental structure to make polar contacts within the annotated active site.  When there is a gap between involved residues of no more than three amino acids in the sequence these residues are also included as part of the motif, but with these positions marked as redesignable.  
Residues are also marked as redesignable when only their backbone atoms appear to be involved in the mechanism.
For the motifs from \emph{de novo} design papers, the motif residues are chosen to be those that were chosen when constructing the putative active site.
 }
\item {Two problems, 7UWL (20 and 30), are segments of a binding interface of IL17-RA that interacts with IL17-RB \citep{wilson2022organizing}.
The native interface involves four segments; two of these segments are included in problem 20 and all four segments are included in problem 30.
These problems were chosen for their difficulty and the potential therapeutic relevance of novel scaffolds that reconstitute this interface.}

\end{itemize}

We prioritized the following characteristics when selecting the motifs above.
\begin{itemize}
    \item{\textbf{Relevance to design}: Most of the test cases are minimal, biochemically active substructures obtained from or characteristic of protein design problems.}
    \item{\textbf{Diversity}: The motifs in the benchmark exhibit a range of characteristics:
    \begin{itemize}
        \item The number of residues in the motifs ranges from as few as five to several dozens.  Because the lengths demanded in a scaffold can be a significant factor for the performance of some methods, for three motifs (4JHW, 6E6R, and 7MQQ) we include two problem variations with different scaffold lengths.
        \item The number of contiguous segments ranges from one to eight.
        \item The number of residues demanded of scaffolds varies from 75 to 225. 
        \item The secondary structure of motifs across problems includes helical, strand, and loop segments, and combinations thereof.
        \end{itemize}}
\item{\textbf{Inclusion of ``Orphan'' motifs}:  Eight ``orphan protein'' motifs are included to ease assessment of possible dependence of measured performance on overfitting by data-driven motif-scaffolding methods and the MotifBench evaluation pipeline.  In particular, evaluations may be artificially inflated for motifs that are structurally conserved (and therefore more highly represented in protein sequence databases and in the PDB) if the corresponding native sequence and then native motif structure are readily predicted by ProteinMPNN and ESMFold as a result of memorization even without a suitable supporting scaffold.  By contrast, motifs extracted from orphan proteins are less likely to be highly represented in these datasets and therefore are less likely to be susceptible to this bias.} 
\end{itemize}

\section{Baseline performance and analysis of reference scaffolds}
\label{section:baseline_section}
We demonstrate MotifBench using the well-established motif-scaffolding method RFdiffusion \citep{watson2022broadly} to assess various aspects of the evaluation pipeline. Key ingredients are summarized below:
\begin{itemize}
    \item \textbf{Performance}: We provide an running example for practitioners to start with MotifBench and demonstrated the challenge of designing cases in MotifBench.
    \item \textbf{Stability}: We showcased MotifBench is robust to the randomness from multiple replicates of both design and evaluation is on a low level.
    \item \textbf{Sensitivity to forward folding method}: We demonstrate a relatively low sensitivity to the choice of two commonly-used forward folding methods, ESMFold and AlphaFold2 (no MSA input), for validating designed scaffolds.
    \item \textbf{Rationality of curated cases}: We show that there exist reasonable solutions for curated cases in MotifBench by running evaluation on reference proteins from which the motifs were defined.
\end{itemize}

Altogether, these findings (1) suggest that substantial improvements could be made to motif-scaffolding methods and (2) point to limitations of the MotifBench V0 evaluation that might be addressed by future versions.

\paragraph{Demonstration of MotifBench and performance evaluation with RFdiffusion scaffolds.}

To assess the feasibility and difficulty of the MotifBench test cases, we evaluated scaffold sets produced with RFdiffusion \citep{watson2022broadly}. We chose RFdiffusion for both its popularity among motif-scaffolding methods and its good performance relative to more recent motif-scaffolding methods \citep[see e.g.][]{wu2024practical,lin2024out,yim2024improved}.

We generated 100 scaffolds for each of the 30 test cases using the open-source RFDiffusion implementation with default hyperparameter settings and contigs under the specification of MotifBench.  Generation required approximately 30 GPU hours across a variety of GPU types on a university cluster. We then evaluated these scaffolds with MotifBench. We provide full details of this generation and code to replicate generation of the scaffolds sets at \url{github.com/blt2114/motif_scaffolding/benchmark/example} and have uploaded the scaffolds and associated metadata in the format required by MotifBench accompanied with all evaluated results to zenodo at \url{https://zenodo.org/records/14731790} for replicability.

RFdiffusion provided at least one solution on 16 of the 30 cases, which indicates greater difficulty of MotifBench test cases as compared to earlier the benchmark set introduced in \citep{watson2022broadly};
on this earlier benchmark set, RFdiffusion has been found by \citet{watson2022broadly, yim2024improved,zhang2023protein} to provide at least one solution in 20 of 24 single-chain test cases.  Across the present test cases, the mean number of unique solutions across problems was 8.83, but 5 or more unique solutions were found for only 7 test cases. The mean novelty across cases was 0.19 and the overall MotifBench score was 28.05. RFdiffusion identified solutions for 7/10 cases for cases in both group 1 and group 2 but 2/10 for group 3 motifs, highlighting challenges for scaffolding motifs including multiple discontinuous segments (\Cref{table:RFdiff_replicates}).


\paragraph{Stability against stochasticity in MotifBench evaluation and scaffold generation.} A challenge in motif-scaffolding evaluation is stochasticity arising from (1) the sequence design step of the evaluation pipeline and (2) the scaffold generation. To assess the stability and reproducibility of MotifBench metrics, we evaluate the variability of MotifBench results across replicates due to each source of stochasticity:
{\begin{itemize}
    \item \textbf{Stochasticity across evaluations on a single scaffold set}. We repeated the evaluation on the initial set of the generated scaffold described above four times. See ``Variability from Benchmark'' in \Cref{table:RFdiff_replicates}.
    \item \textbf{stochasticity across scaffold generations}. We also replicated the RFdiffusion scaffold generation procedure and evaluated thenceforth. See ``Variability from Scaffolds'' in \Cref{table:RFdiff_replicates}.
\end{itemize}}

We observed variability from both sources, with greater variability across replicate scaffold generations in which case both sources of stochasticity contribute. The standard deviation of MotifBench score was 0.39 for multiple evaluations and 0.47 for multiple scaffold sets. Rougnly, this result suggests that differences in the MotifBench score of less than about 1 may not be sufficient to confidently conclude one method outperforms another based on a single evaluation.

\paragraph{Sensitivity of MotifBench to different structure prediction methods.} We explored how the choice of structure prediction method (ESMFold vs. AlphaFold2 without MSA input) affects MotifBench results. While performance was similar for most cases, we observed notable variability for a few cases between the two methods (\Cref{table:RFdiff_folding_method}); this discrepancy has a non-trivial impact on the MotifBench score (28.1 with ESMFold vs. 22.5 with AF2), yet does not imply superiority of one method over the other, as it may reflect either false positives from ESMFold or false negatives from AlphaFold2 (in non-MSA mode) or both.

\paragraph{Evaluation of reference scaffolds from experimental structures and the feasibility of ``unsolved'' problems.}
We next consider the extent to which the failure of RFdiffusion to identify solutions for 14/30 test cases owes to limitations in scaffold generation that could be addressed by improved methods versus limitations of the MotifBench evaluation; though each motif comes from an experimentally characterized structure, it remains possible that neither this experimental \emph{reference} scaffold nor any other scaffold could pass the MotifBench evaluation.  To assess these possibilities we evaluated the reference scaffold for each motif according to MotifBench (\Cref{table:reference_RFdiff}) and compare the overlaps in the test cases for which RFdiffusion succeeds or fails with the cases for which the reference scaffold succeeds or fails in a contingency table (\Cref{table:reference_RFdiff_contingency}).

The reference scaffold was found to be a ``success'' in the majority (20/30) cases (\Cref{table:reference_RFdiff}).  RFdiffusion failed to identify a solution for 6 of these successes, indicating the low success rates are at least partly due to design limitations rather than evaluation. Surprisingly, 2 of these 6 cases for which the reference scaffold is a success but RFdiffusion identifies no solutions are motifs derived from \textit{de novo} proteins that were scaffolded by a pre-deep learning method, 2RKX \citep{rothlisberger2008kemp} and 3B5B \citep{jiang2008novo}.  This result indicates the recent wave of deep-learning methods may not improve uniformly upon techniques used for case-specific designs more than 15 years ago.  

For the remaining 8 unsolved cases the reference scaffold fails to pass the MotifBench evaluation.
Upon inspection, the majority of these failure cases exhibit a high proportion of loop regions, suggesting a limitation of the evaluation pipeline to precisely reconstruct flexible regions in these motifs.
This limitation could owe to the sequence design or structure prediction steps, and may be addressed by improvements in methodologies for these tasks, and could be incorporated into future versions of MotifBench.
However, this failure is not conclusive evidence for the impossibility of identifying successes for these cases even with the MotifBench evaluation; indeed, for 2 of the cases for which the reference scaffold is not a success RFdiffusion nonetheless identifies passing scaffolds.
